---
title: "R Notebook"
output: html_notebook
---

This notebook is code to generate the results for Kelly et al. 2022.


# Required packages

```{r}
```


# Paths to necessary files

```{r filenames}

# custom functions for processing data
eqtl_fncs.f = "./eqtl_fncs.R"

# coloc functions for running credible sets
coloc.f = "./claudia.R"

# map between samples and populations
pop_samps.f = "./misc_data/pop_samples.in5M.txt"

# proteing coding and lincRNA genes from gencode v19
gencode.pc_linc.f = "./misc_data/gencode.v19.genes.v7.patched_contigs.pc_linc.bed.gz"

# maps between gencode IDs and for various gencode versions
v19_genes.f = "./misc_data/gencode.v19.genes.txt"
v26_genes.f = "./misc_data/gencode.v26.genes.txt"
v29_genes.f = "./misc_data/gencode.v29.genes.txt"

gencode.v26.f = "./misc_data/gencode.v26.GRCh38.genes.bed"

# map between gencode ids and gene name
gene_name.f = "./misc_data/gencode.v19.gene_id.gene_name.txt"

# FDR-significant 

```


# Supplementary data

```{r}

set.seed(7)

## load useful functions and packages
source(eqtl_fncs.f)
source(coloc.f)
rename <- dplyr::rename
select <- dplyr::select

# population_data
pop_samps = read.delim(pop_samps.f,
                       header=F, stringsAsFactors=F, col.names=c("sample","pop"))
pops = unique(pop_samps$pop)
names(pops) = pops

# protein_coding and lncRNA genes
gencode.pc_linc = read.delim(gencode.pc_linc.f,
                             header=F,
                             col.names=c("chr","start","end","strand","v19_id","gene_type")) %>%
  mutate(gene_id = strip_ensembl(v19_id))

# mapping between different GENCODE IDs
gene_map = read_tsv(v19_genes.f, col_names="v19_id") %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  merge(read_tsv(v26_genes.f, col_names="v26_id") %>%
          mutate(gene_id = strip_ensembl(v26_id)), all=T) %>%
  merge(read_tsv(v29_genes.f, col_names="v29_id") %>%
          mutate(gene_id = strip_ensembl(v29_id)), all=T)

# pos_map = read_tsv("../geno/5M.imputed.rna_samps.biallelic.hg38.sort.txt.gz",
#                    col_names=c("chr","pos_hg19","pos_hg38"))

gencode.v19.tss = gencode.pc_linc %>%
  mutate(tss = ifelse(strand=="+", start, end)) %>%
  select(chr, tss, v19_id)

gencode.v26.tss = read_tsv(gencode.v26.f,
                           col_names=c("chr","start","end","strand","v26_id","type")) %>%
  mutate(tss = ifelse(strand=="+", start, end)) %>%
  select(chr, tss, v26_id)

# map between gene ids and gene names
gene_name = read_tsv(gene_name.f,
                     col_names=c("v19_id", "gene_name")) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  select(-v19_id)

# ## get sample IDs for each population
# samps.list = lapply(pops, function(pop) readLines(paste0("../misc/pops/in5M/", pop, ".txt")))
# names(samps.list) = pops

# samps = pop_samps$sample

# FDR-significant GTEx v8 results
gtex.v8.fdr05 = read_tsv("../gtex/v8/GTEx_Analysis_v8_eQTL/Whole_Blood.v8.signif_variant_gene_pairs.txt.gz") %>%
  mutate(gene_id = strip_ensembl(gene_id)) %>%
  separate(variant_id, into=c("chr", "pos_hg38", "v8_ref", "v8_alt", "foo"), sep="_")
# add hg19 info
gtex.v8.fdr05 = liftover_call(unique(tabixify(gtex.v8.fdr05$chr,
                                              gtex.v8.fdr05$pos_hg38,
                                              gtex.v8.fdr05$pos_hg38)),
                              "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  separate(new, into=c("chr", "range"), sep=":") %>%
  separate(range, into=c("pos_hg19", "foo"), sep="-") %>%
  separate(old, into=c("chr", "range"), sep=":") %>%
  separate(range, into=c("pos_hg38", "foo"), sep="-") %>%
  select(-foo) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(gtex.v8.fdr05, all.y=T)

# GTEx v8 expression
gtex.v8.med_gex = read_tsv("../misc/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz",
                           skip=2) %>%
  dplyr::rename(v26_id = Name, gene_name = Description) %>%
  mutate(gene_id = strip_ensembl(v26_id))

gtexv8_snps = readLines("../gtex/v8/GTEx_Analysis_v8_tQTL_all_associations_Whole_Blood.tested_snps.txt.gz")

```


# eQTLs mapped for 0-30 PEER factors. Now to find eQTLs with FDR<0.05:

```{r}

fdr=0.05

# gemma_cols = readLines("gemma_cols.lmm4.txt")\
gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allel1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

gemma_stats = data.frame()
for (peer in paste("peer", c(0,1,2,3,4,5,10,15,20,25,30), sep="")){
  
  gemma.f = paste(c("round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.", peer, ".resid.gemma.100kb.lmm4.r2.txt.gz"), collapse="")
  
  gemma = fread(gemma.f, col.names=gemma_cols, na.strings=".") %>%
    filter(r2 > 0.3 | is.na(r2)) %>%
    mutate(gene_id = strip_ensembl(v19_id)) %>%
    filter(gene_id %in% gencode.pc_linc$gene_id) %>%
    group_by(gene_id) %>%
    mutate(bhp = p.adjust(p.value, method="BH"))
  
  gemma.min_bhp = gemma %>%
    group_by(gene_id) %>%
    summarise(min_bhp = min(bhp))
  
  fdr_thresh = gemma.min_bhp %>%
    mutate(bhp_bhp = p.adjust(min_bhp, method="BH")) %>%
    filter(bhp_bhp<=0.05) %>%
    pull(min_bhp) %>%
    max
    
  gemma.fdr05 = gemma %>%
    filter(bhp <= fdr_thresh)
  
  gemma_stats = rbind(
    gemma_stats,
    data.frame(peer = peer,
               neqtl = dim(gemma.fdr05)[1],
               negenes = length(unique(gemma.fdr05$v19_id)),
               pi1 = 1 - pi0est(gemma$p.value)$pi0))
    
}

## save results 
write.table(gemma_stats, "gemma_stats.txt", quote=F, sep="\t", row.names=F)

## read in results to make plot
gemma_stats = read_tsv("gemma_stats.txt")

## plot number of eQTLs
gemma_stats %>%
  gather(key="stat", value="value", -peer) %>%
  mutate(peer24=(peer=="peer24")) %>%
  ggplot(., aes(x=peer, y=value, color=peer24)) + 
  geom_point() + 
  theme_bw(base_size=15) + 
  facet_wrap(~stat, scales="free_y") + 
  scale_color_manual(values=c("black","red")) + 
  theme(legend.position="none", axis.text.x = element_text(angle = 90, hjust = 1))

```


# 20 PEER factors maximizes eQTLs using GEMMA:

```{r}

# eQTL file and column names
gemma.f = "round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.txt.gz"
gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# read in and filter raw eQTL data
gemma = fread(gemma.f, col.names=gemma_cols, na.string=".") %>%
  select(-c(foo, bar)) %>%
  filter(r2 > 0.3 | is.na(r2)) %>% # remove poorly imputed SNPs
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", remove=F, convert=T) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  filter(gene_id %in% gencode.pc_linc$gene_id) %>% # pc and lncRNAs
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  group_by(gene_id) %>%
  mutate(bhp = p.adjust(p.value, method="BH")) # Benjamini-Hochberg correction across SNPs per gene

# get minimum BH-corrected p-value per gene
gemma.min_bhp = gemma %>%
  group_by(gene_id) %>%
  summarise(min_bhp = min(bhp)) 

# threshold for identifying eSNPs
fdr_thresh = gemma.min_bhp %>%
  mutate(bhp_bhp = p.adjust(min_bhp, method="BH")) %>%
  filter(bhp_bhp<=0.05) %>%
  pull(min_bhp) %>%
  max

# data for all SNPs tested for each eQTL gene. Useful downstream
gemma.egenes = gemma %>%
  mutate(fdr_sig = bhp <= fdr_thresh) %>%
  group_by(gene_id) %>%
  filter(sum(fdr_sig) > 0) %>%
  mutate(top_esnp = (p.value==min(p.value)),
         maf = ifelse(af < 0.5, af, 1-af))

# write out useful results
write.table(gemma.egenes, "gemma.egenes.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f gemma.egenes.tsv")
                        
# clean up
rm(gemma, gemma.min_bhp)

```


# Calculate other useful data

```{r}

# map between hg19 and hg38
hg19_to_hg38 = tabix2df("../geno/5M.imputed.rna_samps.biallelic.hg38.sort.txt.gz",
                        unique(tabixify(gemma.egenes$chr,
                                        gemma.egenes$pos_hg19,
                                        gemma.egenes$pos_hg19)),
                        c("chr","pos_hg19","pos_hg38"))

write.table(hg19_to_hg38, "hg19_to_hg38.egenes.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f hg19_to_hg38.egenes.tsv")

# European AF
gemma.egenes.eur_af.tabix = gemma.egenes %>%
  mutate(tabix_id = tabixify(chr, pos_hg19, pos_hg19)) %>%
  pull(tabix_id)

gemma.egenes.eur_af = tabix2df("../misc/ALL.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.EUR.snps.af.txt.gz",
                               unique(tabixify(gemma.egenes$chr,
                                               gemma.egenes$pos_hg19,
                                               gemma.egenes$pos_hg19)),
                               c("chr","pos_hg19","ref_1kg","alt_1kg","eur_af"))

write.table(gemma.egenes.eur_af, "gemma.egenes.eur_af.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f gemma.egenes.eur_af.tsv")

```


# Read in African eQTL and sQTL information and snp-mapping information

```{r}

# read data in
gemma.egenes = read_tsv("./data/gemma.egenes.tsv.gz",
                        col_types="cnncnccnnnnnnnnnnccnlln")

lc.sgenes = read_tsv("./data/")

hg19_to_hg38 = read_tsv("./data/hg19_to_hg38.egenes.tsv.gz")

# hg19_to_hg38 = liftover_call(unique(tabixify(addchr(gemma.egenes$chr),
#                                      gemma.egenes$pos_hg19,
#                                      gemma.egenes$pos_hg19)),
#                             "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
#   separate(new, into=c("chr", "range"), sep=":") %>%
#   separate(range, into=c("pos_hg38", "foo"), sep="-", convert=T) %>%
#   separate(old, into=c("chr", "range"), sep=":") %>%
#   separate(range, into=c("pos_hg19", "foo"), sep="-", convert=T) %>%
#   dplyr::select(-foo) %>%
#   mutate(chr = rmchr(chr))


# gemma.egenes.eur_af = read_tsv("gemma.egenes.eur_af.tsv.gz") %>%
#   mutate(ref_1kg = strsplit(as.character(ref_1kg), ","),
#          alt_1kg = strsplit(as.character(alt_1kg), ","),
#          eur_af = strsplit(as.character(eur_af), ",")) %>%
#   unnest(c(ref_1kg, alt_1kg, eur_af)) %>%
#   mutate(eur_af = as.numeric(eur_af))
# 
# write.table(gemma.egenes.eur_af, "gemma.egenes.eur_af.unnest.tsv", sep="\t", quote=F, row.names=F)
# system("bgzip -f gemma.egenes.eur_af.unnest.tsv")

gemma.egenes.eur_af = read_tsv("./data/gemma.egenes.eur_af.unnest.tsv.gz")

# # nonrmalized gene expression
# gex_norm = read_tsv("../gemma/pheno_resid/peer20/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.txt.gz") %>%
#   mutate(gene_id = strip_ensembl(gene_id))

lc.sgenes = read_tsv("./data/lc.sgenes.tsv.gz", col_types=c("ccnnnccnnnnnnnnnnccnlln")) %>%
  separate(snp_hg19, into=c("chr","pos_hg19"), sep=":") %>%
  separate(intron, into=c("chr1","start_hg19","end_hg19","clust"), sep=":", convert=T) %>%
  unite("intron", chr1, start_hg19, end_hg19, clust, sep=".", remove=F) %>%
  mutate(v19_id1 = v19_id) %>%
  unite("afr_intron", chr1, start_hg19, end_hg19, v19_id1, sep="_") %>%
  mutate(sqtl_hg19 = paste(intron, pos_hg19, sep=":"))

# ## info for cluster-to-gene mapping
# 
#                       col.names=c("clust","chr","start","end","v19_id","mean","median","min","max")) %>%
#   filter(nchar(v19_id) > 1) %>%
#   mutate(gene_id = strip_ensembl(v19_id))

lc.sgenes.eur_af = tabix2df("./misc_data/ALL.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.EUR.snps.af.txt.gz",
                            unique(tabixify(lc.sgenes$chr,
                                            lc.sgenes$pos_hg19,
                                            lc.sgenes$pos_hg19)),
                            c("chr","pos_hg19","ref_1kg","alt_1kg","eur_af"))

# info for clust-to-gene mapping
clust_gene = read_table2("./misc_data/leafcutter_all.clu2gene.txt", na="?",
                        col_names=c("clust","chr","start_hg19","end_hg19","v19_id",paste0("foo",1:4))) %>%
  select(clust, chr, start_hg19, end_hg19, v19_id)

hg19_to_hg38.s = read_tsv("./misc_data/hg19_to_hg38.sgenes.tsv.gz") %>%
  unique

```


# Compare replication between eQTLs and sQTLs

```{r}

# column names
gemma.sort.cols = c("chr","pos_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")
lc.sort.cols = c("clust","chr","pos_hg19","chr_intron","start_intron","end_intron","foo","bar","nmiss","allel1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2","v19_id")

# qtl files
gemma.sort.f = "./data/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.sort.txt.gz"
lc.sort.f = "./data/Sample_ALL.secondpass.leafcutter.covars.peer5.resid.gemma.100kb.lmm4.r2.wgene.sort.txt.gz"

# # data for significant eQTLs in the sQTL scan
# gemma.fdr05.lc.tabix = gemma.egenes %>%
#   filter(fdr_sig) %>%
#   mutate(tabix_id = tabixify(v19_id, pos_hg19, pos_hg19)) %>%
#   pull(tabix_id) %>%
#   unique
# 
# gemma.fdr05.lc = tabix2df(lc.sort.f,
#                           gemma.fdr05.lc.tabix,
#                           lc.sort.cols)

# data for significant sQTLs in the eQTL scan
lc.fdr05.gemma.tabix = lc.sgenes %>%
  filter(fdr_sig) %>%
  mutate(tabix_id = tabixify(v19_id, pos_hg19, pos_hg19)) %>%
  pull(tabix_id) %>%
  unique

lc.fdr05.gemma = tabix2df(gemma.sort.f,
                          lc.fdr05.gemma.tabix,
                          gemma.sort.cols)

# plot p-values
ggplot(lc.fdr05.gemma, aes(x=p.value)) +
  geom_histogram(alpha=0.5) +
  theme_classic(base_size=15) +
  ggtitle("eQTL p-values of eQTLs")

```


# Credible sets for eQTLs and sQTLs

```{r}

# get eQTL credible sets
gemma.cred_set.all = lapply(unique(gemma.egenes$gene_id),
                              function(x){
                                df.tmp = gemma.egenes %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg19,
                                               df.tmp$beta,
                                               df.tmp$se^2,
                                               0.9)})

# get sQTL credible sets
lc.cred_set.all = lapply(unique(lc.sgenes$intron),
                         function(x){
                           df.tmp = lc.sgenes %>%
                             filter(intron==x)
                           get_cred_set.b(df.tmp$sqtl_hg19,
                                          df.tmp$beta,
                                          df.tmp$se^2,
                                          0.9)})

gemma.lc.cred_set = data.frame(cred_set = unlist(lc.cred_set.all)) %>%
  separate(cred_set, into=c("intron", "pos_hg19"), sep=":", convert=T) %>%
  merge(lc.sgenes %>%
          select(intron, pos_hg19, top_ssnp, gene_id),
        all.x=T) %>%
  select(gene_id, pos_hg19, top_ssnp) %>%
  unique %>%
  mutate(sqtl=T) %>%
  merge(data.frame(cred_set = unlist(gemma.cred_set.all)) %>%
          separate(cred_set, into=c("gene_id", "pos_hg19"), sep=":", convert=T) %>%
          mutate(eqtl=T),
        all=T) %>%
  merge(gemma.egenes %>%
          select(gene_id, pos_hg19, top_esnp),
        all.x=T)

gemma.lc.intersect_genes = gemma.lc.cred_set %>%
  filter(sqtl & eqtl) %>%
  pull(gene_id) %>%
  unique

gemma.lc.intersect = gemma.lc.cred_set %>%
  group_by(gene_id) %>%
  summarise(has_eqtl = sum(!is.na(eqtl)) > 0,
            has_sqtl = sum(!is.na(sqtl)) > 0,
            has_top_esnp = sum(top_esnp, na.rm=T) > 0,
            has_top_ssnp = sum(top_ssnp, na.rm=T) > 0)

gemma.lc.intersect.list = list(eGene = which(gemma.lc.intersect$has_eqtl),
     sGene = which(gemma.lc.intersect$has_sqtl),
     top_eSNP = which(gemma.lc.intersect$has_top_esnp),
     top_sSNP = which(gemma.lc.intersect$has_top_ssnp))


# for how many genes is the top SNP the same?
gemma.lc.cred_set %>%
    filter(top_ssnp & top_esnp) %>%
  pull(gene_id) %>% 
  unique %>% 
  length

gemma.lc.cred_set.overlaps = gemma.lc.cred_set %>%
  filter(gene_id %in% gemma.lc.intersect_genes) %>%
  group_by(gene_id) %>%
  summarise(neqtl = sum(eqtl, na.rm=T),
            nsqtl = sum(sqtl, na.rm=T),
            nesqtl = sum(eqtl & sqtl, na.rm=T))

# overlap of credible sets
gemma.cred_set.uniq = data.frame(eqtl_hg19 = unlist(gemma.cred_set.all)) %>%
  separate(eqtl_hg19, into=c("gene_id","pos_hg19"), sep=":") %>%
  unique %>%
  mutate(eqtl=T)

lc.cred_set.uniq = data.frame(sqtl_hg19 = unlist(afr_cred_set.introns.all)) %>%
  separate(sqtl_hg19, into=c("intron","pos_hg19"), sep=":") %>%
  merge(lc.sgenes %>% select(intron, pos_hg19, gene_id)) %>%
  select(gene_id, pos_hg19) %>%
  unique %>%
  mutate(sqtl=T)

gemma.lc.cred_set.uniq = merge(gemma.cred_set.uniq, lc.cred_set.uniq, all=T) %>%
  unique

length(intersect(unique(unlist(gemma.cred_set.all)), lc.cred_set.uniq))/length(union(unique(unlist(gemma.cred_set.all)), lc.cred_set.uniq))

mean(unique(unlist(gemma.cred_set.all)) %in% lc.cred_set.uniq)

```


# Functional enrichment

Get random sample of matched variants

```{r}

## create sampling matrices for two variables. First variable is rows, second variable columns
get_samp_mats <- function(bins){
  
  vars = colnames(bins)
  
  # get min/max for each group
  minmax_idx = bins %>%
    arrange(!!!syms(vars)) %>%
    mutate(idx = 1:dim(bins)[1]) %>%
    group_by(!!!(syms(vars))) %>%
    summarise(min_idx = min(idx),
              max_idx = max(idx))

  # matrix of minimum indices
  min_idx.mat = minmax_idx %>%
    dplyr::select(-max_idx) %>%
    spread(vars[2], min_idx) %>%
    column_to_rownames(vars[1]) %>%
    as.matrix
  
  # matrix of maximum indices
  max_idx.mat = minmax_idx %>%
    dplyr::select(-min_idx) %>%
    spread(vars[2], max_idx) %>%
    column_to_rownames(vars[1]) %>%
    as.matrix
  
  return(list(min_mat = min_idx.mat,
              max_mat = max_idx.mat))
}

# sample indices using matrices
sample_idx <- function(min_mat, max_mat, var1, var2, n=10000){
  
  min_idx = min_mat[var1, var2]
  max_idx = max_mat[var1, var2]
  
  # sample n random indices between min_idx and max_idx
  sample(x=min_idx:max_idx,
         size=n,
         replace=T)
}

## calculate odds ratio
calc_or <- function(a, b, n){
  
  c = n - a
  d = n - b
  
  ## if any entries are 0, apply the Haldane-Anscombe correction
  if (min(c(a,b,c,d)) == 0){
    a = a + 0.5
    b = b + 0.5
    c = c + 0.5
    d = d + 0.5
    }
  
  return(a * d / b / c)
}


maf_bins = 0:10/20
tss_bins = c(-1, 500, 1000, 5000, 10000, 50000, 100000, Inf)
intron_bins = c(-1, 500, 1000, 5000, 10000, 50000, 100000, Inf)

vep_cols = c("chr","pos","maf","max_effect","near_egene","dist_egene","near_sgene","dist_sgene","epi")

vep_stats = read_tsv("./misc_data/5M.imputed.rna_samps.biallelic.maf.max_effect.tss_dist.intron_dist.E062.15state.txt.gz",
                  col_names=vep_cols) %>%
  mutate(snp_hg19 = paste(chr, pos, sep=":")) %>%
  mutate(maf_bin = cut(maf, maf_bins),
         tss_bin = cut(dist_egene, tss_bins),
         intron_bin = cut(dist_sgene, intron_bins))
  
caqtls = read_tsv("./misc_data/elife-39595-supp2.txt.gz") %>%
 dplyr::rename(pos_hg19 = position)
colnames(caqtls) = make.names(colnames(caqtls))


# add whether it is a caqtl
vep_stats$is_caqtl = vep_stats$snp_hg19 %in% paste(caqtls$chr, caqtls$pos_hg19, sep=":")

# add whether it overlaps a TFBS
tfbs_snps = readLines("./misc_data/5M.imputedd.rna_samps.biallelic.GM12878.all_tfs.sort.snps.txt")
vep_stats$is_tfbs = vep_stats$snp_hg19 %in% tfbs_snps
rm(tfbs_snps)


##min/max indices for eQTLs
eqtl_samp_mats = vep_stats %>%
  select(c(maf_bin, tss_bin)) %>%
  get_samp_mats

##min/max indices for eQTLs
sqtl_samp_mats = vep_stats %>%
  select(c(maf_bin, intron_bin)) %>%
  get_samp_mats


################## sampling [e|s]QTL ##################

# get likely causal variants (posterior > 10%)
gemma.post_0.1 = gemma.egenes %>%
  mutate(bf = exp(approx.bf.p(p.value, maf, "quant", 162)$lABF)) %>%
  group_by(gene_id) %>%
  mutate(post_prob = bf / sum(bf)) %>%
  filter(post_prob > 0.1) %>%
  select(gene_id, snp_hg19, post_prob)

lc.post_0.1 = lc.sgenes %>%
  mutate(bf = exp(approx.bf.p(p.value, maf, "quant", 162)$lABF)) %>%
  group_by(intron) %>%
  mutate(post_prob = bf / sum(bf)) %>%
  filter(post_prob > 0.1) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  select(gene_id, intron, snp_hg19, post_prob)


# get the VEP stats for the eQTLs
gemma.vep_stats = gemma.egenes %>%
  filter(fdr_sig) %>%
  select(snp_hg19, gene_id) %>%
  unique %>%
  merge(vep_stats, by="snp_hg19")


# get the VEP stats for the sQTLs
lc.vep_stats = lc.sgenes %>%
  filter(fdr_sig) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  select(intron, snp_hg19) %>%
  unique %>%
  merge(vep_stats, by="snp_hg19")


# all possible VEP categories
vep_cats = sort(unique(vep_stats$max_effect))
epi_cats = sort(unique(vep_stats$epi))

set.seed(7)

# sample variables
gemma.vep_stats.samp = mapply(function(i, j) sample_idx(eqtl_samp_mats$min_mat, eqtl_samp_mats$max_mat, i, j),
                              gemma.vep_stats$maf_bin, gemma.vep_stats$tss_bin)

lc.vep_stats.samp = mapply(function(i, j) sample_idx(sqtl_samp_mats$min_mat, sqtl_samp_mats$max_mat, i, j),
                              lc.vep_stats$maf_bin, lc.vep_stats$intron_bin)



# get vep and epigenetic annotations for eQTLs
# arrange vep_stats again
vep_stats = arrange(vep_stats, maf_bin, tss_bin)
e.null = rbind(
  # add innformation on VEP categories
  apply(gemma.vep_stats.samp, 1,
        function(x) table(vep_stats[x,]$max_effect)[vep_cats]),
  # add information on epigenomic annotations
  apply(gemma.vep_stats.samp, 1,
        function(x) table(vep_stats[x,]$epi)[epi_cats]),
  # add whether it's a caQTL
  apply(gemma.vep_stats.samp, 1,
        function(x) sum(vep_stats[x,]$is_caqtl)),
  # add whether it's a tfbs
  apply(gemma.vep_stats.samp, 1,
        function(x) sum(vep_stats[x,]$is_tfbs)))
rownames(e.null) = c(vep_cats, epi_cats, "caqtl","tfbs")
e.null[is.na(e.null)] = 0


# get vep and epigenetic annotations for sQTLs
vep_stats = arrange(vep_stats, maf_bin, intron_bin)
s.null = rbind(
  # add innformation on VEP categories
  apply(lc.vep_stats.samp, 1,
        function(x) table(vep_stats[x,]$max_effect)[vep_cats]),
  # add information on epigenomic annotations
  apply(lc.vep_stats.samp, 1,
        function(x) table(vep_stats[x,]$epi)[epi_cats]),
  # add whether it's a caQTL
  apply(lc.vep_stats.samp, 1,
        function(x) sum(vep_stats[x,]$is_caqtl)),
  # add whether it's a tfbs
  apply(lc.vep_stats.samp, 1,
        function(x) sum(vep_stats[x,]$is_tfbs)))
rownames(s.null) = c(vep_cats, epi_cats, "caqtl","tfbs")
s.null[is.na(s.null)] = 0


# write out intermediate data
write.table(e.null, "enull.csv", sep=",", quote=F, col.names=F)
write.table(s.null, "snull.csv", sep=",", quote=F, col.names=F)

```


# Start with intermediate data and also sample variants with high causal probability

```{r}
# read in intermediate data
e.null = read_csv("./misc/enull.csv", col_names=F) %>%
  column_to_rownames("X1") %>%
  as.matrix

s.null = read_csv("./misc/snull.csv", col_names=F) %>%
  column_to_rownames("X1") %>%
  as.matrix


# get vep and epigenetic annotations for top sQTLs (posterior prob > 0.1)
e.post_0.1.null = rbind(
  apply(gemma.vep_stats.samp[,gemma.vep_stats$snp_hg19 %in% gemma.post_0.1$snp_hg19], 1,
        function(x) table(vep_stats[x,]$max_effect)[vep_cats]),
  apply(gemma.vep_stats.samp[,gemma.vep_stats$snp_hg19 %in% gemma.post_0.1$snp_hg19], 1,
        function(x) table(vep_stats[x,]$epi)[epi_cats]),
  apply(gemma.vep_stats.samp[,gemma.vep_stats$snp_hg19 %in% gemma.post_0.1$snp_hg19], 1,
        function(x) sum(vep_stats[x,]$is_caqtl)),
  apply(gemma.vep_stats.samp[,gemma.vep_stats$snp_hg19 %in% gemma.post_0.1$snp_hg19], 1,
        function(x) sum(vep_stats[x,]$is_tfbs)))
rownames(e.post_0.1.null) = c(vep_cats, epi_cats, "caqtl", "tfbs")
e.post_0.1.null[is.na(e.post_0.1.null)] = 0

# get vep and epigenetic annotations for top sQTLs (posterior prob > 0.1)
s.post_0.1.null = rbind(
  apply(lc.vep_stats.samp[,lc.vep_stats$snp_hg19 %in% lc.post_0.1$snp_hg19], 1,
        function(x) table(vep_stats[x,]$max_effect)[vep_cats]),
  
  apply(lc.vep_stats.samp[,lc.vep_stats$snp_hg19 %in% lc.post_0.1$snp_hg19], 1,
        function(x) table(vep_stats[x,]$epi)[epi_cats]),
  apply(lc.vep_stats.samp[,lc.vep_stats$snp_hg19 %in% lc.post_0.1$snp_hg19], 1,
        function(x) sum(vep_stats[x,]$is_caqtl)),
  apply(lc.vep_stats.samp[,lc.vep_stats$snp_hg19 %in% lc.post_0.1$snp_hg19], 1,
        function(x) sum(vep_stats[x,]$is_tfbs)))
rownames(s.post_0.1.null) = c(vep_cats, epi_cats, "caqtl", "tfbs")
s.post_0.1.null[is.na(s.post_0.1.null)] = 0


# tabulate observed
e.tab = c(table(c(gemma.vep_stats$max_effect,
                  gemma.vep_stats$epi)),
          sum(gemma.vep_stats$is_caqtl),
          sum(gemma.vep_stats$is_tfbs))
names(e.tab)[c(32,33)] = c("caqtl","tfbs")
s.tab = c(table(c(lc.vep_stats$max_effect,
                lc.vep_stats$epi)),
          sum(lc.vep_stats$is_caqtl),
          sum(lc.vep_stats$is_tfbs))
names(s.tab)[c(32,33)] = c("caqtl","tfbs")

e.post_0.1.tab = c(table(c(gemma.vep_stats %>%
                             filter(snp_hg19 %in% gemma.post_0.1$snp_hg19) %>%
                             pull(max_effect),
                           gemma.vep_stats %>%
                             filter(snp_hg19 %in% gemma.post_0.1$snp_hg19) %>%
                             pull(epi))),
                   gemma.vep_stats %>%
                     filter(snp_hg19 %in% gemma.post_0.1$snp_hg19) %>%
                     pull(is_caqtl) %>%
                     sum,
                   gemma.vep_stats %>%
                     filter(snp_hg19 %in% gemma.post_0.1$snp_hg19) %>%
                     pull(is_tfbs) %>%
                     sum)
names(e.post_0.1.tab)[c(27,28)] = c("caqtl","tfbs")

s.post_0.1.tab = c(table(c(lc.vep_stats %>%
                           filter(snp_hg19 %in% lc.post_0.1$snp_hg19) %>%
                           pull(max_effect),
                         lc.vep_stats %>%
                           filter(snp_hg19 %in% lc.post_0.1$snp_hg19) %>%
                           pull(epi))),
                   lc.vep_stats %>%
                     filter(snp_hg19 %in% lc.post_0.1$snp_hg19) %>%
                     pull(is_caqtl) %>%
                     sum,
                   lc.vep_stats %>%
                     filter(snp_hg19 %in% lc.post_0.1$snp_hg19) %>%
                     pull(is_tfbs) %>%
                     sum)
names(s.post_0.1.tab)[c(28,29)] = c("caqtl","tfbs")

# get null counts of vep categories
get_counts <- function(mat, cat){
  
  counts = apply(mat, 2, function(x) sum(x[grepl(cat, names(x))]))
  counts[is.na(counts)] = 0
  return(counts)
}

# configure outputdataframe
get_out_df <- function(class, cat, or){
  
  data.frame(class = class,
             category = cat,
             or_median = median(or),
             or_025 = quantile(or, 0.025),
             or_975 = quantile(or, 0.975))
}

## categories of annotations
func_cat = c("3_prime_UTR", "5_prime_UTR", "intron", "splice", "synonymous", "missense", "start|stop")
epi_cat = c("Tss","Enh","ReprPC","Tx","Het")

func_enrich = data.frame()
for (cat in c(func_cat, epi_cat, "caqtl", "tfbs")){
  
  # get counts for null
  ecounts.null = get_counts(e.null, cat)
  scounts.null = get_counts(s.null, cat)
  
  ecounts.post_0.1.null = get_counts(e.post_0.1.null, cat)
  scounts.post_0.1.null = get_counts(s.post_0.1.null, cat)
  
  
  # get observect counts
  ecounts = sum(e.tab[grepl(cat, names(e.tab))])
  scounts = sum(s.tab[grepl(cat, names(s.tab))])

  ecounts.post_0.1 = sum(e.post_0.1.tab[grepl(cat, names(e.post_0.1.tab))])
  scounts.post_0.1 = sum(s.post_0.1.tab[grepl(cat, names(s.post_0.1.tab))])
  
  
  # get enrichment
  e.enrich = (ecounts+0.5)/(ecounts.null+0.5)
  s.enrich = (scounts+0.5)/(scounts.null+0.5)
  
  e.post_0.1.enrich = (ecounts.post_0.1+0.5)/(ecounts.post_0.1.null+0.5)
  s.post_0.1.enrich = (scounts.post_0.1+0.5)/(scounts.post_0.1.null+0.5)
    
  
  # get output
  out.df = rbind(get_out_df("eQTL", cat, e.enrich),
                 get_out_df("sQTL", cat, s.enrich),
                 get_out_df("eQTL_post_0.1", cat, e.post_0.1.enrich),
                 get_out_df("sQTL_post_0.1", cat, s.post_0.1.enrich))
  
 func_enrich = rbind(func_enrich, out.df)
}


write.table(func_enrich, "eqtl_sqtl.func_enrich.w_post_0.1.merge_start_stop.txt", quote=F, sep="\t", row.names=F)

library(ggforestplot)

func_enrich %>%
  mutate(category = factor(category, levels=c(epi_cat, func_cat, "caqtl", "tfbs")),
         type = ifelse(grepl("eQTL", class), "eQTL", "sQTL"),
         set = ifelse(grepl("post", class), "post_gt_0.1", "all"),
         gw = factor(ifelse(as.numeric(category) %% 2 == 1, 1, 0))) %>%
  ggplot() +
  geom_stripes(aes(y=category), odd = "#11111111", even = "#00000000") +
  geom_vline(aes(xintercept=0), linetype="dashed", color="gray") +
  geom_pointrange(aes(y=category,
                      x=log2(or_median),
                      xmin=log2(or_025),
                      xmax=log2(or_975),
                      color=type,
                      shape=set),
                  position = position_dodge2(width=0.75, preserve = "single")) +
  theme_bw(base_size=15) +
  theme(panel.grid.major.y = element_blank()) +
  scale_fill_manual(values = c("white", "grey50")) + 
  scale_color_ptol() +
  xlab("log2(Enrich)") + 
  ggtitle("Functional context of tQTLs")


# make horizontal plot
func_enrich %>%
  mutate(category = factor(category, levels=c(epi_cat, func_cat)),
         type = ifelse(grepl("eQTL", class), "eQTL", "sQTL"),
         set = ifelse(grepl("post", class), "post_gt_0.1", "all")) %>%
  ggplot() +
  geom_pointrange(aes(x=log2(or_median),
                      y=category,
                      xmin=log2(or_025),
                      xmax=log2(or_975),
                      color=type,
                      shape=set),
                  position = position_dodge(width = 0.5)) + 
  theme_bw(base_size=15) +
  scale_color_ptol() +
  ylab("log2(Enrich)") + 
  ggtitle("Functional context of tQTLs")

```


# Do the fraction of effects in the same direction with caqtls increase with increasing causal probability, separating those caqtls active in AFR, EUR, and both

```{r}

gemma.egenes.caqtls = gemma.egenes %>%
  filter(fdr_sig) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  group_by(gene_id) %>%
  mutate(pp = bf/sum(bf)) %>%
  select(chr, pos_hg19, allele0, allele1, pp, beta) %>%
  merge(caqtls) %>%
  mutate(pos_al = ifelse(beta > 0, allele0, allele1),
         same_dir = pos_al == consensus.open.allele)

gemma.egenes.caqtls.same_dir = sapply(seq(0, 1, 0.05),
                                      function(x) gemma.egenes.caqtls %>%
                                        filter(pp > x) %>%
                                        pull(same_dir) %>%
                                        mean)


```


# Overlap with GWAS catalog

```{r}

library(rsnps)

# read in GWAS catalogue data
gwas_cat = read_tsv("./misc_data/catalog_011421/gwas_catalog_v1.0.2-associations_e100_r2020-12-15.tsv",
                    col_types = paste0(rep("c", 38), collapse="")) %>%
  # get columns of interest
  select(pubmedid=PUBMEDID, trait=`DISEASE/TRAIT`, n_init=`INITIAL SAMPLE SIZE`, n_rep=`REPLICATION SAMPLE SIZE`,
         chr=CHR_ID, pos_hg38=CHR_POS, strongest_snp=`STRONGEST SNP-RISK ALLELE`, context=`CONTEXT`, risk_af=`RISK ALLELE FREQUENCY`,
         gwas_pval=`P-VALUE`, or_beta=`OR or BETA`, mapped_trait=MAPPED_TRAIT) %>%
  # clean up interaction SNPs, keeping track of whether SNPs they interact with
  # grep for " x " which marks an interaction
  mutate(interaction = ifelse(grepl(" x ", chr) & grepl(" x ", pos_hg38),
                              # clean up the interaction format
                              paste(gsub(" x ", "-", chr),
                                    gsub(" x ", "-", pos_hg38), ":"),
                              NA)) %>%
  # clean up interactions after keeping track of them, for splitting and unnesting later
  mutate(chr = gsub(" x ", ";", chr),
         pos_hg38 = gsub(" x ", ";", pos_hg38)) %>%
  # split chr and position
  mutate(chr = strsplit(chr, ";"),
         pos_hg38 = strsplit(pos_hg38, ";")) %>%
  # unnest
  unnest(c(chr, pos_hg38)) %>%
  mutate(snp_hg38 = paste(chr, pos_hg38, sep=":")) %>%
  merge(read_tsv("./misc_data/gwas_catalog_v1.0.2-associations_e100_r2020-12-15.pos_hg38Tohg19.txt", col_names=c("chr","pos_hg19","pos_hg38")), all.x=T)

write.table(gwas_cat, "./misc_data/gwas_catalog_v1.0.2-associations_e100_r2020-12-15.unnest.tsv", sep="\t", row.names=F, quote=F)

```


# Compare replication in GTEx

First extract eQTL data for GTEx v7 and v8

```{r}

# filenames for gtexv7 and gtexv8. These must be downloaded and generated using 
# scripts provided and tabix indices created
gtex.v7.f = "./misc_data/Whole_Blood.allpairs.sub.txt.gz"
gtex.v7.cols = c("v19_id","chr","pos_hg19","v7_allele1","v7_allele2","tss_dist",
                 "ma_samples","ma_count","gtexv7_maf","gtexv7_pval","gtexv7_slope","gtexv7_se")
gtex.v8.f = "./misc_data/GTEx_Analysis_v8_eQTL_all_associations_Whole_Blood.allpairs.sort.txt.gz"
gtex.v8.cols = c("v26_id","chr","pos_hg38","v8_ref","v8_alt","version","tss_dist",
                 "ma_samples","ma_count","gtexv8_maf","gtexv8_pval","gtexv8_slope","gtexv8_se")


# tabix-formatting for extracting GTEx data
gtex.v7.egenes.tabix = gemma.egenes %>%
  mutate(tabix_id = tabixify(v19_id, pos_hg19, pos_hg19)) %>%
  pull(tabix_id)

gtex.v8.egenes.tabix = merge(gemma.egenes, hg19_to_hg38) %>%
  merge(gene_map) %>%
  filter(!is.na(v26_id)) %>%
  mutate(tabix_id = tabixify(v26_id, pos_hg38, pos_hg38)) %>%
  pull(tabix_id)


# extract GTEx data
gtex.v7.egenes = tabix2df(gtex.v7.f,
                          gtex.v7.egenes.tabix,
                          gtex.v7.cols) %>%
  select(c(v19_id, chr, pos_hg19, v7_allele1, v7_allele2, gtexv7_maf, gtexv7_pval, gtexv7_slope, gtexv7_se))
  
gtex.v8.egenes = tabix2df(gtex.v8.f,
                          gtex.v8.egenes.tabix,
                          gtex.v8.cols) %>%
  select(c(v26_id, chr, pos_hg38, v8_ref, v8_alt, gtexv8_maf, gtexv8_pval, gtexv8_slope, gtexv8_se)) %>%
  mutate(chr = rmchr(chr))


# write out GTEx eGene data for faster processing
write.table(gtex.v7.egenes, "gtex.v7.egenes.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f gtex.v7.egenes.tsv")

gtex.v8.egenes %>%
  mutate(chr = rmchr(chr)) %>% 
  write.table(., "gtex.v8.egenes.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f gtex.v8.egenes.tsv")

```


# Read in GTEx eGene data

```{r}

# read data
gtex.v7.egenes = read_tsv("./misc_data/gtex.v7.egenes.tsv.gz")
gtex.v8.egenes = read_tsv("./misc_data/gtex.v8.egenes.tsv.gz")

gtex.v8.e.sig = read_tsv(".misc_data/GTEx_Analysis_v8_eQTL/Whole_Blood.v8.signif_variant_gene_pairs.txt.gz") %>%
  mutate(v26_id = gene_id,
         gene_id = strip_ensembl(gene_id))

```


# Before checking replication, how many eQTLs can be tested? For how many is it because:
1. The SNP does not liftover
2. The SNP is at low frequency (EUR)
3. The SNP did liftover and is at appreciable frequency but still wasn't tested
4. The Ensembl ID is dropped between v19 and v26

```{r}

library(UpSetR)


# check which eSNPs liftOver
gemma.egenes.tested = gemma.egenes %>%
  filter(fdr_sig) %>%
  merge(hg19_to_hg38, all.x=T) %>%
  mutate(snp_hg38 = paste(chr, pos_hg38, sep=":")) %>%
  mutate(snp_hg38 = ifelse(is.na(pos_hg38), NA, snp_hg38)) %>%
  merge(gene_map, all.x=T) %>%
  select(v26_id, gene_id, snp_hg38, snp_hg19) %>%
  mutate(snp_lifts_over = !is.na(snp_hg38),
         gene_lifts_over = !is.na(v26_id),
         gene_tested_v8 = v26_id %in% gtex.v8.egenes$v26_id,
         snp_tested_v8 = (snp_hg38 %in% paste(rmchr(gtex.v8.egenes$chr), 
                                              gtex.v8.egenes$pos_hg38,
                                              sep=":")))
  
gemma.egenes.tested.list = list(gene_lifts_over = which(gemma.egenes.tested$gene_lifts_over),
     gene_tested_v8 = which(gemma.egenes.tested$gene_tested_v8),
     snp_lifts_over = which(gemma.egenes.tested$snp_lifts_over),
     snp_tested_v8 = which(gemma.egenes.tested$snp_tested_v8))

upset(fromList(gemma.egenes.tested.list), order.by = "freq")

```


# Do eQTLs that aren't tested in GTEx show worse imputation?

```{r}

gemma.egenes.tested %>%
  select(gene_id, snp_hg19, snp_tested_v8) %>%
  merge(gemma.egenes %>%
          select(gene_id, snp_hg19, p.value, maf, r2)) %>%
  mutate(`-log10(p-value)` = -log10(p.value)) %>%
  gather(key=variable, value=stat, c(`-log10(p-value)`, maf, r2)) %>%
  ggplot(aes(x=stat, color=snp_tested_v8)) +
  stat_ecdf() +
  scale_color_ptol() +
  theme_classic(base_size=15) +
  facet_grid(. ~ variable, scales="free_x")
  

```


# Calculate pi0 between African eQTLs and GTEx and compare effect sizes

```{r}

# get pi0 for GTEx v7 and v8
gtex.v7.pi0 = gemma.egenes %>%
  filter(fdr_sig) %>%
  select(v19_id, pos_hg19, allele1, allele0) %>%
  merge(gtex.v7.egenes) %>%
  filter((v7_allele1 == allele1 & v7_allele2 == allele0) |
           (v7_allele1 == allele0 & v7_allele2 == allele1)) %>%
  pull(gtexv7_pval) %>%
  pi0est

gtex.v8.pi0 = gemma.egenes %>%
  filter(fdr_sig) %>% # get significant
  select(v19_id, pos_hg19, allele1, allele0) %>%
  merge(gene_map) %>% # add gene information
  merge(hg19_to_hg38) %>% # add position information
  merge(gtex.v8.egenes) %>% # add GTEx data
  filter((v8_ref == allele1 & v8_alt == allele0) |
           (v8_ref == allele0 & v8_alt == allele1)) %>%
  pull(gtexv8_pval) %>%
  pi0est

gemma.egenes %>%
  filter(top_esnp) %>% # get significant
  group_by(gene_id) %>%
  sample_n(1) %>%
  select(v19_id, pos_hg19, allele1, allele0, beta, se) %>%
  merge(gene_map) %>% # add gene information
  merge(hg19_to_hg38) %>% # add position information
  merge(gtex.v8.egenes %>% select(-chr)) %>% # add GTEx data
  filter((v8_ref == allele1 & v8_alt == allele0) |
           (v8_ref == allele0 & v8_alt == allele1)) %>%
  mutate(gtexv8_slope = ifelse(allele1==v8_ref & allele0==v8_alt, gtexv8_slope, -gtexv8_slope)) %>%
  ggplot(aes(x=beta,
             y=gtexv8_slope)) +
  geom_point(alpha=0.2) +
  theme_classic(base_size=15) +
  ggtitle("Effect sizes of eQTLs")

```


# Check MAF of GTEx p-values that fail to replicate

```{r}

gtex.v7.rep_vs_maf = gemma.egenes %>%
  filter(fdr_sig) %>%
  select(v19_id, pos_hg19, allele1, allele0) %>%
  merge(gtex.v7.egenes) %>%
  filter((v7_allele1 == allele1 & v7_allele2 == allele0) |
           (v7_allele1 == allele0 & v7_allele2 == allele1)) %>%
  ggplot(aes(x=gtexv7_maf, color=gtexv7_pval < 0.01)) +
  stat_ecdf() +
  theme_classic()

gtex.v8.rep_vs_maf = gemma.egenes %>%
  filter(fdr_sig) %>% # get significant
  select(v19_id, pos_hg19, allele1, allele0) %>%
  merge(gene_map) %>% # add gene information
  merge(hg19_to_hg38) %>% # add position information
  merge(gtex.v8.egenes) %>% # add GTEx data
  filter((v8_ref == allele1 & v8_alt == allele0) |
           (v8_ref == allele0 & v8_alt == allele1)) %>%
  ggplot(aes(x=gtexv8_maf, color=gtexv8_pval < 0.01)) +
  stat_ecdf() +
  theme_classic() +
  scale_color_ptol()

```


# Find eQTLs that aren't tested in GTEx but are in 1000G

```{r}

foo = gemma.egenes %>%
  # get only the tag SNPs *not* tested in GTEx v8
  filter(top_esnp) %>%
  # add hg38 info, removing SNPs that don't map
  merge(hg19_to_hg38) %>%
  mutate(snp_hg38 = paste(chr, pos_hg38, sep=":")) %>%
  # get only those SNPs not tested in GTEx v8
  filter(!(snp_hg38 %in% gtexv8_snps)) %>%
  select(gene_id, pos_hg19, pos_hg38, maf) %>%
  # add 1KG EUR AF
  merge(gemma.egenes.eur_af)

```


# Perform GSEA enrichment for genes that show large effect-size differences

```{r}

set.seed(42)

gemma.beta_diffs = gemma.egenes %>%
  select(gene_id, chr, pos_hg19, eqtl_hg19, allele0, allele1, p.value, beta) %>%
  # add hg38 info
  merge(hg19_to_hg38) %>%
  # merge with GTEx v8
  merge(gtex.v8.egenes.range) %>%
  # flip alleles
  filter((allele1 == v8_ref & allele0 == v8_alt) |
           (allele0 == v8_ref & allele1 == v8_alt)) %>%
  mutate(gtexv8_slope = ifelse(allele1 == v8_ref &
                                 allele0 == v8_alt,
                               gtexv8_slope,
                               1 - gtexv8_slope)) %>%
  # get the top SNP per gene that is also tested in GTEx
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1) %>%
  arrange(-beta) %>%
  pull(gene_id)

gemma.beta_diffs.gsea = gost(query = gemma.beta_diffs,
                             organism = "hsapiens",
                             ordered_query = T,
                             custom_bg = gemma.beta_diffs)

```


# Colocalize with GTEx

```{r}

# merge with v7
gemma.gtex.v7.egenes = gemma.egenes %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af)) %>%
  select(v19_id, pos_hg19, allele1, allele0, maf, beta, p.value) %>%
  merge(gtex.v7.egenes) %>%
  filter((v7_allele1 == allele1 & v7_allele2 == allele0) |
           (v7_allele1 == allele0 & v7_allele2 == allele1)) %>%
  mutate(gene_id = strip_ensembl(v19_id))

# merge with v8
gemma.gtex.v8.egenes = gemma.egenes %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af)) %>%
  select(v19_id, pos_hg19, allele1, allele0, maf, p.value) %>%
  merge(gene_map) %>% # add gene information
  merge(hg19_to_hg38) %>% # add position information
  unique %>%
  merge(gtex.v8.egenes.range) %>%
  filter((v8_ref == allele1 & v8_alt == allele0) |
           (v8_ref == allele0 & v8_alt == allele1))

# run colocalization
gemma_gtexv7_h4 = lapply(unique(gemma.gtex.v7.egenes$gene_id), function(gene){
  df.tmp = gemma.gtex.v7.egenes %>%
    filter(gene_id == gene) %>%
    arrange(pos_hg19)
  coloc.abf(dataset1=list(snp=df.tmp$pos_hg19,
                          pvalues=df.tmp$p.value,
                          type="quant",
                          N=162,
                          MAF=df.tmp$maf),
            dataset2=list(snp=df.tmp$pos_hg19,
                          pvalues=df.tmp$gtexv7_pval,
                          type="quant",
                          N=369,
                          MAF=df.tmp$gtexv7_maf))$summary
})

gemma_gtexv7_h4.df = data.frame(gene_id = unique(gemma.gtex.v7.egenes$gene_id),
                                H0 = sapply(gemma_gtexv7_h4, function(x) x[2]),
                                H1 = sapply(gemma_gtexv7_h4, function(x) x[3]),
                                H2 = sapply(gemma_gtexv7_h4, function(x) x[4]),
                                H3 = sapply(gemma_gtexv7_h4, function(x) x[5]),
                                H4 = sapply(gemma_gtexv7_h4, function(x) x[6]),
                                stringsAsFactors=F)

gemma_gtexv7_h4.df.max = gemma_gtexv7_h4.df %>%
  gather(key=hypothesis, value=prob, -gene_id) %>%
  group_by(gene_id) %>%
  filter(prob==max(prob))


# v8
gemma_gtexv8_h4 = lapply(unique(gemma.gtex.v8.egenes$gene_id), function(gene){
  df.tmp = gemma.gtex.v8.egenes %>%
    filter(gene_id == gene) %>%
    arrange(pos_hg38)
  coloc.abf(dataset1=list(snp=df.tmp$pos_hg38,
                          pvalues=df.tmp$p.value,
                          type="quant",
                          N=162,
                          MAF=df.tmp$maf),
            dataset2=list(snp=df.tmp$pos_hg38,
                          pvalues=df.tmp$gtexv8_pval,
                          type="quant",
                          N=670,
                          MAF=df.tmp$gtexv8_maf))$summary
})

gemma_gtexv8_h4.df = data.frame(gene_id = unique(gemma.gtex.v8.egenes$gene_id),
                                H0 = sapply(gemma_gtexv8_h4, function(x) x[2]),
                                H1 = sapply(gemma_gtexv8_h4, function(x) x[3]),
                                H2 = sapply(gemma_gtexv8_h4, function(x) x[4]),
                                H3 = sapply(gemma_gtexv8_h4, function(x) x[5]),
                                H4 = sapply(gemma_gtexv8_h4, function(x) x[6]),
                                stringsAsFactors=F)

gemma_gtexv8_h4.df %>%
  gather(key = Hypothesis, value = frac, -gene_id) %>%
  mutate(indep = gene_id %in% gemma.cond_sig$gene_id) %>%
  ggplot(aes(x=frac, color=Hypothesis)) +
  stat_ecdf() + 
  facet_grid(. ~ indep)

gemma_gtexv8_h4.df.max = gemma_gtexv8_h4.df %>%
  gather(key=hypo, value=prob, -gene_id) %>%
  group_by(gene_id) %>%
  filter(prob==max(prob))

# how does the shared direction of effect size vary across coloc?
gemma.egenes %>%
  filter(top_esnp) %>%
  select(pos_hg19, gene_id, v19_id, allele0, allele1, beta) %>%
  merge(gtex.v7.egenes) %>%
  mutate(gtexv7_slope = ifelse(allele1==v7_allele1 &
                                 allele0==v7_allele2, gtexv7_slope, -gtexv7_slope)) %>%
  merge(gemma_gtexv7_h4.df.max) %>%
  ggplot(aes(x=beta, y=gtexv7_slope, color=hypo)) +
  geom_point(alpha=0.5) +
  theme_classic()

# v8
gemma.egenes %>%
  filter(top_esnp) %>%
  select(pos_hg19, gene_id, v19_id, allele0, allele1, beta) %>%
  merge(gene_map) %>%
  merge(hg19_to_hg38) %>%
  unique %>%
  merge(gtex.v8.egenes) %>%
  mutate(gtexv8_slope = ifelse(allele1==v8_ref &
                                 allele0==v8_alt, gtexv8_slope, -gtexv8_slope)) %>%
  merge(gemma_gtexv8_h4.df.max) %>%
  ggplot(aes(x=beta, y=gtexv8_slope, color=hypo)) +
  geom_point(alpha=0.8) +
  scale_color_ptol() +
  theme_classic()

# do effect sizes correlate better across different color outputs?
gemma.egenes %>%
  filter(top_esnp) %>%
  select(pos_hg19, gene_id, v19_id, allele0, allele1, beta) %>%
  merge(gtex.v7.egenes) %>%
  mutate(gtexv7_slope = ifelse(allele1==v7_allele1 &
                                 allele0==v7_allele2, gtexv7_slope, -gtexv7_slope)) %>%
  merge(gemma_gtexv7_h4.df.max) %>%
  group_by(hypo) %>%
  summarise(cor_beta = cor(beta, gtexv7_slope))



```


# Credible sets in Africans and GTEx

```{r}

gtex.v7.egenes.range.tabix = gencode.v19.tss %>% 
  filter(v19_id %in% gtex.v7.egenes$v19_id) %>%
  mutate(start = tss - 100000,
         end = tss + 100000)

gtex.v8.egenes.range.tabix = gencode.v26.tss %>% 
  filter(v26_id %in% gtex.v8.egenes$v26_id) %>%
  mutate(start = tss - 100000,
         end = tss + 100000) %>%
  mutate(start = ifelse(start<0, 0, start))

# get all GTEx variants with TSS < 100kb and MAF > 0.05
gtex.v7.egenes.range = tabix2df(gtex.v7.f,
                                tabixify(gtex.v7.egenes.range.tabix$v19_id,
                                         gtex.v7.egenes.range.tabix$start,
                                         gtex.v7.egenes.range.tabix$end),
                                gtex.v7.cols) %>%
  # filter(gtexv7_maf > 0.05) %>%
  mutate(gene_id = strip_ensembl(v19_id),
         eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"))

gtex.v8.egenes.range = tabix2df(gtex.v8.f,
                                tabixify(gtex.v8.egenes.range.tabix$v26_id,
                                         gtex.v8.egenes.range.tabix$start,
                                         gtex.v8.egenes.range.tabix$end),
                                gtex.v8.cols) %>%
  # filter(gtexv8_maf >= 0.05) %>%
  mutate(gene_id = strip_ensembl(v26_id),
         eqtl_hg38 = paste(gene_id, pos_hg38, sep=":"))

# add hg19 info
gtex.v8.egenes.range = liftover_snp(addchr(gtex.v8.egenes.range$chr),
                                    gtex.v8.egenes.range$pos_hg38,
                                    "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg19 = pos_new, pos_hg38 = pos_old) %>%
  mutate(chr = rmchr(chr)) %>%
  unique %>%
  merge(gtex.v8.egenes.range, all.y=T)


write.table(gtex.v8.egenes.range, "gtex.v8.egenes.range.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f gtex.v8.egenes.range.tsv")

gtex.v8.egenes.range = read_tsv("./misc_data/gtex.v8.egenes.range.tsv.gz")

# get GTExv7 credible sets
gtex.v7.cred_set.all = lapply(unique(gtex.v7.egenes.range$gene_id),
                              function(x){
                                df.tmp = gtex.v7.egenes.range %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg19,
                                               df.tmp$gtexv7_slope,
                                               df.tmp$gtexv7_se^2, 0.9)})
names(gtex.v7.cred_set.all) = unique(gtex.v7.egenes.range$gene_id)

# get GTExv8 credible sets
gtex.v8.cred_set.all = lapply(unique(gtex.v8.egenes.range$gene_id),
                              function(x){
                                df.tmp = gtex.v8.egenes.range %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg38,
                                               df.tmp$gtexv8_slope,
                                               df.tmp$gtexv8_se^2, 0.9)})
names(gtex.v8.cred_set.all) = unique(gtex.v8.egenes.range$gene_id)

# get African credible sets
gemma.cred_set.all = lapply(unique(gemma.egenes$gene_id),
                              function(x){
                                df.tmp = gemma.egenes %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg19,
                                               df.tmp$beta,
                                               df.tmp$se^2, 0.9)})
names(gemma.cred_set.all) = unique(gemma.egenes$gene_id)


# combine credible set info from Africans and GTEx
gemma.v8.cred_set = data.frame(cred_set = unlist(gemma.cred_set.all)) %>%
  # separate into gene ID and position columns
  separate(cred_set, into=c("gene_id", "pos_hg19"), sep=":", convert=T) %>%
  # add chromosome information
  merge(gemma.egenes %>% select(gene_id, pos_hg19, chr)) %>%
  # add hg38 position information, including only those SNPs that map between hg19 and hg38
  merge(hg19_to_hg38, all.x=T) %>%
  # specify that this is in the African credible set
  mutate(afr=T) %>%
  # add GTEx credible set info
  merge(data.frame(cred_set = unlist(gtex.v8.cred_set.all)) %>%
          # separate as above
          separate(cred_set, into=c("gene_id", "pos_hg38"), sep=":", convert=T) %>%
          # specify these are in the GTEx v8 credible set
          mutate(v8=T),
        # eQTLs (gene_id-pos_hg38 pairs) that overlap will merge, those that don't will not
        all=T) %>%
  unique


# now to compare credible set sizes between Africans and GTEx v8
# color by the SNP threshold until credible sets overlap
gemma.v8.cred_set %>%
  group_by(gene_id) %>%
  summarise(afr_cred_set = sum(afr, na.rm=T),
            v8_cred_set = sum(v8, na.rm=T)) %>%
  merge(gemma.v8.cred_set %>%
          filter(afr & v8) %>%
          group_by(gene_id) %>%
          summarise(min_afr = min(afr_rank),
                 min_v8 = min(v8_rank),
                 thresh = max(min_afr, min_v8)) %>%
          select(gene_id, thresh),
        all.x=T) %>%
  ggplot(aes(x=log10(afr_cred_set),
             y=log10(v8_cred_set),
             color=log10(thresh))) +
  geom_point() +
  theme_classic() +
  scale_color_distiller(palette = "Spectral")
          

# get credible set sizes and intersection sizes
gemma.v8.cred_set.overlaps = gemma.v8.cred_set %>%
  group_by(gene_id) %>%
  summarise(nafr = sum(afr, na.rm=T),
            nv8 = sum(v8, na.rm=T),
            nboth = sum(afr & v8, na.rm=T),
            smaller = min(nafr, nv8))

ggplot(gemma.v8.cred_set.overlaps,
       aes(x=log10(nafr),
           y=log10(nv8),
           color=(smaller-nboth)/smaller)) +
  geom_abline(aes(intercept=0, slope=1), linetype="dashed", alpha=0.5) +
  geom_point(alpha=0.8, size=2) +
  scale_shape_manual(values = c(4, 1)) +
  scale_color_distiller(palette = "Spectral") +
  theme_classic(base_size=15) +
  ggtitle("eQTL credible sets, Africans vs GTEx")


# for how many genes does the intersection of the credible sets contain the top SNP for both?
gemma.v8.intersect_top = gemma.v8.cred_set %>%
  # filter to those eQTLs that are the in the credible sets of both
  filter(afr & v8) %>%
  # for each gene, check whether it contains the top eQTL in neither
  group_by(gene_id) %>%
  summarise(has_top_afr = sum(top_esnp) > 0,
            has_top_v8 = sum(top_esnp_v8) > 0)

# table of how many credible set overlaps contain neither, either or both top SNPs for genes
gemma.v8.intersect_top %>%
  select(has_top_afr, has_top_v8) %>%
  table
            
# for how many genes is the top SNP the same?
gemma.v8.cred_set %>%
    filter(top_esnp & top_esnp_v8) %>%
  pull(gene_id) %>% 
  unique %>% 
  length

# plot cred_set sizes colored by whether they contain both top eSNPs
gemma.v8.cred_set.overlaps = gemma.v8.cred_set %>%
  filter(gene_id %in% gemma.v8.intersect_genes) %>%
  group_by(gene_id) %>%
  summarise(nafr = sum(afr, na.rm=T),
            nv8 = sum(v8, na.rm=T),
            nboth = sum(afr & v8, na.rm=T)) %>%
  merge(gemma.v8.intersect_top)

# Compare credible set sizes
data.frame(gene_id = unique(gtex.v8.egenes.range$gene_id),
           afr_cred_set = sapply(gemma.cred_set.all[unique(gemma.egenes$gene_id) %in% unique(gtex.v8.egenes.range$gene_id)],
                                 function(x) length(x)),
           gtexv8_cred_set = sapply(gtex.v8.cred_set.all, function(x) length(x))) %>%
  merge(gemma_gtexv8_h4.df.max) %>%
  ggplot(aes(x=log10(afr_cred_set), y=log10(gtexv8_cred_set), color=hypothesis)) +
  geom_point(alpha=0.6) +
  geom_abline(aes(intercept=0, slope=1), color="red") +
  ggtitle("eQTL credible set sizes") +
  theme_classic(base_size=15)


# plot loci where the African credible set is smaller and the gtexv7 pvalue is less than 0.001
gemma.egenes %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  select(gene_id, pos_hg19, maf, p.value) %>%
  merge(gtex.v7.egenes.range %>% select(gene_id, pos_hg19, gtexv7_maf, gtexv7_pval)) %>%
  merge(foo) %>%
  merge(gene_name) %>%
  arrange(afr_cred_set / v7_cred_set) %>%
  head(20)

# plot loci where the African credible set is smaller and the EUR variant MAF < 0.01
foo = merge(data.frame(gene_id = unique(gemma.egenes$gene_id),
                       afr_cred_set = sapply(gemma.cred_set.all, function(x) length(x))),
            data.frame(gene_id = unique(gtex.v7.egenes.range$gene_id),
                       v7_cred_set = sapply(gtex.v7.cred_set.all, function(x) length(x))))

bar = merge(data.frame(gene_id = unique(gemma.egenes$gene_id),
                       afr_cred_set = sapply(gemma.cred_set.all, function(x) length(x))),
            data.frame(gene_id = unique(gtex.v8.egenes.range$gene_id),
                       v8_cred_set = sapply(gtex.v8.cred_set.all, function(x) length(x))))
```


# Liftover gtex to eAfr

```{r}

hg38_to_hg19 = read_tsv("../misc/GTEx_Analysis_v8_eQTL_all_associations_Whole_Blood.allpairs.posits.hg19.txt.gz",
                        col_names = c("chr","pos_hg19","pos_hg38"))


gtex.v8.cred_set.liftover = data.frame(gtex_cred_set = unlist(gtex.v8.cred_set.all)) %>%
  separate(gtex_cred_set, into=c("gene_id","pos_hg38"), remove=F, convert=T) %>%
  merge(gtex.v8.egenes %>%
          select(v26_id, chr) %>%
          mutate(gene_id = strip_ensembl(v26_id)) %>%
          select(-v26_id) %>%
          unique) %>%
  merge(hg38_to_hg19, all.x=T)

afr.cred_set.liftover = data.frame(afr_cred_set = unlist(gemma.cred_set.all)) %>%
  separate(afr_cred_set, into=c("gene_id","pos_hg19"), remove=F, convert=T)

credset.liftover = merge(afr.cred_set.liftover, gtex.v8.cred_set.liftover, all=T)
  
credset_intersect = credset.liftover %>%
  group_by(gene_id) %>%
  summarise(afr_size = sum(!is.na(afr_cred_set)),
            gtex_size = sum(!is.na(gtex_cred_set)),
            intersect = sum(!(is.na(afr_cred_set) | is.na(gtex_cred_set))),
            count = n(),
            smaller_set = ifelse(afr_size < gtex_size, afr_size, gtex_size))

ggplot(credset_intersect,
       aes(x=log10(afr_size),
           y=log10(gtex_size),
           size=(smaller_set - intersect)/smaller_set)) +
  geom_point(alpha=0.5) +
  theme_classic(base_size=15)

```


# Are eQTLs found in Africans but not GTEx WB enriched for any tissues?

```{r}

# get all genes tested in Whole Blood
gtex.v8.wb_tested = read_tsv("./misc_data/GTEx_Analysis_v8_eQTL_all_associations_Whole_Blood.tested_genes.txt",
                             col_names=c("v26_id")) %>%
  merge(gene_map) %>%
  filter(!is.na(v26_id))

gtex.v8.all_signif.cols = c("chr","pos_hg38","v8_ref","v8_alt","version","v26_id","tss_distance","ma_samples","ma_count","gtexv8_maf","gtexv8_pval","gtexv8_slope","gtexv8_se","gtexv8_pval_thresh","v8_pval_nominal_threshold","v8_min_pval_nominal","tissue")

gtex.v8.all_signif.tabix = gtex.v8.egenes.range.tabix %>%
  mutate(start = tss - 1000000) %>%
  mutate(end = tss + 1000000) %>%
  mutate(start = ifelse(start < 0, 0, start))


# are there any genes that have an eQTL in Africans but no eQTL in GTEx v8
# all significant eQTL associatios in GTex v8
# these data need to be downloaded, sorted, and tabix'd (see scripts)
gtex.v8.all_signif = tabix2df("./misc_data/GTEx_Analysis_v8_eQTL/ALL.v8.signif_variant_gene_pairs.sort.txt.gz",
                              tabixify(gtex.v8.all_signif.tabix$v26_id,
                                       gtex.v8.all_signif.tabix$start,
                                       gtex.v8.all_signif.tabix$end),
                              gtex.v8.all_signif.cols)

afr_spec_pc_genes = gemma.egenes %>%
  select(gene_id) %>%
  unique %>%
  merge(gene_map) %>%
  filter(!(v26_id %in% unique(gtex.v8.all_signif$v26_id))) %>%
  merge(gene_name) %>%
  filter(!is.na(v26_id)) %>% # must be in gencode v26
  merge(gencode.pc_linc %>%
          select(gene_id, gene_type)) %>%
  pull(gene_id)

# plot gene expression, color by
# 1. is there an eQTL
# 2. does the gene have an eQTL in Africans and not in whole blood
# 3. is untested in all of GTEx

foo = gtex.v8.med_gex %>%
  select(gene_id, gtexv8_tpm=`Whole Blood`) %>%
  merge(gene_map) %>%
  merge(tpm.median %>% dplyr::rename(afr_tpm=median_tpm)) %>%
  mutate(afr_eqtl = gene_id %in% gemma.egenes$gene_id,
         maps_v26 = !is.na(v26_id),
         gtex_eqtl = gene_id %in% strip_ensembl(gtex.v8.all_signif$v26_id),
         gtex_wb_eqtl = gene_id %in% gtex.v8.fdr05$gene_id,
         class = c("no_eqtl","afr_eqtl","maps_v26","gtex_eqtl","gtex_wb_eqtl")[
           1 + as.numeric(afr_eqtl) +
             as.numeric(afr_eqtl & maps_v26) +
             as.numeric(afr_eqtl & maps_v26 & gtex_eqtl) +
             as.numeric(afr_eqtl & maps_v26 & gtex_eqtl & gtex_wb_eqtl)])

```


# Find eQTLs for the subsetted 162 European Americans from GTEx

```{r}

eur162.f = "./misc_data/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz"
eur162.cols = c("snp_hg38","v29_id","chr","pos_hg38","eur162_ref","eur162_alt",
                     "version","tss_dist","eur162_pval","eur162_slope","eur162_maf")

gencode29.pc_linc = read_tsv("../genomes/gencode.v29.pc_linc.bed")

eur162 = fread(eur162.f, col.names=eur162.cols) %>%
  select(-c(snp_hg38, version, tss_dist)) %>%
  merge(., gencode29.pc_linc %>% select(-chr), by="v29_id") %>%
  mutate(pos_hg38 = as.numeric(pos_hg38)) %>%
  mutate(eur162_dist = ifelse(strand=="+", pos_hg38 - start, end - pos_hg38)) %>%
  select(-c(start, end)) %>%
  filter(abs(eur162_dist) < 100000) %>%
  mutate(gene_id = strip_ensembl(v29_id)) %>%
  group_by(gene_id) %>%
  mutate(bhp = p.adjust(eur162_pval, method="BH")) %>%
  mutate(eqtl_hg38 = paste(gene_id, pos_hg38, sep=":"))

eur162.min_bhp = eur162 %>%
  group_by(gene_id) %>%
  summarise(min_bhp = min(bhp))

fdr_thresh = eur162.min_bhp %>%
  mutate(bhp_bhp = p.adjust(min_bhp, method="BH")) %>%
  filter(bhp_bhp<=0.05) %>%
  pull(min_bhp) %>%
  max

eur162.egenes = eur162 %>%
  mutate(fdr_sig = bhp <= fdr_thresh) %>%
  group_by(gene_id) %>%
  filter(sum(fdr_sig) > 0) %>%
  mutate(top_esnp = (eur162_pval==min(eur162_pval)))

# clean up
rm(eur162)

write.table(eur162.egenes, "eur162.egenes.070721.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f eur162.egenes.070721.tsv")

```


# Pi1 with 162 subset

```{r}

# get pi1 of Afr in EUR162
gemma.eur162.fdr05.tabix = gemma.egenes %>%
  filter(fdr_sig) %>%
  select(chr, pos_hg19, gene_id) %>%
  merge(hg19_to_hg38) %>%
  unique %>%
  merge(gene_map) %>%
  filter(!is.na(v29_id)) %>%
  mutate(tabix = tabixify(v29_id, pos_hg38, pos_hg38)) %>%
  pull(tabix)
  
gemma.eur162.fdr05 = tabix2df("./data/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz",
                               gemma.eur162.fdr05.tabix,
                               c("snp_hg38","v29_id","chr","pos_hg38","eur162_ref","eur162_alt",
                                 "version","tss_dist","eur162_pval","eur162_slope","eur162_maf"))

# pi1 of EUR162 in Afr
eur162.fdr05 = eur162.egenes %>%
  filter(fdr_sig)

eur162.gemma.fdr05.tabix = liftover_snp(eur162.fdr05$chr, eur162.fdr05$pos_hg38,
                                         "./misc_data/hg38ToHg19.over.chain") %>%
  dplyr::rename(pos_hg19 = pos_new, pos_hg38 = pos_old) %>%
  merge(eur162.fdr05) %>%
  select(chr, pos_hg19, gene_id) %>%
  unique %>%
  merge(gene_map) %>%
  filter(!is.na(v19_id)) %>%
  mutate(tabix = tabixify(v19_id, pos_hg19, pos_hg19)) %>%
  pull(tabix)

eur162.gemma.fdr05 = tabix2df("./data/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.sort.txt.gz",
                               eur162.gemma.fdr05.tabix,
                               gemma.sort.cols)

# pi1 of EUR162 in the full v8 dataset
eur162.v8.fdr05.tabix = eur162.fdr05 %>%
  merge(gene_map) %>%
  filter(!is.na(v26_id)) %>%
  mutate(tabix = tabixify(v26_id, pos_hg38, pos_hg38)) %>%
  pull(tabix) %>%
  unique

# must generate file GTEx_Analysis_v8_sQTL_all_associations_Whole_Blood.v8.sqtl_allpairs.sort.txt.gz using scripts
eur162.v8.fdr05 = tabix2df("./misc_data/GTEx_Analysis_v8_sQTL_all_associations_Whole_Blood.v8.sqtl_allpairs.sort.txt.gz",
                           eur162.v8.fdr05.tabix,
                           gtex.v8.cols)

```


# Credible sets in Africans vs EAs

```{r message=FALSE}

afr_tested = read_tsv("./misc_data/afr_tested_genes.txt", col_names="gene_id") %>%
  mutate(gene_id = strip_ensembl(gene_id)) %>%
  pull(gene_id)
eur162_tested = read_tsv("./misc_data/eur162_tested_genes.txt", col_names="gene_id") %>%
  mutate(gene_id = strip_ensembl(gene_id)) %>%
  pull
afr_eur162_tested = intersect(afr_tested, eur162_tested)

# egenes with an FDR-significant egene in either study
afr_eur162_any_sig = gene_map %>%
  filter((v19_id %in% gemma.egenes$v19_id |
           v29_id %in% eur162.egenes$v29_id) &
           gene_id %in% afr_eur162_tested) %>%
  pull(gene_id)

# egenes with an FDR-significant egenes in both studies
afr_eur162_both_sig = gemma.egenes %>%
  select(gene_id) %>% unique %>% # get all unique gene ids
  merge(eur162.egenes %>% select(gene_id) %>% unique) %>%
  filter(gene_id %in% afr_eur162_any_sig) %>%
  pull(gene_id) %>% unique

gemma.sort.cols = c("chr","pos_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")
gemma.any_sig = tabix2df("./data/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.sort.txt.gz",
                         tabixify(gene_map %>%
                                    filter(gene_id %in% afr_eur162_any_sig) %>%
                                    pull(v19_id),
                                  "0", "1000000000"),
                         gemma.sort.cols) %>%
  select(-c(foo, bar)) %>%
  filter(r2 > 0.3) %>% # remove poorly imputed SNPs
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af))


eur162.any_sig = tabix2df("./data/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz",
                          tabixify(gene_map %>%
                                    filter(gene_id %in% afr_eur162_any_sig) %>%
                                    pull(v29_id),
                                  "0", "1000000000"),
                          c("snp_hg38","v29_id","chr","pos_hg38","eur162_ref","eur162_alt",
                            "version","tss_dist","eur162_pval","eur162_slope","eur162_maf")) %>%
  select(-c(snp_hg38, version, tss_dist)) %>%
  merge(., gencode29.pc_linc %>% select(-chr), by="v29_id") %>%
  mutate(pos_hg38 = as.numeric(pos_hg38)) %>%
  mutate(eur162_dist = ifelse(strand=="+", pos_hg38 - start, end - pos_hg38)) %>%
  select(-c(start, end)) %>%
  filter(abs(eur162_dist) < 100000) %>%
  mutate(gene_id = strip_ensembl(v29_id)) %>%
  mutate(eqtl_hg38 = paste(gene_id, pos_hg38, sep=":"))


# get credible sets
gemma.any_sig.cred_set = lapply(afr_eur162_any_sig,
                                function(x){
                                  df.tmp = gemma.any_sig %>%
                                    filter(gene_id==x)
                                  get_cred_set.est(ds = list(beta=df.tmp$beta,
                                                             varbeta=df.tmp$se^2,
                                                             MAF=df.tmp$maf,
                                                             N=nrow(df.tmp),
                                                             type="quant"),
                                                   cred_perc = 0.9)})

eur162.any_sig.cred_set = lapply(afr_eur162_any_sig,
                                 function(x){
                                   df.tmp = eur162.any_sig %>%
                                     filter(gene_id==x)
                                   get_cred_set.p(df.tmp$eqtl_hg38,
                                                  df.tmp$eur162_pval,
                                                  df.tmp$eur162_maf, 162, 0.9)})


gemma.eur162.any_sig.intersect = data.frame(afr_cred_set = unlist(gemma.any_sig.cred_set)) %>%
  separate(afr_cred_set, into=c("gene_id","pos_hg19"), convert=T, remove=F) %>%
  merge(hg19_to_hg38, all.x=T) %>%
  merge(data.frame(eur162_cred_set = unlist(eur162.any_sig.cred_set)) %>%
          separate(eur162_cred_set, into=c("gene_id","pos_hg38"), convert=T, remove=F),
        all=T)

gemma.eur162.any_sig.overlap = gemma.eur162.any_sig.intersect %>%
  group_by(gene_id) %>%
  summarise(afr_size = sum(!is.na(afr_cred_set)),
            eur162_size = sum(!is.na(eur162_cred_set)),
            intersect = sum(!(is.na(afr_cred_set) | is.na(eur162_cred_set))),
            smaller_set = ifelse(afr_size < eur162_size, afr_size, eur162_size),
            count = n())


# plot credible set sizes for genes with an eQTL in both datasets
gemma.eur162.any_sig.overlap %>%
  # filter(gene_id %in% afr_eur162_both_sig) %>%
  mutate(frac_uniq = (smaller_set - intersect)/smaller_set) %>%
  ggplot(aes(x=log10(afr_size),
             y=log10(eur162_size),
             color=frac_uniq,
             shape=T)) +
  geom_abline(aes(intercept=0, slope=1), linetype="dashed", alpha=0.5) +
  geom_point(alpha=0.8, size=2) +
  scale_shape_manual(values = c(1)) +
  scale_color_distiller(palette = "Spectral") +
  theme_classic(base_size=15) +
  ggtitle("eQTL credible sets, Africans vs 162 GTEx")

afr_eur162.any_sig.cred_set = data.frame(gene_id = afr_eur162_any_sig,
                                         afr_cred_set = sapply(gemma.any_sig.cred_set, function(x) length(x)),
                                         eur162_cred_set = sapply(eur162.any_sig.cred_set, function(x) length(x))) %>%
  merge(gemma.any_sig %>%
          group_by(gene_id) %>%
          summarise(afr_n = n())) %>%
  merge(eur162.any_sig %>%
          group_by(gene_id) %>%
          summarise(eur162_n = n()))

afr_eur162.any_sig.cred_set %>%
  select(afr_cred_set, eur162_cred_set) %>%
  gather(key=set, value=size) %>%
  ggplot(aes(x=log10(size), color=set)) +
  geom_freqpoly() +
  theme_classic()


# plot of credible sets colored by whether they impute *well*
ggplot(afr_eur162.any_sig.cred_set,
       aes(x=log10(afr_cred_set),
           y=log10(eur162_cred_set))) +
  geom_point(alpha=0.2) +
  geom_abline(aes(intercept=0, slope=1), color="red") +
  theme_classic(base_size=15)

```


# Run coloc between Africans and European Americans

```{r}

# calculate colocalization
gemma.eur162.any_sig.h4 = lapply(afr_eur162_any_sig, function(gene){
    df.tmp = gemma.eur162.any_sig %>%
      filter(gene_id == gene) %>%
      arrange(pos_hg38)
    coloc.abf(dataset1=list(snp=df.tmp$eqtl_hg38,
                            pvalues=df.tmp$p.value,
                            type="quant",
                            N=162,
                            MAF=df.tmp$maf),
              dataset2=list(snp=df.tmp$eqtl_hg38,
                            pvalues=df.tmp$eur162_pval,
                            type="quant",
                            N=162,
                            MAF=df.tmp$eur162_maf))$summary
    })

```


# Look at LD with top SNPs

```{r}

esnps = read_tsv("./misc_data/top_esnps.txt",
                 col_names=c("chr","pos1"))

ld_cols = c("chr","pos1","chr","pos2","n","r2")

afr162_ld = read_tsv("./misc_data/5M.imputed.rna_samps.biallelic.list.hap.ld.gz",
                     col_names=ld_cols) %>%
  rbind(esnps %>%
          mutate(chr_1 = chr,
                 pos2 = pos1,
                 n=324,
                 r2=1))
                 

afr_sub_ld = read_tsv("./misc_data/ALL.phase3_shapeit2_mvncall_integrated_v5a.20130502.afr.sub_162.list.hap.ld.gz",
                      col_names=ld_cols) %>%
  rbind(esnps %>%
          mutate(chr_1 = chr,
                 pos2 = pos1,
                 n=324,
                 r2=1))

eur_sub_ld = read_tsv("./misc_data/ALL.phase3_shapeit2_mvncall_integrated_v5a.20130502.eur.sub_162.list.hap.ld.gz",
                      col_names=ld_cols) %>%
  rbind(esnps %>%
          mutate(chr_1 = chr,
                 pos2 = pos1,
                 n=324,
                 r2=1))

```



# NATURAL SELECTION

## Enrichment of eQTLs and sQTLs among FST outliers

```{r}

maf_bins = 0:10/20
ntag_bins = c(-1, 0, 1, 2, 5, 10, 20, 50, Inf)


tags = read_tsv("./misc_data/5M_rna_pops/5M.imputed.rna_pops.af.tags.list.max_fst.gz",
                    col_types = "cncnnnn") %>%
  select(snp_hg19=snp, ntag, maf, fst, max_fst) %>%
  # filter(!is.na(fst) & !is.na(max_fst)) %>%
  mutate(ntag_bin = cut(ntag, ntag_bins),
         maf_bin = cut(maf, maf_bins)) %>%
  arrange(maf_bin, ntag_bin)

##min/max indices for eQTLs
fst_samp_mats = tags %>%
  select(maf_bin, ntag_bin) %>%
  get_samp_mats

gemma.max_fst = gemma.egenes %>%
  merge(tags, by="snp_hg19") %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1)

gemma.max_fst %>%
  select(eqtl_hg19, ntag_bin, maf_bin, fst, max_fst) %>%
  write.table("gemma.max_fst.tsv", sep="\t", quote=F, col.names=F, row.names=F)


lc.max_fst = lc.sgenes %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(tags, by="snp_hg19") %>%
  group_by(intron) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1)

lc.max_fst %>%
  select(sqtl_hg19, ntag_bin, maf_bin, fst, max_fst) %>%
  write.table("gemma.max_fst.tsv", sep="\t", quote=F, col.names=F, row.names=F)

set.seed(42)

# sample variables
gemma.max_fst.samp = mapply(function(i, j) sample_idx(fst_samp_mats$min_mat, fst_samp_mats$max_mat, i, j),
                            gemma.max_fst$maf_bin, gemma.max_fst$ntag_bin)

lc.max_fst.samp = mapply(function(i, j) sample_idx(fst_samp_mats$min_mat, fst_samp_mats$max_mat, i, j),
                         lc.max_fst$maf_bin, lc.max_fst$ntag_bin)

e.fst.null = apply(gemma.max_fst.samp, 1, function(x) tags[x,]$max_fst)
s.fst.null = apply(lc.max_fst.samp, 1, function(x) tags[x,]$max_fst)


write.table(e.fst.null, "e.fst.null.csv", quote=F, sep=",", row.names=F, col.names=F)
write.table(s.fst.null, "s.fst.null.csv", quote=F, sep=",", row.names=F, col.names=F)


e.fst.null.mean = apply(e.fst.null, 2, mean)
s.fst.null.mean = apply(s.fst.null, 2, mean)


e.fst.null.top99 = sapply(1:ncol(e.fst.null), function(x) mean(e.fst.null[,x] > 0.3627606))
s.fst.null.top99 = sapply(1:ncol(s.fst.null), function(x) mean(s.fst.null[,x] > 0.3627606))


eqtl_fst_outliers = data.frame(frac_outlier = e.fst.null.top99) %>%
  ggplot(aes(x=frac_outlier)) +
  geom_histogram(alpha=0.8) +
  theme_classic(base_size=12) +
  geom_vline(aes(xintercept = mean(gemma.max_fst$max_fst > 0.3627606)), color="red", size=1.5, linetype="dashed", alpha=0.8) +
  ggtitle(expression(paste("eQTL ",F[ST]," outliers")))

sqtl_fst_outliers = data.frame(frac_outlier = s.fst.null.top99) %>%
  ggplot(aes(x=frac_outlier)) +
  geom_histogram(alpha=0.8) +
  theme_classic(base_size=12) +
  geom_vline(aes(xintercept = mean(gemma.max_fst$max_fst > 0.3627606)), color="red", size=1.5, linetype="dashed", alpha=0.8) +
  ggtitle(expression(paste("sQTL ",F[ST]," outliers")))

plot_grid(eqtl_fst_outliers,
          sqtl_fst_outliers,
          nrow=1,
          labels=c("A","B"))



e.fst.null.ranks = sapply(1:nrow(gemma.max_fst), function(x) mean(gemma.max_fst[x,]$max_fst > e.fst.null[x,]))
s.fst.null.ranks = sapply(1:nrow(lc.max_fst), function(x) mean(lc.max_fst[x,]$max_fst > s.fst.null[x,]))


# gsea sorted by fst
gemma.max_fst.genes = gemma.max_fst %>%
  arrange(-max_fst) %>%
  pull(gene_id)
e.fst.go = gost(query = gemma.max_fst.genes,
                organism = "hsapiens",
                # custom_bg = unique(gemma.egenes$gene_id),
                ordered_query = T,
                as_short_link = T)

lc.max_fst.genes = lc.max_fst %>%
  group_by(gene_id) %>%
  filter(max_fst == max(max_fst)) %>%
  arrange(-max_fst) %>%
  pull(gene_id)
s.fst.go = gost(query = lc.max_fst.genes,
                organism = "hsapiens",
                # custom_bg = unique(lc.sgenes$gene_id),
                ordered_query = T)

foo = gconvert(query = lc.max_fst.genes,
               organism = "hsapiens",
               target = "go")


# get all genes corresponding to immune response
library(biomaRt)
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl") #uses human ensembl annotations
#gets gene symbol, transcript_id and go_id for all genes annotated with GO:0006955
immune_response.data <- getBM(attributes=c('ensembl_gene_id', 'go_id'),
                              filters = c('go', 'go_parent_term'), values = c('GO:0006955', 'GO:0006955'), mart = ensembl)


e.fst.null.mean_gt03 = apply(e.fst.null, 2, function(x) mean(x > 0.3))
s.fst.null.mean_gt03 = apply(s.fst.null, 2, function(x) mean(x > 0.3))


data.frame(null = e.fst.null.mean_gt03) %>%
  ggplot(aes(x=null)) +
  geom_histogram(alpha=0.5) +
  geom_vline(data=data.frame(frac_outliers = mean(gemma.max_fst$max_fst > 0.3)),
             aes(xintercept=frac_outliers),
             color="red") +
  theme_classic(base_size=15) +
  ggtitle("Fraction Fst outliers among eQTLs")
      
data.frame(null = s.fst.null.mean_gt03) %>%
  ggplot(aes(x=null)) +
  geom_histogram(alpha=0.5) +
  geom_vline(data=data.frame(frac_outliers = mean(lc.max_fst$max_fst > 0.3)),
             aes(xintercept=frac_outliers),
             color="red") +
  theme_classic(base_size=15) +
  ggtitle("Fraction Fst outliers among sQTLs")


```


## Fst with EUR and tQTLs

```{r}

eur_fst.quantiles = read_tsv("../selection/5M.imputed.rna_samps.biallelic.eur.fst.gz") %>%
  pull(fst) %>%
  quantile(c(0.99, 0.995, 0.999))


# FST values from the full 1KG rna-seq populations
# add EUR AF
gemma.egenes.eur_fst = tabix2df("../selection/5M.imputed.rna_samps.biallelic.eur.fst.gz",
                               unique(tabixify(gemma.egenes$chr, gemma.egenes$pos_hg19, gemma.egenes$pos_hg19)),
                               c("chr","pos_hg19","ref_1kg","alt_1kg","afr_af","eur_af","fst"))

# splicing
lc.sgenes.eur_fst = tabix2df("../selection/5M.imputed.rna_samps.biallelic.eur.fst.gz",
                               unique(tabixify(lc.sgenes$chr, lc.sgenes$pos_hg19, lc.sgenes$pos_hg19)),
                               c("chr","pos_hg19","ref_1kg","alt_1kg","afr_af","eur_af","fst"))


# ## calculate FST
# gemma.egenes.eur_fst = gemma.egenes.eur_af %>%
#   merge(gemma.egenes) %>%
#   filter((ref_1kg==allele1 & alt_1kg==allele0) |
#            (ref_1kg==allele0 & alt_1kg==allele1)) %>%
#   mutate(af = ifelse(ref_1kg==allele1, af, 1-af)) %>%
#   mutate(fst = fst_hud(162, 503, af, eur_af),
#          snp_id = paste(chr, pos_hg19, sep=":")) %>%
#   select(gene_id, snp_id, beta, se, fst, p.value, af, eur_af) %>%
#   mutate(maf = ifelse(af < 0.5, af, 1-af),
#          eur_maf = ifelse(eur_af < 0.5, eur_af, 1-eur_af))
# 
# # splicing
# lc.sgenes.eur_fst = lc.sgenes.eur_af %>%
#   merge(lc.sgenes) %>%
#   filter((ref_1kg==allele1 & alt_1kg==allele0) |
#            (ref_1kg==allele0 & alt_1kg==allele1)) %>%
#   mutate(af = ifelse(ref_1kg==allele1, af, 1-af)) %>%
#   mutate(fst = fst_hud(162, 503, af, eur_af),
#          snp_id = paste(chr, pos_hg19, sep=":")) %>%
#   select(intron, gene_id, snp_id, beta, se, fst, p.value, af, eur_af) %>%
#   mutate(maf = ifelse(af < 0.5, af, 1-af),
#          eur_maf = ifelse(eur_af < 0.5, eur_af, 1-eur_af))


## calculate weighted scores
gemma.egenes.eur_fst.by_gene = gemma.egenes.eur_fst %>%
  merge(gemma.egenes) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  group_by(gene_id) %>%
  summarise(weighted_fst = sum(fst * bf)/sum(bf))


# splicing
lc.sgenes.eur_fst.by_gene = lc.sgenes.eur_fst %>%
  merge(lc.sgenes) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  group_by(gene_id, intron) %>%
  summarise(weighted_fst = sum(fst * bf)/sum(bf))

# write tables
gemma.egenes.eur_fst.by_gene %>%
  filter(weighted_fst > eur_fst.quantiles[1]) %>%
  merge(gene_name) %>%
  as.data.frame %>%
  arrange(-weighted_fst) %>%
  write.csv("../figures/gemma.eur_fst.sel_info.csv", quote=F, row.names=F)

lc.sgenes.eur_fst.by_gene %>%
  filter(weighted_fst > eur_fst.quantiles[1]) %>%
  merge(gene_name) %>%
  as.data.frame %>%
  arrange(-weighted_fst) %>%
  write.csv("../figures/lc.eur_fst.sel_info.csv", quote=F, row.names=F)


# run gene-ontology enrichment of genes sorted by weighted FST
gemma.egenes.eur_fst.by_gene.top100 = gemma.egenes.eur_fst.by_gene %>%
  arrange(-weighted_fst) %>%
  head(100) %>%
  pull(gene_id)
gemma.egenes.eur_fst.by_gene.go = gost(query = gemma.egenes.eur_fst.by_gene.top100,
                                       custom_bg = gex_norm$gene_id,
                                       ordered_query = T,
                                       organism = "hsapiens")

#splicing
lc.sgenes.eur_fst.by_gene.top100 = lc.sgenes.eur_fst.by_gene %>%
  group_by(gene_id) %>%
  filter(weighted_fst == max(weighted_fst)) %>%
  arrange(-weighted_fst) %>%
  head(100) %>%
  pull(gene_id)
lc.sgenes.eur_fst.by_gene.go = gost(query = lc.sgenes.eur_fst.by_gene.top100,
                                       ordered_query = T,
                                       organism = "hsapiens")

# GO terms of selected genes
sig_eur_egenes = gemma.egenes.eur_fst.by_gene %>%
  filter(weighted_fst > eur_fst.quantiles[1]) %>%
  pull(gene_id)
sig_eur_sgenes = lc.sgenes.eur_fst.by_gene %>%
  filter(weighted_fst > eur_fst.quantiles[1]) %>%
  pull(gene_id) %>%
  unique
esgenes.eur_fst.by_gene.go = gost(query = c(sig_eur_egenes, sig_eur_sgenes),
                                  custom_bg = unique(c(gemma.egenes$gene_id, lc.sgenes$gene_id)),
                                  organism = "hsapiens", significant = F)


## plot Fst vs tQTL p-value for selected genes

gemma.eur_fst.sel_info = read_csv("../figures/gemma.eur_fst.sel_info.csv")

gemma.egenes %>%
  filter(gene_id %in% gemma.eur_fst.sel_info$gene_id) %>%
  dplyr::select(gene_id, chr, pos_hg19, p.value) %>%
  merge(gene_name) %>%
  merge(gemma.egenes.eur_fst) %>%
  write_tsv(., "../figures/gemma.sel_hits.eur_fst.pval.tsv")

lc.eur_fst.sel_info = read_csv("../figures/lc.eur_fst.sel_info.csv")

lc.sgenes %>%
  filter(intron %in% lc.eur_fst.sel_info$intron) %>%
  dplyr::select(gene_id, intron, chr, pos_hg19, p.value) %>%
  merge(gene_name) %>%
  merge(lc.sgenes.eur_fst) %>%
  write_tsv(., "../figures/lc.sel_hits.eur_fst.pval.tsv")


```


## Calculate pop-specific Fst and PBS

```{r}

pop_samps.5M = read_tsv("./misc_data/rna_pops.txt",
                        col_names=c("sample","pop"))

pop_map = c("Agaw","Amhara","Argobba","Dizi","Hadza","Mursi","Sabue","Sandawe","Weyto")
names(pop_map) = c("agaw","amhara","argoba","dizi","hadzabe","mursi","sabue","sandawe","weyto")

pop_counts.5M = table(pop_samps.5M$pop) %>%
  as.data.frame
colnames(pop_counts.5M) = c("pop","count")
pop_counts.5M$pop = pop_map[pop_counts.5M$pop]



gemma.frq.tabix = gemma.egenes %>%
  mutate(tabix = tabixify(chr, pos_hg19, pos_hg19)) %>%
  pull(tabix) %>%
  unique

frq.egenes = tabix2df("./misc_data/ALL.5M.imputed.rna_samps.biallelic.frq.gz",
                      gemma.frq.tabix,
                      c("chr","pos_hg19","ref","alt",pops))

# eur.egenes = tabix2df("../misc/1kgenomes.eur_af.txt.gz",
#                       gemma.frq.tabix,
#                       c("chr","pos_hg19","onekg_ref","onekg_alt","eur_af")) %>%
#   mutate(eur_af = as.numeric(eur_af))

ceu.egenes = tabix2df("../misc/1kgenomes.ceu_af.txt.gz",
                      gemma.frq.tabix,
                      c("chr","pos_hg19","onekg_ref","onekg_alt","ceu_af")) %>%
  mutate(ceu_af = as.numeric(ceu_af))

yri.egenes = tabix2df("../misc/1kgenomes.yri_af.txt.gz",
                      gemma.frq.tabix,
                      c("chr","pos_hg19","onekg_ref","onekg_alt","yri_af")) %>%
  mutate(yri_af = as.numeric(yri_af))

# # fst between CEU and YRI
# eur_yri.egenes.fst = merge(eur.egenes, yri.egenes) %>%
#   mutate(fst = fst_hud(n1=503, n2=108, p1=eur_af, p2=yri_af)) %>%
#   select(chr, pos_hg19, fst)

# fst between CEU and YRI
ceu_yri.egenes.fst = merge(ceu.egenes, yri.egenes) %>%
  mutate(fst = fst_hud(n1=99, n2=108, p1=ceu_af, p2=yri_af)) %>%
  select(chr, pos_hg19, fst)

# # fst between EUR and eAFR
# eur.egenes.fst = merge(frq.egenes, eur.egenes) %>%
#   filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
#   mutate(eur_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-eur_af, eur_af)) %>%
#   gather(key=pop, value=af, pops) %>%
#   merge(pop_counts.5M) %>%
#   mutate(fst = fst_hud(n1=count, n2=503, p1=af, p2=eur_af),
#          fst = ifelse(af < eur_af, -fst, fst)) %>%
#   select(chr, pos_hg19, pop, fst) %>%
#   pivot_wider(names_from=pop, values_from=fst)

# fst between CEU and eAFR
ceu.egenes.fst = merge(frq.egenes, ceu.egenes) %>%
  filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
  mutate(ceu_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-ceu_af, ceu_af)) %>%
  gather(key=pop, value=af, pops) %>%
  merge(pop_counts.5M) %>%
  mutate(fst = fst_hud(n1=count, n2=99, p1=af, p2=ceu_af),
         fst = ifelse(af < ceu_af, -fst, fst)) %>%
  select(chr, pos_hg19, pop, fst) %>%
  pivot_wider(names_from=pop, values_from=fst)

# fst between YRI and eAFR
yri.egenes.fst = merge(frq.egenes, yri.egenes) %>%
  filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
  mutate(yri_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-yri_af, yri_af)) %>%
  gather(key=pop, value=af, pops) %>%
  merge(pop_counts.5M) %>%
  mutate(fst = fst_hud(n1=count, n2=108, p1=af, p2=yri_af),
         fst = ifelse(af < yri_af, -fst, fst)) %>%
  select(chr, pos_hg19, pop, fst) %>%
  pivot_wider(names_from=pop, values_from=fst)

egenes.pbs = merge(frq.egenes, yri.egenes) %>%
  merge(ceu.egenes) %>%
  filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
  mutate(yri_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-yri_af, yri_af),
         ceu_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-ceu_af, ceu_af),
         yri_ceu_t = -log(1-fst_wc(n1=108, n2=99, p1=yri_af, p2=ceu_af))) %>%
  gather(key=pop, value=af, pops) %>%
  merge(pop_counts.5M) %>%
  mutate(yri_t = -log(1-fst_wc(n1=count, n2=108, p1=af, p2=yri_af)),
         eur_t = -log(1-fst_wc(n1=count, n2=99, p1=af, p2=ceu_af)),
         pbs = (yri_t + eur_t - yri_ceu_t)/2,
         pbs = ifelse(is.infinite(pbs), 6, pbs),
         pbs = ifelse(pbs < 0, 0, pbs)) %>% # replace with 6 if infinite
  select(chr, pos_hg19, pop, pbs) %>%
  pivot_wider(names_from=pop, values_from=pbs)


# # pop-specific fsts by gene
# eur_yri.egenes.pop_fst = gemma.egenes %>%
#   select(gene_id, chr, pos_hg19, beta, se) %>%
#   mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
#   merge(eur_yri.egenes.fst) %>%
#   group_by(gene_id) %>%
#   summarise(weighted_fst = sum(fst * bf)/sum(bf))
# 
# eur.egenes.pop_fst = gemma.egenes %>%
#   select(gene_id, chr, pos_hg19, beta, se) %>%
#   mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
#   merge(eur.egenes.fst) %>%
#   gather(key=pop, value=fst, pops) %>%
#   group_by(gene_id, pop) %>%
#   summarise(weighted_fst = sum(abs(fst) * bf)/sum(bf)) %>%
#   pivot_wider(names_from=pop, values_from=weighted_fst)

ceu.egenes.pop_fst = gemma.egenes %>%
  select(gene_id, chr, pos_hg19, beta, se) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(ceu.egenes.fst) %>%
  gather(key=pop, value=fst, pops) %>%
  group_by(gene_id, pop) %>%
  summarise(weighted_fst = sum(abs(fst) * bf)/sum(bf)) %>%
  pivot_wider(names_from=pop, values_from=weighted_fst)

yri.egenes.pop_fst = gemma.egenes %>%
  select(gene_id, chr, pos_hg19, beta, se) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(yri.egenes.fst) %>%
  gather(key=pop, value=fst, pops) %>%
  group_by(gene_id, pop) %>%
  summarise(weighted_fst = sum(abs(fst) * bf)/sum(bf)) %>%
  pivot_wider(names_from=pop, values_from=weighted_fst)

egenes.pop_pbs = gemma.egenes %>%
  select(gene_id, chr, pos_hg19, beta, se) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(egenes.pbs) %>%
  gather(key=pop, value=pbs, pops) %>%
  group_by(gene_id, pop) %>%
  summarise(weighted_pbs = sum(abs(pbs) * bf)/sum(bf)) %>%
  pivot_wider(names_from=pop, values_from=weighted_pbs)

write.table(ceu.egenes.pop_fst, "ceu.egenes.pop_fst.tsv", quote=F, sep="\t", row.names=F)
write.table(yri.egenes.pop_fst, "yri.egenes.pop_fst.tsv", quote=F, sep="\t", row.names=F)
write.table(egenes.pop_pbs, "egenes.pop_pbs.tsv", quote=F, sep="\t", row.names=F)


# splicing
lc.frq.tabix = lc.sgenes %>%
  mutate(tabix = tabixify(chr, pos_hg19, pos_hg19)) %>%
  pull(tabix) %>%
  unique

frq.sgenes = tabix2df("/local1/derek/data/Illumina5M/frq/ALL.5M.imputed.rna_samps.biallelic.frq.gz",
                      lc.frq.tabix,
                      c("chr","pos_hg19","ref","alt",pops))

# eur.sgenes = tabix2df("../misc/1kgenomes.eur_af.txt.gz",
#                       lc.frq.tabix,
#                       c("chr","pos_hg19","onekg_ref","onekg_alt","eur_af"))

ceu.sgenes = tabix2df("./misc_data/1kgenomes.ceu_af.txt.gz",
                      lc.frq.tabix,
                      c("chr","pos_hg19","onekg_ref","onekg_alt","ceu_af"))

yri.sgenes = tabix2df("./misc_data/1kgenomes.yri_af.txt.gz",
                      lc.frq.tabix,
                      c("chr","pos_hg19","onekg_ref","onekg_alt","yri_af"))

# eur_yri.sgenes.fst = merge(eur.sgenes, yri.sgenes) %>%
#   mutate(fst = fst_wc(n1=503, n2=108, p1=eur_af, p2=yri_af)) %>%
#   select(chr, pos_hg19, fst)
# 
# eur.sgenes.fst = merge(frq.sgenes, eur.sgenes) %>%
#   filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
#   mutate(eur_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-eur_af, eur_af)) %>%
#   gather(key=pop, value=af, all_of(pops)) %>%
#   merge(pop_counts.5M) %>%
#   mutate(fst = fst_wc(n1=count, n2=503, p1=af, p2=eur_af)) %>%
#   select(chr, pos_hg19, pop, fst) %>%
#   pivot_wider(names_from=pop, values_from=fst)

ceu.sgenes.fst = merge(frq.sgenes, ceu.sgenes) %>%
  filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
  mutate(eur_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-ceu_af, ceu_af)) %>%
  gather(key=pop, value=af, all_of(pops)) %>%
  merge(pop_counts.5M) %>%
  mutate(fst = fst_wc(n1=count, n2=99, p1=af, p2=ceu_af)) %>%
  select(chr, pos_hg19, pop, fst) %>%
  pivot_wider(names_from=pop, values_from=fst)

yri.sgenes.fst = merge(frq.sgenes, yri.sgenes) %>%
  filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
  mutate(eur_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-yri_af, yri_af)) %>%
  gather(key=pop, value=af, all_of(pops)) %>%
  merge(pop_counts.5M) %>%
  mutate(fst = fst_wc(n1=count, n2=108, p1=af, p2=eur_af)) %>%
  select(chr, pos_hg19, pop, fst) %>%
  pivot_wider(names_from=pop, values_from=fst)

sgenes.pbs = merge(frq.sgenes, yri.sgenes) %>%
  merge(ceu.sgenes) %>%
  filter((ref==onekg_ref & alt==onekg_alt) | (ref==onekg_alt & alt==onekg_ref)) %>%
  mutate(yri_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-yri_af, yri_af),
         ceu_af = ifelse(ref==onekg_alt & alt==onekg_ref, 1-ceu_af, ceu_af),
         yri_ceu_t = -log(1-fst_wc(n1=108, n2=99, p1=yri_af, p2=ceu_af))) %>%
  gather(key=pop, value=af, pops) %>%
  merge(pop_counts.5M) %>%
  mutate(yri_t = -log(1-fst_wc(n1=count, n2=108, p1=af, p2=yri_af)),
         eur_t = -log(1-fst_wc(n1=count, n2=99, p1=af, p2=ceu_af)),
         pbs = (yri_t + eur_t - yri_ceu_t)/2,
         pbs = ifelse(is.infinite(pbs), 6, pbs)) %>% # replace with 6 if infinite
  select(chr, pos_hg19, pop, pbs) %>%
  pivot_wider(names_from=pop, values_from=pbs)


# gene scores
# eur_yri.sgenes.pop_fst = lc.sgenes %>%
#   select(chr, pos_hg19, intron, beta, se) %>%
#   mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
#   merge(eur_yri.egenes.fst) %>%
#   group_by(intron) %>%
#   summarise(weighted_fst = sum(fst * bf)/sum(bf))
# 
# eur.sgenes.pop_fst = lc.sgenes %>%
#   select(chr, pos_hg19, intron, beta, se) %>%
#   mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
#   merge(eur.sgenes.fst) %>%
#   gather(key=pop, value=fst, all_of(pops)) %>%
#   group_by(intron, pop) %>%
#   summarise(weighted_fst = sum(fst * bf)/sum(bf)) %>%
#   pivot_wider(names_from=pop, values_from=weighted_fst)

ceu.sgenes.pop_fst = lc.sgenes %>%
  select(chr, pos_hg19, intron, beta, se) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(ceu.sgenes.fst) %>%
  gather(key=pop, value=fst, all_of(pops)) %>%
  group_by(intron, pop) %>%
  summarise(weighted_fst = sum(fst * bf)/sum(bf)) %>%
  pivot_wider(names_from=pop, values_from=weighted_fst)

yri.sgenes.pop_fst = lc.sgenes %>%
  select(chr, pos_hg19, intron, beta, se) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(yri.sgenes.fst) %>%
  gather(key=pop, value=fst, all_of(pops)) %>%
  group_by(intron, pop) %>%
  summarise(weighted_fst = sum(fst * bf)/sum(bf)) %>%
  pivot_wider(names_from=pop, values_from=weighted_fst)

sgenes.pop_pbs = lc.sgenes %>%
  select(intron, chr, pos_hg19, beta, se) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(sgenes.pbs) %>%
  gather(key=pop, value=pbs, pops) %>%
  group_by(intron, pop) %>%
  summarise(weighted_pbs = sum(abs(pbs) * bf)/sum(bf)) %>%
  pivot_wider(names_from=pop, values_from=weighted_pbs) %>%
  merge(lc.sgenes %>% select(intron, gene_id) %>% unique)

write.table(ceu.sgenes.pop_fst, "ceu.sgenes.pop_fst.tsv", quote=F, sep="\t", row.names=F)
write.table(yri.sgenes.pop_fst, "yri.sgenes.pop_fst.tsv", quote=F, sep="\t", row.names=F)
write.table(sgenes.pop_pbs, "sgenes.pop_pbs.tsv", quote=F, sep="\t", row.names=F)

```


## PBS

```{r}

pbs = read_tsv("../selection/ALL.5M.imputed.rna_samps.biallelic.hud.pbs.gz",
               col_types = "nnccnnnnnnnnn")
pbs[is.na(pbs)] = 0

pbs.thresh = apply(pbs[,pops], 2, function(x) quantile(x, c(0.99,0.995,0.999))) %>%
  t %>%
  as.data.frame %>%
  tibble::rownames_to_column("pop")
colnames(pbs.thresh) = c("pop","pbs_99","pbs_995","pbs_999")

write.table(pbs.thresh, "pbs.thresh.txt",
            sep="\t", quote=F, row.names=F)

gemma.weighted_pbs = gemma.egenes %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(pbs %>% mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>% select(-c(chr, pos_hg19, ref, alt))) %>%
  gather(key=pop, value=pbs, pops) %>%
  mutate(pbs = ifelse(pbs < -100, -2, pbs)) %>%
  mutate(pbs = ifelse(pbs > 100, 6, pbs)) %>%
  group_by(gene_id, pop) %>%
  summarise(weighted_pbs = sum(pbs * bf)/sum(bf))

write.table(gemma.weighted_pbs, "gemma.weighted_pbs.tsv",
            sep="\t", quote=F, row.names=F)
  
lc.weighted_pbs = lc.sgenes %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(pbs %>% mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>% select(-c(chr, pos_hg19, ref, alt))) %>%
  gather(key=pop, value=pbs, pops) %>%
  mutate(pbs = ifelse(pbs < -100, -2, pbs)) %>%
  mutate(pbs = ifelse(pbs > 100, 6, pbs)) %>%
  group_by(intron, pop) %>%
  summarise(weighted_pbs = sum(pbs * bf)/sum(bf))

write.table(lc.weighted_pbs, "lc.weighted_pbs.tsv",
            sep="\t", quote=F, row.names=F)

```


## D-stat

```{r}

dstat.dir = read_tsv("/local1/derek/data/Illumina5M/fst/ALL.5M.imputed.rna_samps.biallelic.dstat.directional.txt.gz")

dstat.thresh = apply(dstat.dir[,pops], 2, function(x) quantile(abs(x), c(0.99,0.995,0.999))) %>%
  t %>%
  as.data.frame %>%
  tibble::rownames_to_column("pop")
colnames(dstat.thresh) = c("pop","dstat_99","dstat_995","dstat_999")

write.table(dstat.thresh, "dstat.thresh.txt",
            sep="\t", quote=F, row.names=F)

gemma.weighted_dstat = gemma.egenes %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(dstat.dir %>% mutate(snp_hg19 = paste(chr, pos, sep=":")) %>% select(-c(chr, pos, ref, alt))) %>%
  gather(key=pop, value=dstat, pops) %>%
  group_by(gene_id, pop) %>%
  summarise(weighted_dstat = sum(abs(dstat) * bf/sum(bf)))

write.table(gemma.weighted_dstat, "gemma.weighted_dstat.tsv",
            sep="\t", quote=F, row.names=F)
  
lc.weighted_dstat = lc.sgenes %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(dstat.dir %>% mutate(snp_hg19 = paste(chr, pos, sep=":")) %>% select(-c(chr, pos, ref, alt))) %>%
  gather(key=pop, value=dstat, pops) %>%
  group_by(intron, pop) %>%
  summarise(weighted_dstat = sum(abs(dstat) * bf/sum(bf)))

write.table(lc.weighted_dstat, "lc.weighted_dstat.tsv",
            sep="\t", quote=F, row.names=F)

```


## iHS

```{r}

gemma.egenes.ihs = tabix2df("../ihs/round2/ihs/ALL.5M.imputed.rna_pops.unrelated.r2085.all.ihs.out.100bins.norm.sort.gz",
                            tabixify(gemma.egenes$chr,
                                     gemma.egenes$pos_hg19,
                                     gemma.egenes$pos_hg19),
                            c("chr","pos_hg19","alt_frq","ihh1","ihh0",
                              "ihs_raw","ihs_norm","thresh","pop"))

lc.sgenes.ihs = tabix2df("../ihs/round2/ihs/ALL.5M.imputed.rna_pops.unrelated.r2085.all.ihs.out.100bins.norm.sort.gz",
                         tabixify(lc.sgenes$chr,
                                  lc.sgenes$pos_hg19,
                                  lc.sgenes$pos_hg19),
                         c("chr","pos_hg19","alt_frq","ihh1","ihh0",
                           "ihs_raw","ihs_norm","thresh","pop"))


gemma.weighted_ihs = gemma.egenes %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(gemma.egenes.ihs %>% select(chr, pos_hg19, ihs_norm, pop)) %>%
  group_by(gene_id, pop) %>%
  summarise(weighted_ihs = sum(abs(ihs_norm) * bf/sum(bf)))
  
lc.weighted_ihs = lc.sgenes %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  merge(lc.sgenes.ihs %>% select(chr, pos_hg19, ihs_norm, pop)) %>%
  group_by(intron, pop) %>%
  summarise(weighted_ihs = sum(abs(ihs_norm) * bf/sum(bf)))


gemma.fdr05.top.ihs = tabix2df("../ihs/round2/ihs/ALL.5M.imputed.rna_pops.unrelated.r2085.all.ihs.out.100bins.norm.sort.gz",
                            tabixify(gemma.fdr05.top$chr,
                                     gemma.fdr05.top$pos_hg19,
                                     gemma.fdr05.top$pos_hg19),
                            c("chr","pos_hg19","alt_frq","ihh1","ihh0",
                              "ihs_raw","ihs_norm","thresh","pop")) %>%
  merge(gemma.fdr05.top %>% select(gene_id, chr, pos_hg19)) %>%
  unique


lc.fdr05.top.ihs = tabix2df("../ihs/round2/ihs/ALL.5M.imputed.rna_pops.unrelated.r2085.all.ihs.out.100bins.norm.sort.gz",
                            tabixify(lc.fdr05.top$chr,
                                     lc.fdr05.top$pos_hg19,
                                     lc.fdr05.top$pos_hg19),
                            c("chr","pos_hg19","alt_frq","ihh1","ihh0",
                              "ihs_raw","ihs_norm","thresh","pop")) %>%
  merge(lc.fdr05.top %>% select(intron, gene_id, chr, pos_hg19)) %>%
  unique

```

## genes and introns under selection

```{r}

gemma.weighted_dstat = read_tsv("gemma.weighted_dstat.tsv")
gemma.weighted_pbs = read_tsv("gemma.weighted_pbs.tsv")
gemma.weighted_dstat = read_tsv("gemma.weighted_dstat.tsv")
gemma.weighted_pbs = read_tsv("gemma.weighted_pbs.tsv")

# # plots
# eur.egenes.pop_fst %>%
#   gather(key=pop, value=eur_fst, pops) %>%
#   merge(yri.egenes.pop_fst %>% gather(key=pop, value=yri_fst, pops)) %>%
#   merge(gemma.weighted_dstat) %>%
#   merge(dstat.thresh) %>%
#   mutate(sig = eur_fst > 0.3 & yri_fst > 0.3 & weighted_dstat > quant_99) %>%
#   ggplot(aes(x=eur_fst, y=yri_fst, color=weighted_dstat)) +
#   geom_point() + 
#   theme_classic() +
#   facet_wrap(~ pop)

# # genes and introns under selection
# sel_genes = yri.egenes.pop_fst %>%
#   gather(key=pop, value=yri_fst, pops) %>%
#   merge(ceu.egenes.pop_fst %>% gather(key=pop, value=ceu_fst, pops)) %>%
#   merge(gemma.weighted_dstat) %>%
#   merge(dstat.thresh) %>%
#   mutate(sig = ceu_fst > 0.3 &
#            yri_fst > 0.2 &
#            weighted_dstat > dstat_99) %>%
#   filter(sig) %>%
#   merge(gene_name) %>%
#   arrange(ceu_fst + yri_fst)

sel_genes = gemma.weighted_pbs %>%
  merge(gemma.weighted_dstat) %>%
  merge(dstat.thresh) %>%
  merge(pbs.thresh) %>%
  mutate(sig = weighted_pbs > pbs_995 &
           weighted_dstat > dstat_995) %>%
  filter(sig) %>%
  merge(gene_name) %>%
  arrange(-weighted_dstat)


sel_introns = lc.weighted_pbs %>%
  merge(lc.weighted_dstat) %>%
  merge(pbs.thresh) %>%
  merge(dstat.thresh) %>%
  mutate(sig = weighted_pbs > pbs_995 &
           weighted_dstat > dstat_995) %>%
  filter(sig) %>%
  merge(lc.sgenes %>% select(intron, gene_id) %>% unique) %>%
  merge(gene_name) %>%
  arrange(-weighted_dstat)

# sel_introns = sgenes.pop_pbs %>%
#   gather(key=pop, value=pbs, pops) %>%
#   merge(lc.weighted_dstat) %>%
#   merge(dstat.thresh) %>%
#   mutate(sig = pbs > 0.4 &
#            weighted_dstat > quant_99) %>%
#   filter(sig) %>%
#   merge(lc.sgenes %>% select(intron, gene_id) %>% unique) %>%
#   merge(gene_name) %>%
#   arrange(-pbs)

```


# Get allele frequencies of selected introns and genes

```{r}

# get frequencies of top SNPs for eQTLs
sel_genes.top_snps = gemma.egenes %>%
  filter(top_esnp) %>%
  filter(gene_id %in% sel_genes$gene_id) %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1) %>%
  select(gene_id, chr, pos_hg19)

sel_genes_frqs = tabix2df("/local1/derek/data/Illumina5M/frq/ALL.5M.imputed.rna_samps.biallelic.frq.gz",
                          tabixify(sel_genes.top_snps$chr,
                                   sel_genes.top_snps$pos_hg19,
                                   sel_genes.top_snps$pos_hg19),
                          c("chr","pos_hg19","ref","alt",pops)) %>%
  merge(sel_genes.top_snps) %>%
  merge(sel_genes) %>%
  merge(ceu.egenes %>%
          select(chr, pos_hg19, ceu=ceu_af, onekg_ref, onekg_alt)) %>%
  merge(yri.egenes %>%
          select(chr, pos_hg19, yri=yri_af, onekg_ref, onekg_alt))


# get frequencies of top SNPs for sQTLs
sel_introns.top_snps = lc.sgenes %>%
  filter(top_ssnp) %>%
  filter(intron %in% sel_introns$intron) %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1) %>%
  select(gene_id, intron, chr, pos_hg19)

sel_introns_frqs = tabix2df("/local1/derek/data/Illumina5M/frq/ALL.5M.imputed.rna_samps.biallelic.frq.gz",
                          tabixify(sel_introns.top_snps$chr,
                                   sel_introns.top_snps$pos_hg19,
                                   sel_introns.top_snps$pos_hg19),
                          c("chr","pos_hg19","ref","alt",pops)) %>%
  merge(sel_introns.top_snps) %>%
  merge(sel_introns) %>%
  unique %>%
  merge(ceu.egenes %>%
          select(chr, pos_hg19, ceu=ceu_af, onekg_ref, onekg_alt)) %>%
  merge(yri.egenes %>%
          select(chr, pos_hg19, yri=yri_af, onekg_ref, onekg_alt))


# create bar plots of allele frequencies
sel_genes_frqs %>%
  select(gene_name, pop, c(pops, "ceu", "yri")) %>%
  gather(key=pop_frq, value=alt, c(pops, "ceu", "yri")) %>%
  filter(!(pop %in% c("agaw","amhara","argoba"))) %>%
  mutate(ref = 1-alt) %>%
  unite("target", gene_name, pop, sep=":") %>%
  gather(key=allele, value=frq, c(ref, alt)) %>%
  mutate(pop_frq = factor(pop_frq, levels=c(pops, "yri", "ceu"))) %>%
  ggplot(aes(x=pop_frq, y=frq, fill=allele)) +
  geom_bar(position="stack", stat="identity") +
  theme_classic(base_size=15) +
  scale_fill_ptol() +
  facet_grid(target~.) +
  theme(strip.text.y.right = element_text(angle = 0),
        axis.text.x=element_text(angle=60, hjust=1),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
  ggtitle("Frequency of top eQTLs")


sel_introns_frqs %>%
  select(gene_name, pop, c(pops, "ceu", "yri")) %>%
  gather(key=pop_frq, value=alt, c(pops, "ceu", "yri")) %>%
  mutate(ref = 1-alt) %>%
  unite("target", gene_name, pop, sep=":") %>%
  gather(key=allele, value=frq, c(ref, alt)) %>%
  mutate(pop_frq = factor(pop_frq, levels=c(pops, "yri", "ceu"))) %>%
  ggplot(aes(x=pop_frq, y=frq, fill=allele)) +
  geom_bar(position="stack", stat="identity") +
  theme_classic(base_size=15) +
  scale_fill_ptol() +
  facet_grid(target ~ .) +
  theme(strip.text.y.right = element_text(angle = 0),
        axis.text.x=element_text(angle=60, hjust=1),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
  ggtitle("Frequency of top sQTLs")

```


# Pigmentation GWAS

```{r}

tmem216_snps = data.frame(rsid = c("rs2512809","rs7934229","rs7948623"),
                          chr = 11,
                          pos_hg19 = c(61106525,61209133,61137147))


ddb1_range = "11:60629745-61633624"

gwas_cols = c("chr","pos_hg19","id","gwas_beta","gwas_sd","gwas_pval","gwas_fdr","gwas_pvalgc","gwas_fdrgc","gwas_r2","an","ac","gwas_ref","gwas_alt","hwe","gwas_maf")

pigment = tabix2df("/local1/derek/data/gwas/matt_gwas/5M/25.5M.sort.assoc.gz",
                   ddb1_range,
                   gwas_cols)
  
ld = read_table("tmem216_mapping/11.5M.imputed.ddb1_locus.ld",
                skip=1,
                col_names=c("chr","pos1","foo","chr1","pos2","bar","r2")) %>%
  select(chr,pos1,pos2,r2)

ld = ld %>%
  select(chr, pos1) %>%
  unique %>%
  mutate(pos2=pos1, r2=1) %>%
  rbind(ld)

# locus plot colored by LD with top pigmentation GWAS SNP
plot_rs2512809 = gemma.egenes %>%
  filter(gene_id=="ENSG00000187049") %>%
  merge(pigment %>% select(pos_hg19, gwas_pval)) %>%
  merge(ld %>% filter(pos1==61106525) %>% select(chr, pos_hg19 = pos2, var_r2=r2), all.x=T) %>%
  mutate(var_r2 = ifelse(is.na(var_r2), 0, var_r2)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(p.value), color=var_r2)) +
  geom_point() +
  theme_classic() +
  scale_color_distiller(palette = "Spectral") +
  ggtitle("eQTL vs GWAS pval, rs2512809")

# locus plot colored by LD with top TMEM216 eQTL SNP
plot_rs7934229 = gemma.egenes %>%
  filter(gene_id=="ENSG00000187049") %>%
  merge(pigment %>% select(pos_hg19, gwas_pval)) %>%
  merge(ld %>% filter(pos1==61209133) %>% select(chr, pos_hg19 = pos2, var_r2=r2), all.x=T) %>%
  mutate(var_r2 = ifelse(is.na(var_r2), 0, var_r2)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(p.value), color=var_r2)) +
  geom_point() +
  theme_classic() +
  scale_color_distiller(palette = "Spectral") +
  ggtitle("eQTL vs GWAS pval, rs7934229")

# SNP that looks high in GWAS as well as GWAS
plot_rs7948623 = gemma.egenes %>%
  filter(gene_id=="ENSG00000187049") %>%
  merge(pigment %>% select(pos_hg19, gwas_pval)) %>%
  merge(ld %>% filter(pos1==61137147) %>% select(chr, pos_hg19 = pos2, var_r2=r2), all.x=T) %>%
  mutate(var_r2 = ifelse(is.na(var_r2), 0, var_r2)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(p.value), color=var_r2)) +
  geom_point() +
  theme_classic() +
  scale_color_distiller(palette = "Spectral") +
  ggtitle("eQTL vs GWAS pval, rs7948623")



plot_grid(plot_rs2512809, plot_rs7934229, plot_rs7948623, nrow=1)


# genes in the region
pigment_genes = gencode.pc_linc %>%
  filter((chr==11) &
           (start > 61029745 & start < 61233624) |
           (end > 61029745 & end < 61233624)) %>%
  pull(gene_id)


# v8 WB data for TMEM216
v8_wb_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Whole_Blood.allpairs.sort.txt.gz",
                      paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                      c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))

v8_se_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Skin_Sun_Exposed_Lower_leg.allpairs.sort.txt.gz",
                      paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                      c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))

foo = liftover_snp(addchr(pigment$chr),
                   pigment$pos_hg19,
                   "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  unique %>%
  rename(pos_hg38 = pos_new,
         pos_hg19 = pos_old) %>%
  merge(v8_wb_pigment) %>%
  merge(gene_map) %>%
  merge(gene_name) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment)

bar = liftover_snp(addchr(pigment$chr),
                   pigment$pos_hg19,
                   "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  unique %>%
  rename(pos_hg38 = pos_new,
         pos_hg19 = pos_old) %>%
  merge(v8_se_pigment) %>%
  merge(gene_map) %>%
  merge(gene_name) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment)

bar %>%
  filter(gene_id=="ENSG00000187049") %>%
  merge(ld %>% filter(pos1==61106525) %>% select(chr, pos_hg19 = pos2, var_r2=r2), all.x=T) %>%
  mutate(var_r2 = ifelse(is.na(var_r2), 0, var_r2)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(v8_pval), color=var_r2)) +
  geom_point() +
  theme_classic() +
  scale_color_distiller(palette = "Spectral") +
  ggtitle("eQTL vs GWAS pval, rs2512809")

bar %>%
  filter(gene_id=="ENSG00000187049") %>%
  merge(ld %>% filter(pos1==61209133) %>% select(chr, pos_hg19 = pos2, var_r2=r2), all.x=T) %>%
  mutate(var_r2 = ifelse(is.na(var_r2), 0, var_r2)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(v8_pval), color=var_r2)) +
  geom_point() +
  theme_classic() +
  scale_color_distiller(palette = "Spectral") +
  ggtitle("eQTL vs GWAS pval, rs7934229")



ggplot(foo, aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point() +
  theme_classic(base_size=15) +
  facet_wrap(~ gene_name) +
  ggtitle("Whole Blood")

ggplot(bar, aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point() +
  theme_classic(base_size=15) +
  facet_wrap(~ gene_name) +
  ggtitle("Skin Lower Leg, Sun Exposed")

merge(pigment, gemma.egenes) %>%
  merge(gene_name) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(p.value))) +
  geom_point() +
  theme_classic(base_size=15) +
  facet_wrap(~ gene_name)







## splicing
wb_tmem216 = "61397975:61398261:clu_6182"
se_tmem216 = "61397975:61398261:clu_7959"

v8_wb_pigment.s = tabix2df("../gtex/v8/GTEx_Analysis_v8_sQTL_all_associations_Whole_Blood.v8.sqtl_allpairs.sort.txt.gz",
                      "chr11_61397975_61398261_ENSG00000187049.9:1-100000000",
                      c("v8_intron","chr","pos_hg38","v8_intron_id","variant","tss_dist",
                        "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))

liftover_snp(addchr(pigment$chr),
                   pigment$pos_hg19,
                   "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  unique %>%
  rename(pos_hg38 = pos_new,
         pos_hg19 = pos_old) %>%
  merge(v8_wb_pigment.s) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point() +
  theme_classic(base_size=15)


# roadmap expression
roadmap_exp = read_tsv("/local1/derek/data/epi_roadmap/all/rna_seq/57epigenomes.RPKM.pc.gz")
roadmap_names = read_tsv("/local1/derek/data/epi_roadmap/all/rna_seq/EG.name.txt",
                         col_names=c("id","name"))


roadmap_exp %>%
  filter(gene_id %in% pigment_genes) %>%
  gather(key=id, value=rpkm, -gene_id) %>%
  merge(roadmap_names) %>%
  merge(gene_name) %>%
  filter(gene_name=="TMEM216") %>%
  ggplot(aes(x=name, y=rpkm)) +
  geom_bar(stat="identity") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust=1)) +
  facet_grid(gene_name ~ .)
  
  




# GTEx data inn whole blood and skin exposed
tmem_v7_wb = tabix2df("../misc/Whole_Blood.allpairs.sub.txt.gz",
                      "ENSG00000187049.5:1-100000000",
                      c("gene_id","chr","pos_hg19","allele1","allele2","tss_dist",
                        "ma_samples","ma_count","gtexv7_maf","gtexv7_pval","slope","slope_se")) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":"))

tmem_v7_se = tabix2df("../gtex/v7/Skin_Sun_Exposed_Lower_leg.allpairs.sort.txt.gz",
                      "ENSG00000187049.5:1-100000000",
                      c("gene_id","chr","pos_hg19","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","gtexv7_maf","gtexv7_pval","slope","slope_se")) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":"))

# compare to v7 whole blood
tmem_v7_wb %>%
  mutate(gene_id=strip_ensembl(gene_id)) %>%
  select(gene_id, pos_hg19, gtexv7_pval) %>%
  merge(gemma.egenes) %>%
  ggplot(aes(x=-log10(p.value), y=-log10(gtexv7_pval))) +
  geom_point() +
  theme_classic()

# compare to v7 skin exposed
tmem_v7_se %>%
  mutate(gene_id=strip_ensembl(gene_id)) %>%
  select(gene_id, pos_hg19, gtexv7_pval) %>%
  merge(gemma.egenes) %>%
  ggplot(aes(x=-log10(p.value), y=-log10(gtexv7_pval))) +
  geom_point() +
  theme_classic()

# compare v7 skin exposed to gwas
tmem_v7_se %>%
  mutate(gene_id=strip_ensembl(gene_id)) %>%
  select(gene_id, pos_hg19, gtexv7_pval) %>%
  merge(pigment %>% select(pos_hg19, gwas_pval)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(gtexv7_pval))) +
  geom_point() +
  theme_classic()

```


# EUR-only eQTLs

```{r}

chr11_eur_eqtls = read_parquet("../gtex/v8/EUR_eQTL/GTEx_Analysis_v8_QTLs_GTEx_Analysis_v8_EUR_eQTL_all_associations_Whole_Blood.v8.EUR.allpairs.chr11.parquet") %>%
  separate(variant_id, into=c("chr","pos_hg38","v8_ref","v8_alt","foo"), convert=T) %>%
  mutate(gene_id = strip_ensembl(phenotype_id),
         chr = rmchr(chr)) %>%
  select(-foo)

```


#### CONDITIONAL INDEPENDENCE ####

Regress out all conditionally independent SNPs from GTEx v8

```{r}

# which SNPs were tested in Africans?
afr_snps = readLines("5M.imputed.dose.biallelic.posits.txt.gz")

# read in independent associations
gtex.v8.indep = read_tsv("../gtex/v8/GTEx_Analysis_v8_eQTL_independent/Whole_Blood.v8.independent_eqtls.txt.gz") %>%
  separate(variant_id, into=c("chr","pos_hg38","gtexv8_ref","gtexv8_alt","foo"), sep="_", convert=T) %>%
  mutate(chr = rmchr(chr),
         gene_id = strip_ensembl(gene_id)) %>%
  # restrict to SNPs
  filter(nchar(gtexv8_ref)==1 & nchar(gtexv8_alt)==1) %>%
  dplyr::select(-foo)

gtex.v8.indep.hg19 = liftover_snp(addchr(gtex.v8.indep$chr),
                                  gtex.v8.indep$pos_hg38,
                                  "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  mutate(chr = rmchr(chr))

# get a list of SNPs tested in Africa
gtex.v8.top_in_afr = gtex.v8.egenes %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  group_by(gene_id) %>%
  filter(gtexv8_pval == min(gtexv8_pval)) %>%
  ungroup %>%
  mutate(chr = rmchr(chr))

# because some GTEx SNPs are tested in Africans and vice versa, I am going to
# append the top GTEx SNPs
gtex.v8.indep_top = rbind(gtex.v8.indep %>% select(gene_id, chr, pos_hg38),
                          gtex.v8.top_in_afr %>% select(gene_id, chr, pos_hg38)) %>%
  unique

# add hg19 info
gtex.v8.indep.regress = liftover_call(tabixify(addchr(gtex.v8.indep_top$chr),
                                      gtex.v8.indep_top$pos_hg38,
                                      gtex.v8.indep_top$pos_hg38),
                              "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  separate(new, into=c("chr", "range"), sep=":") %>%
  separate(range, into=c("pos_hg19", "foo"), sep="-") %>%
  separate(old, into=c("chr", "range"), sep=":") %>%
  separate(range, into=c("pos_hg38", "foo"), sep="-") %>%
  dplyr::select(-foo) %>%
  mutate(chr = rmchr(chr)) %>%
  unique %>%
  merge(gtex.v8.indep_top) %>%
  filter(gene_id %in% gemma.egenes$gene_id)




geno_cols = readLines("../misc/geno_cols.txt")
regress_gex = data.frame()
for (i in unique(gtex.v8.indep.regress$gene_id)){
  
  regress_snps = gtex.v8.indep.regress %>%
    filter(gene_id==i)
  
  dose = tabix2df("../geno/5M.imputed.rna_samps.biallelic.vcf.gz",
                  tabixify(regress_snps$chr,
                           regress_snps$pos_hg19,
                           regress_snps$pos_hg19),
                  geno_cols) %>%
    dplyr::select(-c(1,2,6:9)) %>%
    gather(key=sample, value=genotype, -c(ID, REF,ALT)) %>%
    separate(genotype, into=c("geno","dose"), sep=":", convert=T) %>%
    pivot_wider(names_from = ID, values_from = dose, c(ID, sample, dose))
  
  # if (nrow(dose) > 0){
  #   
    gene_info = gex_norm %>%
      filter(gene_id==i)
    
    gex = gene_info %>%
      dplyr::select(-c(1:5)) %>%
      gather(key=sample, value=gex)
    
    dose_gex = merge(dose, gex)
    
    resids = lm(gex ~ ., data=dose_gex %>% dplyr::select(-sample))$residuals
    
    resids.df = data.frame(sample = dose_gex$sample,
                           resid = resids)
    
    out.df = pivot_wider(resids.df, names_from=sample, values_from=resid) %>%
      cbind(gene_info[,1:5], .)
    
    regress_gex = rbind(regress_gex,
                        out.df[,colnames(gene_info)])
  # }  
}


write.table(regress_gex, "Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.indep_v8_resid.txt",
            row.names=F, col.names=F, sep="\t", quote=F)

```


# Read in re-mapped data

```{r}

gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# read in connditional data
gemma = fread("conditional_map/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.indep_v8_resid.txt.gemma.100kb.r2.lmm4.gz", col.names=gemma_cols) %>%
  dplyr::select(-c(foo, bar)) %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af)) %>%
  filter(maf > 0.05) %>%
  # hacky rename so I don't have to define a new 'gemma_cols' variable
  dplyr::rename(gene_id=v19_id) %>%
  # remove poorly imputed SNPs
  filter(r2 > 0.3) %>% 
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", remove=F, convert=T) %>%
  # restrict to pc and lncRNAs
  filter(gene_id %in% gencode.pc_linc$gene_id) %>% 
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  group_by(gene_id) %>%
  # Benjamini-Hochberg correction across SNPs per gene
  mutate(bhp = p.adjust(p.value, method="BH"))

# threshold for calling significant SNPs
bhp_thresh = gemma.egenes %>%
  # FDR significant SNPs
  filter(fdr_sig) %>%
  # get the maximum BH' p-value
  pull(bhp) %>% max

# get those eQTLs that remain FDR-significant after controlling for the top GTEx v8 variant
gemma.cond_sig = gemma %>%
  filter(bhp <= bhp_thresh) %>%
  group_by(gene_id) %>%
  mutate(top_esnp = p.value==min(p.value))

write.table(gemma.cond_sig, "gemma.cond_sig.tsv", sep="\t", quote=F, row.names=F)
system("bgzip -f gemma.cond_sig.tsv")

gemma.cond_sig = read_tsv("gemma.cond_sig.tsv.gz")

```

# Is the recombination rate different for conditionally independent genes

```{r}

cond_dep.tabix = gemma.egenes %>%
  filter(fdr_sig) %>%
  filter(!(gene_id %in% gemma.cond_sig$gene_id)) %>%
  mutate(tabix = tabixify(addchr(chr), pos_hg19, pos_hg19)) %>%
  pull(tabix)

cond_indep.tabix = gemma.egenes %>%
  filter(fdr_sig) %>%
  filter(gene_id %in% gemma.cond_sig$gene_id) %>%
  mutate(tabix = tabixify(chr, pos_hg19, pos_hg19)) %>%
  pull(tabix)

cond_dep.ceu_recomb = tabix2df("../recomb/hg19/CEU/CEU_recombination_map_hg19_chr_17.bed.gz",
                               cond_dep.tabix,
                               c("chr","start","end","rate"))

```


# Compare frequency, p-value, and effect size at top SNPs, coloring by whether they are independent

```{r}

# Is there a frequency difference
gemma.egenes %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  merge(gemma.egenes.eur_af) %>%
  filter(((allele0 == alt_1kg) & (allele1 == ref_1kg)) |
           ((allele1 == alt_1kg) & (allele0 == ref_1kg))) %>%
  mutate(eur_af = ifelse(allele0 == ref_1kg, 1-eur_af, eur_af),
         indep = eqtl_hg19 %in% paste(gemma.cond_sig$gene_id, gemma.cond_sig$pos_hg19, sep=":")) %>%
  ggplot(aes(x=af, y=eur_af, color=indep)) +
  geom_point(alpha=0.5) +
  theme_classic(base_size=15)

gemma.egenes %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  merge(gemma.egenes.eur_af) %>%
  filter(((allele0 == alt_1kg) & (allele1 == ref_1kg)) |
           ((allele1 == alt_1kg) & (allele0 == ref_1kg))) %>%
  mutate(eur_af = ifelse(allele0 == ref_1kg, 1-eur_af, eur_af),
         indep = eqtl_hg19 %in% paste(gemma.cond_sig$gene_id, gemma.cond_sig$pos_hg19, sep=":")) %>%
  ggplot(aes(x=abs(af - eur_af), color=indep)) +
  stat_ecdf() +
  theme_classic(base_size=15)

# ECDF of -log10(p-value)
gemma.egenes %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  merge(gemma.egenes.eur_af) %>%
  filter(((allele0 == alt_1kg) & (allele1 == ref_1kg)) |
           ((allele1 == alt_1kg) & (allele0 == ref_1kg))) %>%
  mutate(eur_af = ifelse(allele0 == ref_1kg, 1-eur_af, eur_af),
         indep = eqtl_hg19 %in% paste(gemma.cond_sig$gene_id, gemma.cond_sig$pos_hg19, sep=":"),
         log_p = -log10(p.value)) %>%
  ggplot(aes(x=log_p, color=indep)) +
  stat_ecdf() +
  theme_classic(base_size=15)

# effect size differences
gtex.v8.egenes %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  select(gene_id, pos_hg38, v8_ref, v8_alt, gtexv8_slope) %>%
  merge(hg19_to_hg38) %>%
  merge(gemma.egenes %>% filter(top_esnp)) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  filter(((allele0 == v8_alt) & (allele1 == v8_ref)) |
           ((allele1 == v8_alt) & (allele0 == v8_ref))) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"),
         gtexv8_slope = ifelse(allele0 == v8_ref, 1-gtexv8_slope, gtexv8_slope),
         indep = eqtl_hg19 %in% paste(gemma.cond_sig$gene_id, gemma.cond_sig$pos_hg19, sep=":")) %>%
  ggplot(aes(x=beta, y=gtexv8_slope, color=indep)) +
  geom_point(alpha=0.5) +
  theme_classic(base_size=15)

gtex.v8.egenes %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  select(gene_id, pos_hg38, v8_ref, v8_alt, gtexv8_slope) %>%
  merge(hg19_to_hg38) %>%
  merge(gemma.fdr05.top) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  filter(((allele0 == v8_alt) & (allele1 == v8_ref)) |
           ((allele1 == v8_alt) & (allele0 == v8_ref))) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"),
         gtexv8_slope = ifelse(allele0 == v8_ref, 1-gtexv8_slope, gtexv8_slope),
         indep = eqtl_hg19 %in% paste(gemma.cond_sig$gene_id, gemma.cond_sig$pos_hg19, sep=":")) %>%
  ggplot(aes(x=abs(beta - gtexv8_slope), color=indep)) +
  stat_ecdf() +
  theme_classic(base_size=15)

gtex.v8.egenes %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  select(gene_id, pos_hg38, v8_ref, v8_alt, gtexv8_slope) %>%
  merge(hg19_to_hg38) %>%
  merge(gemma.fdr05.top) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  filter(((allele0 == v8_alt) & (allele1 == v8_ref)) |
           ((allele1 == v8_alt) & (allele0 == v8_ref))) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"),
         gtexv8_slope = ifelse(allele0 == v8_ref, 1-gtexv8_slope, gtexv8_slope),
         indep = eqtl_hg19 %in% paste(gemma.cond_sig$gene_id, gemma.cond_sig$pos_hg19, sep=":")) %>%
  ggplot(aes(x=indep, y=abs(beta - gtexv8_slope), fill=indep)) +
  geom_violin() +
  theme_classic(base_size=15)

# tss_dist differences (proxy for LD?)
gemma.fdr05.top %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  merge(gencode.v19.tss %>% select(-chr)) %>%
  mutate(tss_dist = pos_hg19 - tss,
         eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"),
         indep = eqtl_hg19 %in% paste(gemma.cond_sig$gene_id, gemma.cond_sig$pos_hg19, sep=":")) %>%
  ggplot(aes(x=tss_dist, color=indep)) +
  stat_ecdf() +
  theme_classic()


```


It appears it's differences in effect sizes that's driving independent signals.
One hypothesis is that it's trans factors (e.g. TFs) that are driving these
differences. I'm going to see if overlap with specific TFs is associated with
replication

```{r}

gemma.gt_01 = gemma.egenes %>%
  mutate(bf = exp(approx.bf.estimates(z=beta/se, V=se^2, type="quant")$lABF)) %>%
  group_by(gene_id) %>%
  mutate(pp = bf / sum(bf)) %>%
  filter(pp > 0.1)

tf_meta = read_tsv("../tf/metadata.tsv") %>%
  select(file=`File accession`, gene_sp=`Experiment target`) %>%
  separate(gene_sp, into=c("gene", "species"), sep="-") %>%
  select(-species) %>%
  unique

np_cols = readLines("../tf/narrowpeak_cols.txt")

gemma.gt_01.tf = tabix2df("../tf/GM12878.all_tfs.sort.bed.gz",
                              tabixify(gemma.gt_01$chr,
                                       gemma.gt_01$pos_hg19-1,
                                       gemma.gt_01$pos_hg19),
                              c(np_cols, "file")) %>%
  unique %>%
  merge(tf_meta)

gemma.gt_01.tf = bed_intersect(r1 = tabixify(gemma.gt_01$chr,
                                           gemma.gt_01$pos_hg19-1,
                                           gemma.gt_01$pos_hg19),
                             r2 = tabixify(gemma.gt_01.tf$chr,
                                      gemma.gt_01.tf$start,
                                      gemma.gt_01.tf$end)) %>%
  separate(r1, into=c("chr", "range"), sep=":") %>%
  separate(range, into=c("foo", "pos_hg19"), sep="-") %>%
  select(-foo) %>%
  separate(r2, into=c("chr", "range"), sep=":") %>%
  separate(range, into=c("start", "end"), sep="-") %>%
  unique %>%
  merge(gemma.gt_01.tf) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":"))

# which tfs are expressed in our samples (median tpm > 0.1)?
tf_exp_afr = tpm.median %>%
  filter(median_tpm > 0.1) %>%
  merge(gene_name) %>%
  filter(gene_name %in% gemma.gt_01.tf$gene) %>%
  pull(gene_name)

# info for genes expressed in Africans
gemma.gt_01.tf.exp = gemma.gt_01.tf %>%
  filter(gene %in% tf_exp_afr)

# fit a logistic regression model
indep_snps = unique(gemma.gt_01.tf$snp_hg19)

# X matrix
indep_tf_logit.df = as.data.frame(mat.or.vec(nr = length(indep_snps), nc = length(tf_exp_afr)))
rownames(indep_tf_logit.df) = indep_snps
colnames(indep_tf_logit.df) = tf_exp_afr

# fill in predictor variables
for (i in 1:nrow(gemma.gt_01.tf.exp)){
  
  indep_tf_logit.df[gemma.gt_01.tf.exp[i,]$snp_hg19,
                    gemma.gt_01.tf.exp[i,]$gene] = 1
}

# indep_tf_logit.df[,"Intercept"] = 1
# add response variable
indep_tf_logit.df$indep = as.numeric(indep_snps %in% gemma.cond_sig$snp_hg19)

library(logistf)
logistf.control(maxit=10000)
# inep_tf_logit_lm = logistf(indep ~ ., data=indep_tf_logit.df)
inep_tf_logit_lm = glm(indep ~ . - 1, data=indep_tf_logit.df, family="binomial")

foo = c()
for (gene in tf_exp_afr){
  
  foo = c(foo, fisher.test(table(indep_tf_logit.df[,c(gene, "indep")]))$p.value)
}

```


Look at GO term enrichment among genes with conditionally independent signals

```{r}

library(gprofiler2)

egenes = unique(gemma.egenes$gene_id)
cond_sig_genes = unique(gemma.cond_sig$gene_id)

cond_sig_go = gost(query = cond_sig_genes,
                   organism = "hsapiens",
                   custom_bg = egenes)

```


coloc between pigmentation and all genes within 1MB of 

```{r}

gtex.v8.se.tmem216.range = tabix2df("../gte")

```










Regress out the most significant SNP in GTEx that's also FDR-significant in Africans

```{r}

# Get the most significant SNP in GTEx per gene thats within 100kb
regress_snps = gtex.v8.egenes.range %>%
  mutate(chr = rmchr(chr)) %>%
  merge(hg19_to_hg38) %>%
  group_by(gene_id) %>%
  filter(gtexv8_pval==min(gtexv8_pval)) %>%
  sample_n(1) %>%
  select(gene_id, chr, pos_hg19)

  

geno_cols = readLines("../misc/geno_cols.txt")
regress_gex = data.frame()
for (i in 1:nrow(regress_snps)){
  
  dose = tabix2df("../geno/5M.imputed.rna_samps.biallelic.vcf.gz",
                tabixify(regress_snps[i,]$chr,
                         regress_snps[i,]$pos_hg19,
                         regress_snps[i,]$pos_hg19),
                geno_cols) %>%
  select(-c(1,2,6:9)) %>%
  gather(key=sample, value=genotype, -c(ID,REF,ALT)) %>%
  separate(genotype, into=c("geno","dose"), sep=":", convert=T) %>%
  select(sample, dose)

  gene_info = gex_norm %>%
    filter(strip_ensembl(gene_id)==regress_snps[i,]$gene_id)
  
  gex = gene_info %>%
    select(-c(1:5)) %>%
    gather(key=sample, value=gex)
  
  dose_gex = merge(dose, gex)
  
  resids = lm(gex ~ dose, data=dose_gex)$residuals
  
  resids.df = data.frame(sample = dose_gex$sample,
                         resid = resids)
  
  out.df = pivot_wider(resids.df, names_from=sample, values_from=resid) %>%
    cbind(gene_info[,1:5], .)
  
  regress_gex = rbind(regress_gex,
                      out.df[,colnames(gene_info)])
  
}

regress_gex = regress_gex %>%
  mutate(gene_id = strip_ensembl(gene_id))

write.table(regress_gex, "Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.top_v8_resid.txt",
            row.names=F, col.names=F, sep="\t", quote=F)
# system("bgzip -f Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.top_v8_resid.txt")
        
```


For each gene:
  1. write phenotype to file
  2. extract genotype information and write to file
  3. run GEMMA
  
```{r}

system("bash run_gemma.resid.sh Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.top_v8_resid.txt")

# gemma call
gemma_call = function(geno.f, pheno.f, grm.f, out.f){
  
  exec_wait(paste0("/home/derkelly/bin/gemma/bin/gemma",
              " -g ", geno.f,
              " -p ", pheno.f,
              " -k ", grm.f,
              " -lmm 4",
              " -o ", out.f))
}

gemma_dir = "/local1/derek/data/transcriptomics/eqtl/afr_wbrna/gemma/"

grm.f = paste0(gemma_dir, "5M.imputed.dose.biallelic.doses.grm.sXX.txt")

system("mkdir -p conditional_map")

for (i in 1:nrow(regress_snps)){

  #### format the phenotype file ####
  pheno = regress_gex %>%
    filter(gene_id==regress_snps[i,]$gene_id) %>%
    gather(key=sample, value=gex, -c(1:5))
  
  # write out phenotype
  write.table(pheno$gex, "conditional_map/regress_gex.tmp.txt", col.names=F, row.names=F, quote=F)
  
  
  #### format the genotype file ####
  snp_range = gencode.v19.tss %>%
    filter(gene_id==regress_snps[i,]$gene_id) %>%
    mutate(tabix = tabixify(rmchr(chr), tss - 100000, tss + 100000)) %>%
    pull(tabix)
    
  # read in bimbam and write to file
  range2df("5M.imputed.dose.biallelic.doses.txt.gz",
           snp_range,
           c("chr","pos_hg19","snp_hg19","ref","alt", samps)) %>%
    select(-c(chr, pos_hg19)) %>%
    write.table("conditional_map/geno.tmp.txt", col.names=F, row.names=F, quote=F)
  
  
  #### run GEMMA ####
  out.f = paste0(regress_snps[i,]$gene_id,
                 ".", regress_snps[i,]$chr, "_", regress_snps[i,]$pos_hg19,
                 ".out.txt")
  
  gemma_call(geno.f=paste0(gemma_dir, "conditional_map/geno.tmp.txt"),
             pheno.f=paste0(gemma_dir, "conditional_map/regress_gex.tmp.txt"),
             grm.f=grm.f,
             out.f="out.tmp")
  
  system2(paste0("tail -n +2 output/out.tmp.txt.assoc.txt", out.f),
          stderr="conditional_map/gemma.err",
          stdout="conditional_map/gemma.out")
}

```


Read in re-mapped data

```{r}

# read in conditional data
gemma = fread("conditional_map/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.top_v8_resid.txt.gemma.100kb.r2.lmm4.gz", col.names=gemma_cols) %>%
  select(-c(foo, bar)) %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af)) %>%
  filter(maf > 0.05) %>%
  # hacky rename so I don't have to define a new 'gemma_cols' variable
  dplyr::rename(gene_id=v19_id) %>%
  # remove poorly imputed SNPs
  filter(r2 > 0.3) %>% 
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", remove=F, convert=T) %>%
  # restrict to pc and lncRNAs
  filter(gene_id %in% gencode.pc_linc$gene_id) %>% 
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  group_by(gene_id) %>%
  # Benjamini-Hochberg correction across SNPs per gene
  mutate(bhp = p.adjust(p.value, method="BH"))

# threshold for calling significant SNPs
bhp_thresh = gemma.egenes %>%
  # FDR significant SNPs
  filter(fdr_sig) %>%
  # get the maximum BH' p-value
  pull(bhp) %>% max

# get those eQTLs that remain FDR-significant after controlling for the top GTEx v8 variant
gemma.cond_sig = gemma %>%
  filter(bhp <= bhp_thresh) %>%
  group_by(gene_id) %>%
  mutate(top_esnp = p.value==min(p.value))

```


For how many genes does the top SNP remain significant after conditioning on the top GTEx SNP?

```{r}

gemma.egenes %>%
  filter(top_esnp) %>%
  select(gene_id, cond_sig) %>%
  group_by(gene_id) %>%
  summarise(n_cond = sum(cond_sig)) %>%
  select(n_cond) %>%
  table

```


How many genes have an FDR-significant SNP at all after correcting? Conversely, for how many does remapping remove all of the eQTL signal

```{r}

gemma.egenes %>%
  filter(cond_sig) %>%
  pull(gene_id) %>%
  unique %>%
  length

```

399 genes still have an FDR_significant SNP after controlling for GTEx SNP


Why are certain SNPs eQTLs in Africans but not Europeans?
1. MAF differences
2. GEx differences
3. trans-factor differences (i.e. are they enriched in particular TF binding sites?)
4. effect size differences?

First, MAF differences. What is the MAF difference between Africans and Europeans at the top eQTL of genes where eQTL signal is removed vs. those genes where the top SNP is not removed

```{r}

gemma.egenes %>%
  filter(top_esnp) %>%
  merge(gemma.egenes.eur_af) %>%
  ggplot(aes(x=eur_af, color=cond_sig)) +
  stat_ecdf() +
  theme_classic()

```


How many "new top SNPs" are FDR-significant in GTEx v8

```{r}

# eQTLs that are the top eQTLs after conditioning on the GTEx v8 SNP
cond_top = gemma.egenes %>%
  filter(cond_sig) %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  pull(eqtl_hg19)

# add info on whether the eQTL is conditionally independent
gemma.egenes$cond_top = gemma.egenes$eqtl_hg19 %in% cond_top


cond_top.in_v8 = gemma.egenes %>%
  filter(cond_top) %>%
  merge(hg19_to_hg38) %>%
  mutate(snp_hg38 = paste(chr, pos_hg38, sep=":"),
         eqtl_hg38 = paste(gene_id, pos_hg38, sep=":"),
         in_v8 = snp_hg38 %in% (paste(gemma.gtex.v8.egenes$chr,
                                      gemma.gtex.v8.egenes$pos_hg38, sep=":")),
         v8_sig = eqtl_hg38 %in% (gtex.v8.all_signif %>%
                                    filter(tissue=="Whole_Blood") %>%
                                    mutate(gene_id = strip_ensembl(v26_id),
                                           eqtl_hg38 = paste(gene_id, pos_hg38, sep=":")) %>%
                                    pull(eqtl_hg38)))

# How many top conditional eQTLs are 1.tested in v8 and 2. FDR-significant in v8
cond_top.in_v8 %>%
  select(in_v8, v8_sig) %>%
  table

# for those SNPs not tested in v8, what is their imputation accuracy and EUR af?
cond_top.in_v8 %>%
  merge(gemma.egenes.eur_af) %>%
  select(in_v8, r2, eur_af) %>%
  gather(key=var, value=val, -in_v8) %>%
  ggplot(aes(x=val, color=in_v8)) + 
  stat_ecdf() +
  facet_grid(. ~ var, scales="free_x") +
  theme_classic()

# for those genes that *are* tested in v8, what is the r2, EUR af of the SNPs that are significant, and the GEx of the gene?
cond_top.in_v8 %>%
  filter(in_v8) %>%
  merge(gemma.egenes.eur_af) %>%
  merge(gtex.v8.med_gex %>%
          mutate(gene_id = strip_ensembl(v26_id)) %>%
          select(gene_id, v8_gex = `Whole Blood`) %>%
          mutate(v8_lgex = log2(v8_gex+0.01))) %>%
  select(v8_sig, v8_lgex, r2, eur_af) %>%
  gather(key=var, value=val, -v8_sig) %>%
  ggplot(aes(x=val, color=v8_sig)) + 
  stat_ecdf() +
  facet_grid(. ~ var, scales="free_x") +
  theme_classic(base_size=15)

```


Possibly causal, conditionally independent SNPs

```{r}

# add probability of being causal
gemma.cond_sig = gemma.cond_sig %>%
  mutate(bf = exp(approx.bf.estimates(z=beta/se, V=se^2, type="quant")$lABF)) %>%
  group_by(gene_id) %>%
  mutate(pp = bf / sum(bf))

# get TF binding information
gemma.

```


How many independent SNPs overlap a GWAS locus?

```{r}

foo = cond_top.in_v8 %>%
  merge(hg19_to_hg38, all.x=T) %>%
  mutate(snp_hg38 = paste(chr, pos_hg38, sep=":")) %>%
  merge(gwas_cat %>% select(snp_hg38, mapped_trait, pubmedid))
  
  
```


Look at EUR MAF

```{r}

# compare MAF in GTEx for replicate and fail to replicate
gemma.egenes %>%
  filter(top_esnp & !is.na(v8_ref)) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  ggplot(aes(x=gtexv8_maf, color=gtexv8_pval<0.01)) +
  stat_ecdf(size=2) +
  theme_classic(base_size=15)

# compare effect sizes
gtex.v8.egenes.tabix %>%
  filter(top_esnp) %>%
  merge(gemma.egenes) %>%
  merge(gemma.egenes.gtexv8) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  mutate(gtexv8_slope = ifelse(v8_ref==allele1 & v8_alt==allele0, gtexv8_slope, -gtexv8_slope)) %>% # reverse if alleles don't match
  mutate(gtexv8_slope = ifelse((v8_ref==allele0 & v8_alt==allele1) |
                                 (v8_ref==allele1 & v8_alt==allele0), gtexv8_slope, NA)) %>%
  ggplot(aes(x=beta, y=gtexv8_slope, color=gtexv8_pval<0.01)) +
  geom_point() +
  theme_classic(base_size=15)

# check colocalization with GTExv8
gtex.v8.egenes = gtex.v8.egenes.tabix %>%
  merge(gemma.egenes) %>%
  merge(gemma.egenes.gtexv8) %>%
  unique

gemma_gtexv8_h4 = lapply(unique(gtex.v8.egenes$gene_id), function(gene){
    df.tmp = gtex.v8.egenes %>%
      filter(gene_id == gene) %>%
      filter((v8_ref == allele0 & v8_alt  == allele1) |
               (v8_ref == allele1 & v8_alt  == allele0)) %>%
      arrange(pos_hg19)
    coloc.abf(dataset1=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$p.value, type="quant", N=162, MAF=df.tmp$maf),
              dataset2=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$gtexv8_pval, type="quant", N=670, MAF=df.tmp$gtexv8_maf))$summary
        # coloc.abf(dataset1=list(snp=df.tmp$snp_hg19, beta=df.tmp$beta, varbeta=df.tmp$se, type="quant", N=162, MAF=df.tmp$maf),
        #       dataset2=list(snp=df.tmp$snp_hg19, beta=df.tmp$gtexv8_slope, varbeta=df.tmp$gtexv8_se, type="quant", N=670, MAF=df.tmp$gtexv8_maf))$summary
    })

gemma_gtexv8_h4.df = data.frame(gene_id = unique(gtex.v8.egenes$gene_id),
                                H0 = sapply(gemma_gtexv8_h4, function(x) x[2]),
                                H1 = sapply(gemma_gtexv8_h4, function(x) x[3]),
                                H2 = sapply(gemma_gtexv8_h4, function(x) x[4]),
                                H3 = sapply(gemma_gtexv8_h4, function(x) x[5]),
                                H4 = sapply(gemma_gtexv8_h4, function(x) x[6]),
                                stringsAsFactors=F)

# TAPBPL shows reversed effect sizes?
gtex.v8.egenes %>%
  filter(gene_id=="ENSG00000139192") %>%
  ggplot(aes(x=pos_hg19)) +
  geom_point(aes(y=beta, color="blue")) +
  geom_point(aes(y=gtexv8_slope, color="red")) +
  theme_classic()

# hypothesis with maximum probability for each gene
gemma_gtexv8_h4.df.max = gemma_gtexv8_h4.df %>%
  gather(key="hypothesis", value="probability", -gene_id) %>%
  group_by(gene_id) %>%
  filter(probability==max(probability))

gemma_gtexv8_h4.df %>%
  gather(key="hypothesis", value="fraction", -gene_id) %>%
  merge(gemma_gtexv8_h4.df %>% gather(key="hypothesis", value="probability", -gene_id)) %>%
  ggplot(aes(x=fraction, fill=hypothesis, color=hypothesis)) +
  geom_histogram() +
  scale_y_log10() +
  theme_classic()


# distribution of allele frequencies

  filter(top_esnp) %>% 
  filter(v8_ref == allele0 & v8_alt  == allele1) %>%
  merge(gemma_gtexv8_h4.df.max) %>% # merge with coloc data
  ggplot() +
  stat_ecdf(aes(x=af, color="red")) +
  stat_ecdf(aes(x=eur_af, color="blue")) +
  theme_classic(base_size=15) +
  facet_wrap(~ hypothesis)


# distribution of expression fc for genes of different hypotheses
merge(gemma_gtexv8_h4.df.max, tpm.median) %>%
  merge(gtex.v8.tpm) %>%
  mutate(tpm_lfc = log2(median_tpm/gtexv8_tpm)) %>%
  ggplot(aes(x=tpm_lfc, color=hypothesis)) +
  stat_ecdf() +
  theme_classic(base_size=15)

library(rWikiPathways)

ebola.genes = getPathway("WP4217") %>%
  str_extract_all("ENSG[0-9]+") %>%
  unlist

library(biomaRt)
# define biomart object
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
# query biomart
results <- getBM(attributes = c("ensembl_gene_id", "ensembl_transcript_id", "go_id"),
                 filters = "ensembl_transcript_id", values = "ENST00000537767",
                 mart = mart)
results

```


Run coloc between Africans and GTEx

```{r}

gemma_gtexv7_h4 = lapply(unique(eur162.egenes$gene_id), function(gene){
    foo = gemma.eur.egenes.sub %>%
      filter(eur_id == gene) %>%
      arrange(pos)
    bar = eur162.egenes %>%
      filter(gene_id == gene) %>%
      arrange(pos)
    coloc.abf(dataset1=list(snp=paste(addchr(foo$chr_hg38), foo$pos_hg38, sep=":"), pvalues=foo$p.value, type="quant", N=162, MAF=foo$maf),
              dataset2=list(snp=paste(bar$chr, bar$pos, sep=":"), pvalues=bar$p.value, type="quant", N=670, MAF=bar$maf))$summary
    })

gemma_eur162_h4.df = data.frame(gene_id = unique(eur162.egenes$gene_id),
                                H0 = sapply(gemma_eur162_h4, function(x) x[2]),
                                H1 = sapply(gemma_eur162_h4, function(x) x[3]),
                                H2 = sapply(gemma_eur162_h4, function(x) x[4]),
                                H3 = sapply(gemma_eur162_h4, function(x) x[5]),
                                H4 = sapply(gemma_eur162_h4, function(x) x[6]))

```


Why don't eQTLs colocalize? frequecy and expression

```{r}

# compare mean FST between different coloc hypotheses
gemma.egenes.gtex %>%
  filter(fdr_sig) %>%
  filter(allele0==onekg_ref & allele1==onekg_alt) %>%
  mutate(fst = fst_wc(503, 162, eur_af, af)) %>%
  group_by(gene_id) %>%
  summarise(mean_fst = mean(fst)) %>%
  merge(gemma_gtexv8_h4.df.max) %>%
  ggplot(aes(x=mean_fst, color=hypothesis)) +
  stat_ecdf() +
  theme_classic()



# expression difference between coloc hypotheses
gemma.egenes.gtex %>%
  filter(top_esnp) %>%
  filter(allele0==onekg_ref & allele1==onekg_alt) %>%
  merge(gtex.v8.tpm) %>%
  merge(tpm.median) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  summarise(abs_logfc = abs(log2(median_tpm / gtexv8_tpm))) %>%
  merge(gemma_gtexv8_h4.df.max) %>%
  ggplot(aes(x=abs_logfc, color=hypothesis)) +
  stat_ecdf() +
  theme_classic()

```


eQTLs in GTEx v8

```{r}

gtex.v8.sig.f = "../gtex/v8/GTEx_Analysis_v8_eQTL/Whole_Blood.v8.signif_variant_gene_pairs.txt.gz"

gtex.v8.sig = read_tsv(gtex.v8.sig.f) %>%
  separate(variant_id, into=c("chr","pos","ref","alt","build")) %>%
  select(-build)

gtex.v8.hg19 = tabix2df("../misc/Whole_Blood.v8.hg38_to_hg19.txt.gz",
                             tabixify(gtex.v8$chr, gtex.v8$pos, gtex.v8$pos),
                             c("chr_hg38","pos_hg38","chr_hg19","pos_hg19")) %>%
  mutate(snp_id = paste(chr_hg38, pos_hg38, sep=":")) %>%
  merge(gtex.v8 %>% add_snpid %>% mutate(snp_id = paste0("chr", snp_id)), ., by="snp_id") %>%
  merge(., afr_gtexv8_genes, by.x="gene_id", by.y="gtexv8_id") %>%
  unique


gtex.v8.gemma = tabix2df("../gemma/gemma.5readsin20.tpm01pop.peer24.sort.txt.gz",
                              tabixify(gtex.v8.hg19$gene_id.y, gtex.v8.hg19$pos_hg19, gtex.v8.hg19$pos_hg19),
                              gemma_sort_cols)

gtex.v8.eur162 = tabix2df("../gtex/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz",
                              tabixify(gtex.v8$gene_id, gtex.v8$pos, gtex.v8$pos),
                              c("snp_id","gene_id","chr","pos","ref","alt","version","tss_dist","p.value","slope","maf"))

```


get pi1 for EUR162?

```{r}

gemma.fdr05.eur162 = tabix2df("../gtex/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz",
                              tabixify(gemma.fdr05.hg38$gtexv8_id, gemma.fdr05.hg38$pos_hg38, gemma.fdr05.hg38$pos_hg38),
                              c("snp_id","gene_id","chr","pos","ref","alt","version","tss_dist","p.value","slope","maf"))

eur162.fdr05.hg19 = tabix2df("../gtex/EUR162_15PEER_5PC_nopermute.hg38_to_hg19.txt.gz",
                             tabixify(eur162.fdr05$chr, eur162.fdr05$pos, eur162.fdr05$pos),
                             c("chr_hg38","pos_hg38","chr_hg19","pos_hg19")) %>%
  mutate(snp_id = paste(chr_hg38, pos_hg38, sep=":")) %>%
  merge(eur162.fdr05 %>% mutate(snp_id = paste(chr, pos, sep=":")), ., by="snp_id") %>%
  merge(., afr_eur162_genes, by.x="gene_id", by.y="eur_id") %>%
  unique

gemma_sort_cols = readLines("../misc/gemma_eqtl.lmm4.sort.cols.txt")

eur162.fdr05.gemma = tabix2df("../gemma/gemma.5readsin20.tpm01pop.peer24.sort.txt.gz",
                              tabixify(eur162.fdr05.hg19$afr_id, eur162.fdr05.hg19$pos_hg19, eur162.fdr05.hg19$pos_hg19),
                              gemma_sort_cols)

gemma.fdr05.eur162.frq = tabix2df("../gtex/eursamples_162_freq.sort.frq.gz",
                                  tabixify(gemma.fdr05.eur162$chr, gemma.fdr05.eur162$pos, gemma.fdr05.eur162$pos),
                                  gtex_frq_cols) %>%
  mutate(af = ifelse(snpid2 == allele1, maf, 1-maf),
         snp_id = paste(foo, pos, sep=":")) %>%
  merge(., gemma.fdr05.eur162 %>% mutate(snp_id = paste(chr, pos, sep=":")), by="snp_id") %>%
  mutate(eqtl_id = paste(gene_id, pos.x, sep=":"))

gemma.fdr05.hg38 %>%
  mutate(eqtl_id = paste(gtexv8_id, pos_hg38, sep=":")) %>%
  merge(., gemma.fdr05.eur162.frq, by="eqtl_id") %>%
  mutate(replicate = p.value.y < 0.05) %>%
  ggplot(., aes(x=replicate, y=abs(af.x - af.y))) +
  geom_violin()
                             
```


Get credible sets for genes ascertained in Africans and European Americans in both datasets

```{r}

eur162.f = "../gtex/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz"

eur162.sort.cols = c("snp_hg38","v29_id","chr","pos_hg38","eur162_ref","eur162_alt",
                     "version","tss_dist","eur162_pval","eur162_slope","eur162_maf")

gemma.egenes.eur162.tabix = gemma.egenes %>%
  filter(!(is.na(v29_id) | is.na(pos_hg38))) %>%
  mutate(tabix_id = tabixify(v29_id, pos_hg38, pos_hg38)) %>%
  pull(tabix_id) %>%
  unique

gemma.egenes.eur162 = tabix2df(eur162.f,
                               gemma.egenes.eur162.tabix,
                               eur162.sort.cols) %>%
  dplyr::select(-tss_dist) %>%
  separate(snp_hg38, into=c("chr", "pos_hg38"), sep=":") %>%
  merge(gemma.egenes, all=T) %>%
  unique

gemma.egenes.eur162 = gemma.egenes.eur162 %>%
  mutate(eur162_fdr_sig = (v29_id %in% eur162.fdr05$gene_id) & pos_hg38 %in% eur162.fdr05$pos_hg38)

gemma.eur162.egenes = gemma.egenes.eur162 %>%
  filter(fdr_sig & eur162_fdr_sig) %>%
  pull(gene_id) %>%
  unique

gtexv8.egenes = tabix2df(gtex.v8.f,
                         paste0(gene_map %>%
                                  filter(gene_id %in% egenes) %>%
                                  pull(v26_id), ":0-1000000000"),
                         gtex.v8.cols) %>%
  filter(abs(tss_dist) < 100000) %>%
  mutate(gene_id = strip_ensembl(v26_id),
         gtexv8_eqtl = paste(gene_id, pos_hg38, sep=":"))


# get credible sets using p-value
get_cred_set.p = function(ids, pvals, maf, sample_size, cred_perc){
  
  bf = approx.bf.p(pvals, maf, "quant", sample_size)
  
  cred_set = order(-bf$lABF)[1:min(which(cumsum(sort(exp(bf$lABF), decreasing=T)/sum(exp(bf$lABF))) > cred_perc))]
  
  ids[cred_set]
}

# get credible sets using regression estiamtes
# get_cred_set.est = function(ids, zs, ses, cred_perc){
#   
#   bf = approx.bf.estimates(z=zs, V=ses, type="quant")
#   
#   cred_set = order(-bf$lABF)[1:min(which(cumsum(sort(exp(bf$lABF), decreasing=T)/sum(exp(bf$lABF))) > cred_perc))]
#   
#   ids[cred_set]
# }

get_cred_set.est = function(ds, cred_perc){
  
  bf = finemap.abf(dataset=ds) %>%
    
  bf = bf[-nrow(bf),]
  cred_set = order(-bf$SNP.PP)[1:min(which(cumsum(sort(bf$SNP.PP, decreasing=T)/sum(bf$SNP.PP)) > cred_perc))]
  
  ids[cred_set]
  }

# # get credible sets using dataset
# get_cred_set_data = function(df){
#   
#   cred_set_idx = df %>%
#     with(get_cred_set.p(p.value, maf, 162, 0.9))
#   
#   return(list(data = df, cred_set_idx = cred_set_idx))
# }

# genes that are significant in both afr and eur162
sig_genes = intersect(unique(gemma.egenes$gene_id),
                      unique(eur162.fdr05$gene_id))

# all_genes = unique(c(gemma.egenes$gene_id,
#                   eur162.fdr05$gene_id))

# credible sets for eGenes with an FDR-significant eQTL in both datasets
afr_cred_set.sig  = lapply(sig_genes,
                           function(x){
                             df.tmp = gemma.egenes %>%
                               filter(gene_id==x)
                             get_cred_set.p(df.tmp$eqtl_hg19,
                                            df.tmp$p.value,
                                            df.tmp$maf, 162, 0.9)})

afr_cred_set.all  = lapply(egenes,
                           function(x){
                             df.tmp = gemma.egenes %>%
                               filter(gene_id==x)
                             get_cred_set.p(df.tmp$eqtl_hg19,
                                            df.tmp$p.value,
                                            df.tmp$maf, 162, 0.9)})

eur162_cred_set.sig  = lapply(sig_genes,
                              function(x){
                                df.tmp = eur162.egenes %>%
                                  filter(gene_id==x)
                                get_cred_set.p(df.tmp$eqtl_hg38,
                                               df.tmp$p.value,
                                               df.tmp$eur162_maf, 162, 0.9)})

cred_set.sig.count = data.frame(gene_id=sig_genes,
                                afr_cred_set = sapply(afr_cred_set.sig, length),
                                eur162_cred_set = sapply(eur162_cred_set.sig, length))


gtexv8_cred_set.all = lapply(egenes[egenes %in% unique(gtexv8.egenes$gene_id)],
                              function(x){
                                df.tmp = gtexv8.egenes %>%
                                  filter(gene_id==x)
                                get_cred_set.p(df.tmp$gtexv8_eqtl,
                                               df.tmp$gtexv8_pval,
                                               df.tmp$gtexv8_maf, 670, 0.9)})

# compare credible sets between Africans and full GTEx v8
afr_gtexv8_cred_set.df = data.frame(gene_id = egenes[egenes %in% unique(gemma.gtexv8.egenes$gene_id)],
                                    afr_cred_set = sapply(afr_cred_set.all[egenes %in% unique(gemma.gtexv8.egenes$gene_id)], length),
                                    gtexv8_cred_set = sapply(gtexv8_cred_set.all, length))

# plot African vs GTEx v8 credible sets, colored by color results
merge(afr_gtexv8_cred_set.df, gemma_gtexv8_h4.df.max) %>%
  ggplot(aes(x=log10(afr_cred_set), y=log10(gtexv8_cred_set), color=hypothesis)) +
  geom_point(alpha = 0.7, size=3) +
  theme_classic()

gemma.eur162.sig_genes = gemma.egenes %>%
  filter(gene_id %in% sig_genes) %>%
  dplyr::select(gene_id, pos_hg38, maf, p.value, allele0, allele1) %>%
  merge(eur162.egenes %>% dplyr::select(gene_id, pos_hg38, eur162_maf, eur162_pval=p.value, eur162_ref, eur162_alt)) %>%
  filter((eur162_ref==allele0 & eur162_alt==allele1) |
           (eur162_ref==allele0 & eur162_alt==allele1)) %>%
  mutate(snp_hg38 = paste(gene_id, pos_hg38, sep=":"))

# assess bayesian colocalization
gemma_eur162_h4 = lapply(unique(gemma.eur162.sig_genes$gene_id), function(gene){
    df.tmp = gemma.eur162.sig_genes %>%
      filter(gene_id == gene) %>%
      arrange(pos_hg38)
    coloc.abf(dataset1=list(snp=df.tmp$snp_hg38, pvalues=df.tmp$p.value, type="quant", N=162, MAF=df.tmp$maf),
              dataset2=list(snp=df.tmp$snp_hg38, pvalues=df.tmp$eur162_pval, type="quant", N=162, MAF=df.tmp$eur162_maf))$summary
    })

gemma_eur162_h4.df = data.frame(gene_id=unique(gemma.eur162.sig_genes$gene_id),
                                H0=sapply(gemma_eur162_h4, function(x) x[2]),
                                H1=sapply(gemma_eur162_h4, function(x) x[3]),
                                H2=sapply(gemma_eur162_h4, function(x) x[4]),
                                H3=sapply(gemma_eur162_h4, function(x) x[5]),
                                H4=sapply(gemma_eur162_h4, function(x) x[6]))

gemma_eur162_h4.df.max = gemma_eur162_h4.df %>%
  gather(key=hypothesis, value=probability, -gene_id) %>%
  group_by(gene_id) %>%
  filter(probability==max(probability))

gemma_eur162_h4.df.max.noh0 = gemma_eur162_h4.df %>%
  dplyr::select(-H0) %>%
  gather(key=hypothesis, value=probability, -gene_id) %>%
  group_by(gene_id) %>%
  filter(probability==max(probability))
                                

# # credible sets for eGenes with an FDR-significant eQTL in either datasets
# afr_cred_set.all  = lapply(unique(gemma.egenes$gene_id),
#                            function(x) gemma.egenes %>%
#                              filter(gene_id==x) %>%
#                              get_cred_set_data)
# afr_cred_set.all.count = sapply(afr_cred_set.all, function(x) length(x$cred_set_idx))
# afr_cred_set.all.frac = afr_cred_set.all.count/sapply(afr_cred_set.all, function(x) dim(x$data)[1])


# eur162_cred_set.rep = parLapply(cl, rep_genes$eur_id, eur_get_cred_set)
# eur162_cred_set.rep.count = sapply(eur162_cred_set.rep, function(x) length(x$cred_set_idx))
# eur162_cred_set.rep.frac = eur162_cred_set.rep.count/sapply(eur162_cred_set.rep, function(x) dim(x$data)[1])

eur162_cred_set.sig = lapply(sig_genes$eur_id,
                           function(x) eur162.egenes %>%
                             filter(gene_id==x) %>%
                             get_cred_set_data)
eur162_cred_set.sig.count = sapply(eur162_cred_set.sig, function(x) length(x$cred_set_idx))
eur162_cred_set.sig.frac = eur162_cred_set.sig.count/sapply(eur162_cred_set.sig, function(x) dim(x$data)[1])

# eur162_cred_set.all  = lapply(all_genes$eur_id,
#                            function(x) eur162.egenes %>%
#                              filter(gene_id==x) %>%
#                              get_cred_set_data)
# eur162_cred_set.all.count = sapply(eur162_cred_set.all, function(x) length(x$cred_set_idx))
# eur162_cred_set.all.frac = eur162_cred_set.all.count/sapply(eur162_cred_set.all, function(x) dim(x$data)[1])


# get African credible sets in hg38
afr_cred_set.sig.hg38 = lapply(afr_cred_set.sig, function(x)
  tabix2df("../geno/5M.imputed.dose.biallelic.hg38.sort.txt.gz",
           tabixify(x$data[x$cred_set_idx,]$chr, x$data[x$cred_set_idx,]$pos, x$data[x$cred_set_idx,]$pos),
           c("chr_hg19","pos_hg19","chr_hg38","pos_hg38")) %>%
    mutate(snp_id = paste(chr_hg19, pos_hg19, sep=":")) %>%
    merge(x$data[x$cred_set_idx,] %>% mutate(snp_id = paste(chr, pos, sep=":")), ., by="snp_id") %>%
    merge(., afr_eur162_genes, by.x="gene_id", by.y="afr_id") %>%
    unique)

# test intersection size of credible sets
afr_eur162_cred_set.intersect =
  mapply(function(x,y) y$data[y$cred_set_idx,] %>%
           mutate(eqtl_id = paste(gene_id, pos, sep=":")) %>%
           merge(., data.frame(eqtl_id=paste(x$eur_id, x$pos_hg38, sep=":")), by="eqtl_id") %>%
           dim(.) %>% .[1],
         afr_cred_set.sig.hg38,
         eur162_cred_set.sig)

# get union of credible sets (for jaccard)
afr_eur162_cred_set.union =
  mapply(function(x,y) y$data[y$cred_set_idx,] %>%
           mutate(eqtl_id = paste(gene_id, pos, sep=":")) %>%
           pull(eqtl_id) %>%
           c(., paste(x$gtexv8_id, x$pos_hg38, sep=":")) %>%
           unique %>% length,
         afr_cred_set.sig.hg38,
         eur162_cred_set.sig)

# jaccard for credible sets
afr_eur162_cred_set.jaccard = afr_eur162_cred_set.intersect/afr_eur162_cred_set.union

# compare credible set sizes, including number overlapping SNPs
cred_set.sig.data = data.frame(gene_id = sig_genes$afr_id,
                               Africa = afr_cred_set.sig.count,
                               GTEx = eur162_cred_set.sig.count,
                               intersect = afr_eur162_cred_set.intersect) %>%
  mutate(small_set = ifelse(Africa < GTEx, Africa, GTEx),
         fraction = intersect/small_set)

ggplot(cred_set.sig.data, aes(x=log10(Africa),
                              y=log10(GTEx),
                              size=1-fraction,
                              color=(intersect==0),
                              shape=(intersect==0))) + 
  geom_point(alpha=0.5) +
  theme_bw(base_size=15) + 
  geom_abline(intercept=0, slope=1, color="red") + 
  scale_shape_manual(values=c(19, 17)) + 
  scale_size_continuous(range = c(3,6)) +
  scale_color_ptol()

# gene that has a credible set size of 1 in Africans and in the dozens in 162 EUR GTEx and has no credible set overlap
data.frame(gene_id = sig_genes$afr_id,
           Africa = afr_cred_set.sig.count,
           GTEx = eur162_cred_set.sig.count,
           intersect = afr_eur162_cred_set.intersect) %>%
  filter(intersect==0 & Africa==1) %>%
  arrange(-GTEx) %>% head

# combine_data for plotting
afr_eur162_cred_set.rep.df = rbind(data.frame(sample="Africa",
                                          log10_count=log10(afr_cred_set.rep.count),
                                          frac=afr_cred_set.rep.frac),
                               data.frame(sample="GTEx",
                                          log10_count=log10(eur162_cred_set.rep.count),
                                          frac=eur162_cred_set.rep.frac))

# ECDF plot
afr_eur162_cred_set.rep.df %>%
  gather(key="stat", value="value", -sample) %>%
  ggplot(., aes(x=value, color=sample)) +
  stat_ecdf(size=1.5) +
  theme_classic(base_size=15) +
  facet_wrap(~ stat, scales="free_x") +
  ggtitle("Credible Set Sizes")

# credible set for eGenes ascertained in Africans or Europeans
afr_cred_set.afr  = parLapply(cl, afr_eur162_genes %>%
                                filter(afr_id %in% gemma.fdr05$gene_id) %>%
                                pull(afr_id), afr_get_cred_set)
afr_cred_set.afr.count = sapply(afr_cred_set.afr, function(x) length(x$cred_set_idx))
afr_cred_set.afr.frac = afr_cred_set.afr.count/sapply(afr_cred_set.afr, function(x) dim(x$data)[1])

eur162_cred_set.eur = parLapply(cl, afr_eur162_genes %>%
                                filter(eur_id %in% eur162.fdr05$gene_id) %>%
                                pull(eur_id), eur_get_cred_set)
eur162_cred_set.eur.count = sapply(eur162_cred_set.eur, function(x) length(x$cred_set_idx))
eur162_cred_set.eur.frac = eur162_cred_set.eur.count/sapply(eur162_cred_set.eur, function(x) dim(x$data)[1])

afr_eur162_cred_set.ind.df = rbind(data.frame(sample="Africa",
                                          log10_count=log10(afr_cred_set.afr.count),
                                          frac=afr_cred_set.afr.frac),
                               data.frame(sample="GTEx",
                                          log10_count=log10(eur162_cred_set.eur.count),
                                          frac=eur162_cred_set.eur.frac))

afr_eur162_cred_set.ind.df %>%
  gather(key="stat", value="value", -sample) %>%
  ggplot(., aes(x=value, color=sample)) +
  stat_ecdf(size=1.5) +
  theme_classic(base_size=15) +
  facet_wrap(~ stat, scales="free_x") +
  ggtitle("Credible Set Sizes")





afr_top20_bf_frac.afr = parSapply(cl, afr_cred_set.afr,
                                  function(x) get_top20_bf_frac.p(pvals=x$data$p.value,
                                                                  maf=x$data$maf,
                                                                  sample_size=162))
                                  

eur162_top20_bf_frac.afr = parSapply(cl, eur162_cred_set.afr,
                                  function(x) get_top20_bf_frac.p(pvals=x$data$p.value,
                                                                  maf=x$data$maf,
                                                                  sample_size=162))

afr_cred_set.eur  = parLapply(cl, afr_eur162_genes %>%
                                filter(eur_id %in% eur162.fdr05$gene_id) %>%
                                pull(afr_id), afr_get_cred_set)

eur162_cred_set.eur = parLapply(cl, afr_eur162_genes %>%
                                filter(eur_id %in% eur162.fdr05$gene_id) %>%
                                pull(eur_id), eur_get_cred_set)


afr_top20_bf_frac.eur = parSapply(cl, afr_cred_set.eur,
                                  function(x) get_top20_bf_frac.p(pvals=x$data$p.value,
                                                                  maf=x$data$maf,
                                                                  sample_size=162))

eur162_top20_bf_frac.eur = parSapply(cl, eur162_cred_set.eur,
                                  function(x) get_top20_bf_frac.p(pvals=x$data$p.value,
                                                                  maf=x$data$maf,
                                                                  sample_size=162))




# plot ECDF of credible set sizes
rbind(data.frame(study="Africa", cred_set=sapply(afr_cred_set.afr, function(x) log10(length(x$cred_set_idx)))),
      data.frame(study="GTEx 162", cred_set=sapply(eur162_cred_set.eur, function(x) log10(length(x$cred_set_idx)))))%>%
  ggplot(., aes(x=cred_set, color=study)) +
  stat_ecdf(size=2) +
  theme_classic(base_size=15)

# plot differences
finemap_boxplot.data = rbind(data.frame(ascertained="AFR",  diff=(afr_top20_bf_frac.afr - eur162_top20_bf_frac.afr)),
      data.frame(ascertained="GTEx", diff=(eur162_top20_bf_frac.eur - afr_top20_bf_frac.eur)))

finemap_boxplot.plot = finemap_boxplot.data%>%
  ggplot(., aes(x=ascertained, y=diff, fill=ascertained)) +
  geom_boxplot(show.legend = FALSE) +
  theme_classic(base_size=15) +
  scale_fill_ptol() + 
  ylab(expression(FBF[20]^{ascertained} - FBF[20]^{other})) + 
  ggtitle(expression("Fine-mapping gain"))

# plot 
finemap_violin.data = rbind(data.frame(ascertained="AFR",  test="AFR",  bf20=afr_top20_bf_frac.afr),
      data.frame(ascertained="GTEx", test="GTEx", bf20=eur162_top20_bf_frac.eur),
      data.frame(ascertained="AFR",  test="GTEx", bf20=eur162_top20_bf_frac.afr),
      data.frame(ascertained="GTEx", test="AFR",  bf20=afr_top20_bf_frac.eur)) %>%
  mutate(group=ifelse(ascertained==test, "ascertained", "other"))

finemap_violin.plot = finemap_violin.plot %>%
  ggplot(., aes(x=test, y=bf20, fill=test)) + 
  geom_violin(show.legend = F) +
  geom_boxplot(width=0.1, outlier.size=-1, show.legend = FALSE) +
  theme_classic(base_size=15) +
  scale_fill_ptol() +
  facet_grid(~ group) +
  ylab(expression(FBF^20)) +
  ggtitle(expression("Fine-mapping in AFR vs. GTEx"[162]))

# p-values
finemap_boxplot.data


# credible sets in GTEx v7
afr_cred_set.sig.v7 = lapply(afr_cred_set.sig, function(x) gemma.fdr05.gtex %>% filter(eqtl_id %in% (x$data[x$cred_set_idx,] %>% mutate(eqtl_id = paste(gene_id, pos, sep=":")) %>% pull(eqtl_id))))

```


Look at dstat among credible sets

```{r}

afr_credset_dstat = gemma.egenes %>%
  filter(eqtl_hg19 %in% unlist(afr_cred_set.all)) %>%
  merge(dstat.dir %>% select(-c(chr, pos, aa, ref, alt))) %>% 
  unique %>%
  gather(key="pop", value="dstat", pops) %>%
  group_by(gene_id, pop) %>%
  summarise(mean_d = mean(abs(dstat)),
            min_d = min(abs(dstat)))

gemma.egenes.cred_set = gemma.egenes %>%
  filter(eqtl_hg19 %in% unlist(afr_cred_set.all))


ihs_cols = c("snp_hg19", "chr","pos","alt_frq","ihh1","ihh0","ihs_raw","ihs_norm","thresh")
credset_ihs = data.frame()
for (population in pops){
  
  ihs.f = paste0("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/ihs/pop_specific/all_5M/", population,
                 ".phasedImputed.R285.PolarizedfromFasta.ihs.out.100bins.norm.txt.gz")
  
  
  credset_ihs = rbind(credset_ihs,
                      tabix2df(ihs.f,
                               tabixify(gemma.egenes.cred_set$chr,
                                        gemma.egenes.cred_set$pos_hg19,
                                        gemma.egenes.cred_set$pos_hg19),
                               ihs_cols) %>%
                        unique %>%
                        merge(gemma.egenes.cred_set) %>%
                        group_by(gene_id) %>%
                        summarise(mean_ihs = mean(ihs_norm)) %>%
                        mutate(pop = population))
  
}

```


Manhattan plots for genes where there is no overlap betwen credible sets

```{r}

bad_genes = cred_set.sig.data %>%
  filter(intersect==0) %>%
  pull(gene_id)

bad_genes.idx = which(sig_genes$afr_id %in% bad_genes)

afr_bad_genes.df = data.frame()
for (idx in bad_genes.idx){
  
  data.tmp = afr_cred_set.sig[[idx]]$data
  
  data.tmp.hg38 = tabix2df("../geno/5M.imputed.dose.biallelic.hg38.sort.txt.gz",
           tabixify(data.tmp$chr, data.tmp$pos, data.tmp$pos),
           c("chr_hg19","pos_hg19","chr_hg38","pos_hg38")) %>%
    mutate(snp_id = paste(chr_hg19, pos_hg19, sep=":")) %>%
    merge(data.tmp, ., by="snp_id") %>%
    merge(., afr_eur162_genes, by.x="gene_id", by.y="afr_id") %>%
    unique %>%
    mutate(pos=pos_hg38, study="Africa") %>%
    select(c(pos, p.value, gene_sub, study))
  
  afr_bad_genes.df = rbind(afr_bad_genes.df, data.tmp.hg38)
}

eur162_bad_genes.df = data.frame()
for (idx in bad_genes.idx){
  
  data.tmp = eur162_cred_set.sig[[idx]]$data %>%
    separate(gene_id, into=c("gene_sub", "foo"), sep="\\.") %>%
    select(pos, p.value, gene_sub)
  data.tmp$study = "GTEx"
  
  eur162_bad_genes.df = rbind(eur162_bad_genes.df, data.tmp)
  }

bad_genes.df = rbind(unique(afr_bad_genes.df), unique(eur162_bad_genes.df)) %>%
  pivot_wider(names_from="study", values_from="p.value")

bad_genes.df$Africa = unlist(sapply(bad_genes.df$Africa, function(x) as.numeric(unlist(x)[1])))
bad_genes.df$GTEx = unlist(sapply(bad_genes.df$GTEx, function(x) as.numeric(unlist(x)[1])))

bad_genes.df %>%
  ggplot(., aes(x=as.numeric(pos))) +
  geom_point(aes(y=-log10(Africa)), color="red") +
  geom_point(aes(y=log10(GTEx)), color="blue") +
  theme_classic() +
  facet_wrap(~ gene_sub, scales="free")

```


Natural Selection

I need genetic map information for running iHS. I'm going to use recombination maps from ftp://ftp.ncbi.nlm.nih.gov/hapmap/recombination/2011-01_phaseII_B37/genetic_map_HapMapII_GRCh37.tar.gz as done by Pala et al. 2017. 

```{r}

snp_data = fread("/local1/derek/data/Illumina5M/5M.imputed.rna_pops.map",
                 col.names=c("chr","snp_id","pos"), header=F)

# interpolate map information
for (i in 1:22){
  
  # read in map data
  map_data.f = paste(c("../ihs/genetic_map_GRCh37/genetic_map_GRCh37_chr",i,".txt"), collapse="")
  map_data = fread(map_data.f, col.names=c("chr","pos","rate","map"))
  
  # filter SNP data for current chromosome
  snp_data.sub = snp_data %>%
    filter(chr==i)
  
  # interpolate SNP data
  map_data.interp = approx(x=map_data$pos,
                           y=map_data$map,
                           xout=snp_data.sub$pos,
                           rule=2)
  
  # add interpolated data
  snp_data.sub$map = map_data.interp$y
  
  # write out interpolated data
  interp.f = paste(c("interp_map/5M.imputed.rna_pops.chr",i,".map"), collapse="")
  write.table(x=snp_data.sub[,c("chr","snp_id","map","pos")],
              file=interp.f,
              sep='\t', col.names=F, row.names=F, quote=F)
}


# repeat for all 5M
```


coloc between Dstat and eQTLs

```{r}

dstat = read_tsv("../fst/Dstat/all_dstat.txt.gz") %>%
  mutate(snp_id = paste(chr, pos, sep=":"))

# convert d-statistics to empirical p-value
dstat.rankp = data.frame(snp_hg19=dstat$snp_id, stringsAsFactors=F)
for (pop in pops){
  
  dstat.rankp[,pop] = rank(-dstat[,pop], ties.method="random")/nrow(dstat)
}

gemma.egenes.dstat = gemma.egenes %>%
  merge(dstat.rankp, by="snp_hg19")

lc.sgenes.dstat = lc.sgenes %>%
  merge(dstat.rankp, by="snp_hg19")


h4_stats = list()
for (pop in pops){
  
  sig_genes = gemma.egenes.dstat[gemma.egenes.dstat[,pop] < 0.001,] %>%
    filter(fdr_sig) %>%
    pull(gene_id) %>%
    unique
  
  sig_introns = lc.sgenes.dstat[lc.sgenes.dstat[,pop] < 0.001,] %>%
    filter(fdr_sig) %>%
    pull(intron) %>%
    unique
  
  sig_coloc.eqtl = sapply(sig_genes, function(gene){
    df.tmp = gemma.egenes.dstat %>% filter(gene_id == gene)
    coloc.abf(dataset1=list(snp=df.tmp$snp_hg19, pvalues=df.tmp[,pop], type="quant", N=162),
              dataset2=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$p.value, type="quant", N=162), MAF=df.tmp$maf)$summary
    }) %>% t %>% as.data.frame
  
  sig_coloc.eqtl$gene_id = rownames(sig_coloc.eqtl)
  
  sig_coloc.sqtl = sapply(sig_introns, function(intr){
    df.tmp = lc.sgenes.dstat %>% filter(intron == intr)
    coloc.abf(dataset1=list(snp=df.tmp$snp_hg19, pvalues=df.tmp[,pop], type="quant", N=162),
              dataset2=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$p.value, type="quant", N=162), MAF=df.tmp$maf)$summary
    }) %>% t %>% as.data.frame
  
  sig_coloc.sqtl$intron = rownames(sig_coloc.sqtl)
  
  h4_stats[[pop]] = list(eqtl = sig_coloc.eqtl %>% merge(gene_name),
                         sqtl = sig_coloc.sqtl %>%
                           merge(lc.fdr05.top %>% dplyr::select(c(intron, v19_id))) %>%
                           merge(gene_map) %>%
                           merge(gene_name))
}

h4_stats.sig = data.frame()
for (x in pops){
  h4_stats.sig = rbind(h4_stats.sig,
        h4_stats[[x]]$eqtl %>%
          gather(key="hypothesis", value="prob",
                 c(PP.H0.abf,PP.H1.abf,PP.H2.abf,PP.H3.abf,PP.H4.abf)) %>%
          group_by(gene_id) %>%
          filter(prob==max(prob)) %>%
          filter(hypothesis=="PP.H4.abf") %>%
          ungroup %>%
          mutate(pop=x, analysis="eqtl", intron=NA) %>%
          dplyr::select(c(gene_id, gene_name, intron, prob, pop, analysis)),
        h4_stats[[x]]$sqtl %>%
          gather(key="hypothesis", value="prob",
                 c(PP.H0.abf,PP.H1.abf,PP.H2.abf,PP.H3.abf,PP.H4.abf)) %>%
          group_by(gene_id) %>%
          filter(prob==max(prob)) %>%
          filter(hypothesis=="PP.H4.abf") %>%
          ungroup %>%
          mutate(pop=x, analysis="sqtl") %>%
          dplyr::select(c(gene_id, gene_name, intron, prob, pop, analysis)))
}

gemma.egenes.dstat %>% filter(top_esnp) %>% merge(h4_stats.sig %>% filter(analysis=="eqtl")) %>% gather(key="dstat_pop", value="dstat", pops) %>% filter(pop==dstat_pop) %>% unique %>% arrange(dstat)

h4_stats.sig.scat = list()
h4_stats.sig.manhat = list()
for(population in pops){
  
  stats = h4_stats.sig %>%
    filter(pop==population)
  
  estats = stats %>% 
    filter(analysis=="eqtl")
  
  sstats = stats %>%
    filter(analysis=="sqtl")
  
  eqtl.tmp = gemma.egenes.dstat %>%
    filter(gene_id %in% estats$gene_id) %>%
    merge(gene_name)
  
  sqtl.tmp = lc.sgenes.dstat %>%
    filter(intron %in% sstats$intron) %>%
    merge(gene_name)
  
  eplot.df = data.frame()
  etext.df = data.frame()
  if (nrow(eqtl.tmp) > 0){
    
    eplot.df = data.frame(dstat = eqtl.tmp[,population],
                          p.value = eqtl.tmp$p.value,
                          pos = as.numeric(eqtl.tmp$pos_hg19),
                          gene_id = eqtl.tmp$gene_id) %>%
      merge(gene_name) %>%
      mutate(name = paste0(gene_name,"_eqtl"))
    
    etext.df = data.frame(name = paste0(estats$gene_name,"_eqtl"),
                          H4 = round(estats$prob, digits=2),
                          x = 0.5,
                          y = eplot.df %>%
                            group_by(gene_id) %>%
                            summarise(y = -log10(min(p.value)) * 0.75) %>%
                            pull(y))
  }
  
  splot.df = data.frame()
  stext.df = data.frame()
  if (nrow(sqtl.tmp) > 0){
    
    splot.df = data.frame(dstat = sqtl.tmp[,population],
                          p.value = sqtl.tmp$p.value,
                          pos = as.numeric(sqtl.tmp$pos_hg19),
                          gene_id = sqtl.tmp$gene_id) %>%
      merge(gene_name) %>%
      mutate(name = paste0(gene_name,"_sqtl"))
    
    stext.df = data.frame(name = paste0(sstats$gene_name,"_sqtl"),
                          H4 = round(sstats$prob, digits=2),
                          x = 0.5,
                          y = splot.df %>%
                            group_by(gene_id) %>%
                            summarise(y = -log10(min(p.value)) * 0.75) %>%
                            pull(y))
  }
  
  plot.df = rbind(eplot.df, splot.df)
  text.df = rbind(etext.df, stext.df)
  
  h4_stats.sig.scat[[population]] =  ggplot(plot.df, aes(x=-log10(dstat),
                                                  y=-log10(p.value))) +
    geom_point() +
    geom_text(data=text.df, mapping=aes(x=x, y=y, label=H4)) +
    theme_classic() +
    facet_wrap(~ name, scales="free")
  
  h4_stats.sig.manhat[[population]] =  ggplot(plot.df) +
    geom_point(aes(x=pos, y=-log10(p.value), color="red")) +
    geom_point(aes(x=pos, y=log10(dstat), color="blue")) +
    theme_classic() +
    facet_wrap(~ name, scales="free")
  
}



```


## GWAS eqtl

```{r}

traits = read_tsv("/local1/derek/data/gwas/matt_gwas/5M/trait_units", col_names=F)
hits_5M = read_tsv("/local1/derek/data/gwas/matt_gwas/5M/FDR_5M.snpinfo")
hits_Meta = read_tsv("/local1/derek/data/gwas/matt_gwas/1M_5M/FDR_Meta.snpinfo")

gwas_eqtl_hits = rbind(hits_5M %>% dplyr::select(c(Trait, Chr, Pos, pval, pval_GC)),
                       hits_Meta %>% dplyr::select(c(Trait, Chr, Pos, pval, pval_GC))) %>%
  mutate(snp_hg19 = paste(Chr, Pos, sep=":")) %>%
  filter(snp_hg19 %in% c(gemma.egenes %>% filter(fdr_sig) %>% pull(snp_hg19)))

coloc_genes = gemma.egenes %>%
  filter(fdr_sig) %>%
  filter(snp_hg19 %in% gwas_eqtl_hits$snp_hg19) %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  pull(gene_id)
  

coloc_data = data.frame()
for (gene in coloc_genes){
  
  name = gene_name %>%
    filter(gene_id==gene) %>%
    pull(gene_name)
  
  eqtl_stats = gemma.egenes %>%
    filter(gene_id == gene) %>%
    dplyr::select(c(gene_id, chr, pos_hg19, snp_hg19, maf, p.value))
  
  test_traits = gwas_eqtl_hits %>%
    mutate(snp_hg19 = paste(Chr, Pos, sep=":")) %>%
    filter(snp_hg19 %in% eqtl_stats$snp_hg19) %>%
    pull(Trait) %>%
    unique
  
  for (trait in test_traits){
    
    # if (trait %in% c(1,7,8,10,11,12,13)){
    #   gwas_stats = tabix2df(paste0("/local1/derek/data/gwas/matt_gwas/1M_5M/", trait, ".1M_5M.sort.assoc.gz"),
    #                         tabixify(eqtl_stats$chr, eqtl_stats$pos_hg19, eqtl_stats$pos_hg19),
    #                         c("chr","pos_hg19","snp_hg19","beta","stddev","gwas_p.value","fdr","pval_GC","FDR_GC","Num_Study","pvalue_RE",
    #                           "Beta_RE","StdDev_RE","pvalue_RE2","Stat1_RE2","Stat2_RE2","I_square","Q","pvalue_Q","tau_square","pvalue_1M","pvalue_5M")) %>%
    #     dplyr::select(c(snp_hg19, gwas_maf=MAF, n_gwas=`Num_Alleles_Alt(AC)`, gwas_p.value)) %>%
    #     mutate(n_gwas = n_gwas/2)
    # 
    #   } else{
      gwas_stats = tabix2df(paste0("/local1/derek/data/gwas/matt_gwas/5M/", trait, ".5M.sort.assoc.gz"),
                            tabixify(eqtl_stats$chr, eqtl_stats$pos_hg19, eqtl_stats$pos_hg19),
                            c("chr","pos_hg19","snp_hg19","beta","stddev","gwas_p.value","FDR","pval_GC","FDR_GC",
                              "Imputation_R2(R2)","Num_Alleles_Total(AN)","Num_Alleles_Alt(AC)","Ref","Alt","HWE_pval","MAF")) %>%
        dplyr::select(c(snp_hg19, gwas_maf=MAF, n_gwas=`Num_Alleles_Alt(AC)`, gwas_p.value)) %>%
        mutate(n_gwas = n_gwas/2)
      # 
      # }    
    
    data = merge(eqtl_stats, gwas_stats)
    data$pheno = paste0(name, "_", traits[trait,]$X3, "_eqtl")
    data$intron = NA
    
    coloc_data = rbind(coloc_data, data)
    
  }
}



gwas_sqtl_hits = rbind(hits_5M %>% dplyr::select(c(Trait, Chr, Pos, pval, pval_GC)),
                       hits_Meta %>% dplyr::select(c(Trait, Chr, Pos, pval, pval_GC))) %>%
  mutate(snp_hg19 = paste(Chr, Pos, sep=":")) %>%
  filter(snp_hg19 %in% c(lc.sgenes %>% filter(fdr_sig) %>% pull(snp_hg19)))

coloc_introns = lc.sgenes %>%
  filter(fdr_sig) %>%
  filter(snp_hg19 %in% gwas_sqtl_hits$snp_hg19) %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  pull(intron)


for (int in coloc_introns){
  
  name = lc.sgenes %>%
    filter(top_ssnp & intron==int) %>%
    merge(gene_name) %>%
    pull(gene_name) %>%
    unique
  
  sqtl_stats = lc.sgenes %>%
    filter(intron==int) %>%
    mutate(maf = ifelse(af < 0.5, af, 1-af)) %>%
    dplyr::select(c(gene_id, chr, pos_hg19, snp_hg19, maf, p.value))
  
  test_traits = gwas_eqtl_hits %>%
    mutate(snp_hg19 = paste(Chr, Pos, sep=":")) %>%
    filter(snp_hg19 %in% sqtl_stats$snp_hg19) %>%
    pull(Trait) %>%
    unique
  
  # for (trait in test_traits){
  #   
  #   if (trait %in% c(1,7,8,10,11,12,13)){
  #     gwas_stats = tabix2df(paste0("/local1/derek/data/gwas/matt_gwas/1M_5M/", trait, ".1M_5M.sort.assoc.gz"),
  #                           tabixify(sqtl_stats$chr, sqtl_stats$pos_hg19, sqtl_stats$pos_hg19),
  #                           c("chr","pos_hg19","snp_hg19","beta","stddev","gwas_p.value","fdr","pval_GC","FDR_GC","Num_Study","pvalue_RE",
  #                             "Beta_RE","StdDev_RE","pvalue_RE2","Stat1_RE2","Stat2_RE2","I_square","Q","pvalue_Q","tau_square","pvalue_1M","pvalue_5M")) %>%
  #       dplyr::select(c(snp_hg19, gwas_maf=MAF, n_gwas=`Num_Alleles_Alt(AC)`, gwas_p.value)) %>%
  #       mutate(n_gwas = n_gwas/2)
  #     
  #   } else{
      gwas_stats = tabix2df(paste0("/local1/derek/data/gwas/matt_gwas/5M/", trait, ".5M.sort.assoc.gz"),
                            tabixify(sqtl_stats$chr, sqtl_stats$pos_hg19, sqtl_stats$pos_hg19),
                            c("chr","pos_hg19","snp_hg19","beta","stddev","gwas_p.value","FDR","pval_GC","FDR_GC",
                              "Imputation_R2(R2)","Num_Alleles_Total(AN)","Num_Alleles_Alt(AC)","Ref","Alt","HWE_pval","MAF")) %>%
        dplyr::select(c(snp_hg19, gwas_maf=MAF, n_gwas=`Num_Alleles_Alt(AC)`, gwas_p.value)) %>%
        mutate(n_gwas = n_gwas/2)
    #   
    # }    
    
    data = merge(sqtl_stats, gwas_stats)
    data$pheno = paste0(name, "_", traits[trait,]$X3, "_sqtl")
    
    coloc_data = rbind(coloc_data, data)
    
  }
}


esqtl_gwas_coloc = sapply(unique(coloc_data$pheno), function(phe){
    df.tmp = coloc_data %>% filter(pheno==phe)
    coloc.abf(dataset1=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$p.value, type="quant", N=162, MAF=df.tmp$maf),
              dataset2=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$gwas_p.value, type="quant", N=df.tmp$n_gwas, MAF=df.tmp$gwas_maf))$summary
    }) %>% t %>% as.data.frame


esqtl_gwas_coloc$pheno = rownames(esqtl_gwas_coloc)


ggplot(coloc_data, aes(x=-log10(gwas_p.value), y=-log10(p.value))) +
  geom_point() +
  theme_classic() +
  facet_wrap(~ pheno, scales="free")  

ggplot(coloc_plots[[1]], aes(x=-log10(gwas_p.value), -log10(p.value))) +
      geom_point() +
      theme_classic()

```



```{r}

epi_dir = "/local1/derek/data/epi_roadmap/all/rna_seq/"
epi_names = read_tsv(paste0(epi_dir, "EG.name.txt"), col_names = c("id","ctype"))
epi_gex = read_tsv(paste0(epi_dir, "57epigenomes.RPKM.pc.gz"))

```


Look at TMEM216 in GTEx v7

```{r}

tmem_v7 = read_tsv("../gtex/TMEM216.ALL.v7.signif_variant_gene_pairs.txt",
                   col_names = c("variant_id","gene_id","tss_distance","ma_samples",
                                 "ma_count","maf","pval_nominal","slope","slope_se",
                                 "pval_nominal_threshold","min_pval_nominal","pval_beta","tissue")) %>%
  separate(variant_id, into=c("chr","pos","ref","alt","genome"), sep="_") %>%
  mutate(snp_id = paste(chr, pos, sep=":"))

tmem_v7_wb = tabix2df("../misc/Whole_Blood.allpairs.sub.txt.gz",
                      "ENSG00000187049.5:1-100000000",
                      c("gene_id","chr","pos","allele1","allele2","tss_dist",
                        "ma_samples","ma_count","maf","p.value","slope","slope_se")) %>%
  mutate(snp_id = paste(chr, pos, sep=":"))

tmem_v7_se = tabix2df("../gtex/v7/Skin_Sun_Exposed_Lower_leg.allpairs.sort.txt.gz",
                      "ENSG00000187049.5:1-100000000",
                      c("gene_id","chr","pos","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","maf","p.value","slope","slope_se")) %>%
  mutate(snp_id = paste(chr, pos, sep=":"))

tmem_gwas = tabix2df("/local1/derek/data/gwas/matt_gwas/5M/25.5M.sort.assoc.gz",
                     tabixify(tmem_v7_se$chr, tmem_v7_se$pos, tmem_v7_se$pos),
                     c("chr","pos","snp_id","beta","stddev","gwas_p.value","FDR",
                       "pval_GC","FDR_GC","Imputation_R2(R2)","Num_Alleles_Total(AN)",
                       "Num_Alleles_Alt(AC)","Ref","Alt","HWE_pval","MAF"))

tmem_gwas %>%
  select(snp_id, gwas_p.value) %>%
  merge(., tmem_v7_wb) %>%
  ggplot(aes(x=-log10(gwas_p.value), y=-log10(p.value))) +
  geom_point() +
  theme_classic()

tmem_gwas %>%
  select(snp_id, gwas_p.value) %>%
  merge(., tmem_v7_se) %>%
  ggplot(aes(x=-log10(gwas_p.value), y=-log10(p.value))) +
  geom_point() +
  theme_classic()


meta_cols = readLines("/local1/derek/data/gtex/GTEx_Analysis_v7_eQTL/metasoft_cols.txt")

meta = read_tsv("/local1/derek/data/gtex/GTEx_Analysis_v7_eQTL/GTEx_Analysis_v7.metasoft.txt.gz",
                col_names=meta_cols)


```


Conditional GEMMA

```{r}
old_path = Sys.getenv("PATH")
Sys.setenv(LD_LIBRARY_PATH = paste("/usr/local/lib64/", old_path, sep = ":"))

pheno_resid = read_tsv("pheno_resid/peer24/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer24.resid.txt.gz")
geno_file = "../geno/5M.imputed.dose.biallelic.genos.w_posit.txt.gz"

mursi.dstat.top = gemma.fdr05.dstat %>%
  group_by(gene_id) %>% 
  filter(abs(mursi)==max(abs(mursi))) %>%
  sample_n(1) %>%
  arrange(-abs(mursi))

genos = tabix2df(geno_file, tabixify(mursi.dstat.top$chr[7], mursi.dstat.top$pos[7], mursi.dstat.top$pos[7]), c("chr","pos",samps))

genos.long = genos %>%
  select(-c(1:2)) %>%
  gather(key=sample, value=gtype)

pheno = pheno_resid %>%
  filter(gene_id==mursi.dstat.top$gene_id[7])

pheno.long = pheno %>%
  select(-c(1:5)) %>%
  gather(key=sample, value=gex)

## residualize against genotype
pheno.lm = merge(genos.long, pheno.long, sort=F) %>%
  lm(gex ~ gtype, .)

resid.df = data.frame(chr = pheno$chr,
                      start = pheno$start,
                      end = pheno$end,
                      strand = pheno$strand,
                      gene_id = pheno$gene_id,
                      sample = samps,
                      resid = pheno.lm$residuals) %>%
  pivot_wider(names_from=sample, values_from=resid)

resid.f = paste(c("resid_test/", pheno$gene_id, "_", paste(genos$chr, genos$pos, sep=":"), "_resid"), collapse="")

write.table(resid.df, resid.f, row.names=F, col.names=F, quote=F, sep="\t")

## run GEMMA
system(paste0("bash run_gemma.sh ", resid.f))

gemma.resid = read_tsv(paste(resid.f, "gemma.100kb.lmm4", sep="."),
                       col_names = gemma_cols, col_types="ciciiccnnnnnnnnn") %>%
  separate(snp_id, into=c("chr","pos"), sep=":", remove=F) %>%
  mutate(eqtl_id = paste(gene_id, pos, sep=":"))

gemma.orig = gemma %>% filter(gene_id==pheno$gene_id)

rbind(data.frame(pos = as.numeric(gemma.orig$pos),
                 p.value = gemma.orig$p.value,
                 version = "original"),
      data.frame(pos = as.numeric(gemma.resid$pos),
                 p.value = gemma.resid$p.value,
                 version = paste(genos$chr, genos$pos, sep=":"))) %>%
  ggplot(., aes(x=pos, y=-log10(p.value), color=version)) +
  geom_point() +
  theme_classic()



```


Look at iHS and eQTLs

```{r}
ihs_cols = c("chr","pos","alt_frq","ihh1","ihh0","ihs_raw","ihs_norm","thresh")
gemma.egenes.ihs = data.frame(snp_id = gemma.egenes$snp_hg19,
                              stringsAsFactors=F)

for (pop in pops){
  
  ihs.f = paste(c("../ihs/pop_specific/", pop, ".5M.imputed.dose.biallelic.ihs.out.100bins.norm.gz"), collapse="")
  
  ihs.df = tabix2df(ihs.f, tabixify(gemma.egenes.ihs$chr, gemma.egenes.ihs$pos_hg19, gemma.egenes.ihs$pos_hg19), ihs_cols) %>%
    mutate(snp_id = paste(chr, pos_hg19, sep=":"))
  colnames(ihs.df)[colnames(ihs.df)=="ihs_norm"] = paste0(pop, "_ihs")
    
  gemma.egenes.ihs = ihs.df[,c("snp_id", paste0(pop, "_ihs"))] %>%
    merge(gemma.egenes.ihs, ., by="snp_id", all.x="T")
}

gemma.egenes.ihs = merge(gemma.egenes, gemma.egenes.ihs) %>%
  unique %>%
  mutate(eqtl_id = paste(gene_id, pos, sep=":"))

h4_ihs = list()
for (pop in paste0(pops, "_ihs")){
  
  sig_genes = gemma.egenes.ihs[pnorm(abs(gemma.egenes.ihs[,pop]), lower.tail=F)*2 < 0.001,] %>%
    filter(eqtl_id %in% gemma.fdr05$eqtl_id) %>%
    pull(gene_id) %>%
    unique
  
  # sig_clust
  #   filter(sqtl_id %in% lc.fdr05$sqtl_id) %>%
  #   pull(clust) %>%
  #   unique
  
  sig_coloc.eqtl = sapply(sig_genes, function(gene){
    df.tmp = gemma.egenes.ihs %>%
      filter(gene_id == gene)
    ihs_pvals = pnorm(abs(df.tmp[,pop]), lower.tail=F)*2
    df.tmp = df.tmp[!is.na(ihs_pvals),]
    ihs_pvals = ihs_pvals[!is.na(ihs_pvals)]
    coloc.abf(dataset1=list(snp=df.tmp$snp_id, pvalues=ihs_pvals, type="quant", N=162),
              dataset2=list(snp=df.tmp$snp_id, pvalues=df.tmp$p.value, type="quant", N=162), MAF=df.tmp$maf)$summary
    }) %>% t %>% as.data.frame
  
  sig_coloc.eqtl$gene_id = rownames(sig_coloc.eqtl)
  
  # sig_coloc.sqtl = sapply(sig_clusts, function(cluster){
  #   df.tmp = lc.sgenes.dstat %>% filter(clust == cluster)
  #   coloc.abf(dataset1=list(snp=df.tmp$snp_id, pvalues=df.tmp[,pop], type="quant", N=162),
  #             dataset2=list(snp=df.tmp$snp_id, pvalues=df.tmp$p.value, type="quant", N=162), MAF=df.tmp$maf)$summary
  #   }) %>% t %>% as.data.frame
  # 
  # sig_coloc.sqtl$clust = rownames(sig_coloc.sqtl)
  
  # h4_stats[[pop]] = list(eqtl = sig_coloc.eqtl %>% merge(gene_name),
  #                        sqtl = sig_coloc.sqtl %>% merge(lc.fdr05.top %>% select(c(clust, gene_id)), by="clust") %>%
  #                          merge(gene_name))

  h4_ihs[[pop]] = list(eqtl = sig_coloc.eqtl %>% merge(gene_name))
}
```


Enrichment of eQTLs and sQTLs

```{r}

maf_bins = 1:10/20
ntag_bins = c(-1, 2, 5, 10, 20, 50, Inf)

tags = read_tsv("../eur_fst/5M.imputed.fix_ref.r205.tags.list.max_fst.05.txt.gz") %>%
  filter(!is.na(fst)) %>%
  select(chr, pos, ntag, maf, fst, max_fst) %>%
  unite("snp_hg19", chr, pos, sep=":")

tags = tags %>%
  filter(!is.na(fst)) %>%
  mutate(ntag_bin = cut(ntag, ntag_bins),
         maf_bin = cut(maf, maf_bins)) %>%
  arrange(maf_bin, ntag_bin)

##min/max indices for eQTLs
fst_samp_mats = tags %>%
  dplyr::select(maf_bin, ntag_bin) %>%
  get_samp_mats

gemma.max_fst = gemma.egenes %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  group_by(gene_id) %>%
  mutate(post_p = bf/sum(bf)) %>%
  filter(top_esnp & post_p > 0.5) %>%
  sample_n(1) %>%
  merge(tags, by="snp_hg19")

lc.max_fst = lc.sgenes %>%
  mutate(bf = exp(approx.bf.estimates(beta/se, se^2, "quant")$lABF)) %>%
  group_by(intron) %>%
  mutate(post_p = bf/sum(bf)) %>%
  filter(top_ssnp & post_p > 0.5) %>%
  sample_n(1) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(tags, by="snp_hg19")

set.seed(7)

# sample variables
gemma.max_fst.samp = mapply(function(i, j) sample_idx(fst_samp_mats$min_mat, fst_samp_mats$max_mat, i, j),
                            gemma.max_fst$maf_bin, gemma.max_fst$ntag_bin)

lc.max_fst.samp = mapply(function(i, j) sample_idx(fst_samp_mats$min_mat, fst_samp_mats$max_mat, i, j),
                         lc.max_fst$maf_bin, lc.max_fst$ntag_bin)

e.fst.null = apply(gemma.max_fst.samp, 1, function(x) tags[x,]$max_fst)

s.fst.null = apply(lc.max_fst.samp, 1, function(x) tags[x,]$max_fst)


e.fst.null.mean_gt03 = apply(e.fst.null, 2, function(x) mean(x > 0.3))
s.fst.null.mean_gt03 = apply(s.fst.null, 2, function(x) mean(x > 0.3))


data.frame(null = e.fst.null.mean_gt03) %>%
  ggplot(aes(x=null)) +
  geom_histogram(alpha=0.5) +
  geom_vline(data=data.frame(frac_outliers = mean(gemma.max_fst$max_fst > 0.3)),
             aes(xintercept=frac_outliers),
             color="red") +
  theme_classic(base_size=15) +
  ggtitle("Fraction Fst outliers among eQTLs")
      
data.frame(null = s.fst.null.mean_gt03) %>%
  ggplot(aes(x=null)) +
  geom_histogram(alpha=0.5) +
  geom_vline(data=data.frame(frac_outliers = mean(lc.max_fst$max_fst > 0.3)),
             aes(xintercept=frac_outliers),
             color="red") +
  theme_classic(base_size=15) +
  ggtitle("Fraction Fst outliers among sQTLs")




# look at enrichment among immune genes
immport = read_tsv("../misc/immport_ensembl_gene_transcript_id.txt",
                   col_names=c("gene_id","trans_id","gene_name"))

sum(gemma.max_fst[which(gemma.max_fst$gene_id %in% immport$gene_id),]$max_fst > 0.3)

e.fst.imm.null = apply(gemma.max_fst.samp[,gemma.max_fst$gene_id %in% immport$gene_id], 1, function(x) tags[x,]$max_fst)

s.fst.imm.null = apply(lc.max_fst.samp[,lc.max_fst$gene_id %in% immport$gene_id], 1, function(x) tags[x,]$max_fst)

# consider credible sets and enrichment in FST outliers
afr_credset_fst = gemma.egenes %>%
  filter(eqtl_hg19 %in% unlist(afr_cred_set.all)) %>%
  merge(tags %>% select(snp_hg19 = snp, fst)) %>%
  unique %>%
  group_by(gene_id) %>%
  summarise(mean_fst = mean(fst, na.rm=T),
            n = n(),
            min_fst = min(fst, na.rm=T))

gemma.egenes.cred_set = gemma.egenes %>%
  filter(eqtl_hg19 %in% unlist(afr_cred_set.all))


# FST weighted by bayes factors
gemma.weighted_fst = gemma.egenes %>%
  mutate(bf = exp(approx.bf.p(p.value, maf, "quant", 162)$lABF)) %>%
  merge(tags %>% select(snp_hg19=snp, fst)) %>%
  filter(!is.na(fst)) %>%
  group_by(gene_id) %>%
  summarise(weighted_fst = sum(fst * bf/sum(bf)))

lc.weighted_fst = lc.sgenes %>%
  mutate(bf = exp(approx.bf.p(p.value, maf, "quant", 162)$lABF)) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(tags %>% select(snp_hg19=snp, fst)) %>%
  filter(!is.na(fst)) %>%
  group_by(intron) %>%
  summarise(weighted_fst = sum(fst * bf/sum(bf)))


gemma.weighted_dstat = gemma.egenes %>%
  mutate(bf = exp(approx.bf.p(p.value, maf, "quant", 162)$lABF)) %>%
  merge(dstat.dir %>% mutate(snp_hg19 = paste(chr, pos, sep=":")) %>% select(-c(chr, pos, aa, ref, alt))) %>%
  gather(key=pop, value=dstat, pops) %>%
  group_by(gene_id, pop) %>%
  summarise(weighted_dstat = sum(abs(dstat) * bf/sum(bf)))
  
lc.weighted_dstat = lc.sgenes %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  mutate(bf = exp(approx.bf.p(p.value, maf, "quant", 162)$lABF)) %>%
  merge(dstat.dir %>% mutate(snp_hg19 = paste(chr, pos, sep=":")) %>% select(-c(chr, pos, aa, ref, alt))) %>%
  gather(key=pop, value=dstat, pops) %>%
  group_by(intron, pop) %>%
  summarise(weighted_dstat = sum(abs(dstat) * bf/sum(bf)))

# gene ontology
library(gprofiler2)

fst_sig_genes = unique(c(gemma.weighted_fst %>%
                    filter(weighted_fst > 0.3) %>%
                    pull(gene_id),
                  lc.weighted_fst %>%
                    filter(weighted_fst > 0.3) %>%
                    merge(lc.fdr05 %>% select(intron, gene_id) %>% unique) %>%
                    pull(gene_id)))

gemma.sig_fst.go = gost(query = fst_sig_genes, custom_bg = unique(c(gemma.fdr05$gene_id, lc.fdr05$gene_id)))


dstat.thresh = apply(dstat.dir[,pops], 2, function(x) quantile(abs(x), c(0.99,0.999))) %>%
  as.data.frame %>%
  t
colnames(dstat.thresh) = c("quant_99","quant_999")
dstat.thresh = dstat.thresh %>%
  as.data.frame %>%
  rownames_to_column("pop")


dstat.go = gost(query = unique(c(lc.weighted_dstat %>% merge(intron_name) %>% merge(dstat.thresh) %>% filter(weighted_dstat > quant_99) %>% pull(gene_id),
                  gemma.weighted_dstat %>% merge(dstat.thresh) %>% filter(weighted_dstat > quant_99) %>% pull(gene_id))),
                custom_bg = unique(c(lc.fdr05$gene_id, gemma.fdr05$gene_id)),
                sources=c("GO:BP","KEGG","REAC"), significant = F)



ihs_cols = c("snp_hg19", "chr","pos","alt_frq","ihh1","ihh0","ihs_raw","ihs_norm","thresh")
credset_ihs = data.frame()
for (population in pops){
  
  ihs.f = paste0("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/ihs/pop_specific/all_5M/", population,
                 ".phasedImputed.R285.PolarizedfromFasta.ihs.out.100bins.norm.txt.gz")
  
  
  credset_ihs = rbind(credset_ihs,
                      tabix2df(ihs.f,
                               tabixify(gemma.egenes.cred_set$chr,
                                        gemma.egenes.cred_set$pos_hg19,
                                        gemma.egenes.cred_set$pos_hg19),
                               ihs_cols) %>%
                        unique %>%
                        merge(gemma.egenes.cred_set) %>%
                        group_by(gene_id) %>%
                        summarise(mean_ihs = mean(ihs_norm)) %>%
                        mutate(pop = population))
  
}


```


Functions for plots eQTLs and sQTLs

```{r}

plot_eqtl = function(gene, snp){
  
  chr = unlist(strsplit(snp, ":"))[1]
  pos = unlist(strsplit(snp, ":"))[2]
  
  geno_cols = readLines("../misc/geno_cols.txt")
  samps = geno_cols[-c(1:9)]
  
  gtypes = tabix2df("../geno/5M.imputed.rna_samps.biallelic.vcf.gz",
                    tabixify(chr, pos, pos),
                    geno_cols) %>%
    gather(key=sample, value=genotype, samps)
  
  ref = unique(gtypes$REF)
  alt = unique(gtypes$ALT)
  
  if (is.logical(ref)){ ref="T" } 
  if (is.logical(alt)){ alt="T" }
  
  genotypes = c(paste(ref, ref, sep="/"), paste(ref, alt, sep="/"), paste(alt, alt, sep="/"))
  
   gtypes = gtypes%>%
    separate(genotype, into=c("gtype","dose"), sep=":", convert=T) %>%
    separate(gtype, into=c("allele1","allele2"), sep="\\|", convert=T) %>%
    mutate(allele_count = allele1 + allele2) %>%
    mutate(genotype = factor(genotypes[allele_count+1], levels=genotypes)) %>%
    select(sample, genotype, dose)
  
  gex = read_tsv("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/gemma/pheno_resid/peer20/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.txt.gz") %>%
    mutate(gene_id = strip_ensembl(gene_id)) %>%
    filter(gene_id==gene) %>%
    gather(key=sample, value=tpm, -c(chr, start, end, strand, gene_id)) %>%
    select(sample, tpm)
  
  eqtl_plot = merge(gtypes, gex) %>%
    ggplot(aes(x=genotype, y=tpm)) +
    geom_boxplot() + 
    geom_beeswarm() +
    theme_classic(base_size=15)
  
  return(eqtl_plot)
}




```


Diagram GWAS

```{r}

diagram = read_delim("../misc/diagram.mega-meta.txt", delim=" ")
diagram = read_tsv("../misc/diagram.mega-meta.posits.hg18_to_hg19.txt.gz",
                       col_names = c("CHR","POS","pos_hg19")) %>%
  merge(diagram)

```


posterior probability of individual variants

```{r}

gemma.egenes.post_p = gemma.egenes %>%
  mutate(bf = exp(approx.bf.p(p.value, maf, "quant", 162)$lABF)) %>%
  group_by(gene_id) %>%
  mutate(post_p = bf / sum(bf)) %>%
  select(gene_id, maf, snp_hg19, post_p)

```


Colocalization between eQTLs and sQTLs

```{r}

# genes that have a significant eQTL or sQTL
esgenes = unique(union(gemma.egenes$v19_id, lc.sgenes$v19_id))

# QTL data for both scans
gemma.lc.esgenes = gemma.egenes %>%
  filter(v19_id %in% esgenes) %>%
  select(v19_id, gene_id, maf, chr, pos_hg19, eqtl_beta = beta, eqtl_pval = p.value) %>%
  merge(lc.sgenes %>%
          filter(v19_id %in% esgenes) %>%
          select(v19_id, pos_hg19, sqtl_beta = beta, sqtl_pval = p.value)) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  mutate(pheno = paste(gene_id, intron, sep=":")) %>%
  unique

# all significant gene-intron pairs
esphenos = unique(gemma.lc.esgenes$pheno)

# coloc for 
gemma.lc.esgenes.coloc = lapply(esphenos, function(espheno){
    df.tmp = gemma.lc.esgenes %>%
      filter(pheno == espheno) %>%
      arrange(pos_hg19)
    coloc.abf(dataset1=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$eqtl_pval, type="quant", N=162, MAF=df.tmp$maf),
              dataset2=list(snp=df.tmp$snp_hg19, pvalues=df.tmp$sqtl_pval, type="quant", N=162, MAF=df.tmp$maf))$summary
    })

gemma.lc.h4.df = data.frame(pheno = esphenos,
                            H0 = sapply(gemma.lc.esgenes.coloc, function(x) x[2]),
                            H1 = sapply(gemma.lc.esgenes.coloc, function(x) x[3]),
                            H2 = sapply(gemma.lc.esgenes.coloc, function(x) x[4]),
                            H3 = sapply(gemma.lc.esgenes.coloc, function(x) x[5]),
                            H4 = sapply(gemma.lc.esgenes.coloc, function(x) x[6]),
                            stringsAsFactors=F)

# hypothesis with maximum probability for each gene
gemma.lc.h4.df.max = gemma.lc.h4.df %>%
  gather(key="hypothesis", value="probability", -pheno) %>%
  group_by(pheno) %>%
  filter(probability==max(probability)) %>%
  separate(pheno, into=c("gene_id","intron"), sep=":")


gemma.lc.h4.df.max %>%
  group_by(gene_id) %>%
  filter(probability==max(probability)) %>%
  group_by(hypothesis) %>%
  summarise(n=n()) %>%
  ggplot(aes(x="eQTL/sQTL coloc", y=n, fill=hypothesis)) +
  geom_bar(position="stack", stat="identity") +
  theme_classic(base_size=15)
```

