---
title: "figures"
author: "derkelly"
date: "11/6/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Global Options

```{r}
library(tidyverse)
library(ggplot2)
library(ggthemes)
library(RColorBrewer)
library(stringr)

source("../analysis/main/eqtl_fncs.R")
source("../analysis/main/coloc/R/claudia.R")

pop_palette = brewer.pal(n = 12, name = "Paired")
names(pop_palette) = c("Agaw","CEU","Amhara","CHB","Argoba","YRI","Sandawe","Mursi","Chabu","Hadza", "Weyto","Dizi")

sample_info = read_csv("../misc/Table_TishkoffDB.csv") %>%
  as.data.frame %>%
  dplyr::rename(sample=lab_id)

# population_data
pop_samps = read_tsv("../misc/pop_samples.in5M.txt",
                     col_names=c("sample","pop")) %>%
  mutate(pop = ifelse(pop=="sabue", "chabu", pop),
         pop = ifelse(pop=="hadzabe","hadza",pop))

long_lat = merge(pop_samps, sample_info) %>%
  dplyr::select(c(pop, long=Longitude_Consensus, lat=Latitude_Consensus)) %>%
  unique %>%
  mutate(pop = ifelse(pop=="sabue", "chabu", pop))

pops_1kg = read_tsv("/local1/derek/data/1kgenomes/1khg_sample_pop.tsv",
                    col_names=c("sample","pop"))

# diet information
sdiet = data.frame(pop = c("agaw", "amhara", "argoba", "dizi", "hadzabe", "mursi", "sabue", "sandawe", "weyto"),
                  diet = c("agr", "agr", "agr", "pas", "hg", "pas", "hg", "fhg", "fhg"))

gtexv8_snps = readLines("../gtex/v8/GTEx_Analysis_v8_tQTL_all_associations_Whole_Blood.tested_snps.txt.gz")

pop_map = c("Agaw","Amhara","Argoba","Dizi","Hadza","Mursi","Chabu","Sandawe","Weyto")
names(pop_map) = c("agaw","amhara","argoba","dizi","hadzabe","mursi","sabue","sandawe","weyto")


## intron info
intron_map = read_tsv("../leafcutter/gemma/intron_map.tsv")

# gene info
gene_map = read_tsv("../misc/gencode.v19.genes.txt", col_names="v19_id") %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  merge(read_tsv("../misc/gencode.v26.genes.txt", col_names="v26_id") %>%
          mutate(gene_id = strip_ensembl(v26_id)), all=T) %>%
  merge(read_tsv("../misc/gencode.v29.genes.txt", col_names="v29_id") %>%
          mutate(gene_id = strip_ensembl(v29_id)), all=T)

```


## Figure 1

```{r figure 1}
library(sf)
library(raster)
library(dplyr)
library(spData)
library(tmap)
select <- dplyr::select
filter <- dplyr::filter

# get sf object for Ethiopia and Tanzania
countries = world %>%
  filter(iso_a2 %in% c("ET","TZ"))


###################################################
####################### MAP #######################
###################################################

world_points <- st_centroid(world)
world_points <- cbind(world, st_coordinates(st_centroid(world$geometry)))

pd <- position_dodge(0.4)

fig1.map.plot = ggplot(data = world) +
  geom_sf(fill = "gray95") +
  # geom_point(data = long_lat, aes(x=long, y=lat, color=pop), size=5) +
  geom_point(data = long_lat %>% mutate(Population = str_to_title(pop)),
             aes(x=long, y=lat, color=Population), size=5) +
  geom_text(data=data.frame(long=c(40.5,35), lat=c(7.5,-7.5), country=c("Ethiopia", "Tanzania")),
            aes(x=long, y=lat, label=country)) +
  #     color = "darkblue", fontface = "bold", check_overlap = FALSE) +
  # annotate(geom = "text", x = -90, y = 26, label = "Gulf of Mexico", 
  #     fontface = "italic", color = "grey22", size = 6) +
  coord_sf(xlim = c(26, 52), ylim = c(-14, 17), expand = FALSE) +
  theme_bw(base_size=12) +
  scale_color_manual(values=pop_palette) +
  # ggtitle("A") +
  theme(plot.title = element_text(size = 30, vjust=3),
        legend.position = "none")



###################################################
####################### PCA #######################
###################################################

var_explained = read_table("../pca/5M_3deg_1kg_merged.intersection.rna_samps.noATGCsnps.MAC2.R2.99.ld_prune.eval",
                           col_names="eval") %>%
  mutate(var_explained = eval/sum(eval))

pcs = read.delim("../pca/5M_3deg_1kg_merged.intersection.rna_samps.noATGCsnps.MAC2.R2.99.ld_prune.evec",
                 sep="", header=F, skip = 1, na.strings="???", 
                 col.names=c("samps", paste("PC", 1:10, sep=""), "foo")) %>%
  separate(samps, into=c("sample","bar"), sep=":") %>%
  select(-c(foo,bar)) %>%
  merge(., rbind(pop_samps, pops_1kg), by="sample") %>%
  mutate(pop = ifelse(pop=="hadzabe", "hadza", pop),
         pop = ifelse(pop=="sabue", "chabu", pop)) %>%
  mutate(population = factorize_pops(pop)) %>%
  mutate(PC2 = -PC2)

fig1.pca12.plot = ggplot(pcs, aes(x=PC1, y=PC2, color=population)) +
  geom_point(aes(size=2, alpha=0.8)) +
  theme_bw(base_size=12) +
  # scale_color_calc() +
  coord_fixed() +
  scale_color_manual(values=pop_palette) +
  scale_x_continuous(breaks = seq(-0.2, 0.2, 0.1)) +
  scale_size_continuous(guide=F) +
  scale_alpha_continuous(guide=F) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  # ggtitle("B") +
  theme(plot.title = element_text(size = 30, vjust=3))


## supplemtal_plot

pcs.1_to_4.df = rbind(pcs %>%
                        select(population, coord1 = PC1, coord2 = PC2) %>%
                        mutate(PC1 = "PC1", PC2 = "PC2"),
                      pcs %>%
                        select(population, coord1 = PC1, coord2 = PC3) %>%
                        mutate(PC1 = "PC1", PC2 = "PC3"),
                      pcs %>%
                        select(population, coord1 = PC1, coord2 = PC4) %>%
                        mutate(PC1 = "PC1", PC2 = "PC4"),
                      pcs %>%
                        select(population, coord1 = PC2, coord2 = PC3) %>%
                        mutate(PC1 = "PC2", PC2 = "PC3"),
                      pcs %>%
                        select(population, coord1 = PC2, coord2 = PC4) %>%
                        mutate(PC1 = "PC2", PC2 = "PC4"),
                      pcs %>%
                        select(population, coord1 = PC3, coord2 = PC4) %>%
                        mutate(PC1 = "PC3", PC2 = "PC4")) %>%
  mutate(PC1 = factor(PC1, levels = paste0("PC", 1:4)),
         PC2 = factor(PC2, levels = paste0("PC", 4:1)))

pca1_to_4.plot = ggplot(pcs.1_to_4.df, aes(x=coord2, y=coord1, color=population)) +
  geom_point(aes(alpha=0.8), size=2) +
  facet_grid(PC2 ~ PC1, scales="free", switch="y") + 
  scale_x_continuous(breaks = seq(-0.2, 0.2, 0.1)) +
  scale_y_continuous(breaks = seq(-0.2, 0.2, 0.1)) +
  scale_color_manual(values=pop_palette) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  theme_bw(base_size=12) +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        strip.text.y = element_text(size = 12))
  # scale_color_calc() +
  coord_fixed() +
  scale_color_manual(values=pop_palette) +
  scale_size_continuous(guide=F) +
  scale_alpha_continuous(guide=F) +
  # ggtitle("B") +
  theme(plot.title = element_text(size = 30, vjust=3))

fig1.pca34.plot = ggplot(pcs, aes(x=PC3, y=PC4, color=population)) +
  geom_point(aes(size=2, alpha=0.8)) +
  theme_bw(base_size=12) +
  # scale_color_calc() +
  coord_fixed() +
  scale_color_manual(values=pop_palette) +
  scale_x_continuous(breaks = seq(-0.2, 0.2, 0.1)) +
  scale_size_continuous(guide=F) +
  scale_alpha_continuous(guide=F) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  # ggtitle("B") +
  theme(plot.title = element_text(size = 30, vjust=3))


#### Run my own PCA ####
samps_keep = readLines("../pca/5M_3deg_1kg_20YRI_CEU_CHB.txt")

sub_cols = readLines("../pca/5M_1kg_merged.intersection.rna_samps.noATGCsnps.MAC2.R2.99.sub.cols.txt")

sub_genos = read_tsv("../pca/5M_1kg_merged.intersection.rna_samps.noATGCsnps.MAC2.R2.99.sub.counts.txt.gz",
                     col_names = sub_cols)

sub_genos.pca = prcomp(t(as.matrix(sub_genos[,samps_keep])), center=T, scale=T, retx=T)

all_genos.pca = predict(sub_genos.pca, t(as.matrix(sub_genos[,-c(1:5)])))

all_genos.pcs = as.data.frame(all_genos.pca) %>%
  rownames_to_column("sample") %>%
  merge(., rbind(pop_samps, pops_1kg), by="sample") %>%
  mutate(population = factorize_pops(pop)) %>%
  mutate(PC2 = -PC2)

fig1.sub_genos.pc12.plot = ggplot(all_genos.pcs, aes(x=PC1, y=PC2, color=population)) +
  geom_point(aes(size=2, alpha=0.8)) +
  theme_bw(base_size=12) +
  # scale_color_calc() +
  coord_fixed() +
  scale_color_manual(values=pop_palette) +
  # scale_x_continuous(breaks = seq(-0.2, 0.2, 0.1)) +
  scale_size_continuous(guide="none") +
  scale_alpha_continuous(guide="none") +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  # ggtitle("B") +
  theme(plot.title = element_text(size = 30, vjust=3))

fig1.sub_genos.pc34.plot = ggplot(sub_genos.pcs, aes(x=PC3, y=PC4, color=population)) +
  geom_point(aes(size=2, alpha=0.8)) +
  theme_bw(base_size=12) +
  # scale_color_calc() +
  coord_fixed() +
  scale_color_manual(values=pop_palette) +
  # scale_x_continuous(breaks = seq(-0.2, 0.2, 0.1)) +
  scale_size_continuous(guide=F) +
  scale_alpha_continuous(guide=F) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  # ggtitle("B") +
  theme(plot.title = element_text(size = 30, vjust=3))


#########################################################
####################### ADMIXTURE #######################
#########################################################


admix_pops = read_table("../admixture/5M_3deg_1kg_merged.intersection.rna_samps.noATGCsnps.MAC2.R2.99.ld_prune.fam",
                        col_names = c("sample",1:5)) %>%
  select(sample)

admix_pops = merge(admix_pops, rbind(pop_samps, pops_1kg), sort=F)

Q = data.frame()

for (i in 2:8){
  
  Q.tmp = read_table(paste0("../admixture/5M_3deg_1kg_merged.intersection.rna_samps.noATGCsnps.MAC2.R2.99.ld_prune.",i,".Q"),
                 col_names=paste0("K",1:i))
  Q.tmp$sample = admix_pops$sample
  Q.tmp$pop    = admix_pops$pop
  Q.tmp$k = i
  Q = Q.tmp %>%
    gather(key=K, value=fraction, -c(sample, pop, k)) %>%
    rbind(Q, .)
}

## I need to reassign "K"s so that colors match up. I'm going to use k-means 
## clustering to cluster the colors based on their patterns across individuals
Q.wide = Q %>%
  unite("K", k, K) %>% # make each "K" unique and keep track of which run they come from
  pivot_wider(names_from="sample", values_from="fraction", -pop)

Q.k = Q.wide %>%
  select(-K) %>%
  as.matrix %>%
  kmeans(centers=3)

admix_palette = pop_palette
names(admix_palette) = paste0("K", 1:12)
sadmix_palette = admix_palette[c(6,4,9,7,2,10,1,3,5,8,11,12)]
names(sadmix_palette) = paste0("K", 1:12)

sfig1.admix.plot = Q.wide %>%
  separate(K, into=c("nK", "K"), sep="_") %>%           # separate nK and K
  mutate(nK = paste0("K=", nK)) %>%
  # mutate(K = paste0("K", Q.k$cluster),
  mutate(K = factor(K, levels=paste0("K", 1:12)), # add new K assignments and make a factor
         nK = factor(nK, levels=paste0("K=",c(2:12)))) %>%           # make a factor for plotting
  gather(key="sample", value="fraction", -c(nK, K)) %>% # reshape
  merge(admix_pops) %>%                                 # add population data
  mutate(population = factorize_pops(pop)) %>%
  ggplot(aes(x=sample, y=fraction, fill=K, color=K)) +           # plot
  geom_bar(stat="identity", position="stack") +
  theme_void(base_size=12) +
  facet_grid(nK ~ population, scales="free_x") +
  theme(panel.spacing = unit(0.1, "lines"),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        strip.background = element_blank(),
        plot.margin = unit(c(0,10,0,10), units="points"),
        legend.position = "none") +
  # ggtitle("C") +
  # theme(aspect.ratio=1.8) +
  scale_fill_manual(values=sadmix_palette) +
  scale_color_manual(values=sadmix_palette)
  
nK6_palette = pop_palette[c("YRI","CHB","Chabu","Sandawe","CEU","Hadza")]
names(nK6_palette) = paste0("K", 1:6)

# only K=6
fig1.admix.plot = Q.wide %>%
  # separate nK and K
  separate(K, into=c("nK", "K"), sep="_") %>%
  # restrict to K=6
  filter(nK == 6) %>%
  # mutate(K = paste0("K", Q.k$cluster),
  # add new K assignments and make a factor
  mutate(K = factor(K, levels=paste0("K", 1:12)),
         # make a factor for plotting
         nK = factor(nK, levels=c(2))) %>%
  # reshape
  gather(key="sample", value="fraction", -c(nK, K)) %>%
  # add population data
  merge(admix_pops) %>%
  mutate(population = factorize_pops(pop)) %>%
  # plot
  ggplot(aes(x=sample, y=fraction, fill=K, color=K)) +
  geom_bar(stat="identity", position="stack") +
  theme_void(base_size=12) +
  facet_grid(nK ~ population, scales="free_x", switch="x") +
  theme(panel.spacing = unit(0.1, "lines"),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        strip.background = element_blank(),
        strip.text.y = element_blank(),
        plot.margin = unit(c(0,10,0,10), units="points"),
        legend.position = "none") +
  scale_fill_manual(values=nK6_palette) +
  scale_color_manual(values=nK6_palette) +
  # ggtitle("C") +
  theme(aspect.ratio=1.8)


#### combine plots ####
  
fig1.top = plot_grid(fig1.map.plot, fig1.pca.plot,
                     labels = c('A', 'B'),
                     label_size = 20,
                     rel_widths = c(2, 3))

fig1 = plot_grid(fig1.top,
                 fig1.admix.plot,
                 labels = c('', 'C'), label_size = 20,
                 ncol = 1,
                 rel_heights = c(3,2))

ggsave("fig1.061521.svg", fig1)

```


## Figure 2

```{r figure 2}

#### enrichment near target feature ####

# TSS enrichment of top eQTLs
gencode.v19.tss = read_tsv("../genomes/gencode.v19.genes.v7.patched_contigs.pc_linc.bed.gz",
                           col_names=c("chr","start","end","strand","v19_id","type")) %>%
  mutate(tss = ifelse(strand=="+", start, end),
         chr = rmchr(chr)) %>%
  select(chr, tss, v19_id)


set.seed(42)
gemma.fdr05.top = read_tsv("../gemma/gemma.egenes.tsv.gz",
                        col_types="cnncnccnnnnnnnnnnccnlln") %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1)

fig2.eqtl_tss = gemma.fdr05.top %>%
  merge(gencode.v19.tss) %>%
  mutate(tss_dist = (pos_hg19 - tss)/1000) %>%
  ggplot(aes(x=tss_dist)) +
  geom_histogram(bins=50, alpha=0.6) +
  geom_vline(aes(xintercept=0), color="red", linetype="dashed", size=0.7) +
  theme_classic(base_size=12) +
  xlab("eQTL distance from TSS (kb)") #+
  #coord_fixed(ratio=1000)
  


# Splicing enrichment of top sQTLs

library(cowplot)

# protein_coding and lncRNA genes
gencode.pc_linc = read_tsv("../genomes/gencode.v19.genes.v7.patched_contigs.pc_linc.bed.gz",
                           col_names=c("chr","start","end","strand","gene_id","gene_type")) %>%
  mutate(gene_id = strip_ensembl(gene_id),
         chr = rmchr(chr))

set.seed(42)
lc.fdr05.top = read_tsv("../leafcutter/gemma/lc.sgenes.tsv.gz",
                        col_types=c("ccnnnccnnnnnnnnnnccnlln")) %>%
  filter(top_ssnp) %>%
  group_by(intron) %>%
  sample_n(1) %>%
  separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T) %>%
  separate(intron, into=c("chr1","start_hg19","end_hg19","clust"), sep=":", convert=T) %>%
  unite("intron", chr1, start_hg19, end_hg19, clust, sep=".", remove=F) %>%
  mutate(v19_id1 = v19_id) %>%
  unite("afr_intron", chr1, start_hg19, end_hg19, v19_id1, sep="_") %>%
  mutate(sqtl_hg19 = paste(intron, pos_hg19, sep=":"))
  

# variant distribution plotting
lc.top.plot.df = lc.fdr05.top %>%
  separate(intron, into=c("chr","intron_start","intron_end","clust"), sep="\\.", convert=T) %>%
  merge(gencode.pc_linc) %>%
  mutate(five_prime_dist = ifelse(strand=="+",
                                  pos_hg19 - intron_start,
                                  intron_end - pos_hg19),
         three_prime_dist = ifelse(strand=="+",
                                   pos_hg19 - intron_end,
                                   intron_start - pos_hg19))

lc.top.upstream.hist = lc.top.plot.df %>%
  filter(five_prime_dist < 0) %>%
  ggplot(., aes(x=five_prime_dist/1000)) +
  geom_histogram(bins=20, alpha=0.6) + 
  theme_classic(base_size=12) + 
  theme(axis.line.y=element_blank(),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) + 
  xlab("5' distance (kb)")

lc.top.intron.hist = lc.top.plot.df %>%
  filter(five_prime_dist >= 0 & three_prime_dist <= 0) %>%
  mutate(intron_dist = five_prime_dist / (five_prime_dist - three_prime_dist)) %>%
  ggplot(., aes(x=intron_dist)) +
  geom_histogram(breaks = 0:10/10, alpha=0.6) +
  theme_classic(base_size=12) + 
  theme(axis.line.y=element_blank(),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
  xlab("intron fraction")

lc.top.downstream.hist = lc.top.plot.df %>%
  filter(three_prime_dist > 0) %>%
  ggplot(., aes(x=three_prime_dist/1000)) +
  geom_histogram(bins=20, alpha=0.6) + 
  theme_classic(base_size=12) + 
  theme(axis.line.y=element_blank(),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) + 
  xlab("3' distance (kb)")

fig2.sqtl_tss = plot_grid(lc.top.upstream.hist,
                          lc.top.intron.hist,
                          lc.top.downstream.hist,
                          nrow=1)




#### functional enrichment ####

# functional enrichment of eQTLs and sQTLs
library(ggforestplot)

func_enrich = read_tsv("../gemma/eqtl_sqtl.func_enrich.w_post_0.1.merge_start_stop.txt")

cats = c("3' UTR","5' UTR","Intron","Splice","Synonymous","Missense","Start|Stop","TSS","Enh","ReprPC","Tx","Het","caQTL","TFBS")
names(cats) = c("3_prime_UTR","5_prime_UTR","intron","splice","synonymous","missense","start|stop","Tss","Enh","ReprPC","Tx","Het","caqtl","tfbs")

func_enrich = func_enrich %>%
  mutate(category = cats[category])

func_cat = c("3' UTR", "5' UTR", "Intron", "Splice", "Synonymous", "Missense", "Start|Stop")
epi_cat = c("TSS","Enh","ReprPC","Tx","Het")



fig2.func_vert = func_enrich %>%
  mutate(category = factor(category, levels=c(epi_cat, func_cat, "caQTL", "TFBS")),
         type = ifelse(grepl("eQTL", class), "eQTL", "sQTL"),
         type = factor(type, levels = c("eQTL", "sQTL")),
         set = ifelse(grepl("post", class), "post_gt_0.1", "all"),
         gw = factor(ifelse(as.numeric(category) %% 2 == 1, 1, 0))) %>%
  ggplot() +
  geom_stripes(aes(y=category), odd = "#11111111", even = "#00000000") +
  geom_vline(aes(xintercept=0), linetype="dashed", color="gray") +
  geom_pointrange(aes(y=category,
                      x=log2(or_median),
                      xmin=log2(or_025),
                      xmax=log2(or_975),
                      color=type,
                      shape=set),
                  position = position_dodge2(width=0.75, preserve = "single")) +
  theme_bw(base_size=10) +
  theme(panel.grid.major.y = element_blank()) +
  scale_fill_manual(values = c("white", "grey50")) + 
  scale_color_ptol() +
  xlab("log2(Enrich)") + 
  ggtitle("Functional context of tQTLs")


# make horizontal plot
fig2.func_horz = func_enrich %>%
  mutate(category = factor(category, levels=c("caQTL", "TFBS", epi_cat, func_cat)),
         type = ifelse(grepl("eQTL", class), "eQTL", "sQTL"),
         type = factor(type, levels = c("eQTL", "sQTL")),
         set = ifelse(grepl("post", class), "PP > 0.1", "FDR < 0.05")) %>%
  ggplot() +
  geom_hline(aes(yintercept=0), linetype="dashed", color="gray") +
  geom_pointrange(aes(y=log2(or_median),
                      x=category,
                      ymin=log2(or_025),
                      ymax=log2(or_975),
                      color=type,
                      shape=set),
                  position = position_dodge(width = 0.7)) +
  geom_rect(data=data.frame(x1=c(0:6 * 2 + 0.5),
                       x2=c(0:6 * 2 + 1.5),
                       y1=rep(-5, 7),
                       y2=rep(5, 7)),
            aes(xmin=x1, xmax=x2, ymin=y1, ymax=y2),
            fill="#11111111") +
  coord_cartesian(xlim = c(1,14), ylim = c(-2.75, 3.25)) +
  theme_bw(base_size=15) +
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust=1),
        axis.title.x=element_blank(),
        panel.grid.major.x = element_blank()) +
  scale_color_ptol() +
  ylab(expression(~ log[2](Enrich)))# +
  # ggtitle("Functional context of tQTLs")


#### combine plots ####

fig2.top = plot_grid(fig2.eqtl_tss, fig2.sqtl_tss,
                     labels = c('A', 'B'),
                     label_size = 20,
                     rel_widths = c(2,3),
                     scale = 0.9)

fig2 = plot_grid(fig2.top,
                 fig2.func_horz,
                 labels = c('', 'C'), label_size = 20,
                 ncol = 1,
                 scale=0.9)

ggsave("fig2.041921.svg", fig2)

```


## Figure 3

```{r}

#### maf ####

# Look at GTEx v8 MAF for eQTLs and sQTLs with MAF < 0.01
gtex.v7.egenes = read_tsv("../gemma/gtex.v7.egenes.tsv.gz")

gtex.v8.egenes = read_tsv("../gemma/gtex.v8.egenes.tsv.gz") %>%
  mutate(gene_id = strip_ensembl(v26_id))



eqtl_maf.df = liftover_snp(addchr(gemma.fdr05.top$chr),
                           gemma.fdr05.top$pos_hg19, 
             "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old,
         pos_hg38 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  unique %>%
  merge(gemma.fdr05.top %>% select(gene_id, chr, pos_hg19)) %>%
  select(gene_id, pos_hg38) %>%
  merge(gtex.v8.egenes) %>%
  select(gtexv8_maf, gtexv8_pval)

fig3.eqtl_maf = eqtl_maf.df %>%
  mutate(thresh = factor(gtexv8_pval < 0.01, levels=c(T,F))) %>%
  ggplot(aes(x=gtexv8_maf, color=thresh)) +
  stat_ecdf(size=1.2) +
  theme_classic(base_size=15) +
  coord_fixed(ratio = 0.5) +
  scale_color_ptol(expression(~ p[v8] < 0.01)) +
  xlab(expression(~ MAF[v8])) +
  ylab(expression(~ Fn(MAF[v8]))) +
  ggtitle("eQTL")


## p1
gemma.fdr05 = gemma.egenes %>%
  filter(fdr_sig)

eqtl_v7_pval.df = gemma.fdr05 %>%
  select(gene_id, chr, pos_hg19) %>%
  merge(gtex.v7.egenes) %>%
  select(gtexv7_maf, gtexv7_pval)

eqtl_v8_pval.df = liftover_snp(addchr(gemma.fdr05$chr),
                            gemma.fdr05$pos_hg19, 
                            "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old,
         pos_hg38 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  unique %>%
  merge(gemma.fdr05 %>% select(gene_id, chr, pos_hg19)) %>%
  select(gene_id, pos_hg38) %>%
  merge(gtex.v8.egenes) %>%
  select(gtexv8_maf, gtexv8_pval)


## splicing
lc.sgenes.gtex = read_tsv("../leafcutter/gemma/lc.sgenes.gtex.tsv.gz")

sqtl_maf.df = liftover_snp(addchr(lc.fdr05.top$chr),
             lc.fdr05.top$pos_hg19,
             "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old,
         pos_hg38 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  unique %>%
  merge(lc.fdr05.top) %>%
  merge(intron_map) %>%
  select(gtexv8_intron, pos_hg38) %>%
  merge(lc.sgenes.gtex, all.X=T) %>%
  select(gtexv8_maf, gtexv8_pval)

fig3.sqtl_maf = sqtl_maf.df %>%
  mutate(thresh = factor(gtexv8_pval < 0.01, levels=c(T,F))) %>%
  ggplot(aes(x=gtexv8_maf, color=thresh)) +
  stat_ecdf(size=1.2) +
  theme_classic(base_size=15) +
  coord_fixed(ratio = 0.5) +
  scale_color_ptol(expression(~ p[v8] < 0.01)) +
  xlab(expression(~ MAF[v8])) +
  ylab(expression(~ Fn(MAF[v8]))) +
  ggtitle("sQTL")

## p1
intron_map = read_tsv("../leafcutter/gemma/intron_map.tsv")

hg19_to_hg38.s = read_tsv("../leafcutter/gemma/hg19_to_hg38.sgenes.tsv.gz") %>%
  unique

sqtl_pval.df =  lc.sgenes %>%
  filter(fdr_sig) %>%
  select(intron, chr, pos_hg19) %>%
  merge(hg19_to_hg38.s) %>%
  merge(intron_map) %>%
  merge(lc.sgenes.gtex) %>%
  select(gtexv8_maf, gtexv8_pval)


fig3S.sqtl_pval = sqtl_pval.df %>%
  ggplot(aes(x=gtexv8_pval)) +
  geom_histogram(alpha=0.8) +
  theme_classic(base_size=15) +
  xlab("GTEx v8 p-value") +
  annotate("text", x=0.5, y=25000, parse=T, label = as.character(expression(~pi[1] == 0.91)), size=5) +
  ggtitle("sQTL")

## combine pi1 plots
fig3S = rbind(eqtl_v7_pval.df %>%
                rename(maf = gtexv7_maf, p.value = gtexv7_pval) %>%
                mutate(class = "GTEx v7 eQTL"),
              eqtl_v8_pval.df %>%
                rename(maf = gtexv8_maf, p.value = gtexv8_pval) %>%
                mutate(class = "GTEx v8 eQTL"),
              sqtl_pval.df %>%
                rename(maf = gtexv8_maf, p.value = gtexv8_pval) %>%
                mutate(class = "GTEx v8 sQTL")) %>%
  ggplot(aes(x=p.value, y=..density..)) +
  geom_histogram(alpha=0.8) +
  theme_classic(base_size=15) +
  xlab("p-value") +
  geom_text(data=data.frame(x = 0.5, y=15, class = c("GTEx v7 eQTL",
                                                        "GTEx v8 eQTL",
                                                        "GTEx v8 sQTL"),
                            label = c(as.character(expression(~pi[1] == 0.83)),
                                      as.character(expression(~pi[1] == 0.88)),
                                      as.character(expression(~pi[1] == 0.91)))),
            aes(x=x, y=y, label=label), parse=T, size=5) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20),
        panel.spacing = unit(2, "lines")) +
  facet_wrap(~ class)

# combine maf plots


# plot together
fig3.maf_diff = rbind(eqtl_maf.df %>% mutate(class = "eQTL"),
      sqtl_maf.df %>% mutate(class = "sQTL")) %>%
  mutate(`GTEx p-value < 0.01` = factor(gtexv8_pval < 0.01, levels=c(T,F))) %>%
  ggplot(aes(x=gtexv8_maf, color=`GTEx p-value < 0.01`)) +
  stat_ecdf() +
  theme_classic(base_size=15) +
  facet_grid(. ~ class) +
  coord_fixed(ratio = 0.5) +
  scale_color_ptol()


#### effect sizes ####

# compare effect sizes, coloring by whether the QTL is independent
gtex.v8.egenes.range = read_tsv("../gemma/gtex.v8.egenes.range.tsv.gz") %>%
  mutate(chr = rmchr(chr))


# independent GTEx eQTLs
gemma.cond_sig = read_tsv("../gemma/gemma.cond_sig.tsv.gz")


# plot effect sizes, coloring by whether the credible sets overlap
set.seed(42)

gemma.beta_plot.df = gemma.egenes %>%
  select(gene_id, chr, pos_hg19, eqtl_hg19, allele0, allele1, p.value, beta) %>%
  # add hg38 info
  merge(hg19_to_hg38) %>%
  # merge with GTEx v8
  merge(gtex.v8.egenes.range) %>%
  # flip alleles
  filter((allele1 == v8_ref & allele0 == v8_alt) |
           (allele0 == v8_ref & allele1 == v8_alt)) %>%
  mutate(gtexv8_slope = ifelse(allele1 == v8_ref &
                                 allele0 == v8_alt,
                               gtexv8_slope,
                               1 - gtexv8_slope)) %>%
  # get the top SNP per gene that is also tested in GTEx
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1) %>%
  mutate(shared = !(eqtl_hg19 %in% gemma.cond_sig$eqtl_hg19)) %>%
  ungroup %>%
  select(gene_id, beta, gtexv8_slope, p.value, gtexv8_pval, shared)
  
fig3.eqtl_beta = gemma.beta_plot.df %>%
  mutate(shared = factor(shared, levels=c(T,F))) %>%
  ggplot(aes(x=beta, y=gtexv8_slope, color=shared)) +
  geom_point(alpha=0.2) +
  geom_smooth(method="lm") +
  coord_fixed() +
  scale_x_continuous(breaks = c(-2,-1,0,1,2)) +
  xlab(expression(~ beta[Afr])) +
  ylab(expression(~ beta[v8])) +
  theme_bw(base_size=15) +
  scale_color_ptol() +
  ggtitle("eQTL")


# check sQTLs
gtex.v8.sgenes.range = read_tsv("../leafcutter/gemma/gtex.v8.sgenes.range.tsv.gz") %>%
  mutate(chr = rmchr(chr))

# independent GTEx sQTLs
lc.cond_sig = read_tsv("../leafcutter/gemma/lc.cond_sig.tsv.gz") %>%
  mutate(intron = gsub(":", ".", intron),
         sqtl_hg19 = paste(intron, pos_hg19, sep=":"))

hg19_to_hg38.s = read_tsv("../leafcutter/gemma/hg19_to_hg38.sgenes.tsv.gz")

intron_map = read_tsv("../leafcutter/gemma/intron_map.tsv")

# plot effect sizes, coloring by whether the african allele is independent
set.seed(42)

lc.beta_plot.df = lc.sgenes %>%
  select(intron, gene_id, chr, pos_hg19, sqtl_hg19, allele0, allele1, p.value, beta) %>%
  # add hg38 info
  merge(hg19_to_hg38.s) %>%
  # add intron info %>%
  merge(intron_map) %>%
  # merge with GTEx v8
  merge(gtex.v8.sgenes.range) %>%
  # flip alleles
  filter((allele1 == v8_ref & allele0 == v8_alt) |
           (allele0 == v8_ref & allele1 == v8_alt)) %>%
  mutate(gtexv8_slope = ifelse(allele1 == v8_ref &
                                 allele0 == v8_alt,
                               gtexv8_slope,
                               1 - gtexv8_slope)) %>%
  # get the top SNP per gene that is also tested in GTEx
  group_by(intron) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1) %>%
  mutate(shared = !(sqtl_hg19 %in% lc.cond_sig$sqtl_hg19)) %>%
  ungroup %>%
  select(intron, gene_id, beta, gtexv8_slope, p.value, gtexv8_pval, shared)

fig3.sqtl_beta = lc.beta_plot.df %>%
  mutate(shared = factor(shared, levels=c(T,F))) %>%
  ggplot(aes(x=beta, y=gtexv8_slope, color=shared)) +
  geom_point(alpha=0.2, size=0.8) +
  geom_smooth(method="lm") +
  # geom_text_repel(data = lc.beta_plot.df %>%
  #                   filter(gene_id == "ENSG00000198502") %>%
  #                   merge(gene_name),
  #                 aes(x=beta, y=gtexv8_slope, label=gene_name),
  #                 color="black",
  #                 nudge_y = 1) +
  xlab(expression(~ beta[Afr])) +
  ylab(expression(~ beta[v8])) +
  theme_bw(base_size=15) +
  coord_fixed() +
  scale_color_ptol() +
  ggtitle("sQTL")
  

#### combine plots ####

fig3.top.plots = plot_grid(fig3.eqtl_maf +
                       theme(legend.position="none"),
                     fig3.sqtl_maf +
                       theme(legend.position="none") +
                       ylab(NULL),
                     align="vh")

fig3.top.legend = get_legend(fig3.eqtl_maf + theme(legend.box.margin = margin(0, 0, 0, 0)))

# add legend back
fig3.top = plot_grid(fig3.top.plots, fig3.top.legend, rel_widths = c(2,0.5))


## bottom
fig3.bot.plots = plot_grid(fig3.eqtl_beta +
                             theme(legend.position="none"),
                           fig3.sqtl_beta +
                             theme(legend.position="none"),
                           align="vh")

fig3.bot.legend = get_legend(fig3.eqtl_beta + theme(legend.box.margin = margin(0, 10, 0, 0)))

fig3.bottom = plot_grid(fig3.bot.plots, fig3.bot.legend, rel_widths = c(1,0.2))

fig3 = plot_grid(fig3.top,
                 fig3.bottom,
                 labels = c('A', 'B'), label_size = 20,
                 ncol = 1,
                 rel_widths = c(4,3))

ggsave("fig3.052021.svg", fig3)




# both plotted together
gemma.lc.beta_plot.df = rbind(gemma.beta_plot.df %>%
                                mutate(class = "eQTL"),
                              lc.beta_plot.df %>%
                                mutate(class = "sQTL")) %>%
  mutate(shared = factor(shared, levels=c(T,F)))

gemma.lc.beta_plot.df %>%
  ggplot(aes(x=beta, y=gtexv8_slope, color=shared)) +
  geom_point(alpha=0.5, size=0.7) +
  geom_smooth(method="lm") +
  geom_text_repel(data = gemma.lc.beta_plot.df %>%
                    filter(gene_id == "ENSG00000198502") %>%
                    merge(gene_name) %>%
                    mutate(class="sQTL"),
                  aes(x=beta, y=gtexv8_slope, label=gene_name)) +
  theme_bw(base_size=15) +
  facet_wrap(~ class, scales="free") +
  scale_color_ptol()

# test if there is a significant interaction between independence and effect size correlation
gemma.lc.beta_plot.lm = lm(gtexv8_slope ~ beta + independent + beta * independent,
                           gemma.lc.beta_plot.df)

```


### look at distribution of eQTLs p-values for EUR MAF < 0.01

```{r}

# eQTL file and column names
gemma.f = "../gemma/round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.txt.gz"
gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# read in and filter raw eQTL data
gemma.pvals = fread(gemma.f, col.names=gemma_cols, na.string=".") %>%
  filter(r2 > 0.3 | is.na(r2)) %>% # remove poorly imputed SNPs
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", convert=T) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  filter(gene_id %in% gencode.pc_linc$gene_id) %>%
  select(chr, pos_hg19, gene_id, p.value)

gemma.pvals.eur_af = tabix2df("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/misc/1kgenomes.eur_af.txt.gz",
                         unique(tabixify(gemma.pvals$chr,
                                         gemma.pvals$pos_hg19,
                                         gemma.pvals$pos_hg19)),
                         c("chr","pos_hg19","ref","alt","eur_af")) %>%
  merge(gemma.pvals)

rm(gemma.pvals)


# sQTL info
lc_cols = c("intron","snp_hg19","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2","v19_id")

lc.f = "../leafcutter/gemma/round2/Sample_ALL.secondpass.leafcutter.covars.peer5.resid.gemma.100kb.lmm4.r2.wgene.txt.gz"

lc.pvals = fread(lc.f, col.names=lc_cols, na.strings=c(".","?")) %>%
  filter(!is.na(v19_id)) %>%
  filter(af >= 0.05 & af <= 0.95) %>% # SNPs with MAF >= 0.05
  filter(r2 > 0.3 | is.na(r2)) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  filter(gene_id %in% gencode.pc_linc$gene_id) %>%
  separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T) %>%
  select(chr, pos_hg19, intron, gene_id, p.value)

lc.pvals.eur_af = tabix2df("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/misc/1kgenomes.eur_af.txt.gz",
                         unique(tabixify(lc.pvals$chr,
                                         lc.pvals$pos_hg19,
                                         lc.pvals$pos_hg19)),
                         c("chr","pos_hg19","ref","alt","eur_af")) %>%
  merge(lc.pvals)

rm(lc.pvals)


# eur162 eQTLs
eur162.f = "../gtex/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz"
eur162.cols = c("snp_hg38","v29_id","chr","pos_hg38","eur162_ref","eur162_alt",
                     "version","tss_dist","eur162_pval","eur162_slope","eur162_maf")
gencode29 = read_tsv("../genomes/gencode.v29.genes.bed",
                     col_names=c("chr","start","end","strand","v29_id"))

eur162.pvals = fread(eur162.f, col.names=eur162.cols) %>%
  merge(., gencode29 %>% select(-chr), by="v29_id") %>%
  mutate(pos_hg38 = as.numeric(pos_hg38)) %>%
  mutate(eur162_dist = ifelse(strand=="+", pos_hg38 - start, end - pos_hg38)) %>%
  filter(abs(eur162_dist) < 100000) %>%
  mutate(gene_id = strip_ensembl(v29_id)) %>%
  select(chr, pos_hg38, gene_id, eur162_pval)

eur162.pvals.hg19 = liftover_snp(eur162.pvals$chr,
                                 eur162.pvals$pos_hg38,
                                 "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  unique %>%
  mutate(chr = rmchr(chr)) %>%
  rename(pos_hg19 = pos_new, pos_hg38 = pos_old)

eur162.pvals.afr_af = tabix2df("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/misc/1kgenomes.afr_af.txt.gz",
                         unique(tabixify(eur162.pvals.hg19$chr,
                                         eur162.pvals.hg19$pos_hg19,
                                         eur162.pvals.hg19$pos_hg19)),
                         c("chr","pos_hg19","ref","alt","afr_af")) %>%
  merge(eur162.pvals.hg19) %>%
  merge(eur162.pvals %>% mutate(chr = rmchr(chr)))

rm(eur162.pvals)

```


### test r2 between top SNP in Africans and top SNP in Europeans

```{r}

set.seed(42)
gemma.fdr05.top = gemma.egenes %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1)

egenes_tss = gencode.v19.tss %>%
  filter(v19_id %in% gemma.egenes$v19_id) %>%
  merge(gene_map)

egenes.range.1kg_pos = tabix2df("/local1/derek/data/1kgenomes/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz",
                           tabixify(rmchr(egenes_tss$chr),
                                    egenes_tss$tss - 100000,
                                    egenes_tss$tss + 100000),
                           c("chr","pos_hg19","rsid","ref","alt","qual","filt","info")) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":"))

set.seed(42)
gemma.top.in_1kg = gemma.egenes %>%
  filter(snp_hg19 %in% egenes.range.1kg_pos$snp_hg19) %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1)

afr_eld = read_tsv("../ld/round_061721/AFR.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.eqtl.ld")
eur_eld = read_tsv("../ld/round_061721/EUR.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.eqtl.ld")

afr_eur_eld = merge(afr_eld %>% rename(afr_r2 = r2),
                    eur_eld %>% rename(eur_r2 = r2))

snp_emap = merge(gemma.top.in_1kg %>%
                  select(chr, pos_hg19, gene_id),
                afr_eur_eld %>%
                  select(chr = chr_a, pos_hg19 = pos_a, rsid = snp_a) %>%
                  unique)

afr_eur_eld.cor = afr_eur_eld %>%
  group_by(snp_a) %>%
  summarise(r2_cor = cor(afr_r2, eur_r2)) %>%
  merge(snp_emap %>% select(snp_a = rsid, pos_hg19, gene_id))

afr_eur_eld.jaccard = afr_eur_eld %>%
  group_by(snp_a) %>%
  summarise(r2_jaccard = sum(afr_r2>0.5 & eur_r2>0.5)/sum(afr_r2>0.5 | eur_r2>0.5)) %>%
  merge(snp_map %>% select(snp_a = rsid, gene_id))

gtex.v8.egenes.sig_snp_gene = gtex.v8.egenes %>%
  filter(gtexv8_pval < 0.01) %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  merge(hg19_to_hg38) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  pull(eqtl_hg19)

afr_eur_eld.cor %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  mutate(independent = !(eqtl_hg19 %in% gtex.v8.egenes.sig_snp_gene)) %>%
  ggplot(aes(x = r2_cor, color=independent)) +
  stat_ecdf() +
  theme_classic()

afr_eur_ld.jaccard %>%
  mutate(shared = !(gene_id %in% gemma.cond_sig$gene_id)) %>%
  ggplot(aes(x = r2_jaccard, color=shared)) +
  stat_ecdf() +
  theme_classic()


## sQTL

set.seed(42)
lc.fdr05.top = lc.sgenes %>%
  filter(top_ssnp) %>%
  group_by(intron) %>%
  sample_n(1)

sintrons_range = lc.fdr05.top %>%
  separate(afr_intron, into = c("chr","start","end","v19_id"), sep="_", convert=T) %>%
  select(chr, start, end)

sintrons.range.1kg_pos = tabix2df("/local1/derek/data/1kgenomes/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz",
                           tabixify(sintrons_range$chr,
                                    sintrons_range$start - 100000,
                                    sintrons_range$end + 100000),
                           c("chr","pos_hg19","rsid","ref","alt","qual","filt","info")) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":"))

set.seed(42)
lc.top.in_1kg = lc.sgenes %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  filter(snp_hg19 %in% sintrons.range.1kg_pos$snp_hg19) %>%
  group_by(intron) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1)

afr_sld = read_tsv("../ld/round_061721/AFR.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.sqtl.ld")
eur_sld = read_tsv("../ld/round_061721/EUR.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.sqtl.ld")

afr_eur_sld = merge(afr_sld %>% rename(afr_r2 = r2),
                    eur_sld %>% rename(eur_r2 = r2))

rm(afr_sld)
rm(eur_sld)

snp_smap = merge(lc.top.in_1kg %>%
                   select(chr, pos_hg19, intron, gene_id),
                 afr_eur_sld %>%
                   select(chr = chr_a, pos_hg19 = pos_a, rsid = snp_a) %>%
                   unique)

afr_eur_sld.cor = afr_eur_sld %>%
  group_by(snp_a) %>%
  summarise(r2_cor = cor(afr_r2, eur_r2)) %>%
  merge(snp_smap %>% select(snp_a = rsid, pos_hg19, intron, gene_id))

afr_eur_sld.jaccard = afr_eur_sld %>%
  group_by(snp_a) %>%
  summarise(r2_jaccard = sum(afr_r2>0.2 & eur_r2>0.2)/sum(afr_r2>0.2 | eur_r2>0.2)) %>%
  merge(snp_map %>% select(snp_a = rsid, pos_hg19, gene_id))

gtex.v8.sgenes.sig_snp_intron = lc.sgenes.gtex %>%
  filter(gtexv8_pval < 0.01) %>%
  merge(intron_map) %>%
  merge(hg19_to_hg38.s) %>%
  mutate(sqtl_hg19 = paste(intron, pos_hg19, sep=":")) %>%
  pull(sqtl_hg19)

afr_eur_sld.cor %>%
  mutate(sqtl_hg19 = paste(intron, pos_hg19, sep=":")) %>%
  mutate(independent = !(sqtl_hg19 %in% gtex.v8.sgenes.sig_snp_intron)) %>%
  ggplot(aes(x = r2_cor, color=independent)) +
  stat_ecdf() +
  theme_classic()

afr_eur_sld.jaccard %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  mutate(shared = !(eqtl_hg19 %in% gemma.cond_sig$eqtl_hg19)) %>%
  ggplot(aes(x = r2_jaccard, color=shared)) +
  stat_ecdf() +
  theme_classic()


#### plot together ####
r2cor.plot = rbind(afr_eur_eld.cor %>%
        mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
        mutate(shared = !(eqtl_hg19 %in% gemma.cond_sig$eqtl_hg19),
               class = "eQTL") %>%
        select(-eqtl_hg19),
      afr_eur_sld.cor %>%
        mutate(sqtl_hg19 = paste(intron, pos_hg19, sep=":")) %>%
        mutate(shared = !(sqtl_hg19 %in% lc.cond_sig$sqtl_hg19),
               class = "sQTL") %>%
        select(-c(sqtl_hg19, intron))) %>%
  mutate(shared = factor(shared, levels=c(T,F))) %>%
  ggplot(aes(x = r2_cor, color = shared)) +
  stat_ecdf() +
  theme_classic(base_size=15) +
  scale_color_ptol() +
  xlab(expression(rho*"("*r[AFR]^{2}*","*r[EUR]^{2}*")")) +
  theme(axis.title.y=element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  facet_wrap(~ class)


deltamaf.plot = rbind(gemma.fdr05.top.eur_af %>%
                        mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"),
                               shared = !(eqtl_hg19 %in% gemma.cond_sig$eqtl_hg19),
                               class = "eQTL") %>%
                        select(-eqtl_hg19),
                      lc.fdr05.top.eur_af %>%
                        mutate(sqtl_hg19 = paste(intron, pos_hg19, sep=":"),
                               shared = !(sqtl_hg19 %in% lc.cond_sig$sqtl_hg19),
                               class = "sQTL") %>%
                        select(-c(intron, sqtl_hg19))) %>%
  mutate(shared = factor(shared, levels=c(T,F))) %>%
  ggplot(aes(x = abs(af - eur_af), color = shared)) +
  stat_ecdf() +
  theme_classic(base_size=15) +
  scale_color_ptol() +
  xlab(expression("|"*Delta*"AF"*"|")) +
  theme(axis.title.y=element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  facet_wrap(~ class)


deltamaf_r2cor.plot = plot_grid(deltamaf.plot,
                              r2cor.plot,
                              ncol=1,
                              labels=c("A","B"))

```


Compare haplotypes between African and CEU samples

```{r}

foo = gemma.fdr05.top %>%
  filter(gene_id=="ENSG00000187049") %>%
  mutate(range = tabixify(chr, pos_hg19 - 10000, pos_hg19 + 10000)) %>%
  pull(range)

hap_samps = read_table("../haplo/5M_1kg_merged.intersection.rna_ceu.ColReordered.full.noATGCsnps.MAF01.R2.99.sample", skip=2, col_names = c("sample","foo","missing")) %>%
  select(sample)

ceu_samps = read_tsv("/local1/derek/data/1kgenomes/1khg_sample_pop.tsv", col_names = c("sample","pop")) %>%
  filter(pop=="CEU") %>%
  pull(sample)

hap_samps.rna_ceu = hap_samps %>%
  filter(sample %in% c(pop_samps$sample, ceu_samps))

hap_cols = c("chr","id","pos_hg19","ref","alt",
             expand.grid(c(1,2), hap_samps$sample) %>%
               mutate(col = paste(Var2, Var1, sep="_")) %>%
               pull(col))

haplo = range2df("../haplo/5M_1kg_merged.intersection.rna_ceu.ColReordered.full.noATGCsnps.MAF01.R2.99.tab.hap.gz",
                 foo,
                 hap_cols)

ref_alt = haplo[haplo$pos_hg19==61209133,-c(1:5)] == 1

```


## Figure 4

```{r}

gencode29 = read_tsv("../genomes/gencode.v29.genes.bed",
                     col_names=c("chr","start","end","strand","v29_id"))
gemma.egenes = read_tsv("../gemma/gemma.egenes.tsv.gz",
                        col_types="cnncnccnnnnnnnnnnccnlln")

# read in EUR162 data
eur162.egenes = read_tsv("../gemma/eur162.egenes.070721.tsv.gz")

eur162.egenes = liftover_snp(eur162.egenes$chr,
                             eur162.egenes$pos_hg38,
                             "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  dplyr::rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  unique %>%
  merge(eur162.egenes, all.y=T) %>%
  mutate(chr = rmchr(chr))

#### Compare credible set sizes between Africans and 162 European Americans from GTEx ####

afr_tested = read_tsv("../misc/afr_tested_genes.txt", col_names="gene_id") %>%
  mutate(gene_id = strip_ensembl(gene_id)) %>%
  pull(gene_id)
eur162_tested = read_tsv("../misc/eur162_tested_genes.txt", col_names="gene_id") %>%
  filter(gene_id %in% gencode29.pc_linc$v29_id) %>%
  mutate(gene_id = strip_ensembl(gene_id)) %>%
  pull(gene_id)
afr_eur162_tested = intersect(afr_tested, eur162_tested)

# egenes with an FDR-significant egene in either study
afr_eur162_any_sig = gene_map %>%
  filter((v19_id %in% gemma.egenes$v19_id |
           v29_id %in% eur162.egenes$v29_id) &
           gene_id %in% afr_eur162_tested) %>%
  pull(gene_id)

# egenes with an FDR-significant egenes in both studies
afr_eur162_both_sig = gemma.egenes %>%
  select(gene_id) %>% unique %>% # get all unique gene ids
  merge(eur162.egenes %>% select(gene_id) %>% unique) %>%
  filter(gene_id %in% afr_eur162_any_sig) %>%
  pull(gene_id) %>% unique

gemma.sort.cols = c("chr","pos_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

gemma.any_sig = tabix2df("../gemma/round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.sort.txt.gz",
                         tabixify(gene_map %>%
                                    filter(gene_id %in% afr_eur162_any_sig) %>%
                                    pull(v19_id),
                                  "0", "1000000000"),
                         gemma.sort.cols) %>%
  select(-c(foo, bar)) %>%
  filter(r2 > 0.3) %>% # remove poorly imputed SNPs
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":")) %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af))


eur162.any_sig = tabix2df("../gtex/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz",
                          tabixify(gene_map %>%
                                    filter(gene_id %in% afr_eur162_any_sig) %>%
                                    pull(v29_id),
                                  "0", "1000000000"),
                          c("snp_hg38","v29_id","chr","pos_hg38","eur162_ref","eur162_alt",
                            "version","tss_dist","eur162_pval","eur162_slope","eur162_maf")) %>%
  select(-c(snp_hg38, version, tss_dist)) %>%
  merge(., gencode29 %>% select(-chr), by="v29_id") %>%
  mutate(pos_hg38 = as.numeric(pos_hg38)) %>%
  mutate(eur162_dist = ifelse(strand=="+", pos_hg38 - start, end - pos_hg38)) %>%
  select(-c(start, end)) %>%
  filter(abs(eur162_dist) < 100000) %>%
  mutate(gene_id = strip_ensembl(v29_id)) %>%
  mutate(eqtl_hg38 = paste(gene_id, pos_hg38, sep=":"))


# assign genes to shared, afr-specific, and eur162-specific
afr_eur162_class = data.frame(gene_id = afr_eur162_any_sig) %>%
  mutate(afr_sig = gene_id %in% gemma.egenes$gene_id,
         eur162_sig = gene_id %in% eur162.egenes$gene_id) %>%
  group_by(gene_id) %>%
  summarise(afr_eur162_sig = interaction(afr_sig, eur162_sig))

# venn diagram
shared_go = gost(query = afr_eur162_class %>%
                   filter(afr_eur162_sig=="TRUE.TRUE") %>%
                   pull(gene_id),
                 custom_bg = afr_eur162_class$gene_id)

afr_go = gost(query = afr_eur162_class %>%
                filter(afr_eur162_sig=="TRUE.FALSE") %>%
                pull(gene_id),
              custom_bg = afr_eur162_class$gene_id)

eur162_go = gost(query = afr_eur162_class %>%
                filter(afr_eur162_sig=="FALSE.TRUE") %>%
                pull(gene_id),
              custom_bg = afr_eur162_class$gene_id)

# Do African-specific eQTLs have higher MAF?
gemma.egenes %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  merge(gemma.egenes.eur_af) %>%
  merge(afr_eur162_class) %>%
  mutate(eur_maf = ifelse(eur_af < 0.5, eur_af, 1-eur_af)) %>%
  ggplot(aes(x=maf, y=eur_maf, color=afr_eur162_sig)) +
  geom_point() +
  theme_bw()

gemma.any_sig %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1) %>%
  merge(hg19_to_hg38) %>%
  merge(eur162.any_sig %>% select(gene_id, pos_hg38, eur162_slope)) %>%
  merge(afr_eur162_class) %>%
  ggplot(aes(x=beta, y=eur162_slope, color=afr_eur162_sig)) +
  geom_point() +
  theme_bw()
  


# get credible sets
gemma.any_sig.cred_set = lapply(afr_eur162_any_sig,
                                function(x){
                                  df.tmp = gemma.any_sig %>%
                                    filter(gene_id==x)
                                  get_cred_set.p(df.tmp$eqtl_hg19,
                                                 df.tmp$p.value,
                                                 df.tmp$maf, 162, 0.9)})

eur162.any_sig.cred_set = lapply(afr_eur162_any_sig,
                                 function(x){
                                   df.tmp = eur162.any_sig %>%
                                     filter(gene_id==x)
                                   get_cred_set.p(df.tmp$eqtl_hg38,
                                                  df.tmp$eur162_pval,
                                                  df.tmp$eur162_maf, 162, 0.9)})


gemma.any_sig.long = data.frame(afr_cred_set = unlist(gemma.any_sig.cred_set)) %>%
  separate(afr_cred_set, into=c("gene_id","pos_hg19"), convert=T, remove=F) %>%
  merge(gemma.any_sig %>% select(chr, pos_hg19, gene_id) %>% unique)


gemma.eur162.any_sig.intersect = liftover_snp(addchr(gemma.any_sig.long$chr),
                                              gemma.any_sig.long$pos_hg19,
                                              "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  dplyr::rename(pos_hg19 = pos_old, pos_hg38 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(gemma.any_sig.long, all=T) %>%
  merge(data.frame(eur162_cred_set = unlist(eur162.any_sig.cred_set)) %>%
          separate(eur162_cred_set, into=c("gene_id","pos_hg38"), convert=T, remove=F),
        all=T)

gemma.eur162.any_sig.overlap = gemma.eur162.any_sig.intersect %>%
  group_by(gene_id) %>%
  summarise(afr_size = sum(!is.na(afr_cred_set)),
            eur162_size = sum(!is.na(eur162_cred_set)),
            intersect = sum(!(is.na(afr_cred_set) | is.na(eur162_cred_set))),
            smaller_set = ifelse(afr_size < eur162_size, afr_size, eur162_size),
            count = n())

library(scales)

breaks <- 10^(0:4)
minor_breaks <- rep(1:9, 5)*(10^rep(0:4, each=9))

# plot credible set sizes for genes with an eQTL in both datasets
fig4.cred_sets = gemma.eur162.any_sig.overlap %>%
  # filter(gene_id %in% afr_eur162_both_sig) %>%
  mutate(frac_uniq = intersect/smaller_set) %>%
  sample_frac(1L) %>%
  ggplot(aes(x=afr_size,
             y=eur162_size,
             color=frac_uniq)) +
  geom_abline(aes(intercept=0, slope=1), linetype="dashed", alpha=0.8) +
  geom_point(alpha=0.5, size=2) +
  scale_x_log10(breaks = breaks, minor_breaks = minor_breaks) +
  scale_y_log10(breaks = breaks, minor_breaks = minor_breaks) +
  annotation_logticks() +
  coord_fixed() +
  scale_color_distiller(palette = "Spectral", direction=1) +
  xlab(expression(paste("|",CS[Afr],"|"))) +
  ylab(expression(paste("|",CS[EA162],"|"))) +
  theme_bw(base_size=15) +
  labs(color = expression(frac(paste("|",A,intersect(B),"|"), paste("|",A,"|"))))


gemma.eur162.any_sig.overlap %>%
  select(afr_size, eur162_size) %>%
  gather(key=group, value=cred_set) %>%
  ggplot(aes(x=cred_set, color=group)) +
  stat_ecdf() +
  scale_x_log10(breaks = breaks, minor_breaks = minor_breaks) +
  theme_bw(base_size=15)


## TSS e
gencode29 = read_tsv("../genomes/gencode.v29.genes.bed",
                     col_names=c("chr","start","end","strand","v29_id")) %>%
  mutate(gene_id = strip_ensembl(v29_id))

set.seed(42)
eur162.fdr05.top.tss_dist = eur162.egenes %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  merge(gencode29 %>%
          select(gene_id, strand, start, end)) %>%
  mutate(tss_dist = ifelse(strand=="+", pos_hg38 - start, end - pos_hg38)) %>%
  select(gene_id, tss_dist)
  
gemma.fdr05.top.tss_dist = gemma.fdr05.top %>%
  merge(gencode.pc_linc %>%
          select(gene_id, strand, start, end)) %>%
  mutate(tss_dist = ifelse(strand=="+", pos_hg19 - start, end - pos_hg19)) %>%
  select(gene_id, tss_dist)

rbind(eur162.fdr05.top.tss_dist %>% mutate(group="ea162"),
      gemma.fdr05.top.tss_dist %>% mutate(group="afr")) %>%
  ggplot(aes(x=tss_dist, color=group)) +
  geom_freqpoly() +
  theme_classic()




#### Find tQTLs with improved fine mapping in eAFR ####

# eQTLs
gtex.v8.cred_set.all = lapply(unique(gtex.v8.egenes.range$gene_id),
                              function(x){
                                df.tmp = gtex.v8.egenes.range %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg38,
                                               df.tmp$gtexv8_slope,
                                               df.tmp$gtexv8_se^2, 0.9)})
names(gtex.v8.cred_set.all) = unique(gtex.v8.egenes.range$gene_id)

# sQTLs
gtex.v8.cred_set.all.s = lapply(unique(gtex.v8.sgenes.range$gtexv8_intron),
                              function(x){
                                df.tmp = gtex.v8.sgenes.range %>%
                                  filter(gtexv8_intron==x)
                                get_cred_set.b(df.tmp$sqtl_hg38,
                                               df.tmp$gtexv8_slope,
                                               df.tmp$gtexv8_se^2, 0.9)})
names(gtex.v8.cred_set.all.s) = unique(gtex.v8.sgenes.range$gtexv8_intron)

# get African credible sets
gemma.cred_set.all = lapply(unique(gemma.egenes$gene_id),
                              function(x){
                                df.tmp = gemma.egenes %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg19,
                                               df.tmp$beta,
                                               df.tmp$se^2, 0.9)})
names(gemma.cred_set.all) = unique(gemma.egenes$gene_id)

lc.cred_set.all = lapply(unique(lc.sgenes$intron),
                              function(x){
                                df.tmp = lc.sgenes %>%
                                  filter(intron==x)
                                get_cred_set.b(df.tmp$sqtl_hg19,
                                               df.tmp$beta,
                                               df.tmp$se^2, 0.9)})
names(lc.cred_set.all) = unique(lc.sgenes$intron)


# combine eQTL credible set info from Africans and GTEx
gemma.v8.cred_set = data.frame(cred_set = unlist(gemma.cred_set.all)) %>%
  # separate into gene ID and position columns
  separate(cred_set, into=c("gene_id", "pos_hg19"), sep=":", convert=T) %>%
  # add chromosome information
  merge(gemma.egenes %>% select(gene_id, pos_hg19, chr)) %>%
  # add hg38 position information, including only those SNPs that map between hg19 and hg38
  merge(hg19_to_hg38, all.x=T) %>%
  # specify that this is in the African credible set
  mutate(afr=T) %>%
  # add GTEx credible set info
  merge(data.frame(cred_set = unlist(gtex.v8.cred_set.all)) %>%
          # separate as above
          separate(cred_set, into=c("gene_id", "pos_hg38"), sep=":", convert=T) %>%
          # specify these are in the GTEx v8 credible set
          mutate(v8=T),
        # eQTLs (gene_id-pos_hg38 pairs) that overlap will merge, those that don't will not
        all=T) %>%
  unique


# combine sQTLcredible set info from Africans and GTEx
lc.v8.cred_set = data.frame(cred_set = unlist(lc.cred_set.all)) %>%
  # separate into gene ID and position columns
  separate(cred_set, into=c("intron", "pos_hg19"), sep=":", convert=T) %>%
  # # add chromosome information
  # merge(
  # add hg38 position information, including only those SNPs that map between hg19 and hg38
  merge(hg19_to_hg38.s, all.x=T) %>%
  # only introns tested in both
  merge(intron_map) %>%
  unique %>%
  # specify that this is in the African credible set
  mutate(afr=T) %>%
  # add GTEx credible set info
  merge(data.frame(cred_set = unlist(gtex.v8.cred_set.all.s)) %>%
          # separate as above
          separate(cred_set, into=c("gtexv8_intron", "pos_hg38"), sep=":", convert=T) %>%
          # specify these are in the GTEx v8 credible set
          mutate(v8=T),
        # sQTLs (intron-pos_hg38 pairs) that overlap will merge, those that don't will not
        all=T) %>%
  unique

# get genes whose credible sets overlap
gemma.v8.cred_set.overlap = gemma.v8.cred_set %>%
  group_by(gene_id) %>%
  summarise(afr_size = sum(!is.na(afr)),
            v8_size = sum(!is.na(v8)),
            intersect = sum(!(is.na(afr) | is.na(v8))),
            smaller_set = ifelse(afr_size < v8_size, afr_size, v8_size),
            count = n())

lc.v8.cred_set.overlap = lc.v8.cred_set %>%
  group_by(gtexv8_intron) %>%
  summarise(afr_size = sum(!is.na(afr)),
            v8_size = sum(!is.na(v8)),
            intersect = sum(!(is.na(afr) | is.na(v8))),
            smaller_set = ifelse(afr_size < v8_size, afr_size, v8_size),
            count = n())

# get GWAS catalogue
gwas_cat = read_tsv("/local1/derek/data/gwas/catalog_011421/gwas_catalog_v1.0.2-associations_e100_r2020-12-15.unnest.tsv") %>%
  filter(chr %in% as.character(1:22)) %>%
  mutate(chr = as.numeric(chr))

# add hg19 info
gwas_cat = liftover_snp(addchr(gwas_cat$chr), gwas_cat$pos_hg38,
                        "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  dplyr::rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(gwas_cat)

gemma.egenes.eur_af = read_tsv("../gemma/gemma.egenes.eur_af.tsv.gz")

# get genes mapped to 1 variant and overlapping a GWAS SNP
gemma.v8.1_snp_gwas = gemma.v8.cred_set.overlap %>%
  filter(afr_size==1) %>%
  merge(gemma.fdr05.top %>%
          select(gene_id, chr, pos_hg19, af)) %>%
  merge(gemma.egenes.eur_af) %>% 
  merge(gwas_cat %>% select(chr, pos_hg19, pos_hg38, trait, pubmedid, strongest_snp, gwas_pval)) %>%
  merge(gene_name) %>%
  filter(afr_size < v8_size & afr_size==1) %>%
  unique

gemma.v8.afr_lt_v8_gwas = gemma.v8.cred_set.overlap %>%
  filter(afr_size < v8_size) %>%
  merge(gemma.fdr05.top %>%
          select(gene_id, chr, pos_hg19, af)) %>%
  merge(gemma.egenes.eur_af) %>% 
  merge(gwas_cat %>% select(chr, pos_hg19, pos_hg38, trait, pubmedid, strongest_snp, gwas_pval)) %>%
  merge(gene_name) %>%
  unique

lc.v8.1_snp_gwas = lc.v8.cred_set.overlap %>%
  filter(afr_size==1) %>%
  merge(lc.fdr05.top %>%
          merge(intron_map) %>%
          select(gtexv8_intron, gene_id, chr, pos_hg19, af)) %>%
  merge(gwas_cat %>% select(chr, pos_hg19, pos_hg38, trait, pubmedid, strongest_snp, gwas_pval)) %>%
  merge(gene_name) %>%
  filter(afr_size < v8_size & afr_size==1) %>%
  unique

top_snp_info = gemma.v8.1_snp_gwas %>%
  select(strongest_snp) %>%
  separate(strongest_snp, into=c("rsid","allele"), sep="-") %>%
  pull(rsid) %>%
  ncbi_snp_query

top_snp_info.s = lc.v8.1_snp_gwas %>%
  select(strongest_snp) %>%
  separate(strongest_snp, into=c("rsid","allele"), sep="-") %>%
  pull(rsid) %>%
  ncbi_snp_query

gemma.v8.1_snp_gwas = top_snp_info %>%
  select(rsid = query,
         chr = chromosome,
         pos_hg38_top = bp) %>%
  merge(gemma.v8.1_snp_gwas %>%
          separate(strongest_snp, into=c("rsid","foo"), sep="-")) %>%
  unique

lc.v8.1_snp_gwas = top_snp_info.s %>%
  select(rsid = query,
         chr = chromosome,
         pos_hg38_top = bp) %>%
  merge(lc.v8.1_snp_gwas %>%
          separate(strongest_snp, into=c("rsid","foo"), sep="-")) %>%
  unique

write.table(gemma.v8.1_snp_gwas, "gemma.v8.1_snp_gwas.csv", sep=",", quote=F, row.names=F)
write.table(lc.v8.1_snp_gwas, "lc.v8.1_snp_gwas.csv", sep=",", quote=F, row.names=F)


# plot all of these genes
gemma.v8.1_snp_gwas.loci = gemma.egenes %>%
  filter(gene_id %in% gemma.v8.1_snp_gwas$gene_id) %>%
  merge(hg19_to_hg38) %>%
  select(gene_id, pos_hg38, afr_pval = p.value) %>%
  merge(gtex.v8.egenes %>% select(gene_id, pos_hg38, gtexv8_pval)) %>%
  merge(gene_name)
  
ggplot() +
  geom_point(data=gemma.v8.1_snp_gwas.loci, aes(x=pos_hg38, y=-log10(afr_pval)), color="blue") +
  geom_point(data=gemma.v8.1_snp_gwas.loci, aes(x=pos_hg38, y=log10(gtexv8_pval)), color="red") +
  theme_classic() + 
  facet_wrap(~ gene_name, scales="free")


# create separate plots for each gene
gemma.v8.1_snp_gwas.plots = list()
for (x in unique(gemma.v8.1_snp_gwas$gene_id)){
  
  plot.df = gemma.egenes %>%
    filter(gene_id==x) %>%
    merge(hg19_to_hg38) %>%
    select(gene_id, pos_hg38, afr_pval = p.value) %>%
    merge(gtex.v8.egenes %>% select(gene_id, pos_hg38, gtexv8_pval))
  
  name.tmp = gene_name %>%
    filter(gene_id==x) %>%
    pull(gene_name)
  
  gemma.v8.1_snp_gwas.plots[[x]] = ggplot() +
    geom_point(data=plot.df, aes(x=pos_hg38, y=-log10(afr_pval)), color="blue") +
    geom_point(data=plot.df, aes(x=pos_hg38, y=log10(gtexv8_pval)), color="red") +
    theme_classic() +
    ggtitle(name.tmp)
  
}


lc.v8.1_snp_gwas.plots = list()
for (x in unique(lc.v8.1_snp_gwas$gtexv8_intron)){
  
  plot.df = gtex.v8.sgenes.range %>%
    filter(gtexv8_intron==x) %>%
    select(gtexv8_intron, pos_hg38, gtexv8_pval) %>%
    merge(hg19_to_hg38.s) %>%
    merge(intron_map) %>%
    merge(lc.sgenes %>% select(intron, pos_hg19, afr_pval = p.value))
  
  name.tmp = lc.v8.1_snp_gwas %>%
    filter(gtexv8_intron==x) %>%
    pull(gene_name)
  
  lc.v8.1_snp_gwas.plots[[x]] = ggplot() +
    geom_point(data=plot.df, aes(x=pos_hg38, y=-log10(afr_pval)), color="blue") +
    geom_point(data=plot.df, aes(x=pos_hg38, y=log10(gtexv8_pval)), color="red") +
    theme_classic() +
    ggtitle(name.tmp)
  
}
  

#### colocalization between eAfr and GTEx v8
  



#### TLR1 plot ####


# generate plots for TLR1
tlr1_
  filter(gene_id=="ENSG00000174125") %>%
  mutate(cred_set = eqtl_hg38 %in% gtex.v8.cred_set.all[["ENSG00000174125"]]) %>%
  ggplot(aes(x=pos_hg38, y=-log10(gtexv8_pval), color=cred_set)) +
  geom_point() +
  xlab("Genomic Position") +
  ylab(expression(paste(-log[10](P[v8])))) +
  theme_classic(base_size=15) +
  theme(legend.position="none") +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  ylim(10,0)

tlr1_afr = gemma.egenes %>%
  filter(gene_id=="ENSG00000174125") %>%
  merge(hg19_to_hg38) %>%
  mutate(cred_set = eqtl_hg19 %in% gemma.cred_set.all[["ENSG00000174125"]]) %>%
  ggplot(aes(x=pos_hg38, y=-log10(p.value), color=cred_set)) +
  geom_point() +
  ylab(expression(paste(-log[10](P[Afr])))) +
  theme_classic(base_size=15) +
  theme(legend.position="none",
        axis.line.x=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  ylim(0,10)

# plot together
plot_grid(tlr1_afr, tlr1_eur162, ncol = 1, align = 'v')

# colocalization plot
tlr1_afr_v_v8 = gemma.egenes %>%
  filter(gene_id=="ENSG00000174125") %>%
  merge(hg19_to_hg38) %>%
  select(gene_id, pos_hg38, p.value) %>%
  merge(gtex.v8.egenes.range %>% select(gene_id, pos_hg38, gtexv8_pval)) %>%
  ggplot(aes(x=-log10(p.value), y=-log10(gtexv8_pval))) +
  geom_point() +
  coord_fixed() +
  theme_classic(base_size=15) +
  xlab(expression(paste(-log[10](P[Afr])))) +
  ylab(expression(paste(-log[10](P[v8])))) +
  scale_color_manual(values="#a1a1a1")
  

#### THRA plot ####

# generate plots for TLR1
thra_eur162 = gtex.v8.egenes.range %>%
  filter(gene_id=="ENSG00000126351") %>%
  mutate(cred_set = eqtl_hg38 %in% gtex.v8.cred_set.all[["ENSG00000126351"]]) %>%
  ggplot(aes(x=pos_hg38, y=-log10(gtexv8_pval), color=cred_set)) +
  geom_point() +
  xlab("Genomic Position") +
  ylab(expression(paste(-log[10](P[v8])))) +
  theme_classic(base_size=15) +
  theme(legend.position="none") +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  scale_x_continuous(expand = c(0,0)) +
  ylim(10,0)

thra_afr = gemma.egenes %>%
  filter(gene_id=="ENSG00000126351") %>%
  merge(hg19_to_hg38) %>%
  mutate(cred_set = eqtl_hg19 %in% gemma.cred_set.all[["ENSG00000126351"]]) %>%
  ggplot(aes(x=pos_hg38, y=-log10(p.value), color=cred_set)) +
  geom_point() +
  ylab(expression(paste(-log[10](P[Afr])))) +
  theme_classic(base_size=15) +
  theme(legend.position="none",
        axis.line.x=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  scale_x_continuous(expand = c(0,0)) +
  ylim(0,10)

# plot together
plot_grid(thra_afr, thra_eur162, ncol = 1, align = 'v')

# colocalization plot
thra_afr_v_v8 = gemma.egenes %>%
  filter(gene_id=="ENSG00000126351") %>%
  merge(hg19_to_hg38) %>%
  select(gene_id, pos_hg38, p.value) %>%
  merge(gtex.v8.egenes.range %>% select(gene_id, pos_hg38, gtexv8_pval)) %>%
  ggplot(aes(x=-log10(p.value), y=-log10(gtexv8_pval))) +
  geom_point() +
  coord_fixed() +
  theme_classic(base_size=15) +
  theme(legend.title = element_blank()) +
  xlab(expression(paste(-log[10](P[Afr])))) +
  ylab(expression(paste(-log[10](P[v8])))) +
  scale_color_manual(values="#a1a1a1")


#### NR1D1 plot ####

# get recombination info
nr1d1_range = gencode.v19.tss %>%
  filter(v19_id == "ENSG00000126368.5") %>%
  mutate(tabix = tabixify(chr, tss-110000, tss+110000)) %>%
  pull(tabix)

nr1d1.ceu_recomb = tabix2df("../recomb/hg19/CEU/CEU_recombination_map_hg19_chr_17.bed.gz",
                            nr1d1_range,
                            c("chr","start","end","rate"))

nr1d1.yri_recomb = tabix2df("../recomb/hg19/YRI/YRI_recombination_map_hg19_chr_17.bed.gz",
                            nr1d1_range,
                            c("chr","start","end","rate"))


# GWAS data

nr1d1.ms = tabix2df("../gwas/discovery_metav3.0.meta.tsv.gz",
                    rmchr(nr1d1_range),
                    c("chr","pos_hg19","rsid","a1","a2","n","gwas_pval","gwas_or"))

gtex.v8.egenes.range.nr1d1 = gtex.v8.egenes.range %>% filter(gene_id=="ENSG00000126368")
gtex.v8.egenes.range.nr1d1 = liftover_snp(addchr(gtex.v8.egenes.range.nr1d1$chr),
                                          gtex.v8.egenes.range.nr1d1$pos_hg38,
                                          "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg19 = pos_new, pos_hg38 = pos_old) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(gtex.v8.egenes.range.nr1d1)
gtex.v8.egenes.range.nr1d1.ms = gtex.v8.egenes.range.nr1d1 %>% merge(nr1d1.ms)

gemma.egenes.nr1d1.ms = gemma.egenes %>%
  filter(gene_id=="ENSG00000126368") %>%
  merge(nr1d1.ms)

# coloc
gemma.egenes.nr1d1.ms.coloc = coloc.abf(dataset1=list(pvalues = gemma.egenes.nr1d1.ms$gwas_pval,
                                                      MAF = gemma.egenes.nr1d1.ms$gwas_maf,
                                                      type="quant",
                                                      N=115803),
                                        dataset2=list(beta = gemma.pigment.df$afr_beta,
                                                      varbeta = gemma.pigment.df$afr_se^2,
                                                      MAF = gemma.pigment.df$afr_maf,
                                                      type="quant",
                                                      N=162))

  

# convert recombination rate bed format to segments for plotting
make_segments = function(df){
  
  rbind(data.frame(x = df$start, y = df$rate,
                   xend = df$end, yend = df$rate),
        data.frame(x = head(df$end,-1), y = head(df$rate,-1),
                   xend = tail(df$start,-1), yend = tail(df$rate,-1)))
}

# used for plotting in scientific notation
scientific_10 <- function(x) {
  parse(text=gsub("e", " %*% 10^", scales::scientific_format()(x)))
}

nr1d1.ceu_recomb.segs = make_segments(nr1d1.ceu_recomb)
nr1d1.yri_recomb.segs = make_segments(nr1d1.yri_recomb)

nr1d1_xlims =  c(gemma.egenes %>% filter(gene_id=="ENSG00000126368") %>% pull(pos_hg19) %>% min,
                 gemma.egenes %>% filter(gene_id=="ENSG00000126368") %>% pull(pos_hg19) %>% max)

gtex.v8.egenes.range.nr1d1 = gtex.v8.egenes.range %>%
  filter(gene_id=="ENSG00000126368")

# generate plots for NR1D1
nr1d1_eur162 = liftover_snp(addchr(gtex.v8.egenes.range.nr1d1$chr),
                            gtex.v8.egenes.range.nr1d1$pos_hg38,
                            "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(gtex.v8.egenes.range.nr1d1) %>%
  mutate(cred_set = eqtl_hg38 %in% gtex.v8.cred_set.all[["ENSG00000126368"]]) %>%
  ggplot(aes(x=pos_hg19, y=-log10(gtexv8_pval))) +
  geom_segment(data=nr1d1.ceu_recomb.segs,
               aes(x=x, y=y*5*10**7, xend=xend, yend=yend*5*10**7),
               color="blue") +
  geom_point(aes(color=cred_set)) +
  # xlab("Genomic Position") +
  ylab(expression(paste(-log[10](P[v8])))) +
  theme_classic(base_size=12) +
  theme(legend.position="none",
        axis.line.x=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  scale_y_continuous(limits = c(10,0), trans = "reverse",
                     sec.axis = sec_axis(~./5/10**7,
                                         name = "Recomb. Rate",
                                         label=scientific_10)) +
  xlim(nr1d1_xlims) +
  scale_x_continuous(expand = c(0,0))
  # ylim(10,0)


nr1d1_afr = gemma.egenes %>%
  filter(gene_id=="ENSG00000126368") %>%
  mutate(cred_set = eqtl_hg19 %in% gemma.cred_set.all[["ENSG00000126368"]]) %>%
  ggplot(aes(x=pos_hg19, y=-log10(p.value))) +
  geom_segment(data=nr1d1.yri_recomb.segs,
               aes(x=x, y=y*5*10**7, xend=xend, yend=yend*5*10**7),
               color="blue") +
  geom_point(aes(color=cred_set)) +
  ylab(expression(paste(-log[10](P[Afr])))) +
  theme_classic(base_size=12) +
  theme(legend.position="none",
        axis.line.x=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  xlim(nr1d1_xlims) +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_continuous(limits = c(0,10),
                     sec.axis = sec_axis(~./5/10**7,
                                         name = "Recomb. Rate",
                                         label=scientific_10))

# plot together
plot_grid(nr1d1_afr, nr1d1_eur162, ncol = 1, align = 'v')



#### combine plots ####

nr1d1_locus = read_image_svg("nr1d1_locus.hg19.svg")
nr1d1_locus = rsvg("nr1d1_locus.hg19.svg", width = 1000)
image_read(nr1d1_locus)

foo = ggdraw() + draw_image(nr1d1_locus)

plot_grid(fig4.cred_sets, foo)
          


###### TNFRSF1A plot ######
tnf_range = gencode.v19.tss %>%
  filter(v19_id == "ENSG00000067182.3") %>%
  mutate(tabix = tabixify(chr, tss-110000, tss+110000)) %>%
  pull(tabix)

tnf.ceu_recomb = tabix2df("../recomb/hg19/CEU/CEU_recombination_map_hg19_chr_12.bed.gz",
                          tnf_range,
                          c("chr","start","end","rate"))

tnf.yri_recomb = tabix2df("../recomb/hg19/YRI/YRI_recombination_map_hg19_chr_12.bed.gz",
                          tnf_range,
                          c("chr","start","end","rate"))


# GWAS data

tnf.ms = tabix2df("../gwas/discovery_metav3.0.meta.tsv.gz",
                  rmchr(tnf_range),
                  c("chr","pos_hg19","rsid","a1","a2","n","gwas_pval","gwas_or"))

tnf.bc = tabix2df("../gwas/cordell_2015_26394269_pbc_efo1001486_1_gwas.sumstats.sort.tsv.gz",
                  rmchr(tnf_range),
                  c("chr","pos_hg19","rsid","other_allele","effect_allele","gwas_pval","gwas_beta","gwas_se","gwas_or","or_lower","or_upper"))


gtex.v8.egenes.range.tnf = gtex.v8.egenes.range %>% filter(gene_id=="ENSG00000067182")
gtex.v8.egenes.range.tnf = liftover_snp(addchr(gtex.v8.egenes.range.tnf$chr),
                                          gtex.v8.egenes.range.tnf$pos_hg38,
                                          "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg19 = pos_new, pos_hg38 = pos_old) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(gtex.v8.egenes.range.tnf)
gtex.v8.egenes.range.tnf.ms = gtex.v8.egenes.range.tnf %>% merge(tnf.ms)
gtex.v8.egenes.range.tnf.bc = gtex.v8.egenes.range.tnf %>% merge(tnf.bc)

gtex.v8.heart.tnf = read_tsv("../gtex/v8/GTEx_Analysis_v8_QTLs_GTEx_Analysis_v8_eQTL_all_associations_Heart_Atrial_Appendage.allpairs.TNFRSF1A.txt.gz", col_names=c("v26_id","variant_id","tss_distance","ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se")) %>%
  separate(variant_id, into=c("chr","pos_hg38","v8_ref","v8_alt","foo")) %>%
  select(-foo) %>%
  mutate(gene_id = strip_ensembl(v26_id))
gtex.v8.heart.tnf.ms = liftover_snp(gtex.v8.heart.tnf$chr,
                                    gtex.v8.heart.tnf$pos_hg38,
                                    "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  unique %>%
  merge(gtex.v8.heart.tnf) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(tnf.ms)

gemma.egenes.tnf.ms = gemma.egenes %>%
  filter(gene_id=="ENSG00000067182") %>%
  merge(tnf.ms)

gemma.egenes.tnf.bc = gemma.egenes %>%
  filter(gene_id=="ENSG00000067182") %>%
  merge(tnf.bc)


## plot TNF
tnf.ceu_recomb.segs = make_segments(tnf.ceu_recomb)
tnf.yri_recomb.segs = make_segments(tnf.yri_recomb)

tnf_xlims =  c(gemma.egenes %>% filter(gene_id=="ENSG00000067182") %>% pull(pos_hg19) %>% min,
                 gemma.egenes %>% filter(gene_id=="ENSG00000067182") %>% pull(pos_hg19) %>% max)

gtex.v8.egenes.range.tnf = gtex.v8.egenes.range %>%
  filter(gene_id=="ENSG00000067182")

# generate plots for tnf
tnf_v8 = liftover_snp(addchr(gtex.v8.egenes.range.tnf$chr),
                            gtex.v8.egenes.range.tnf$pos_hg38,
                            "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(gtex.v8.egenes.range.tnf) %>%
  mutate(cred_set = eqtl_hg38 %in% gtex.v8.cred_set.all[["ENSG00000067182"]]) %>%
  ggplot(aes(x=pos_hg19, y=-log10(gtexv8_pval))) +
  geom_segment(data=tnf.ceu_recomb.segs,
               aes(x=x, y=y*5*10**7, xend=xend, yend=yend*5*10**7),
               color="blue") +
  geom_point(aes(color=cred_set)) +
  # xlab("Genomic Position") +
  ylab(expression(paste(-log[10](P[v8])))) +
  theme_classic(base_size=12) +
  theme(legend.position="none",
        axis.line.x=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  scale_y_continuous(limits = c(13,0), trans = "reverse",
                     sec.axis = sec_axis(~./5/10**7,
                                         name = "Recomb. Rate",
                                         label=scientific_10)) +
  xlim(tnf_xlims) +
  scale_x_continuous(expand = c(0,0))
  # ylim(10,0)


tnf_afr = gemma.egenes %>%
  filter(gene_id=="ENSG00000067182") %>%
  mutate(cred_set = eqtl_hg19 %in% gemma.cred_set.all[["ENSG00000067182"]]) %>%
  ggplot(aes(x=pos_hg19, y=-log10(p.value))) +
  geom_segment(data=tnf.yri_recomb.segs,
               aes(x=x, y=y*5*10**7, xend=xend, yend=yend*5*10**7),
               color="blue") +
  geom_point(aes(color=cred_set)) +
  ylab(expression(paste(-log[10](P[Afr])))) +
  theme_classic(base_size=12) +
  theme(legend.position="none",
        axis.line.x=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B")) +
  xlim(tnf_xlims) +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_continuous(limits = c(0,13),
                     sec.axis = sec_axis(~./5/10**7,
                                         name = "Recomb. Rate",
                                         label=scientific_10))

# plot together
plot_grid(tnf_afr, tnf_v8, ncol = 1, align = 'v')


# read in gene model data
# gencode19 = read_tsv("../genomes/gencode.v19.genes.v7.patched_contigs.gtf.gz", skip=6, col_names=c("chr","source","feature","start","end","score","strand","frame","attribute")) %>%
#   mutate(gene_id = str_extract(attribute, "ENSG[:digit:]+")) %>%
#   merge(gene_name) %>%
#   select(-attribute)


# gencode19.gtex = read_tsv("../genomes/gencode.v19.genes.v7.patched_contigs.exon.bed.gz",
#                           col_names = c("chr","start","end","strand","v19_id")) %>%
#   mutate(model = "exon") %>%
#   rbind(read_tsv("../genomes/gencode.v19.genes.v7.patched_contigs.intron.bed.gz",
#                  col_names = c("chr","start","end","strand","v19_id")) %>%
#                    mutate(model = "gap",
#                           strand = "*")) %>%
#   mutate(gene_id = strip_ensembl(v19_id)) %>%
#   merge(gene_name)
# 
# 
# gencode19.gr = GRanges(
#     seqnames = rmchr(gencode19.gtex$chr),
#     ranges = IRanges(start = gencode19.gtex$start,
#                      end   = gencode19.gtex$end),
#     strand = gencode19.gtex$strand,
#     gene_id = gencode19.gtex$gene_id,
#     gene_name = gencode19.gtex$gene_name,
#     model = gencode19.gtex$model)
# 
# gencode19.grange = split(gencode19.gr, gencode19.gr$gene_name)
# names(gencode19.grange) = unique(gencode19.gr$gene_name)
# rm(gencode19.gr)
# 
# 
# 
# 
# 
# 
# 
# gemma.egenes.ranges = gencode.v19.tss %>%
#   filter(v19_id %in% gemma.egenes$v19_id) %>%
#   group_by(v19_id) %>%
#   summarise(chr = unique(chr),
#             start = tss - 100000,
#             end = tss + 100000) %>%
#   mutate(start = ifelse(start<0, 0, start),
#          tabix_id = tabixify(chr, start, end)) %>%
#   pull(tabix_id)



library(EnsDb.Hsapiens.v75)
ensdb <- EnsDb.Hsapiens.v75

thra.range = gencode.v19.tss %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  dplyr::filter(gene_id=="ENSG00000126351") %>%
  summarise(chr = rmchr(unique(chr)),
            start = tss - 100000,
            end = tss + 100000) %>%
  mutate(start = ifelse(start < 0, 0, start))


thra.gr <- GRanges(seqnames = thra.range$chr,
                   IRanges(thra.range$start, thra.range$end),
                   strand = "*")

autoplot(ensdb, GRangesFilter(thra.gr), names.expr = "gene_name")


gencode19.egenes = tabix2df("/local1/derek/data/genomes/gencode.v19.annotation.sort.gtf.gz",
                     gemma.egenes.ranges,
                     c("chr","source","feature","start","end","score","strand","frame","attribute")) %>%
  mutate(gene_id = str_extract(attribute, "ENSG[:digit:]+"),
         tx_id = str_extract(attribute, "ENST[:digit:]+\\.[:digit:]*"),
         type = str_extract(attribute, "(?<=gene_type \")[:alpha:]+")) %>%
  merge(gene_name) %>%
  dplyr::select(-attribute) %>%
  mutate(feature = tolower(feature)) %>%
  filter(feature %in% c("exon","utr") &
           type %in% c("protein","lincRNA")) %>%
  unique

gencode19.gr = GRanges(
    seqnames = gencode19.egenes$chr,
    ranges = IRanges(start = gencode19.egenes$start,
                     end   = gencode19.egenes$end),
    strand = gencode19.egenes$strand,
    tx_id = gencode19.egenes$tx_id,
    tx_name = gencode19.egenes$gene_name,
    gene_id = gencode19.egenes$gene_id,
    model = gencode19.egenes$feature)

# gencode19.gaps = gaps(gencode19.gr[gencode19.gr$model=="exon",])
# strand(gencode19.gaps) = "*"
# gencode19.gaps
# gencode19.gr = c(gencode19.gr, ))
# gencode19.gr = gencode19.gr[start(gencode19.gr) > 1,]


for (tx_id in unique(gencode19.gr$tx_id)){
  
  print(tx_id)
  
  gene_info.df = gencode19.gr[gencode19.gr$tx_id==tx_id,]
  
  gap.df = gaps(gencode19.gr[gencode19.gr$tx_id==tx_id & 
                               gencode19.gr$model=="exon",])[-1,]
  
  if (length(gap.df) > 0){
    
    gap.df$tx_id = unique(gene_info.df$tx_id)
    gap.df$tx_name = unique(gene_info.df$tx_name)
    gap.df$gene_id = unique(gene_info.df$gene_id)
    strand(gap.df) = "*"
    gap.df$model = "gap"
    
  }
  
  gencode19.gr = c(gencode19.gr,
                       gap.df)
}



gencode19.grange = split(gencode19.gr, gencode19.gr$tx_id)
names(gencode19.grange) = unique(gencode19.gr$tx_id)
rm(gencode19.gr)


gencode19.grange[[1:3]]








ggplot(gemma.v8.cred_set.overlap, aes(x=afr_size, y=v8_size)) +
  geom_abline(aes(intercept=0, slope=1), linetype="dashed", alpha=0.8) +
  geom_point(alpha=0.5, size=2) +
  scale_x_log10(breaks = breaks, minor_breaks = minor_breaks) +
  scale_y_log10(breaks = breaks, minor_breaks = minor_breaks) +
  annotation_logticks() +
  coord_fixed() +
  scale_color_distiller(palette = "Spectral") +
  xlab(expression(paste("|",CS[Afr],"|"))) +
  ylab(expression(paste("|",CS[v8],"|"))) +
  theme_bw(base_size=15) +
  labs(color = expression(frac(paste("|",A,"\\",B,"|"), paste("|",A,"|")))) +
  ggtitle("eQTL credible sets, Africans vs GTEx v8")

```


## Figure 5

```{r figure 5}

clust_gene = read_table2("../leafcutter/gemma/leafcutter_all.clu2gene.txt", na="?",
                        col_names=c("clust","chr","start_hg19","end_hg19","v19_id",paste0("foo",1:4))) %>%
  select(clust, chr, start_hg19, end_hg19, v19_id)

gemma.weighted_pbs = read_tsv("../gemma/gemma.weighted_pbs.tsv")
gemma.weighted_dstat = read_tsv("../gemma/gemma.weighted_dstat.tsv")

lc.weighted_pbs = read_tsv("../gemma/lc.weighted_pbs.tsv")
lc.weighted_dstat = read_tsv("../gemma/lc.weighted_dstat.tsv")

dstat.thresh = read_tsv("../gemma/dstat.thresh.txt")
pbs.thresh = read_tsv("../gemma/pbs.thresh.txt")

#### selection hits ####

gemma.sel_info = merge(gemma.weighted_pbs,
                       gemma.weighted_dstat) %>%
  merge(dstat.thresh) %>%
  merge(pbs.thresh) %>%
  mutate(sig = weighted_pbs > pbs_995 &
           weighted_dstat > dstat_995) %>%
  merge(gene_name) %>%
  mutate(gene_name = ifelse(sig, gene_name, ""),
         pop = pop_map[pop])

lc.sel_info = merge(lc.weighted_pbs,
                    lc.weighted_dstat) %>%
  merge(dstat.thresh) %>%
  merge(pbs.thresh) %>%
  mutate(sig = weighted_pbs > pbs_995 &
           weighted_dstat > dstat_995) %>%
  merge(clust_gene %>%
          unite("intron", chr, start_hg19, end_hg19, clust, sep=".")) %>%
  merge(gene_map %>% select(v19_id, gene_id)) %>%
  merge(gene_name) %>%
  mutate(gene_name = ifelse(sig, gene_name, ""),
         pop = pop_map[pop]) %>%
  group_by(gene_id, pop) %>%
  filter(weighted_dstat==max(weighted_dstat))

## plots ##
gemma.sel_plot = gemma.sel_info %>%
  ggplot(aes(x = weighted_pbs,
             y = weighted_dstat,
             color = sig)) +
  geom_vline(aes(xintercept=pbs_995), linetype="dashed", color="red", alpha=0.5) +
  geom_hline(aes(yintercept=dstat_995), linetype="dashed", color="red", alpha=0.5) +
  geom_point() +
  xlab(expression(~ PBS[gene])) +
  ylab(expression(~ d[gene])) +
  geom_text_repel(aes(x = weighted_pbs,
                      y = weighted_dstat,
                      label = gene_name),
                  size = 3,
                  force = 10,
                  max.overlaps = Inf) +
  theme_bw(base_size=12) +
  coord_fixed(ratio=max(lc.sel_info$weighted_pbs)/max(lc.sel_info$weighted_dstat)) +
  facet_wrap(~ pop) +
  theme(legend.position="none",
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12)) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B"))

lc.sel_plot = lc.sel_info %>%
  ggplot(aes(x = weighted_pbs,
             y = weighted_dstat,
             color = sig)) +
  geom_vline(aes(xintercept=pbs_995), linetype="dashed", color="red", alpha=0.5) +
  geom_hline(aes(yintercept=dstat_995), linetype="dashed", color="red", alpha=0.5) +
  geom_point() +
  xlab(expression(~ PBS[gene])) +
  ylab(expression(~ d[gene])) +
  geom_text_repel(aes(x = weighted_pbs,
                      y = weighted_dstat,
                      label = gene_name),
                  size = 3,
                  force = 10,
                  box.padding = 0.2,
                  max.overlaps = Inf) +
  theme_bw(base_size=12) +
  coord_fixed(ratio=max(lc.sel_info$weighted_pbs)/max(lc.sel_info$weighted_dstat)) +
  facet_wrap(~ pop) +
  theme(legend.position="none",
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12)) +
  scale_color_manual(values=c("#a1a1a1", "#C5000B"))

## write tables
gemma.sel_info %>%
  filter(sig) %>%
  select(pop, gene_id, gene_name, weighted_pbs, weighted_dstat) %>%
  arrange(pop, gene_name) %>%
  write.csv("gemma.sel_info.csv", quote=F, row.names=F)

lc.sel_info %>%
  filter(sig) %>%
  select(pop, intron, gene_id, gene_name, weighted_pbs, weighted_dstat) %>%
  mutate(intron = gsub("\\.",":",intron)) %>%
  arrange(pop, gene_name, intron) %>%
  write.csv("lc.sel_info.csv", quote=F, row.names=F)




#### TMEM216 plot ####

ddb1_range = "11:59625865-62637504"

# gwas_cols = c("chr","pos_hg19","id","gwas_beta","gwas_sd","gwas_pval","gwas_fdr","gwas_pvalgc","gwas_fdrgc","gwas_r2","an","ac","gwas_ref","gwas_alt","hwe","gwas_maf")
# gwas_cols = c("rsid","chr","pos_hg19","gwas_beta","gwas_se","gwas_pval")

pigment = read_tsv("../pigmentation/5M.hg19.FlipRef.DDB1.1Mb.mergedrefs.mm.dose.ps.maf.txt.gz",
                   col_names = c("snp_hg19","gwas_beta","gwas_se","gwas_pval","gwas_maf"),
                   col_types="cnnnn") %>%
  separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T, remove=F)

pigment.sig = pigment %>%
  filter(gwas_pval < 5 * 10^-8)

pigment_genes = gencode.pc_linc %>%
  filter(chr==11) %>%
  filter((start > 59625865 & start < 62637504) |
           (end > 59625865 & end < 62637504)) %>%
  pull(gene_id)

pigment_genes.v26 = gene_map %>%
  filter(gene_id %in% pigment_genes) %>%
  pull(v26_id)
pigment_genes.v26 = pigment_genes.v26[!is.na(pigment_genes.v26)]

ld = read_table("../gemma/tmem216_mapping/11.5M.imputed.ddb1_locus.ld",
                skip=1,
                col_names=c("chr","pos1","foo","chr1","pos2","bar","r2")) %>%
  select(chr,pos1,pos2,r2)

tmem216_v8 = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL/ALL.v8.signif_variant_gene_pairs.sort.txt.gz",
                   "ENSG00000187049.9:0-1000000000",
                   c("chr","pos_hg38","ref","alt","version","v26_id","dist_tss","ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se","v8_pval_thresh","v8_min_pval","v8_pval_beta","tissue"))

# tmem216 eQTLs that are also pigmentation snps
pigment.tmem216_v8 = liftover_snp(addchr(pigment.sig$chr),
             pigment.sig$pos_hg19,
             "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg38 = pos_new, pos_hg19 = pos_old) %>%
  merge(tmem216_v8)

# get GTEx v8 data
v8_wb_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Whole_Blood.allpairs.sort.txt.gz",
                         paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                         c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                           "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))


v8_se_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Skin_Sun_Exposed_Lower_leg.allpairs.sort.txt.gz",
                      paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                      c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))


#### combine plots ####
fig5.top = plot_grid(gemma.sel_plot,
                 lc.sel_plot,
                 nrow = 1,
                 labels = c('A', 'B'),
                 label_size = 20)








# plot sQTL
mica_nonsyn = c("6:31378358","6:31378425","6:31378977","6:31379109","6:31379795","6:31379807","6:31379823","6:31380198")

intron3_snp = "6:31378977"

mica_gtypes = tabix2df("../geno/5M.imputed.rna_samps.biallelic.vcf.gz",
                      "6:31271356-31471356",
                      c("chr","pos_hg19","snp_hg19","ref","alt","qual","filter","info","format",samps)) %>%
  select(-c(1,2,4:9)) %>%
  gather(key=sample, value=genotype, -snp_hg19) %>%
  separate(genotype, into=c("gtype","dose"), sep=":", convert=T) %>%
  separate(gtype, into=c("allele1", "allele2"), sep="\\|", convert=T)


snp = "6:31378650-31378650"
intron1 = "6:31378574:31378849:clu_34541_NA"
intron2 = "6:31378574:31378986:clu_34541_NA"



sqtl_plot = function(intron, snp){
  
  gtype = tabix2df("../geno/5M.imputed.rna_samps.biallelic.vcf.gz", snp,
                      c("chr","pos_hg19","snp_hg19","ref","alt","qual","filter","info","format",samps)) %>%
    select(-c(1:9)) %>%
    gather(key=sample, value=genotype) %>%
    separate(genotype, into=c("gtype","dose"), sep=":", convert=T)
  
  int_exp = lc.norm %>%
    filter(ID==intron) %>%
    select(-c(1:4)) %>%
    gather(key=sample, value=int_exp)
  
  merge(gtype, int_exp) %>%
    ggplot(aes(x=dose, y=int_exp)) +
    geom_point() +
    theme_classic(base_size=15)
  
}

```

## Figure 6

# colocalize v8 eQTL and pigmentation

```{r}

library(arrow)

v8_wb_eur_tmem216 = read_parquet("../gtex/v8/EUR_eQTL/GTEx_Analysis_v8_QTLs_GTEx_Analysis_v8_EUR_eQTL_all_associations_Whole_Blood.v8.EUR.allpairs.chr11.parquet") %>%
  filter(phenotype_id=="ENSG00000187049.9") %>%
  separate(variant_id, into=c("chr","pos_hg38","v8_ref","v8_alt","version"), sep="_")
  
v8_se_eur_tmem216 = read_parquet("../gtex/v8/EUR_eQTL/GTEx_Analysis_v8_QTLs_GTEx_Analysis_v8_EUR_eQTL_all_associations_Skin_Sun_Exposed_Lower_leg.v8.EUR.allpairs.chr11.parquet") %>%
  filter(phenotype_id=="ENSG00000187049.9") %>%
  separate(variant_id, into=c("chr","pos_hg38","v8_ref","v8_alt","version"), sep="_")

# add hg19 info
v8_wb_eur_tmem216 = liftover_snp(v8_wb_eur_tmem216$chr,
                                 v8_wb_eur_tmem216$pos_hg38,
                                 "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  dplyr::rename(pos_hg19 = pos_new, pos_hg38 = pos_old) %>%
  merge(v8_wb_eur_tmem216) 

v8_se_eur_tmem216 = liftover_snp(v8_se_eur_tmem216$chr,
             v8_se_eur_tmem216$pos_hg38,
             "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  dplyr::rename(pos_hg19 = pos_new, pos_hg38 = pos_old) %>%
  merge(v8_se_eur_tmem216)

#### TMEM216 plots ####
pigment_genes = gencode.pc_linc %>%
  filter(chr=="chr11") %>%
  filter((start > 59625865 & start < 62637504) |
           (end > 59625865 & end < 62637504)) %>%
  pull(gene_id)

pigment_genes.v26 = gene_map %>%
  filter(gene_id %in% pigment_genes) %>%
  pull(v26_id)
pigment_genes.v26 = pigment_genes.v26[!is.na(pigment_genes.v26)]

#### get data for TMEM216 plot ####

## get genetype information for calculating r2

cols_1kg = readLines("/local1/derek/data/1kgenomes/vcf/1kg_cols.txt")
pop_samps.1kg = read_tsv("/local1/derek/data/1kgenomes/1khg_sample_pop_superpop.tsv", col_names=c("sample","pop","superpop")) %>%
  filter(sample %in% cols_1kg)

gtype_tmem216.eur.df = tabix2df("/local1/derek/data/1kgenomes/vcf/ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz",
                              "11:61029745-61233624",
                              cols_1kg) %>%
  unite("snp_hg19", CHROM, POS, sep=":") %>%
  select(snp_hg19, ID, pop_samps.1kg %>% filter(superpop=="EUR") %>% pull(sample)) %>%
  gather(key=sample, value=gtype, -c(snp_hg19, ID)) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  mutate(all_cnt = gtype1 + gtype2) %>%
  select(-c(gtype1, gtype2)) %>%
  group_by(ID) %>%
  filter(mean(all_cnt) > 0 & mean(all_cnt) < 2) %>%
  ungroup %>%
  pivot_wider(names_from=sample, values_from=all_cnt)

gtype_tmem216.afr.df = tabix2df("/local1/derek/data/1kgenomes/vcf/ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz",
                              "11:61029745-61233624",
                              cols_1kg) %>%
  unite("snp_hg19", CHROM, POS, sep=":") %>%
  select(snp_hg19, ID, pop_samps.1kg %>% filter(superpop=="AFR") %>% pull(sample)) %>%
  gather(key=sample, value=gtype, -c(snp_hg19, ID)) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  mutate(all_cnt = gtype1 + gtype2) %>%
  select(-c(gtype1, gtype2)) %>%
  group_by(ID) %>%
  filter(mean(all_cnt) > 0 & mean(all_cnt) < 2) %>%
  ungroup %>%
  pivot_wider(names_from=sample, values_from=all_cnt)

cols_5M.rna_pops  = readLines("/local1/derek/data/Illumina5M/5M_cols.rna_pops.txt")
gtype_tmem216.eafr.df = tabix2df("/local1/derek/data/Illumina5M/5M.imputed.rna_pops.vcf.gz",
                              "11:61029745-61233624",
                              cols_5M.rna_pops) %>%
  select(snp_hg19 = ID, pop_samps.5M$sample) %>%
  gather(key=sample, value=genotype, -snp_hg19) %>%
  separate(genotype, into=c("gtype","dose"), sep=":", convert=T) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  mutate(all_cnt = gtype1 + gtype2) %>%
  select(-c(gtype1, gtype2, dose)) %>%
  pivot_wider(names_from=sample, values_from=all_cnt)

## calculate r2 for select SNPs
tmem_rsmap = data.frame(rsid = c("rs7948623","rs11230664","rs3741265"),
                        snp = c("11:61137147","11:61076372","11:61165280"))
rs7948623 = "11:61137147"
rs11230664 = "11:61076372"
rs3741265 = "11:61165280"

# eur
r2_tmem216.eur.rs7948623 = gtype_tmem216.eur.df %>%
  select(-c(snp_hg19, ID)) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.eur.df$snp_hg19) %>%
  filter(snp==rs7948623) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.eur.df$snp_hg19,
         ID = gtype_tmem216.eur.df$ID)

r2_tmem216.eur.rs11230664 = gtype_tmem216.eur.df %>%
  select(-c(snp_hg19, ID)) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.eur.df$snp_hg19) %>%
  filter(snp==rs11230664) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.eur.df$snp_hg19,
         ID = gtype_tmem216.eur.df$ID)

r2_tmem216.eur.rs3741265 = gtype_tmem216.eur.df %>%
  select(-c(snp_hg19, ID)) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.eur.df$snp_hg19) %>%
  filter(snp==rs3741265) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.eur.df$snp_hg19,
         ID = gtype_tmem216.eur.df$ID)


# afr
r2_tmem216.afr.rs7948623 = gtype_tmem216.afr.df %>%
  select(-c(snp_hg19, ID)) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.afr.df$snp_hg19) %>%
  filter(snp==rs7948623) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.afr.df$snp_hg19,
         ID = gtype_tmem216.afr.df$ID)

r2_tmem216.afr.rs11230664 = gtype_tmem216.afr.df %>%
  select(-c(snp_hg19, ID)) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.afr.df$snp_hg19) %>%
  filter(snp==rs11230664) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.afr.df$snp_hg19,
         ID = gtype_tmem216.afr.df$ID)

r2_tmem216.afr.rs3741265 = gtype_tmem216.afr.df %>%
  select(-c(snp_hg19, ID)) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.afr.df$snp_hg19) %>%
  filter(snp==rs3741265) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.afr.df$snp_hg19,
         ID = gtype_tmem216.afr.df$ID)


# eafr
r2_tmem216.eafr.rs7948623 = gtype_tmem216.eafr.df %>%
  select(-snp_hg19) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.eafr.df$snp_hg19) %>%
  filter(snp==rs7948623) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.eafr.df$snp_hg19)

r2_tmem216.eafr.rs11230664 = gtype_tmem216.eafr.df %>%
  select(-snp_hg19) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.eafr.df$snp_hg19) %>%
  filter(snp==rs11230664) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.eafr.df$snp_hg19)

r2_tmem216.eafr.rs3741265 = gtype_tmem216.eafr.df %>%
  select(-snp_hg19) %>%
  as.matrix %>%
  t %>%
  cor %>%
  .**2 %>%
  as.data.frame %>%
  mutate(snp = gtype_tmem216.eafr.df$snp_hg19) %>%
  filter(snp==rs3741265) %>%
  gather(key=snp_hg19, value=r2, -snp) %>%
  mutate(snp_hg19 = gtype_tmem216.eafr.df$snp_hg19)
  

#### read in eQTL data ####
v8_wb_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Whole_Blood.allpairs.sort.txt.gz",
                      paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                      c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))

v8_se_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Skin_Sun_Exposed_Lower_leg.allpairs.sort.txt.gz",
                      paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                      c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))


# create plotting and coloc dfs
tmem216.eur.wb.df = liftover_snp(v8_wb_pigment$chr,
                                 v8_wb_pigment$pos_hg38,
                                 "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  unique %>%
  dplyr::rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  merge(v8_wb_pigment %>%
          select(gene_id = v26_id, chr, pos_hg38, v8_maf, v8_slope, v8_se, v8_pval)) %>%
  mutate(chr = rmchr(chr)) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(pigment)
tmem216.eur.wb.df = rbind(tmem216.eur.wb.df %>%
                            merge(r2_tmem216.eur.rs7948623 %>% select(snp, snp_hg19, r2)),
                          tmem216.eur.wb.df %>%
                            merge(r2_tmem216.eur.rs11230664 %>% select(snp, snp_hg19, r2)),
                          tmem216.eur.wb.df %>%
                            merge(r2_tmem216.eur.rs3741265 %>% select(snp, snp_hg19, r2)))


tmem216.eur.se.df = liftover_snp(v8_se_pigment$chr,
                                 v8_se_pigment$pos_hg38,
                                 "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  unique %>%
  dplyr::rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  merge(v8_se_pigment %>%
          select(gene_id = v26_id, chr, pos_hg38, v8_maf, v8_slope, v8_se, v8_pval)) %>%
  mutate(chr = rmchr(chr)) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(pigment)
tmem216.eur.se.df = rbind(tmem216.eur.se.df %>%
                            merge(r2_tmem216.eur.rs7948623 %>% select(snp, snp_hg19, r2)),
                          tmem216.eur.se.df %>%
                            merge(r2_tmem216.eur.rs11230664 %>% select(snp, snp_hg19, r2)),
                          tmem216.eur.se.df %>%
                            merge(r2_tmem216.eur.rs3741265 %>% select(snp, snp_hg19, r2)))



tmem216.eafr.df = gemma.egenes %>%
  filter(gene_id=="ENSG00000187049") %>%
  select(chr, pos_hg19, snp_hg19, maf, beta, se, afr_pval = p.value) %>%
  merge(pigment)
tmem216.eafr.df = rbind(tmem216.eafr.df %>%
                            merge(r2_tmem216.eafr.rs7948623 %>% select(snp, snp_hg19, r2)),
                          tmem216.eafr.df %>%
                            merge(r2_tmem216.eafr.rs11230664 %>% select(snp, snp_hg19, r2)),
                        tmem216.eafr.df %>%
                            merge(r2_tmem216.eafr.rs3741265 %>% select(snp, snp_hg19, r2)))




tmem216.eur.se.df = liftover_snp(v8_se_pigment$chr,
                                  v8_se_pigment$pos_hg38,
                                  "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  unique %>%
  dplyr::rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  merge(v8_se_pigment %>%
          select(gene_id = v26_id, chr, pos_hg38, v8_maf, v8_slope, v8_se, v8_pval)) %>%
  mutate(chr = rmchr(chr)) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(pigment)
tmem216.eur.se.df = rbind(tmem216.eur.se.df %>%
                              merge(r2_tmem216.eur.rs7948623 %>% select(snp, snp_hg19, r2)),
                            tmem216.eur.se.df %>%
                              merge(r2_tmem216.eur.rs11230664 %>% select(snp, snp_hg19, r2)),
                            tmem216.eur.se.df %>%
                              merge(r2_tmem216.eur.rs3741265 %>% select(snp, snp_hg19, r2)))

tmem216.eur.se.df %>%
  merge(tmem_rsmap) %>%
  filter(gene_id=="ENSG00000162144.9") %>%
  mutate(rsid = factor(rsid, levels=c("rs7948623","rs11230664","rs3741265"))) %>%
  merge(pigment) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(v8_pval), color=r2)) +
  geom_point() +
  theme_classic(base_size=15) +
  scale_color_distiller(palette = "Spectral") +
  facet_wrap(~ rsid)
  




# run coloc w/ susie

# get LD



# run coloc
tmem216.eur.wb.coloc.df = tmem216.eur.wb.df %>%
  filter(snp == "11:61137147" & gene_id=="ENSG00000187049.9") %>%
  filter(gwas_maf > 0.001)
tmem216.eur.wb.coloc = coloc.abf(dataset1=list(beta = tmem216.eur.wb.coloc.df$gwas_beta,
                                               varbeta = tmem216.eur.wb.coloc.df$gwas_se^2,
                                               MAF = tmem216.eur.wb.coloc.df$gwas_maf,
                                               type="quant",
                                               N=1570),
                                 dataset2=list(beta = tmem216.eur.wb.coloc.df$v8_slope,
                                               varbeta = tmem216.eur.wb.coloc.df$v8_se^2,
                                               MAF = tmem216.eur.wb.coloc.df$v8_maf,
                                               type="quant",
                                               N=670))


tmem216.eur.se.coloc.df = tmem216.eur.se.df %>%
  filter(snp == "11:61137147" & gene_id=="ENSG00000187049.9") %>%
  filter(gwas_maf > 0.001)
tmem216.eur.se.coloc = coloc.abf(dataset1=list(beta = tmem216.eur.se.coloc.df$gwas_beta,
                                               varbeta = tmem216.eur.se.coloc.df$gwas_se^2,
                                               MAF = tmem216.eur.se.coloc.df$gwas_maf,
                                               type="quant",
                                               N=1570),
                                 dataset2=list(beta = tmem216.eur.se.coloc.df$v8_slope,
                                               varbeta = tmem216.eur.se.coloc.df$v8_se^2,
                                               MAF = tmem216.eur.se.coloc.df$v8_maf,
                                               type="quant",
                                               N=670))

tmem216.eafr.coloc.df = tmem216.eafr.df %>%
  filter(snp == "11:61137147") %>%
  filter(gwas_maf > 0.001)
tmem216.eafr.coloc = coloc.abf(dataset1=list(beta = tmem216.eafr.coloc.df$gwas_beta,
                                             varbeta = tmem216.eafr.coloc.df$gwas_se^2,
                                             MAF = tmem216.eafr.coloc.df$gwas_maf,
                                             type="quant",
                                             N=1570),
                               dataset2=list(beta = tmem216.eafr.coloc.df$beta,
                                             varbeta = tmem216.eafr.coloc.df$se^2,
                                             MAF = tmem216.eafr.coloc.df$maf,
                                             type="quant",
                                             N=162))


#### plot figure ####
tmem216.pp4 = data.frame(tissue = factor(c(rep("v8 Whole Blood",3),
                                           rep("v8 Sun Exposed Skin",3),
                                           rep("East Africa",3)),
                                         levels=c("East Africa","v8 Whole Blood","v8 Sun Exposed Skin")),
                         rsid = factor(rep(c("rs7948623","rs11230664","rs3741265"), 3),
                                       levels=c("rs7948623","rs11230664","rs3741265")),
                         labels = c(rep("PP4 = 0.07",3),
                                    rep("PP4 = 0.12",3),
                                    rep("PP4 = 0.95",3)),
                         x = rep(2.5, 9),
                         y = rep(9.2,9))

fig5.tmem216.plot.df = rbind(tmem216.eur.wb.df %>%
        filter(gene_id=="ENSG00000187049.9") %>%
        select(eqtl_pval = v8_pval, gwas_pval, r2, snp) %>%
        mutate(tissue = "v8 Whole Blood") %>%
        merge(tmem_rsmap),
      tmem216.eur.se.df %>%
        filter(gene_id=="ENSG00000187049.9") %>%
        select(eqtl_pval = v8_pval, gwas_pval, r2, snp) %>%
        mutate(tissue = "v8 Sun Exposed Skin") %>%
        merge(tmem_rsmap),
      tmem216.eafr.df %>%
        select(eqtl_pval = afr_pval, gwas_pval, r2, snp) %>%
        mutate(tissue = "East Africa") %>%
        merge(tmem_rsmap)) %>%
  arrange(r2) %>%
  mutate(rsid = factor(rsid, levels=c("rs7948623","rs11230664","rs3741265")),
         tissue = factor(tissue, levels=c("East Africa","v8 Whole Blood","v8 Sun Exposed Skin")))

write_csv(fig5.tmem216.plot.df, "fig5.tmem216.plot.df.csv")


## start here
fig5.tmem216.plot = ggplot(fig5.tmem216.plot.df) +
  geom_point(aes(x=-log10(gwas_pval), y=-log10(eqtl_pval), color=r2)) +
  theme_bw(base_size=12) +
  coord_fixed(ratio = 2/3) +
  ylab(expression(~ -log[10](P[eQTL]))) +
  xlab(expression(~ -log[10](P[GWAS]))) +
  labs(color = expression(~ r^2)) +
  scale_x_continuous(
    labels = scales::number_format(accuracy = 1)) +
  scale_color_distiller(palette = "Spectral") +
  facet_grid(rsid ~ tissue) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        strip.text.y = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  geom_text(data = tmem216.pp4, aes(x=x, y=y, label=labels))


#### plot together ####
fig5 = plot_grid(fig5.top,
                 fig5.tmem216.plot,
                 ncol = 1,
                 labels = c('', 'C'),
                 label_size = 20,
                 rel_widths = c(1,0.8))







ENSG00000187049
ENSG00000187049.9

# v8 all whole blood
pigment.v8_wb.df = liftover_snp(addchr(pigment$chr), pigment$pos_hg19,
                                "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old, pos_hg38 = pos_new) %>%
  merge(v8_wb_pigment %>%
          filter(v26_id=="ENSG00000187049.9") %>%
          select(chr, pos_hg38, v8_maf, v8_slope, v8_se, v8_pval)) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment %>% select(chr, pos_hg19, gwas_maf, gwas_beta, gwas_sd, gwas_pval))



ggplot(pigment.v8_wb.df,
       aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point() +
  theme_classic()

# v8 all skin exposed
pigment.v8_se.df = liftover_snp(addchr(pigment$chr), pigment$pos_hg19,
                                "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old, pos_hg38 = pos_new) %>%
  merge(v8_se_pigment %>%
          filter(v26_id=="ENSG00000187049.9") %>%
          select(chr, pos_hg38, v8_maf, v8_slope, v8_se, v8_pval)) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment %>% select(chr, pos_hg19, gwas_maf, gwas_beta, gwas_sd, gwas_pval))

pigment.v8_se.coloc = coloc.abf(dataset1=list(beta = pigment.v8_se.df$gwas_beta,
                                              varbeta = pigment.v8_se.df$gwas_sd^2,
                                              MAF = pigment.v8_se.df$
                                              type="quant",
                                              N=1570),
                                dataset2=list(beta = pigment.v8_se.df$v8_slope,
                                              varbeta = pigment.v8_se.df$v8_se^2,
                                              MAF = pigment.v8_se.df$v8_maf,
                                              type="quant",
                                              N=670))

ggplot(pigment.v8_se.df,
       aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point() +
  theme_classic()

# EUR only whole blood
liftover_snp(addchr(pigment$chr), pigment$pos_hg19, "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old, pos_hg38 = pos_new) %>%
  merge(v8_wb_eur_tmem216 %>% select(chr, pos_hg38, v8_pval = pval_nominal)) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment %>% select(chr, pos_hg19, gwas_pval)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point() +
  theme_classic()

# EUR only skin exposed
liftover_snp(addchr(pigment$chr), pigment$pos_hg19, "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old, pos_hg38 = pos_new) %>%
  merge(v8_se_eur_tmem216 %>% select(chr, pos_hg38, v8_pval=pval_nominal, v8_maf=maf)) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment %>% select(chr, pos_hg19, gwas_pval)) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point(size=2) +
  theme_classic()

# Afr
gemma.pigment.df = gemma.egenes %>%
  filter(gene_id=="ENSG00000187049") %>%
  select(chr, pos_hg19, afr_maf=maf, afr_beta=beta, afr_se=se, afr_pval=p.value) %>%
  merge(pigment %>% select(chr, pos_hg19, gwas_maf, gwas_beta, gwas_sd, gwas_pval))

gemma.pigment.coloc = coloc.abf(dataset1=list(beta = gemma.pigment.df$gwas_beta,
                                              varbeta = gemma.pigment.df$gwas_sd^2,
                                              MAF = gemma.pigment.df$gwas_maf,
                                              type="quant",
                                              N=1570),
                                dataset2=list(beta = gemma.pigment.df$afr_beta,
                                              varbeta = gemma.pigment.df$afr_se^2,
                                              MAF = gemma.pigment.df$afr_maf,
                                              type="quant",
                                              N=162))

ggplot(gemma.pigment.df, aes(x=-log10(gwas_pval), y=-log10(afr_pval))) +
  geom_point() +
  theme_classic()


```


# Supplemental Figures

## gene expression comparison between Africans and GTEx v8

```{r}

tested_v8 = read_tsv("../misc/gtex_genes.txt", c("v26_id", "gene_id"))

tpm.median = read_tsv("../gex/Sample_ALL.secondpass.rsem.genes.results.tpm") %>%
  select(-`transcript_id(s)`) %>%
  mutate(gene_id = strip_ensembl(gene_id)) %>%
  gather(key = sample, value = tpm, -gene_id) %>%
  group_by(gene_id) %>%
  summarise(median_tpm = median(tpm))

tpm.med.gtex = read_tsv("../gtex/v8/gex/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.whole_blood.wgs.gct.gz") %>%
  mutate(gene_id = strip_ensembl(Name)) %>%
  select(-c(Name, Description)) %>%
  gather(key = sample, value = tpm, -gene_id) %>%
  group_by(gene_id) %>%
  summarise(median_tpm = median(tpm))

# from https://slowkow.com/notes/ggplot2-color-by-density/
#
# Get density of points in 2 dimensions.
# @param x A numeric vector.
# @param y A numeric vector.
# @param n Create a square n by n grid to compute density.
# @return The density within each square.
get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}


library(MASS)
library(viridis)


tpm.medians = tpm.median %>%
  rename(afr_tpm = median_tpm) %>%
  merge(tpm.med.gtex %>%
          rename(gtex_tpm = median_tpm)) %>%
  filter(afr_tpm > 0 & gtex_tpm > 0)

tpm.medians$density <- get_density(log10(tpm.medians$afr_tpm+1),
                                   log10(tpm.medians$gtex_tpm+1),
                                   h = c(1, 1),
                                   n = 100)

ggplot(tpm.medians, aes(x=log2(afr_tpm+1), y=log2(gtex_tpm+1), color=density)) +
  geom_point() +
  coord_fixed() +
  theme_bw(base_size=15) +
  scale_color_viridis()

testable_afr = gost(query = genes.filt[genes.filt %in% gtex.v8.tpm$gene_id & !(genes.filt %in% tested_v8$gene_id)],
                    custom_bg = genes.filt[genes.filt %in% gtex.v8.tpm$gene_id])

testable_v8 = gost(query = tested_v8$gene_id[tested_v8$gene_id %in% tpm$gene_id & !(tested_v8$gene_id %in% genes.filt)],
                   custom_bg = tested_v8$gene_id[tested_v8$gene_id %in% tpm$gene_id])

```

## celltypes that non-replicating eQTLs are enriched in

```{r}



```



## NR1D1 overlap

```{r}

tf_meta = read_tsv("../tf/metadata.tsv") %>%
  rename(id = `File accession`)

nr1e1.tfs = tabix2df("../tf/GM12878.all_tfs.sort.bed.gz",
                     "17:38252660-38252660",
                     c("chr","start","end","name","score","strand","signal_val",
                       "p_value","q_value","peak","id"))

foo = tf_meta %>%
  select(id, `Experiment target`) %>%
  merge(nr1e1.tfs)

```


## Tables

```{r}

# table of covariate data ####
sex_age = read_table("../gemma/covariates.sex_age", col_names=c("group","sample","foo","sex","age")) %>%
  select(c(sample, sex, age))

date = read_tsv("../gemma/delivery_date.txt", col_names=c("sample","date"))

covar_data.df = Reduce(merge, list(sample_info %>% select(sample, population=Ethnicity),
                                   date,
                                   sex_age)) %>%
  mutate(population = ifelse(population=="Argobba", "Argoba", population),
         population = ifelse(population=="Sabue", "Chabu", population))


# table of colocalizing and independent genes ####

## read in eQTL data for East Africans and European Americans from GTEx v8 ####
if (!exists("gemma.egenes")){
  gemma.egenes = read_tsv("../gemma/gemma.egenes.tsv.gz",
                          col_types="cnncnccnnnnnnnnnnccnlln")
}

if (!exists("lc.sgenes")){
  lc.sgenes = read_tsv("../leafcutter/gemma/lc.sgenes.tsv.gz", col_types=c("ccnnnccnnnnnnnnnnccnlln")) %>%
    separate(snp_hg19, into=c("chr","pos_hg19"), sep=":") %>%
    separate(intron, into=c("chr1","start_hg19","end_hg19","clust"), sep=":", convert=T) %>%
    unite("intron", chr1, start_hg19, end_hg19, clust, sep=".", remove=F) %>%
    mutate(v19_id1 = v19_id) %>%
    unite("afr_intron", chr1, start_hg19, end_hg19, v19_id1, sep="_") %>%
    mutate(sqtl_hg19 = paste(intron, pos_hg19, sep=":"))
}

if (!exists("gtex.v8.egenes.range")){
  gtex.v8.egenes.range = read_tsv("../gemma/gtex.v8.egenes.range.tsv.gz") %>%
    mutate(chr = rmchr(chr))
}

if (!exists("gtex.v8.sgenes.range")){
  gtex.v8.sgenes.range = read_tsv("../leafcutter/gemma/gtex.v8.sgenes.range.tsv.gz")
}

if (!exists("gemma.cond_sig")){
  gemma.cond_sig = read_tsv("../gemma/gemma.cond_sig.tsv.gz")
}

if (!exists("lc.cond_sig")){
  lc.cond_sig = read_tsv("../leafcutter/gemma/lc.cond_sig.tsv.gz")
}

if (!exists("hg19_to_hg38")){
  hg19_to_hg38 = read_tsv("../gemma/hg19_to_hg38.egenes.tsv.gz")
}

if (!exists("hg19_to_h38.s")){
  hg19_to_hg38.s = read_tsv("../leafcutter/gemma/hg19_to_hg38.sgenes.tsv.gz") %>%
    unique
}

## merge datasets ####
gemma.egenes.v8.merge = gemma.egenes %>%
  select(gene_id, chr, pos_hg19, beta, se, maf) %>%
  merge(hg19_to_hg38) %>%
  merge(gene_map) %>%
  merge(gtex.v8.egenes.range %>% select(gene_id, pos_hg38, gtexv8_slope, gtexv8_se, gtexv8_maf))
  
lc.sgenes.v8.merge = lc.sgenes %>%
  merge(hg19_to_hg38.s) %>%
  merge(intron_map) %>%
  select(intron, afr_intron, gtexv8_intron, chr, pos_hg38, beta, se, maf) %>%
  merge(gtex.v8.sgenes.range %>% select(gtexv8_intron, chr, pos_hg38, gtexv8_slope, gtexv8_se, gtexv8_maf) %>% mutate(chr = rmchr(chr)))


## colocalize ####

gemma_v8_h4 = lapply(unique(gemma.egenes.v8.merge$gene_id), function(x){
  df = gemma.egenes.v8.merge %>%
    filter(gene_id==x)
  coloc.abf(dataset1=list(snp=df$pos_hg38, beta=df$beta, varbeta=df$se^2, type="quant", N=162, MAF=df$maf),
            dataset2=list(snp=df$pos_hg38, beta=df$gtexv8_slope, varbeta=df$gtexv8_se^2, type="quant", N=162, MAF=df$gtexv8_maf))$summary
})

lc_v8_h4 = lapply(unique(lc.sgenes.v8.merge$intron), function(x){
  df = lc.sgenes.v8.merge %>%
    filter(intron==x)
  coloc.abf(dataset1=list(snp=df$pos_hg38, beta=df$beta, varbeta=df$se^2, type="quant", N=162, MAF=df$maf),
            dataset2=list(snp=df$pos_hg38, beta=df$gtexv8_slope, varbeta=df$gtexv8_se^2, type="quant", N=162, MAF=df$gtexv8_maf))$summary
})


## create dataframes ####

gemma_v8_h4.df = data.frame(gene_id = unique(gemma.egenes.v8.merge$gene_id),
                          H0 = sapply(gemma_v8_h4, function(x) x[2]),
                          H1 = sapply(gemma_v8_h4, function(x) x[3]),
                          H2 = sapply(gemma_v8_h4, function(x) x[4]),
                          H3 = sapply(gemma_v8_h4, function(x) x[5]),
                          H4 = sapply(gemma_v8_h4, function(x) x[6]))

lc_v8_h4.df = data.frame(intron = unique(lc.sgenes.v8.merge$intron),
                          H0 = sapply(lc_v8_h4, function(x) x[2]),
                          H1 = sapply(lc_v8_h4, function(x) x[3]),
                          H2 = sapply(lc_v8_h4, function(x) x[4]),
                          H3 = sapply(lc_v8_h4, function(x) x[5]),
                          H4 = sapply(lc_v8_h4, function(x) x[6])) %>%
  merge(lc.sgenes %>% select(c(intron, gene_id)) %>% unique, .)

gemma_v8_h4.df %>%
  mutate(independent = gene_id %in% gemma.cond_sig$gene_id) %>%
  write.table(., "gemma.v8.coloc.cond_sig.tsv", sep="\t", row.names=F, quote=F)

lc_v8_h4.df %>%
  mutate(independent = intron %in% lc.cond_sig$intron) %>%
  write.table(., "lc.v8.coloc.cond_sig.tsv", sep="\t", row.names=F, quote=F)

```


## p-values of SNPs unique to our study

```{r}

snps_1kg = readLines("../geno/ALL.phase3_shapeit2_mvncall_integrated_v5a.20130502.in_rna.snps.txt")

gemma.f = "../gemma/round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.txt.gz"
gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# read in and filter raw eQTL data
gemma.uniq = fread(gemma.f, col.names=gemma_cols, na.string=".") %>%
  select(-c(foo, bar)) %>%
  filter(!(snp_hg19 %in% snps_1kg)) %>%
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", remove=F, convert=T) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  filter(gene_id %in% gencode.pc_linc$gene_id) %>% # pc and lncRNAs
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"))

```


## caQTL direction

```{r}

caqtls = read_tsv("../atac/elife-39595-supp2.txt.gz") %>%
 dplyr::rename(pos_hg19 = position)
colnames(caqtls) = make.names(colnames(caqtls))

# get data relating positive eQTLs to open allele
gemma.fdr05.caqtls = gemma.egenes %>%
  filter(fdr_sig) %>%
  # polarize beta to reflect the minor allele
  mutate(min_beta = ifelse(af < 0.5, beta, -beta),
         # get the minor allele
         min_allele = ifelse(af < 0.5, allele0, allele1)) %>%
  merge(caqtls %>% select(chr, pos_hg19, consensus.open.allele, consensus.closed.allele)) %>%
  mutate(min_pos = min_beta > 0,
         min_open = min_allele == consensus.open.allele) %>%
  select(min_beta, min_pos, min_open)


# gemma.caqtls.plot = gemma.fdr05.caqtls %>%
#   mutate(min_class = c("closed","open")[min_open + 1]) %>%
#   ggplot(aes(x=min_class, y=min_beta, fill=min_class)) +
#   geom_violin() +
#   theme_classic(base_size = 15)


# run chis-square test  
gemma.fdr05.caqtls.fisher = gemma.fdr05.caqtls %>%
  select(min_pos, min_open) %>%
  table %>%
  fisher.test



## repeat for sQTLs
# get data relating positive sQTLs to open allele
lc.fdr05.caqtls = lc.sgenes %>%
  filter(fdr_sig) %>%
  # polarize beta to reflect the minor allele
  mutate(min_beta = ifelse(af < 0.5, beta, -beta),
         # get the minor allele
         min_allele = ifelse(af < 0.5, allele0, allele1)) %>%
  merge(caqtls %>% select(chr, pos_hg19, consensus.open.allele, consensus.closed.allele)) %>%
  mutate(min_pos = min_beta > 0,
         min_open = min_allele == consensus.open.allele) %>%
  select(min_beta, min_pos, min_open)


# lc.caqtls.plot = gemma.fdr05.caqtls %>%
#   mutate(min_class = c("closed","open")[min_open + 1]) %>%
#   ggplot(aes(x=min_class, y=min_beta, fill=min_class)) +
#   geom_violin() +
#   theme_classic(base_size = 15)


# run chis-square test  
lc.fdr05.caqtls.fisher = lc.fdr05.caqtls %>%
  select(min_pos, min_open) %>%
  table %>%
  fisher.test

```


## eQTL/sQTL effect size directions

```{r}

# column names
gemma.sort.cols = c("chr","pos_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# qtl files
gemma.sort.f = "../gemma/round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.sort.txt.gz"

set.seed(42)

lc.fdr05.top = lc.sgenes %>%
  filter(top_ssnp) %>%
  group_by(intron) %>%
  sample_n(1)

lc.top.gemma.tabix = lc.fdr05.top %>%
  mutate(tabix_id = tabixify(v19_id, pos_hg19, pos_hg19)) %>%
  pull(tabix_id) %>%
  unique

lc.top.gemma = tabix2df(gemma.sort.f,
                        lc.top.gemma.tabix,
                        gemma.sort.cols)

lc.top.gemma.tbl = lc.fdr05.top %>%
  select(v19_id, pos_hg19, sqtl_beta = beta) %>%
  merge(lc.top.gemma %>% select(v19_id, pos_hg19, eqtl_beta = beta)) #%>%
  merge(lc.frac.rank)

ggplot(lc.top.gemma.tbl, aes(x=sqtl_beta, y=eqtl_beta)) +
  geom_point() +
  theme_classic(base_size=15)
    
```


## eQTL p-value distribution of FDR-significant sQTLs

```{r}

lc.fdr05.gemma.tabix = lc.sgenes %>%
  filter(fdr_sig) %>%
  mutate(tabix_id = tabixify(v19_id, pos_hg19, pos_hg19)) %>%
  pull(tabix_id) %>%
  unique

lc.fdr05.gemma = tabix2df(gemma.sort.f,
                        lc.fdr05.gemma.tabix,
                        gemma.sort.cols)

pi0est(lc.fdr05.gemma$p.value)

eqtl_sqtl_pi1.plot = ggplot(lc.fdr05.gemma, aes(x=p.value)) +
  geom_histogram(alpha = 0.8) +
  theme_classic(base_size=15) +
  annotate("text", x=0.5, y=10000, parse=T, label = as.character(expression(~pi[1] == 0.61)), size=5) +
  xlab("eQTL p-value")

```


## true positives among EUR MAF < 0.01 in Africans and vice versa

```{r}

## Europeans

eur162.f = "../gtex/EUR162_15PEER_5PC_nopermute.wfrq.sort.txt.gz"
eur162.cols = c("snp_hg38","v29_id","chr","pos_hg38","eur162_ref","eur162_alt",
                     "version","tss_dist","eur162_pval","eur162_slope","eur162_maf")
gencode29.pc_linc = read_tsv("../genomes/gencode.v29.pc_linc.bed")

eur162.info = fread(eur162.f, col.names=eur162.cols) %>%
  filter(v29_id %in% gencode29.pc_linc$v29_id) %>%
  select(-c(snp_hg38, version, tss_dist)) %>%
  merge(., gencode29.pc_linc %>% select(-chr), by="v29_id") %>%
  mutate(pos_hg38 = as.numeric(pos_hg38)) %>%
  mutate(eur162_dist = ifelse(strand=="+", pos_hg38 - start, end - pos_hg38)) %>%
  select(-c(start, end)) %>%
  filter(abs(eur162_dist) < 100000) %>%
  mutate(gene_id = strip_ensembl(v29_id)) %>%
  select(chr, pos_hg38, gene_id, eur162_pval)


eur162.info = liftover_snp(eur162$chr,
                           eur162$pos_hg38,
                           "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  unique %>%
  merge(eur162.info ) %>%
  mutate(chr = rmchr(chr))

rm(eur162)

eur162.afr_af = tabix2df("../misc/1kgenomes.afr_af.txt.gz",
                         unique(tabixify(eur162.info$chr,
                                         eur162.info$pos_hg19,
                                         eur162.info$pos_hg19)),
                         c("chr","pos_hg19","ref","alt","eur_af"))


## Africans
gemma.f = "../gemma/round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.txt.gz"
gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# read in and filter raw eQTL data
gemma.info = fread(gemma.f, col.names=gemma_cols, na.string=".") %>%
  select(-c(foo, bar)) %>%
  filter(r2 > 0.3 | is.na(r2)) %>% # remove poorly imputed SNPs
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", remove=F, convert=T) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  filter(gene_id %in% gencode.pc_linc$gene_id) %>% # pc and lncRNAs
  select(chr, pos_hg19, gene_id, p.value)
  
eur162.afr_af = tabix2df("../misc/1kgenomes.afr_af.txt.gz",
                         unique(tabixify(gemma.info$chr,
                                         gemma.info$pos_hg19,
                                         gemma.info$pos_hg19)),
                         c("chr","pos_hg19","ref","alt","eur_af"))

```


## mapping statistics

```{r}

map_stats = read_tsv("../misc/mapping_stats.060521.txt")

map_stats %>%
  mutate(log10_depth = log10(nreads),
         frac_map = (uniq_mapped + multi_mapped)/nreads,
         frac_map_uniq = uniq_mapped/nreads) %>%
  select(`log10(depth)` = log10_depth,
         `fraction mapped` = frac_map,
         `fraction uniquely mapped` = frac_map_uniq) %>%
  gather(key = stat, value = val) %>%
  mutate(stat = factor(stat, levels=c("log10(depth)",
                                      "fraction mapped",
                                      "fraction uniquely mapped"))) %>%
  ggplot(aes(x=val)) +
  geom_histogram(alpha=0.8) +
  theme_classic(base_size=15) +
  theme(axis.title.x=element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20),
        panel.spacing.x=unit(2, "lines")) +
  facet_wrap(~ stat, scales = "free_x")
  

```


## freq diff of top tQTLs for shared and conditionally independent SNPs

```{r}

set.seed(42)
gemma.fdr05.top = gemma.egenes %>%
  group_by(gene_id) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1)

lc.fdr05.top = lc.sgenes %>%
  group_by(intron) %>%
  filter(p.value==min(p.value)) %>%
  sample_n(1)

gemma.fdr05.top.eur_af = tabix2df("../misc/ALL.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.EUR.snps.af.txt.gz",
                                  tabixify(gemma.fdr05.top$chr,
                                           gemma.fdr05.top$pos_hg19,
                                           gemma.fdr05.top$pos_hg19),
                                  c("chr","pos_hg19","ref_1kg","alt_1kg","eur_af")) %>%
  merge(gemma.fdr05.top %>%
          select(gene_id, chr, pos_hg19, allele0, allele1, af), all.y=T) %>%
  filter(ref_1kg==allele1 & alt_1kg==allele0 | ref_1kg==allele0 & alt_1kg==allele1) %>%
  mutate(eur_af = ifelse(ref_1kg==allele1, eur_af, 1-eur_af))

lc.fdr05.top.eur_af = tabix2df("../misc/ALL.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.EUR.snps.af.txt.gz",
                                  tabixify(lc.fdr05.top$chr,
                                           lc.fdr05.top$pos_hg19,
                                           lc.fdr05.top$pos_hg19),
                                  c("chr","pos_hg19","ref_1kg","alt_1kg","eur_af")) %>%
  merge(lc.fdr05.top %>%
          select(intron, gene_id, chr, pos_hg19, allele0, allele1, af), all.y=T) %>%
  filter(ref_1kg==allele1 & alt_1kg==allele0 | ref_1kg==allele0 & alt_1kg==allele1) %>%
  mutate(eur_af = ifelse(ref_1kg==allele1, eur_af, 1-eur_af))



deltamaf.plot = rbind(gemma.fdr05.top.eur_af %>%
        mutate(shared = !(gene_id %in% gemma.cond_sig$gene_id)) %>%
        mutate(class = "eQTL"),
      lc.fdr05.top.eur_af %>%
        mutate(shared = !(intron %in% lc.cond_sig$intron)) %>%
        mutate(class = "sQTL") %>%
        select(-intron)) %>%
  mutate(shared = factor(shared, levels=c(T,F))) %>%
  ggplot(aes(x = abs(af - eur_af), color = shared)) +
  stat_ecdf() +
  theme_classic(base_size=15) +
  scale_color_ptol() +
  xlab(expression("|"*Delta*"AF"*"|")) +
  theme(axis.title.y=element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  facet_wrap(~ class)

```


## Manhattan plots of non-replicating tQTLs

```{r}

hg19_to_hg38 = read_tsv("../gemma/hg19_to_hg38.egenes.tsv.gz")

indep_eqtl = c("INPP5K","TMEM140","ACSM3","CNTNAP3","PPP1R14C","PDZK1TP1","GPR56","TRAM2")
indep_sqtl = c("ADAM8","ICAM2","LINC00694","MAPK1")

gemma.egenes %>%
  merge(gene_name) %>%
  filter(gene_name %in% indep_eqtl) %>%
  merge(hg19_to_hg38) %>%
  merge(gtex.v8.egenes.range %>% select(gene_id, pos_hg38, gtexv8_pval), all.x = T) %>%
  ggplot(aes(x=pos_hg38)) +
  geom_point(aes(y = -log10(p.value))) +
  geom_point(aes(y = log10(gtexv8_pval))) +
  facet_wrap(~ gene_name, scales = "free")

```

## Afr vs 162 EA, compared with eQTLGen consortium

```{r}

eur162.fdr05 = eur162.egenes %>%
  filter(fdr_sig) %>%
  filter(!is.na(pos_hg19))

eur162.fdr05.egen = tabix2df("../misc/2019-12-11-cis-eQTLsFDR-ProbeLevel-CohortInfoRemoved-BonferroniAdded.sort.txt.gz",
                             tabixify(eur162.fdr05$gene_id, eur162.fdr05$pos_hg19, eur162.fdr05$pos_hg19),
                             c("egen_pval","rsid","chr","pos_hg19","allele0","allele1","zscore","gene_id","gene_name",
                               "gene_chr","gene_pos","n_cohorts","n_samples","fdr","bon_pval"))

gemma.fdr05.egen = tabix2df("../misc/2019-12-11-cis-eQTLsFDR-ProbeLevel-CohortInfoRemoved-BonferroniAdded.sort.txt.gz",
                             tabixify(gemma.fdr05$gene_id, gemma.fdr05$pos_hg19, gemma.fdr05$pos_hg19),
                             c("egen_pval","rsid","chr","pos_hg19","allele0","allele1","zscore","gene_id","gene_name",
                               "gene_chr","gene_pos","n_cohorts","n_samples","fdr","bon_pval"))

gemma.eur162.fdr05.egen = gemma.fdr05 %>%
  select(gene_id, pos_hg19) %>%
  mutate(afr=T) %>%
  merge(eur162.fdr05 %>%
          select(gene_id, pos_hg19) %>%
          mutate(eur162=T), all=T) %>%
  merge(unique(rbind(gemma.fdr05.egen, eur162.fdr05.egen)))
gemma.eur162.fdr05.egen[is.na(gemma.eur162.fdr05.egen)] = F

```


## tQTL pval vs fst

```{r}

gemma.sel_hits.eur_fst.pval = read_tsv("gemma.sel_hits.eur_fst.pval.tsv")

gemma.eur_fst.sel_plot = ggplot(gemma.sel_hits.eur_fst.pval, aes(x=-log10(p.value), y=fst)) +
  geom_point(aes(alpha=0.8)) +
  theme_bw(base_size=12) +
  facet_wrap(~ gene_name, scales = "free_x") +
  theme(legend.position="none",
        strip.background = element_blank()) +
  xlab(-log[10]~(p[eQTL])) +
  ylab(expression(F[ST]))


lc.eur_fst.sel_info = read_csv("lc.eur_fst.sel_info.csv")
lc.eur_fst.sel_info.keep = lc.eur_fst.sel_info %>%
  group_by(gene_name) %>%
  filter(weighted_fst == max(weighted_fst))

lc.sel_hits.eur_fst.pval = read_tsv("lc.sel_hits.eur_fst.pval.tsv")

lc.eur_fst.sel_plot = lc.sel_hits.eur_fst.pval %>%
  filter(intron %in% lc.eur_fst.sel_info.keep$intron) %>%
  ggplot(aes(x=-log10(p.value), y=fst)) +
  geom_point(aes(alpha=0.8)) +
  theme_bw(base_size=12) +
  facet_wrap(~ gene_name, scales = "free_x") +
  theme(legend.position="none",
        strip.background = element_blank()) +
  xlab(-log[10]~(p[sQTL])) +
  ylab(expression(F[ST]))

library(cowplot)

plot_grid(gemma.eur_fst.sel_plot, lc.eur_fst.sel_plot, nrow=1, labels = c("A","B"), rel_widths = c(6,5))

```


## effect size and frequency of tQTLs

```{r}

gemma.beta_maf.plot = gemma.fdr05.top %>%
  mutate(MAF = ifelse(af < 0.5, af, 1-af)) %>%
  ggplot(aes(x=MAF, y=beta)) +
  geom_point(alpha=0.5) +
  ylab(expression(~ beta)) +
  ggtitle("eQTL effect size vs. MAF") +
  theme_bw(base_size=15)


lc.beta_maf.plot = lc.fdr05.top %>%
  mutate(MAF = ifelse(af < 0.5, af, 1-af)) %>%
  ggplot(aes(x=MAF, y=beta)) +
  geom_point(alpha=0.5) +
  ylab(expression(~ beta)) +
  ggtitle("sQTL effect size vs. MAF") +
  theme_bw(base_size=15)

plot_grid(gemma.beta_maf.plot,
          lc.beta_maf.plot,
          nrow=1,
          labels=c("A","B"))




# plot GTEx
gtex.v8.fdr05 %>%
  group_by(gene_id) %>%
  filter(pval_nominal==min(pval_nominal)) %>%
  sample_n(1) %>%
  ggplot(aes(x=maf, y=slope)) +
  geom_point(alpha=0.5) +
  ylab(expression(~ beta)) +
  ggtitle("GTEx eQTL effect size vs. MAF") +
  theme_bw(base_size=15)

eur162.egenes %>%
  filter(top_esnp) %>%
  group_by(gene_id) %>%
  sample_n(1) %>%
  ggplot(aes(x=eur162_maf, y=eur162_slope)) +
  geom_point(alpha=0.5) +
  ylab(expression(~ beta)) +
  ggtitle("GTEx eQTL effect size vs. MAF") +
  theme_bw(base_size=15)

```


## GWAS overlap

```{r message = FALSE}

gemma.f = "../gemma/round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.txt.gz"
gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# get statistics for all SNPs
gemma.all.egenes = fread(gemma.f, col.names=gemma_cols, na.string=".") %>%
  select(-c(foo, bar)) %>%
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", remove=F, convert=T) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  filter(gene_id %in% gemma.egenes$gene_id) %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af))

pheno_codes = read_tsv("../gwas/pheno_codes.tsv")

gemma.fdr05 = gemma.egenes %>%
  filter(fdr_sig)

gemma.fdr05.sig_ukbb = tabix2df("../gwas/pan_ukbb.sig_results.sort.tsv.gz",
                                unique(tabixify(gemma.fdr05$chr,
                                                gemma.fdr05$pos_hg19,
                                                gemma.fdr05$pos_hg19)),
                                c("chr","pos_hg19","ref","alt","pval_meta","pheno"))

gemma.fdr05.sig_ukbb.range = gemma.egenes %>%
  filter(fdr_sig) %>%
  select(chr, pos_hg19, gene_id, p.value) %>%
  merge(gemma.fdr05.sig_ukbb) %>%
  select(gene_id, pheno) %>%
  unique %>%
  merge(gencode.v19.tss %>% mutate(gene_id = strip_ensembl(v19_id))) %>%
  mutate(chr = rmchr(chr),
         start = tss - 100000,
         end = tss + 100000) %>%
  mutate(range = tabixify(chr, start, end)) %>%
  select(gene_id, pheno, range)

write_tsv(gemma.fdr05.sig_ukbb.range, "gemma.fdr05.sig_ukbb.range.tsv", col_names = F)


lc.fdr05.sig_ukbb.range = lc.sgenes %>%
  filter(fdr_sig) %>%
  select(chr, pos_hg19, intron, gene_id, p.value) %>%
  merge(gemma.fdr05.sig_ukbb) %>%
  select(intron, gene_id, pheno) %>%
  unique %>%
  separate(intron, into=c("chr","start","end","clust"), sep="\\.", remove=F, convert=T) %>%
  mutate(start = start - 100000,
         end = end + 100000,
         range = tabixify(chr, start, end)) %>%
  select(intron, gene_id, pheno, range)

write_tsv(lc.fdr05.sig_ukbb.range, "lc.fdr05.sig_ukbb.range.tsv", col_names = F)


#### run coloc ####

gemma.01.cols = c("v19_id","foo","chr","pos_hg19","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score")
gemma.01.egenes = tabix2df("../gemma/round3/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.gz",
                           paste(unique(gemma.egenes$v19_id), "0-1000000000", sep=":"),
                           gemma.01.cols) %>%
  mutate(gene_id = strip_ensembl(v19_id),
         maf = ifelse(af < 0.5, af, 1-af))



gemma.fdr05.sig_ukbb.range.tested = data.frame()
for (i in 1:nrow(gemma.fdr05.sig_ukbb.range)){
  
  range_info = gemma.fdr05.sig_ukbb.range[i,]
  
  pheno_file = paste0("../gwas/ranges/eqtl/", range_info$gene_id, "_", range_info$pheno, ".tsv")

  ukbb = read_tsv(pheno_file)
  
  if (range_info$pheno=="FEV1FVC"){
     
    ukbb.top_var = ukbb %>%
      filter(!is.na(pval_EUR)) %>%
      filter(pval_EUR == min(pval_EUR))
  } else{

    ukbb.top_var = ukbb %>%
      filter(!is.na(pval_meta)) %>%
      filter(pval_meta == min(pval_meta))
  }
  
  if (paste(ukbb.top_var$chr, ukbb.top_var$pos, sep=":") %in% gemma.egenes$snp_hg19){
    
    gemma.fdr05.sig_ukbb.range.tested = rbind(gemma.fdr05.sig_ukbb.range.tested,
                                              range_info)
  }
}

# eQTLs
gemma.01.ukbb.coloc.df = data.frame()
# eur162.ukbb.coloc.df = data.frame()
gtex.ukbb.coloc.df = data.frame()
for (i in 1:nrow(gemma.fdr05.sig_ukbb.range.tested)){
  
  range_info = gemma.fdr05.sig_ukbb.range.tested[i,]
  
  pheno_info = pheno_codes %>% filter(phenocode==range_info$pheno)
  
  n_case_ctrl = pheno_info %>%
    select(starts_with("n_")) %>%
    select(-contains("full")) %>%
    gather(key = pop, value = n) %>%
    separate(pop, into=c("foo","group","pop")) %>%
    group_by(group) %>%
    summarise(n = sum(n, na.rm=T)) %>%
    pivot_wider(names_from = group, values_from = n)
  
  pheno_file = paste0("../gwas/ranges/eqtl/", range_info$gene_id, "_", range_info$pheno, ".tsv")

  ukbb = read_tsv(pheno_file)
  
  gemma.01.ukbb = gemma.01.egenes %>%
    filter(gene_id==range_info$gene_id) %>%
    select(chr, pos = pos_hg19, alt = allele1, ref = allele0, beta, se, maf, p.value) %>%
    merge(ukbb)
  
  # eur162.ukbb = eur162.egenes %>%
  #   mutate(chr = as.numeric(chr)) %>%
  #   filter(gene_id==range_info$gene_id) %>%
  #   select(chr, pos = pos_hg19, alt = eur162_alt, ref = eur162_ref, eur162_pval, eur162_maf) %>%
  #   merge(ukbb)
  
  gtex.ukbb = gtex.v8.egenes.range %>%
    filter(gene_id==range_info$gene_id) %>%
    select(chr, pos = pos_hg19, alt = v8_alt, ref = v8_ref, gtexv8_slope, gtexv8_se, gtexv8_maf, gtexv8_pval) %>%
    merge(ukbb) %>%
    filter(gtexv8_maf >= 0.05)
  
  # gemma.01.ukbb = gemma.01.ukbb %>% filter(pos %in% gtex.ukbb$pos)
  # gtex.ukbb = gtex.ukbb %>% filter(pos %in% gemma.ukbb$pos)
  
  # FEV1FVC is EUR only, so rename to be used later and add MAF
  if (range_info$pheno=="FEV1FVC"){
    
    gemma.01.ukbb = gemma.01.ukbb %>%
      rename(af_meta = af_EUR, beta_meta = beta_EUR, se_meta = se_EUR) %>%
      mutate(ukbb_maf = ifelse(af_meta < 0.5, af_meta, 1-af_meta))
    
    # eur162.ukbb = eur162.ukbb %>%
    #   rename(af_meta = af_EUR, pval_meta = pval_EUR) %>%
    #   mutate(ukbb_maf = ifelse(af_meta < 0.5, af_meta, 1-af_meta))
    
    gtex.ukbb = gtex.ukbb %>%
      rename(af_meta = af_EUR, beta_meta = beta_EUR, se_meta = se_EUR) %>%
      mutate(ukbb_maf = ifelse(af_meta < 0.5, af_meta, 1-af_meta))
    
    type_study = "quant"
    
  } else{
    
    # these are case-control studies, so
    if (range_info$pheno %in% c("401", "401.1", "1747")){
        
      ukbb.af = ukbb %>%
        select(chr, pos , starts_with("af")) %>%
        select(-contains("meta")) %>%
        gather(key=pop, value=af, -c(chr, pos)) %>%
        separate(pop, into=c("foo","bar","pop"), "_") %>%
        select(-c(foo, bar))
      
      ukbb.n_case = pheno_info %>%
        select(contains("cases")) %>%
        select(-contains("full")) %>%
        gather(key = pop, value = n_case) %>%
        separate(pop, into=c("foo","bar","pop")) %>%
        select(-c(foo, bar))
      
      ukbb.n_control = pheno_info %>%
        select(contains("control")) %>%
        select(-contains("full")) %>%
        gather(key = pop, value = n_control) %>%
        separate(pop, into=c("foo","bar","pop")) %>%
        select(-c(foo, bar))
      
      ukbb.maf = merge(ukbb.af, ukbb.n_case) %>%
        merge(ukbb.n_control) %>%
        mutate(n_case = ifelse(is.na(n_case), 0, n_case),
               n_control = ifelse(is.na(n_control), 0, n_control)) %>%         
        mutate(af = af*(n_case+ n_control)/(n_case + n_control)) %>%
        mutate(ukbb_maf = ifelse(af < 0.5, af, 1-af)) %>%
        select(pos, chr, ukbb_maf)
      
      gemma.01.ukbb = merge(gemma.01.ukbb, ukbb.maf)
      # eur162.ukbb = merge(eur162.ukbb, ukbb.maf)
      gtex.ukbb = merge(gtex.ukbb, ukbb.maf)
      
      # set type to case-control
      type_study = "cc"
      
    } else{
      
      gemma.01.ukbb = gemma.01.ukbb %>%
        mutate(ukbb_maf = ifelse(af_meta < 0.5, af_meta, 1-af_meta))

      # eur162.ukbb = eur162.ukbb %>%
      #   mutate(ukbb_maf = ifelse(af_meta < 0.5, af_meta, 1-af_meta))

      gtex.ukbb = gtex.ukbb %>%
        mutate(ukbb_maf = ifelse(af_meta < 0.5, af_meta, 1-af_meta))

      type_study = "quant"
      
    }

  }
  
  gemma.01.ukbb = gemma.01.ukbb %>%
    select(beta, se, maf, beta_meta, se_meta, ukbb_maf)
  gemma.01.ukbb = gemma.01.ukbb[complete.cases(gemma.01.ukbb),]
  # eur162.ukbb = eur162.ukbb[complete.cases(eur162.ukbb),]
  gtex.ukbb = gtex.ukbb %>%
    select(gtexv8_slope, gtexv8_se, gtexv8_maf, beta_meta, se_meta, ukbb_maf)
  gtex.ukbb = gtex.ukbb[complete.cases(gtex.ukbb),]
  
  
  gemma.01.ukbb.coloc = coloc.abf(dataset1=list(beta = gemma.01.ukbb$beta_meta,
                                             varbeta = gemma.01.ukbb$se_meta^2,
                                             MAF = gemma.01.ukbb$ukbb_maf,
                                             type=type_study,
                                             s=n_case_ctrl$cases/c(n_case_ctrl$cases + n_case_ctrl$controls),
                                             N=pheno_info$n_cases_full_cohort_both_sexes),
                               dataset2=list(beta = gemma.01.ukbb$beta,
                                             varbeta = gemma.01.ukbb$se^2,
                                             MAF = gemma.01.ukbb$maf,
                                             type="quant",
                                             N=162))

  gemma.01.ukbb.coloc.df = rbind(gemma.01.ukbb.coloc.df,
                              data.frame(gene_id = range_info$gene_id,
                                         pheno = range_info$pheno,
                                         H0 = gemma.01.ukbb.coloc$summary[2],
                                         H1 = gemma.01.ukbb.coloc$summary[3],
                                         H2 = gemma.01.ukbb.coloc$summary[4],
                                         H3 = gemma.01.ukbb.coloc$summary[5],
                                         H4 = gemma.01.ukbb.coloc$summary[6]))
  
  # if (nrow(eur162.ukbb) > 0){
  # 
  #   eur162.ukbb.coloc = coloc.abf(dataset1=list(pvalues = eur162.ukbb$pval_meta,
  #                                               MAF = eur162.ukbb$ukbb_maf,
  #                                               type=type_study,
  #                                               s=n_case_ctrl$cases/c(n_case_ctrl$cases + n_case_ctrl$controls),
  #                                               N=pheno_info$n_cases_full_cohort_both_sexes),
  #                                 dataset2=list(pvalues = eur162.ukbb$eur162_pval,
  #                                               MAF = eur162.ukbb$eur162_maf,
  #                                               type="quant",
  #                                               N=162))
  # 
  #   
  #   eur162.ukbb.coloc.df = rbind(eur162.ukbb.coloc.df,
  #                              data.frame(gene_id = range_info$gene_id,
  #                                         pheno = range_info$pheno,
  #                                         H0 = eur162.ukbb.coloc$summary[2],
  #                                         H1 = eur162.ukbb.coloc$summary[3],
  #                                         H2 = eur162.ukbb.coloc$summary[4],
  #                                         H3 = eur162.ukbb.coloc$summary[5],
  #                                         H4 = eur162.ukbb.coloc$summary[6]))
  # }
  
  if (nrow(gtex.ukbb) > 0){

    gtex.ukbb.coloc = coloc.abf(dataset1=list(beta = gtex.ukbb$beta_meta,
                                              varbeta = gtex.ukbb$se_meta^2,
                                              MAF = gtex.ukbb$ukbb_maf,
                                              type=type_study,
                                              s=n_case_ctrl$cases/c(n_case_ctrl$cases + n_case_ctrl$controls),
                                              N=pheno_info$n_cases_full_cohort_both_sexes),
                                dataset2=list(beta = gtex.ukbb$gtexv8_slope,
                                              varbeta = gtex.ukbb$gtexv8_se^2,
                                              MAF = gtex.ukbb$gtexv8_maf,
                                              type="quant",
                                              N=670))


    gtex.ukbb.coloc.df = rbind(gtex.ukbb.coloc.df,
                               data.frame(gene_id = range_info$gene_id,
                                          pheno = range_info$pheno,
                                          H0 = gtex.ukbb.coloc$summary[2],
                                          H1 = gtex.ukbb.coloc$summary[3],
                                          H2 = gtex.ukbb.coloc$summary[4],
                                          H3 = gtex.ukbb.coloc$summary[5],
                                          H4 = gtex.ukbb.coloc$summary[6]))
  }

}

write_tsv(gemma.01.ukbb.coloc.df, "gemma.ukbb.coloc.maf01.tsv")
write_tsv(gtex.ukbb.coloc.df, "gtex.ukbb.coloc.maf01.tsv")

gtex.ukbb.coloc.df = read_tsv("gtex.ukbb.coloc.tsv")



lc_cols = c("intron","snp_hg19","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2","v19_id")

lc.f = "../leafcutter/gemma/round2/Sample_ALL.secondpass.leafcutter.covars.peer5.resid.gemma.100kb.lmm4.r2.wgene.txt.gz"

lc.01.sgenes = fread(lc.f, col.names=lc_cols, na.strings=c(".","?")) %>%
  select(-c(foo, bar)) %>%
  mutate(intron = gsub(":",".",intron)) %>%
  filter(intron %in% lc.sgenes$intron) %>%
  mutate(gene_id = strip_ensembl(v19_id),
         maf = ifelse(af < 0.5, af, 1-af)) %>%
  separate(snp_hg19, into = c("chr","pos_hg19"), sep=":", convert=T, remove=F) %>%
  as.data.frame

# sQTLs
lc.01.ukbb.coloc.df = data.frame()
for (i in 1:nrow(lc.fdr05.sig_ukbb.range)){
  
  range_info = lc.fdr05.sig_ukbb.range[i,]
  
  pheno_info = pheno_codes %>% filter(phenocode==range_info$pheno)
  
  pheno_file = paste0("../gwas/ranges/sqtl/", range_info$intron, "_", range_info$pheno, ".tsv")
  
  ukbb = read_tsv(pheno_file)
  
  lc.01.ukbb = lc.01.sgenes %>%
    filter(intron==range_info$intron) %>%
    select(chr, pos = pos_hg19, alt = allele0, ref = allele1, beta, se, maf, p.value) %>%
    merge(ukbb)
  
  if (range_info$pheno=="FEV1FVC"){
    
    lc.01.ukbb = lc.01.ukbb %>%
      rename(beta_meta = beta_EUR, se_meta = se_EUR)
    
    ukbb.maf = (lc.01.ukbb$af_EUR * pheno_info$n_cases_full_cohort_both_sexes)/
      pheno_info$n_cases_full_cohort_both_sexes
    
    lc.01.ukbb$ukbb_maf = ukbb.maf
    
  } else{
    
    if (range_info$pheno=="401" |
        range_info$pheno=="401.1" |
        range_info$pheno=="1747"){
      
      ukbb.af = lc.01.ukbb %>%
        select(chr, pos , starts_with("af")) %>%
        select(-contains("meta")) %>%
        gather(key=pop, value=af, -c(chr, pos)) %>%
        separate(pop, into=c("foo","bar","pop"), "_") %>%
        select(-c(foo, bar))
      
    } else{
      
      ukbb.af = lc.01.ukbb %>%
        select(chr, pos , starts_with("af")) %>%
        select(-contains("meta")) %>%
        gather(key=pop, value=af, -c(chr, pos)) %>%
        separate(pop, into=c("foo","pop"), "_") %>%
        select(-foo)
      
    }
    
    ukbb.n_case = pheno_info %>%
      select(contains("cases")) %>%
      select(-contains("full")) %>%
      gather(key = pop, value = n_case) %>%
      separate(pop, into=c("foo","bar","pop")) %>%
      select(-c(foo, bar))
    
    ukbb.n_control = pheno_info %>%
      select(contains("control")) %>%
      select(-contains("full")) %>%
      gather(key = pop, value = n_control) %>%
      separate(pop, into=c("foo","bar","pop")) %>%
      select(-c(foo, bar))
    
    ukbb.maf = merge(ukbb.af, ukbb.n_case) %>%
      merge(ukbb.n_control) %>%
      mutate(n_case = ifelse(is.na(n_case), 0, n_case),
             n_control = ifelse(is.na(n_control), 0, n_control)) %>%         
      mutate(af = af*(n_case+ n_control)/(n_case + n_control)) %>%
      mutate(ukbb_maf = ifelse(af < 0.5, af, 1-af)) %>%
      select(pos, chr, ukbb_maf)
    
    lc.01.ukbb = merge(lc.01.ukbb, ukbb.maf)
    
  }
  
  lc.01.ukbb = lc.01.ukbb %>%
    select(beta, se, maf, beta_meta, se_meta, ukbb_maf)
  lc.01.ukbb = lc.01.ukbb[complete.cases(lc.01.ukbb),]
  
  
  lc.01.ukbb.coloc = coloc.abf(dataset1=list(beta = lc.01.ukbb$beta_meta,
                                             varbeta = lc.01.ukbb$se_meta^2,
                                             MAF = lc.01.ukbb$ukbb_maf,
                                             type="quant",
                                             N=pheno_info$n_cases_full_cohort_both_sexes),
                               dataset2=list(beta = lc.01.ukbb$beta,
                                             varbeta = lc.01.ukbb$se^2,
                                             MAF = lc.01.ukbb$maf,
                                             type="quant",
                                             N=162))
  
  lc.01.ukbb.coloc.df = rbind(lc.01.ukbb.coloc.df,
                              data.frame(intron = range_info$intron,
                                         gene_id = range_info$gene_id,
                                         pheno = range_info$pheno,
                                         H0 = lc.01.ukbb.coloc$summary[2],
                                         H1 = lc.01.ukbb.coloc$summary[3],
                                         H2 = lc.01.ukbb.coloc$summary[4],
                                         H3 = lc.01.ukbb.coloc$summary[5],
                                         H4 = lc.01.ukbb.coloc$summary[6]))
  
}

lc.01.ukbb.coloc.df = write_tsv(lc.01.ukbb.coloc.df, "lc.01.ukbb.coloc.maf01.tsv")

```


## colocalization of nearby pigmentation eGenees

```{r}

pigment = read_tsv("../pigmentation/5M.hg19.FlipRef.DDB1.1Mb.mergedrefs.mm.dose.ps.maf.txt.gz",
                   col_names = c("snp_hg19","gwas_beta","gwas_se","gwas_pval","gwas_maf"),
                   col_types="cnnnn") %>%
  group_by(snp_hg19) %>%
  # get unique SNPs
  mutate(n = n()) %>% 
  filter(n==1) %>%
  select(-n) %>%
  separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T, remove=F)

pigment = liftover_snp(addchr(pigment$chr),
                       pigment$pos_hg19,
                       "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  rename(pos_hg19 = pos_old, pos_hg38 = pos_new) %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment)

pigment.sig = pigment %>%
  filter(gwas_pval < 5 * 10^-8)

pigment_genes = gencode.pc_linc %>%
  filter(chr==11) %>%
  filter((start > 60137147 & start < 62137147) |
           (end > 60137147 & end < 62137147)) %>%
  pull(gene_id)

pigment_genes.v26 = gene_map %>%
  filter(gene_id %in% pigment_genes) %>%
  pull(v26_id)
pigment_genes.v26 = pigment_genes.v26[!is.na(pigment_genes.v26)]

# tmem216 eQTLs that are also pigmentation snps
pigment.tmem216_v8 = liftover_snp(addchr(pigment.sig$chr),
             pigment.sig$pos_hg19,
             "/local1/derek/data/genomes/hg19ToHg38.over.chain") %>%
  dplyr::rename(pos_hg38 = pos_new, pos_hg19 = pos_old) %>%
  merge(tmem216_v8)

# get GTEx v8 data
v8_wb_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Whole_Blood.allpairs.sort.txt.gz",
                         paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                         c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                           "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))


v8_se_pigment = tabix2df("../gtex/v8/GTEx_Analysis_v8_eQTL_all_associations_Skin_Sun_Exposed_Lower_leg.allpairs.sort.txt.gz",
                      paste(gene_map[gene_map$gene_id %in% pigment_genes,]$v26_id, "1-100000000", sep=":"),
                      c("v26_id","chr","pos_hg38","allele1","allele2","foo","tss_dist",
                        "ma_samples","ma_count","v8_maf","v8_pval","v8_slope","v8_se"))

# get eGenes
v8_wb_egenes = read_tsv("../gtex/v8/GTEx_Analysis_v8_eQTL/Whole_Blood.v8.egenes.txt.gz")

v8_se_egenes = read_tsv("../gtex/v8/GTEx_Analysis_v8_eQTL/Skin_Sun_Exposed_Lower_leg.v8.egenes.txt.gz")

# get 1kgp from CEU (to generate LD)
cols_1kg = readLines("/local1/derek/data/1kgenomes/vcf/1kg_cols.txt")

ceu_samps = read_tsv("/local1/derek/data/1kgenomes/1khg_sample_pop.tsv",
                     col_names = c("sample", "pop")) %>%
  filter(pop == "CEU" &
           sample %in% cols_1kg) %>%
  pull(sample)

v8_ceu_geno = tabix2df("/local1/derek/data/1kgenomes/vcf/ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz",
                       tabixify("11", unique(v8_se_pigment$pos_hg38),
                                unique(v8_se_pigment$pos_hg38)),
                       cols_1kg) %>%
  select(c(CHROM, POS, REF, ALT, ceu_samps)) %>%
  unite("snp_id", CHROM, POS, REF, ALT) %>%
  unique %>%
  gather(key = sample, value = genotype, -snp_id) %>%
  mutate(genotype = ifelse(genotype=="0|0", "0", genotype),
         genotype = ifelse(genotype=="0|1", "1", genotype),
         genotype = ifelse(genotype=="1|0", "1", genotype),
         genotype = ifelse(genotype=="1|1", "2", genotype),
         genotype = as.numeric(genotype)) %>%
  pivot_wider(names_from = sample, values_from = genotype) %>%
  separate(snp_id, into=c("chr", "pos_hg19", "ref", "alt"), sep="_", convert=T)


# plot
v8_wb_pigment = v8_wb_pigment %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment) %>%
  merge(gene_map) %>%
  merge(gene_name) %>%
  merge(v8_wb_egenes %>% select(v26_id = gene_id, pval_nominal_threshold)) %>%
  group_by(gene_id) %>%
  filter(min(v8_pval) < pval_nominal_threshold)

v8_wb_pigment.plot = ggplot(v8_wb_pigment, aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point(alpha=0.5) +
  geom_hline(aes(yintercept = -log10(pval_nominal_threshold)), linetype="dashed", color="red") +
  geom_vline(aes(xintercept = -log10(5*10^-8)), linetype="dashed", color="red") +
  theme_bw(base_size=12) +
  theme(legend.position="none",
        axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 10)) +
  ylab(expression(~ -log[10](P[eQTL]))) +
  xlab(expression(~ -log[10](P[GWAS]))) +
  ggtitle("Pigmentation GWAS vs. GTEx v8 Whole Blood eQTL") +
  facet_wrap(~ gene_name, scales="free_y")

v8_wb_pigment.coloc.df = data.frame()
for (x in unique(v8_wb_pigment$gene_id)){
  
  df = v8_wb_pigment %>%
    filter(gene_id==x)
  
  v8_wb_pigment.coloc = coloc.abf(dataset1=list(beta = df$gwas_beta,
                                                varbeta = df$gwas_se^2,
                                                MAF = df$gwas_maf,
                                                type="quant",
                                                N=1570),
                                  dataset2=list(beta = df$v8_slope,
                                                varbeta = df$v8_se^2,
                                                MAF = df$v8_maf,
                                                type="quant",
                                                N=670))
  
  v8_wb_pigment.coloc.df = rbind(v8_wb_pigment.coloc.df,
                                 data.frame(gene_id = x,
                                            H4 = v8_wb_pigment.coloc$summary[6]))
  
}

v8_se_pigment = v8_se_pigment %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment) %>%
  merge(gene_map) %>%
  merge(gene_name) %>%
  merge(v8_se_egenes %>% select(v26_id = gene_id, pval_nominal_threshold)) %>%
  group_by(gene_id) %>%
  filter(min(v8_pval) < pval_nominal_threshold)

v8_se_pigment.plot = ggplot(v8_se_pigment, aes(x=-log10(gwas_pval), y=-log10(v8_pval))) +
  geom_point(alpha=0.5) +
  geom_hline(aes(yintercept = -log10(pval_nominal_threshold)), linetype="dashed", color="red") +
  geom_vline(aes(xintercept = -log10(5*10^-8)), linetype="dashed", color="red") +
  theme_bw(base_size=12) +
  theme(legend.position="none",
        axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 10)) +
  ylab(expression(~ -log[10](P[eQTL]))) +
  xlab(expression(~ -log[10](P[GWAS]))) +
  ggtitle("Pigmentation GWAS vs. GTEx v8 Sun Exposed Skin eQTL") +
  facet_wrap(~ gene_name, scales="free_y")

v8_se_pigment.coloc.df = data.frame()
for (x in unique(v8_se_pigment$gene_id)){
  
  df = v8_se_pigment %>%
    filter(gene_id==x)
  
  v8_se_pigment.coloc = coloc.abf(dataset1=list(beta = df$gwas_beta,
                                                varbeta = df$gwas_se^2,
                                                MAF = df$gwas_maf,
                                                type="quant",
                                                N=1570),
                                  dataset2=list(beta = df$v8_slope,
                                                varbeta = df$v8_se^2,
                                                MAF = df$v8_maf,
                                                type="quant",
                                                N=670))
  
  v8_se_pigment.coloc.df = rbind(v8_se_pigment.coloc.df,
                                 data.frame(gene_id = x,
                                            H4 = v8_se_pigment.coloc$summary[6]))
  
}

# Africans
gemma.genes.pigment = gemma.egenes %>%
  merge(pigment) %>%
  merge(gene_name)
  
ggplot(gemma.genes.pigment, aes(x=-log10(gwas_pval), y=-log10(p.value))) +
  geom_point() +
  # geom_hline(aes(yintercept = -log10(pval_nominal_threshold)), linetype="dashed", color="red") +
  geom_vline(aes(xintercept = -log10(5*10^-8)), linetype="dashed", color="red") +
  theme_bw(base_size=12) +
  facet_wrap(~ gene_name, scales="free_y")

v8_wb_pigment.gf = gtable_frame(ggplotGrob(v8_wb_pigment.plot))
v8_se_pigment.gf = gtable_frame(ggplotGrob(v8_se_pigment.plot))

# v8_wb_se_pigment.gf = gtable_frame(gtable_rbind(v8_wb_pigment.gf, v8_se_pigment.gf))
               # width = unit(2, "null"),
               # height = unit(1, "null"))

gtable_cbind(v8_wb_pigment.gf, v8_se_pigment.gf)


grid.arrange(grobs = lapply(
  list(v8_wb_pigment.plot, v8_se_pigment.plot),
  set_panel_size,
  width = unit(2.5, "cm"),
  height = unit(2.5, "cm")
),
ncol=2,
labels=c("A","B"))

## combine plots
v8_wb_se_pigment.plot = plot_grid(v8_wb_pigment.plot,
                                  v8_se_pigment.plot,
                                  align = 'v',
                                  labels = c('A', 'B'),
                                  label_size = 12)



## plot African genes

# eQTL file and column names
gemma.f = "../gemma/round2/Sample_ALL.secondpass.rsem.genes.results.tpm.covars.peer20.resid.gemma.100kb.lmm4.r2.txt.gz"
gemma_cols = c("snp_hg19","v19_id","foo","bar","nmiss","allele1","allele0","af","beta","se","logl_H1","l_remle","l_mle","p_wald","p.value","p_score","r2")

# read in and filter raw eQTL data
gemma.pigment = fread(gemma.f, col.names=gemma_cols, na.string=".") %>%
  select(-c(foo, bar)) %>%
  mutate(gene_id = strip_ensembl(v19_id)) %>%
  filter(gene_id %in% pigment_genes) %>%
  separate(snp_hg19, into=c("chr", "pos_hg19"), sep=":", remove=F, convert=T) %>%
  filter(gene_id %in% gencode.pc_linc$gene_id) %>% # pc and lncRNAs
  mutate(eqtl_hg19 = paste(gene_id, pos_hg19, sep=":"))


gemma.pigment %>%
  as.data.frame %>%
  merge(pigment) %>%
  merge(gene_name) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(p.value))) +
  geom_point() +
  theme_classic() +
  facet_wrap(~ gene_name)


gemma.maxp = gemma.egenes %>%
  filter(fdr_sig) %>%
  group_by(gene_id) %>%
  summarise(maxp = max(p.value))

gemma_pigment.plot = gemma.egenes %>%
  filter(gene_id %in% pigment_genes) %>%
  merge(pigment) %>%
  merge(gene_name) %>%
  merge(gemma.maxp) %>%
  ggplot(aes(x=-log10(gwas_pval), y=-log10(p.value))) +
  geom_point(alpha=0.5) +
  geom_hline(aes(yintercept = -log10(maxp)), linetype="dashed", color="red") +
  geom_vline(aes(xintercept = -log10(5*10^-8)), linetype="dashed", color="red") +
  theme_bw(base_size=12) +
  theme(legend.position="none",
        axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 10)) +
  ylab(expression(~ -log[10](P[eQTL]))) +
  xlab(expression(~ -log[10](P[GWAS]))) +
  ggtitle("Pigmentation GWAS vs. African Whole Blood eQTL") +
  facet_wrap(~ gene_name, scales="free_y")
  
grid.arrange(grobs = lapply(
  list(gemma_pigment.plot),
  set_panel_size,
  width = unit(5, "cm"),
  height = unit(5, "cm")
))

```


## colocalizatio of nearby pigmentation sGenes

```{r}

v8_wb_sgenes = read_tsv("../gtex/v8/GTEx_Analysis_v8_sQTL/Whole_Blood.v8.sgenes.txt.gz")
v8_se_sgenes = read_tsv("../gtex/v8/GTEx_Analysis_v8_sQTL/Skin_Sun_Exposed_Lower_leg.v8.sgenes.txt.gz")

gtex.v8.sqtl.f = "../../gtex/v8/GTEx_Analysis_v8_sQTL_all_associations_Whole_Blood.v8.sqtl_allpairs.sort.txt.gz"
gtex.v8.sqtl.cols = c("gtexv8_intron","chr","pos_hg38","gtexv8_clust","gtexv8_snp","tss_dist","ma_samples","ma_count","gtexv8_maf","gtexv8_pval","gtexv8_slope","gtexv8_se")

v8_wb_pigment.s = read_tsv("../gtex/v8/GTEx_Analysis_v8_sQTL_all_associations_Whole_Blood.v8.sqtl_allpairs.chr11.txt.gz",
                           col_names = c("gtexv8_intron","chr","pos_hg38","phenotype_id","gtexv8_snp","tss_dist",
                                         "ma_samples","ma_count","gtexv8_maf","gtexv8_pval","gtexv8_slope","gtexv8_se")) %>%
  filter(phenotype_id %in% v8_wb_sgenes$phenotype_id) %>%
  separate(gtexv8_intron, into=c("chr","start","end","v26_id"), sep="_", convert=T, remove=F) %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  filter(gene_id %in% pigment_genes)

v8_se_pigment.s = read_tsv("../gtex/v8/GTEx_Analysis_v8_QTLs_GTEx_Analysis_v8_sQTL_all_associations_Skin_Sun_Exposed_Lower_leg.v8.chr11.txt.gz",
                           col_names = c("phenotype_id","variant_id","tss_dist","ma_samples","ma_count",
                                         "gtexv8_maf","gtexv8_pval","gtexv8_slope","gtexv8_se")) %>%
  filter(phenotype_id %in% v8_se_sgenes$phenotype_id) %>%
  separate(phenotype_id, into=c("chr","start","end","clust","v26_id"), sep=":", convert=T, remove=F) %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  filter(gene_id %in% pigment_genes) %>%
  separate(variant_id, into=c("chr","pos_hg38","v8_ref","v8_alt","foo"), sep="_", convert=T) %>%
  select(-foo)

v8_wb_pigment.s = v8_wb_pigment.s %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment) %>%
  merge(gene_name) %>%
  merge(v8_wb_sgenes %>% select(phenotype_id, pval_nominal_threshold))
  group_by(phenotype_id) %>%
  filter(min(gtexv8_pval) < pval_nominal_threshold)


v8_se_pigment.s = v8_se_pigment.s %>%
  mutate(chr = rmchr(chr)) %>%
  merge(pigment) %>%
  merge(gene_name) %>%
  merge(v8_se_sgenes %>% select(phenotype_id, pval_nominal_threshold)) %>%
  group_by(phenotype_id) %>%
  filter(min(gtexv8_pval) < pval_nominal_threshold)


v8_wb_pigment.s.plot = ggplot(v8_wb_pigment.s, aes(x=-log10(gwas_pval), y=-log10(gtexv8_pval))) +
  geom_point(alpha=0.5) +
  geom_hline(aes(yintercept = -log10(pval_nominal_threshold)), linetype="dashed", color="red") +
  geom_vline(aes(xintercept = -log10(5*10^-8)), linetype="dashed", color="red") +
  theme_bw(base_size=12) +
  theme(legend.position="none",
        plot.title = element_text(size = 12),
        axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 10)) +
  ylab(expression(~ -log[10](P[sQTL]))) +
  xlab(expression(~ -log[10](P[GWAS]))) +
  ggtitle("Pigmentation GWAS vs. GTEx v8 Whole Blood sQTL") +
  facet_wrap(~ gene_name, scales="free_y")

v8_se_pigment.s.plot = ggplot(v8_se_pigment.s, aes(x=-log10(gwas_pval), y=-log10(gtexv8_pval))) +
  geom_point(alpha=0.5) +
  geom_hline(aes(yintercept = -log10(pval_nominal_threshold)), linetype="dashed", color="red") +
  geom_vline(aes(xintercept = -log10(5*10^-8)), linetype="dashed", color="red") +
  theme_bw(base_size=12) +
  theme(legend.position="none",
        plot.title = element_text(size = 12),
        axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        strip.background = element_blank(),
        strip.text.x = element_text(size = 10)) +
  ylab(expression(~ -log[10](P[sQTL]))) +
  xlab(expression(~ -log[10](P[GWAS]))) +
  ggtitle("Pigmentation GWAS vs. GTEx v8 Sun Exposed Skin sQTL") +
  facet_wrap(~ gene_name, scales="free_y")


v8_wb_se_pigment.s.plot = plot_grid(v8_wb_pigment.s.plot,
                                    v8_se_pigment.s.plot,
                                    labels = c('A', 'B'),
                                    rel_widths = c(3,4),
                                    label_size = 12)

grid.arrange(grobs = lapply(
  list(v8_wb_pigment.s.plot, v8_se_pigment.s.plot),
  set_panel_size,
  width = unit(2.5, "cm"),
  height = unit(2.5, "cm")
),
ncol=2,
labels=c("A","B"))

```


## susie colocalization at CYB561A3

```{r}

# things that are needed by Susie:
# 1. 

tmem216.eur.se.df = liftover_snp(v8_se_pigment$chr,
                                  v8_se_pigment$pos_hg38,
                                  "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  unique %>%
  dplyr::rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  merge(v8_se_pigment %>%
          select(gene_id = v26_id, chr, pos_hg38, v8_maf, v8_slope, v8_se, v8_pval)) %>%
  mutate(chr = rmchr(chr)) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(pigment)

cyb561a3_range = gencode.pc_linc %>%
  filter(gene_id=="ENSG00000162144") %>%
  mutate(tss = ifelse(strand=="+", start, end)) %>%
  summarise(start = tss - 1000000,
            end = tss + 1000000)

cyb561a3.eur.se.df = tmem216.eur.se.df %>%
  filter(gene_id=="ENSG00000162144.9") %>%
  filter(pos_hg19 < max(cyb561a3_range) &
           pos_hg19 > min(cyb561a3_range)) %>%
  # eliminate fixed snps
  filter(gwas_maf > 0.001 & v8_maf > 0) %>%
  # restrict to biallelic SNPs
  group_by(snp_hg19) %>%
  mutate(nsnps = n()) %>%
  filter(nsnps==1) %>%
  select(-nsnps)


# European LD
gtype_cyb561a3.eur.df = tabix2df("/local1/derek/data/1kgenomes/vcf/ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz",
                              paste0("11:",min(cyb561a3.eur.se.df$pos_hg19),
                                     "-",max(cyb561a3.eur.se.df$pos_hg19)),
                              cols_1kg) %>%
  unite("snp_hg19", CHROM, POS, sep=":") %>%
  # get EUR individuals
  select(snp_hg19, ID, pop_samps.1kg %>% filter(superpop=="EUR") %>% pull(sample)) %>%
  # only SNPs in eQTL/GWAS data
  filter(snp_hg19 %in% cyb561a3.eur.se.df$snp_hg19) %>%
  # restrict to biallelic SNPs
  group_by(snp_hg19) %>%
  mutate(nsnps = n()) %>%
  filter(nsnps==1) %>%
  select(-nsnps) %>%
  # convert genotypes into allele counts
  gather(key=sample, value=gtype, -c(snp_hg19, ID)) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  mutate(all_cnt = gtype1 + gtype2) %>%
  select(-c(gtype1, gtype2)) %>%
  # remove fixed variants
  group_by(ID) %>%
  filter(mean(all_cnt) > 0 & mean(all_cnt) < 2) %>%
  ungroup %>%
  pivot_wider(names_from=sample, values_from=all_cnt)

# East African LD
cols_5M  = readLines("/local1/derek/data/Illumina5M/5M_cols.txt")
gtype_tmem216.eafr.df = tabix2df("/local1/derek/data/Illumina5M/imputed/11.5M.imputed.vcf.gz",
                              paste0("11:",min(cyb561a3.eur.se.df$pos_hg19),
                                     "-",max(cyb561a3.eur.se.df$pos_hg19)),
                              cols_5M) %>%
  select(snp_hg19 = ID, pop_samps.5M$sample) %>%
  filter(snp_hg19 %in% cyb561a3.eur.se.df$snp_hg19) %>%
  gather(key=sample, value=genotype, -snp_hg19) %>%
  separate(genotype, into=c("gtype","dose"), sep=":", convert=T) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  mutate(all_cnt = gtype1 + gtype2) %>%
  select(-c(gtype1, gtype2, dose)) %>% 
  # remove fixed variants
  group_by(snp_hg19) %>%
  filter(mean(all_cnt) > 0 & mean(all_cnt) < 2) %>%
  ungroup %>%
  pivot_wider(names_from=sample, values_from=all_cnt)

# restrict to SNPs shared by all
snps_keep = intersect(intersect(gtype_cyb561a3.eur.df$snp_hg19,
                                cyb561a3.eur.se.df$snp_hg19),
                      gtype_tmem216.eafr.df$snp_hg19)

gtype_cyb561a3.eur.df = gtype_cyb561a3.eur.df %>%
  filter(snp_hg19 %in% snps_keep)
                      
cyb561a3.eur.se.df = cyb561a3.eur.se.df %>%
  filter(snp_hg19 %in% snps_keep)

gtype_tmem216.eafr.df = gtype_tmem216.eafr.df %>%
  filter(snp_hg19 %in% snps_keep)

# calculate EUR LD
cyb561a3.r2_eur = gtype_cyb561a3.eur.df %>%
  ungroup %>%
  select(-ID) %>%
  column_to_rownames("snp_hg19") %>%
  as.matrix %>%
  t %>%
  cor# %>%
  .**2

# calculate EAfr LD
r2_eafr = gtype_tmem216.eafr.df %>%
  column_to_rownames("snp_hg19") %>%
  as.matrix %>%
  t %>%
  cor# %>%
  .**2


# data for susie
pigment_cyb561a3.coloc = list(
  beta = cyb561a3.eur.se.df$gwas_beta,
  varbeta = cyb561a3.eur.se.df$gwas_se^2,
  N = 1570,
  type = "quant",
  MAF = cyb561a3.eur.se.df$gwas_maf,
  LD = r2_eafr,
  snp = cyb561a3.eur.se.df$snp_hg19
)

v8_cyb561a3.coloc = list(
  beta = cyb561a3.eur.se.df$v8_slope,
  varbeta = cyb561a3.eur.se.df$v8_se^2,
  N = 605,
  type = "quant",
  MAF = cyb561a3.eur.se.df$v8_maf,
  LD = cyb561a3.r2_eur,
  snp = cyb561a3.eur.se.df$snp_hg19
)

# test coloc
pigment_v8.coloc <- coloc.abf(dataset1=pigment_cyb561a3.coloc,
                              dataset2=v8_cyb561a3.coloc)

# run susie on each
pigment_cyb561a3.susie = runsusie(pigment_cyb561a3.coloc)
v8_cyb561a3.susie = runsusie(v8_cyb561a3.coloc, coverage=0.1)

pigment_v8.susie = coloc.susie(pigment_cyb561a3.susie,
                                         v8_cyb561a3.susie)


```


## susie colocalization at TMEM216

```{r}

# things that are needed by Susie:
# 1. 

tmem216.eur.se.df = liftover_snp(v8_se_pigment$chr,
                                  v8_se_pigment$pos_hg38,
                                  "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  unique %>%
  dplyr::rename(pos_hg38 = pos_old, pos_hg19 = pos_new) %>%
  merge(v8_se_pigment %>%
          select(gene_id = v26_id, chr, pos_hg38, v8_maf, v8_slope, v8_se, v8_pval)) %>%
  mutate(chr = rmchr(chr)) %>%
  mutate(snp_hg19 = paste(chr, pos_hg19, sep=":")) %>%
  merge(pigment)

tmem216_range = gencode.pc_linc %>%
  filter(gene_id=="ENSG00000187049") %>%
  mutate(tss = ifelse(strand=="+", start, end)) %>%
  summarise(start = tss - 1000000,
            end = tss + 1000000)

tmem216.eur.se.df = tmem216.eur.se.df %>%
  filter(gene_id=="ENSG00000187049.9") %>%
  filter(pos_hg19 < max(tmem216_range) &
           pos_hg19 > min(tmem216_range)) %>%
  # eliminate fixed snps
  filter(gwas_maf > 0.001 & v8_maf > 0) %>%
  # restrict to biallelic SNPs
  group_by(snp_hg19) %>%
  mutate(nsnps = n()) %>%
  filter(nsnps==1) %>%
  select(-nsnps)


# European LD
gtype_tmem216.eur.df = tabix2df("/local1/derek/data/1kgenomes/vcf/ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz",
                              paste0("11:",min(tmem216.eur.se.df$pos_hg19),
                                     "-",max(tmem216.eur.se.df$pos_hg19)),
                              cols_1kg) %>%
  unite("snp_hg19", CHROM, POS, sep=":") %>%
  # get EUR individuals
  select(snp_hg19, ID, pop_samps.1kg %>% filter(superpop=="EUR") %>% pull(sample)) %>%
  # only SNPs in eQTL/GWAS data
  filter(snp_hg19 %in% tmem216.eur.se.df$snp_hg19) %>%
  # restrict to biallelic SNPs
  group_by(snp_hg19) %>%
  mutate(nsnps = n()) %>%
  filter(nsnps==1) %>%
  select(-nsnps) %>%
  # convert genotypes into allele counts
  gather(key=sample, value=gtype, -c(snp_hg19, ID)) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  mutate(all_cnt = gtype1 + gtype2) %>%
  select(-c(gtype1, gtype2)) %>%
  # remove fixed variants
  group_by(ID) %>%
  filter(mean(all_cnt) > 0 & mean(all_cnt) < 2) %>%
  ungroup %>%
  pivot_wider(names_from=sample, values_from=all_cnt)

# East African LD
cols_5M  = readLines("/local1/derek/data/Illumina5M/5M_cols.txt")
gtype_tmem216.eafr.df = tabix2df("/local1/derek/data/Illumina5M/imputed/11.5M.imputed.vcf.gz",
                              paste0("11:",min(tmem216.eur.se.df$pos_hg19),
                                     "-",max(tmem216.eur.se.df$pos_hg19)),
                              cols_5M) %>%
  select(snp_hg19 = ID, pop_samps.5M$sample) %>%
  filter(snp_hg19 %in% tmem216.eur.se.df$snp_hg19) %>%
  gather(key=sample, value=genotype, -snp_hg19) %>%
  separate(genotype, into=c("gtype","dose"), sep=":", convert=T) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  mutate(all_cnt = gtype1 + gtype2) %>%
  select(-c(gtype1, gtype2, dose)) %>% 
  # remove fixed variants
  group_by(snp_hg19) %>%
  filter(mean(all_cnt) > 0 & mean(all_cnt) < 2) %>%
  ungroup %>%
  pivot_wider(names_from=sample, values_from=all_cnt)

# restrict to SNPs shared by all
snps_keep = intersect(intersect(intersect(gtype_tmem216.eur.df$snp_hg19,
                                          tmem216.eur.se.df$snp_hg19),
                                gtype_tmem216.eafr.df$snp_hg19),
                      tmem216.eafr.df$snp_hg19)

gtype_tmem216.eur.df = gtype_tmem216.eur.df %>%
  filter(snp_hg19 %in% snps_keep)
                      
tmem216.eur.se.df = tmem216.eur.se.df %>%
  filter(snp_hg19 %in% snps_keep)

gtype_tmem216.eafr.df = gtype_tmem216.eafr.df %>%
  filter(snp_hg19 %in% snps_keep)

# calculate EUR LD
tmem216.r2_eur = gtype_tmem216.eur.df %>%
  ungroup %>%
  select(-ID) %>%
  column_to_rownames("snp_hg19") %>%
  as.matrix %>%
  t %>%
  cor #%>%
  .**2

# calculate EAfr LD
r2_eafr = gtype_tmem216.eafr.df %>%
  column_to_rownames("snp_hg19") %>%
  as.matrix %>%
  t %>%
  cor #%>%
  .**2


# data for susie
pigment_tmem216.coloc = list(
  beta = tmem216.eur.se.df$gwas_beta,
  varbeta = tmem216.eur.se.df$gwas_se^2,
  N = 1570,
  type = "quant",
  MAF = tmem216.eur.se.df$gwas_maf,
  LD = r2_eafr,
  snp = tmem216.eur.se.df$snp_hg19
)

v8_tmem216.coloc = list(
  beta = tmem216.eur.se.df$v8_slope,
  varbeta = tmem216.eur.se.df$v8_se^2,
  N = 605,
  type = "quant",
  MAF = tmem216.eur.se.df$v8_maf,
  LD = tmem216.r2_eur,
  snp = tmem216.eur.se.df$snp_hg19
)

tmem216.eafr.uniq.df = tmem216.eafr.df %>%
  filter(snp=="11:61137147")
eafr_tmem216.coloc = list(
  beta = tmem216.eafr.uniq.df$beta,
  varbeta = tmem216.eafr.uniq.df$se^2,
  N = 162,
  type = "quant",
  MAF = tmem216.eafr.uniq.df$maf,
  LD = r2_eafr,
  snp = tmem216.eafr.uniq.df$snp_hg19
)

# test coloc
pigment_v8.coloc <- coloc.abf(dataset1=pigment_tmem216.coloc,
                              dataset2=v8_tmem216.coloc)

pigment_eafr.coloc <- coloc.abf(dataset1=pigment_tmem216.coloc,
                              dataset2=eafr_tmem216.coloc)

# run susie on each
pigment_tmem216.susie = runsusie(pigment_tmem216.coloc, coverage=0.1)
v8_tmem216.susie = runsusie(v8_tmem216.coloc, coverage=0.1)
eafr_tmem216.susie = runsusie(eafr_tmem216.coloc, coverage=0.1)

pigment_v8.susie = coloc.susie(pigment_tmem216.susie,
                                         v8_tmem216.susie)


```


## colocalization of PBS, d-statistic, and eQTL p-valules at TMEM216

```{r}

gemma.tmem216 = gemma.egenes %>%
  filter(gene_id == "ENSG00000187049")

dstat.tmem216 = tabix2df("/local1/derek/data/Illumina5M/fst/ALL.5M.imputed.rna_samps.biallelic.dstat.directional.txt.gz",
                         tabixify(gemma.tmem216$chr,
                                  gemma.tmem216$pos_hg19,
                                  gemma.tmem216$pos_hg19),
                         colnames = c("chr","pos_hg19","ref","alt",pops))

pbs.tmem216 = tabix2df("../selection/ALL.5M.imputed.rna_samps.biallelic.hud.pbs.gz",
                       tabixify(gemma.tmem216$chr,
                                gemma.tmem216$pos_hg19,
                                gemma.tmem216$pos_hg19),
                       colnames = c("chr","pos_hg19","ref","alt",pops))

tmem216.d_eqtl = gemma.tmem216 %>%
  select(chr, pos_hg19, p.value) %>%
  merge(r2_tmem216.eafr.rs7948623 %>%
          select(snp_hg19, rs7948623 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(r2_tmem216.eafr.rs11230664 %>%
          select(snp_hg19, rs11230664 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(r2_tmem216.eafr.rs3741265 %>%
          select(snp_hg19, rs3741265 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(dstat.tmem216 %>% select(chr, pos_hg19, d = mursi)) %>%
  gather(key = snp, value = r2, c(rs7948623, rs11230664, rs3741265)) %>%
  mutate(snp = factor(snp, levels = c("rs7948623", "rs11230664", "rs3741265"))) %>%
  ggplot(aes(x=abs(d), y=-log10(p.value), color=r2)) + 
  geom_point() +
  ylab(expression(~ -log[10](P[eQTL]))) +
  xlab(expression(~ d[Mursi])) +
  theme_bw(base_size=15) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  labs(color = expression(~ r^2)) +
  scale_color_distiller(palette = "Spectral") +
  facet_wrap(~ snp)

tmem216.pbs_eqtl = gemma.tmem216 %>%
  select(chr, pos_hg19, p.value) %>%
  merge(r2_tmem216.eafr.rs7948623 %>%
          select(snp_hg19, rs7948623 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(r2_tmem216.eafr.rs11230664 %>%
          select(snp_hg19, rs11230664 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
    merge(r2_tmem216.eafr.rs3741265 %>%
          select(snp_hg19, rs3741265 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(pbs.tmem216 %>% select(chr, pos_hg19, pbs = mursi)) %>%
  gather(key = snp, value = r2, c(rs7948623, rs11230664, rs3741265)) %>%
  mutate(snp = factor(snp, levels = c("rs7948623", "rs11230664", "rs3741265"))) %>%
  ggplot(aes(x=pbs, y=-log10(p.value), color=r2)) + 
  geom_point() +
  ylab(expression(~ -log[10](P[eQTL]))) +
  xlab(expression(~ PBS[Mursi])) +
  theme_bw(base_size=15) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  labs(color = expression(~ r^2)) +
  scale_color_distiller(palette = "Spectral") +
  facet_wrap(~ snp)


# plot together
tmem216.eqtl.legend = get_legend(tmem216.pbs_eqtl + theme(legend.box.margin = margin(0, 0, 0, 0)))

# add legend back
tmem216.d_pbs_eqtl.plot = plot_grid(plot_grid(tmem216.pbs_eqtl + theme(legend.position="none"),
                                              tmem216.d_eqtl + theme(legend.position="none"),
                                              ncol=1),
                                    tmem216.eqtl.legend, nrow=1, rel_widths = c(4,1))



```


## d/pbs and gwas p-values

```{r}

tmem216.d_pigment = pigment %>%
  select(chr, pos_hg19, gwas_pval) %>%
  merge(r2_tmem216.eafr.rs7948623 %>%
          select(snp_hg19, rs7948623 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(r2_tmem216.eafr.rs11230664 %>%
          select(snp_hg19, rs11230664 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(r2_tmem216.eafr.rs3741265 %>%
          select(snp_hg19, rs3741265 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(dstat.tmem216 %>% select(chr, pos_hg19, d = mursi)) %>%
  gather(key = snp, value = r2, c(rs7948623, rs11230664, rs3741265)) %>%
  mutate(snp = factor(snp, levels = c("rs7948623", "rs11230664", "rs3741265"))) %>%
  ggplot(aes(x=abs(d), y=-log10(gwas_pval), color=r2)) + 
  geom_hline(aes(yintercept = -log10(5*10^(-8))), linetype="dashed", color="red", alpha=0.5) +
  geom_point() +
  ylab(expression(~ -log[10](P[GWAS]))) +
  xlab(expression(~ d[Mursi])) +
  theme_bw(base_size=15) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  labs(color = expression(~ r^2)) +
  scale_color_distiller(palette = "Spectral") +
  facet_wrap(~ snp)

tmem216.pbs_pigment = pigment %>%
  select(chr, pos_hg19, gwas_pval) %>%
  merge(r2_tmem216.eafr.rs7948623 %>%
          select(snp_hg19, rs7948623 = r2) %>% 
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(r2_tmem216.eafr.rs11230664 %>%
          select(snp_hg19, rs11230664 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(r2_tmem216.eafr.rs3741265 %>%
          select(snp_hg19, rs3741265 = r2) %>%
          separate(snp_hg19, into=c("chr","pos_hg19"), sep=":", convert=T)) %>%
  merge(pbs.tmem216 %>% select(chr, pos_hg19, pbs = mursi)) %>%
  gather(key = snp, value = r2, c(rs7948623, rs11230664, rs3741265)) %>%
  mutate(snp = factor(snp, levels = c("rs7948623", "rs11230664", "rs3741265"))) %>%
  ggplot(aes(x=pbs, y=-log10(gwas_pval), color=r2)) + 
  geom_hline(aes(yintercept = -log10(5*10^(-8))), linetype="dashed", color="red", alpha=0.5) +
  geom_point() +
  ylab(expression(~ -log[10](P[GWAS]))) +
  xlab(expression(~ PBS[Mursi])) +
  theme_bw(base_size=15) +
  theme(strip.background = element_blank(),
        strip.text.x = element_text(size = 12),
        panel.spacing.x=unit(2, "lines")) +
  labs(color = expression(~ r^2)) +
  scale_color_distiller(palette = "Spectral") +
  facet_wrap(~ snp)


# plot together
tmem216.pigment.legend = get_legend(tmem216.pbs_pigment + theme(legend.box.margin = margin(0, 0, 0, 0)))

# add legend back
tmem216.d_pbs_pigment.plot = plot_grid(plot_grid(tmem216.pbs_pigment + theme(legend.position="none"),
                                              tmem216.d_pigment + theme(legend.position="none"),
                                              ncol=1),
                                    tmem216.pigment.legend, nrow=1, rel_widths = c(4,1))

```


## frequencies of TMEM216 variants

```{r}

rs64 = "T light"
rs65 = "A light"
rs23 = "A light"


light_map = data.frame(rsid = rep(c("rs11230664","rs3741265","rs7948623"), each=2),
                       alt_ref = rep(c("aaf","raf"), 3),
                       allele = c("Dark","Light","Light","Dark","Light","Dark"))

tmem_rsmap

tmem216.1kg_frq = tabix2df("../misc/1kgenomes.superpop_af.txt.gz",
                           c("11:61137147-61137147",
                             "11:61076372-61076372",
                             "11:61165280-61165280"),
                           c("chr","pos_hg19","EAS","AMR","AFR","EUR","SAS")) %>%
  gather(key=pop, value=aaf, -c(chr, pos_hg19))

tmem216.5M_frq = tabix2df("/local1/derek/data/Illumina5M/frq/ALL.5M.imputed.rna_samps.biallelic.frq.gz",
                           c("11:61137147-61137147",
                             "11:61076372-61076372",
                             "11:61165280-61165280"),
                           c("chr","pos_hg19","ref","alt",pops)) %>%
  select(-c(ref,alt)) %>%
  gather(key=pop, value=aaf, -c(chr, pos_hg19)) %>%
  mutate(pop = pop_map[pop])

tmem_frqs.plot = rbind(tmem216.1kg_frq, tmem216.5M_frq) %>%
  unite("snp", chr, pos_hg19, sep=":") %>%
  merge(tmem_rsmap) %>%
  mutate(raf = 1 - aaf) %>%
  gather(key = alt_ref, value = frq, -c(snp, pop, rsid)) %>%
  merge(light_map) %>%
  select(-alt_ref) %>%
  mutate(pop = factor(pop, levels=c("Mursi","Chabu","Dizi","Agaw",
                                    "Amhara","Argoba","Hadza","Sandawe",
                                    "Weyto","AFR","SAS","EAS","AMR","EUR")),
         rsid = factor(rsid, levels = c("rs7948623","rs11230664","rs3741265")),
         allele = factor(allele, levels = c("Light","Dark"))) %>%
  ggplot(aes(x=pop, y=frq, fill=allele)) +
  geom_bar(stat="identity") +
  scale_fill_ptol() +
  theme_classic(base_size=15) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggtitle("Global frequencies of TMEM216 SNPs") +
  facet_grid(rsid ~ .)
  


```


## sQTL TMEM216

```{r}

# read in splicing data for TMEM216
v8_se_tmem216.s = read_tsv("GTEx_Analysis_v8_QTLs_GTEx_Analysis_v8_sQTL_all_associations_Skin_Sun_Exposed_Lower_leg.v8.TMEM216_sqtl.txt",
                           col_names = c("phenotype_id","variant_id","tss_distance","v8_ma_samples",
                                         "v8_ma_count","v8_maf","v8_pval","v8_slope","v8_se")) %>%
  separate(variant_id, into=c("chr","pos_hg38","v8_ref","v8_alt","foo"), sep="_", convert=T) %>%
  select(-foo) %>%
  mutate(chr = rmchr(chr))

## add hg19 info
v8_se_tmem216.s = liftover_snp(addchr(v8_se_tmem216.s$chr),
                               v8_se_tmem216.s$pos_hg38,
                               "/local1/derek/data/genomes/hg38ToHg19.over.chain") %>%
  mutate(chr = rmchr(chr)) %>%
  rename(pos_hg19 = pos_new, pos_hg38 = pos_old) %>%
  merge(v8_se_tmem216.s)

v8_se_pig_genes.s = read_tsv("../gtex/v8/GTEx_Analysis_v8_QTLs_GTEx_Analysis_v8_sQTL_all_associations_Skin_Sun_Exposed_Lower_leg.v8.chr11.txt.gz",
                         col_names = c("phenotype_id","variant_id","tss_distance","v8_ma_samples",
                                         "v8_ma_count","v8_maf","v8_pval","v8_slope","v8_se")) %>%
  separate(phenotype_id, into=c("chr","start","end","clust","v26_id"), sep=":", remove=F, convert=T) %>%
  filter(v26_id %in% pigment_genes.v26) %>%
  separate(variant_id, into=c("chr","pos_hg38","v8_ref","v8_alt","genome")) %>%
  select(-c(start, end, clust, genome))


```




Credible Sets

```{r}

# get GTEx v8 credible sets
gtex.v8.cred_set.all = lapply(unique(gtex.v8.egenes.range$gene_id),
                              function(x){
                                df.tmp = gtex.v8.egenes.range %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg38,
                                               df.tmp$gtexv8_slope,
                                               df.tmp$gtexv8_se^2, 0.9)})
names(gtex.v8.cred_set.all) = unique(gtex.v8.egenes.range$gene_id)

# get African credible sets
gemma.cred_set.all = lapply(unique(gemma.egenes$gene_id),
                              function(x){
                                df.tmp = gemma.egenes %>%
                                  filter(gene_id==x)
                                get_cred_set.b(df.tmp$eqtl_hg19,
                                               df.tmp$beta,
                                               df.tmp$se^2, 0.9)})
names(gemma.cred_set.all) = unique(gemma.egenes$gene_id)

# combine credible set info from Africans and GTEx
gemma.v8.cred_set = data.frame(cred_set = unlist(gemma.cred_set.all)) %>%
  # separate into gene ID and position columns
  separate(cred_set, into=c("gene_id", "pos_hg19"), sep=":", convert=T) %>%
  # add chromosome information
  merge(gemma.egenes %>% select(gene_id, pos_hg19, chr)) %>%
  # add hg38 position information, including only those SNPs that map between hg19 and hg38
  merge(hg19_to_hg38, all.x=T) %>%
  # specify that this is in the African credible set
  mutate(afr=T) %>%
  # add GTEx credible set info
  merge(data.frame(cred_set = unlist(gtex.v8.cred_set.all)) %>%
          # separate as above
          separate(cred_set, into=c("gene_id", "pos_hg38"), sep=":", convert=T) %>%
          # specify these are in the GTEx v8 credible set
          mutate(v8=T),
        # eQTLs (gene_id-pos_hg38 pairs) that overlap will merge, those that don't will not
        all=T) %>%
  unique
  

```




Locus plot

```{r}

foo = gemma.egenes %>%
  filter(gene_id=="ENSG00000150991")

bar = gtex.v8.egenes %>%
  mutate(gene_id = strip_ensembl(v26_id)) %>%
  filter(gene_id=="ENSG00000150991") %>%
  merge(hg19_to_hg38)
  

recomb_yri = tabix2df("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/misc/YRI_recombination_map_hg19.bed.gz",
                      tabixify(addchr(foo$chr), min(foo$pos_hg19), max(foo$pos_hg19)),
                      c("chr","start","end","recomb"))

recomb_ceu = tabix2df("/local1/derek/data/transcriptomics/eqtl/afr_wbrna/misc/CEU_recombination_map_hg19.bed.gz",
                      tabixify(addchr(bar$chr), min(bar$pos_hg19), max(bar$pos_hg19)),
                      c("chr","start","end","recomb"))


ggplot(recomb_yri, aes(x=end, y=recomb)) +
  geom_line() +
  geom_point(data=foo, aes(x = as.numeric(pos_hg19), y=-log10(p.value)/10^7)) +
  theme_classic()


    scale_y_continuous(
    
    # Features of the first axis
    name = "First Axis",
    
    # Add a second axis and specify its features
    sec.axis = sec_axis(~.*coeff, name="Second Axis")
  )
```



FST clustering

```{r}

fst_1M = read_tsv("../fst/all_all.weir.fst.1M_samp.clean.txt")

fst_1M.mean = fst_1M %>%
  select(-c(1,2)) %>%
  as.matrix %>%
  apply(.,2,mean)

pops = unique(pop_samps$pop)

fst_1M.mean.mat = data.frame(pops = names(fst_1M.mean),
                            fst = fst_1M.mean) %>%
  separate(pops, into=c("pop1","pop2"), sep="_") %>%
  rbind(., data.frame(pop1=pops, pop2=pops, fst=0)) %>%
  arrange(pop1, pop2) %>%
  pivot_wider(names_from="pop2", values_from="fst") %>%
  column_to_rownames("pop1") %>%
  as.matrix
fst_1M.mean.mat[is.na(fst_1M.mean.mat)] = 0
fst_1M.mean.mat = fst_1M.mean.mat + t(fst_1M.mean.mat)

hclust(as.dist(fst_1M.mean.mat), "complete")

```

Mean Fst vs. transcriptional distance

```{r}

mean_fst = read_tsv("../fst/mean_fst.all.txt", col_names=c("pop1","pop2","mean_fst")) %>%
  rbind(., data.frame(pop1=pops, pop2=pops, mean_fst=0))

# because mean_fst only has unique 
mean_fst.mat = mean_fst %>%
  mutate(pop3 = pop2,
         pop2 = pop1,
         pop1 = pop3) %>% # swqp populations
  select(-pop3) %>%
  rbind(., mean_fst) %>%  # bind it to the original mean_fst
  unique %>%
  pivot_wider(names_from=pop2, values_from=mean_fst) %>%
  column_to_rownames("pop1") %>%
  as.matrix

# how different are the transcriptomes of these populations
combat.tmm.mean_pop = combat.wctypes.tmm %>%
  as.data.frame %>%
  rownames_to_column("gene_id") %>%
  gather(key="sample", value="tpm", -gene_id) %>% # make long
  merge(pop_samps.all) %>%            # add population
  group_by(pop, gene_id) %>%          #
  summarise(mean_tpm = mean(tpm)) %>% # get the mean tpm for each population
  pivot_wider(names_from="pop", values_from="mean_tpm")

# %>% # make wide
#   column_to_rownames("gene_id") %>%
#   as.matrix %>%
#   cor

```

TSS enrichment in African cohort vs. GTEx

```{r}

gencode29 = read_tsv("../genomes/gencode.v29.genes.bed",
                     col_names=c("chr","start","end","strand","v29_id")) %>%
  mutate(gene_id = strip_ensembl(v29_id))

gemma_tssdist = gemma.egenes %>%
  filter(top_esnp) %>%
  merge(gencode.pc_linc, by="gene_id") %>%
  mutate(tss_dist = ifelse(strand=="+", pos_hg19 - start, end - pos_hg19)) %>%
  select(tss_dist) %>%
  mutate(study="Africa")

eur162_tssdist = eur162.egenes %>%
  ungroup %>%
  filter(top_esnp) %>%
  select(tss_dist = eur162_dist) %>%
  mutate(study="GTEx")

rbind(gemma_tssdist, eur162_tssdist) %>%
  ggplot(., aes(x=tss_dist, color=study)) +
  geom_freqpoly(aes(y=..density..)) +
  theme_classic(base_size=15) +
  scale_x_continuous(limits = c(-100000, 100000)) +
  theme(axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
  ggtitle("Genomic distance from TSS")

```

beta vs MAF

```{r}

lc.fdr05.top %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af)) %>%
  ggplot(., aes(x=maf, y=beta)) +
  geom_point(alpha=0.5, size=2) +
  theme_classic(base_size=15) + 
  ggtitle("sQTL beta vs. MAF")

gemma.fdr05.top %>%
  mutate(maf = ifelse(af < 0.5, af, 1-af)) %>%
  ggplot(., aes(x=maf, y=beta)) +
  geom_point(alpha=0.5, size=2) +
  theme_classic(base_size=15) + 
  ggtitle("eQTL beta vs. MAF")


```

Manhattan plot, stolen and adapted from http://www.danielroelfs.com/coding/manhattan_plots/

```{r}

nCHR = length(unique(gwas.dat$CHR))
gwas.dat$BPcum = NA
s = 0
nbp = c()
nbp[1] = 0
for (i in 2:22){
  nbp[i] = nbp[i-1] + max(gemma.fdr05.pcadapt[gemma.fdr05.pcadapt$chr==i,]$pos)
}

gemma.fdr05.pcadapt = gemma.fdr05.pcadapt %>%
  mutate(bp_cum = pos + nbp[chr])

axis.set  =  pcadapt %>% 
  group_by(chr) %>% 
  summarize(center = (max(BPcum) + min(BPcum)) / 2)
ylim = abs(floor(log10(min(pcadapt$pcadapt)))) + 2
sig = 0.05 / nrow(pcadapt)

manhplot = ggplot(gwas.dat, aes(x=pos, y=-log10(pcadapt))) +
  geom_point(aes(color=(chr %% 2)), alpha = 0.75, size = 1.25) +
  scale_color_manual(values = c("darkslateblue","cadetblue")) +
  scale_x_continuous(label = axis.set$CHR, breaks = axis.set$center) +
  scale_y_continuous(expand = c(0,0), limits = c(0,ylim)) +
  geom_hline(yintercept = -log10(sig), color = "grey40", linetype = "dashed") + 
  labs(x = NULL, y = "-log10(p)") + 
  theme_minimal() +
  theme( 
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(angle = 60, size = 8, vjust = 0.5)
  )

```

Co-localization of D-statistic and eQTL signals

```{r}

ihs_cols = c("chr","pos","alt_frq","ihh1","ihh0","ihs_raw","ihs_norm","thresh")

threshs = lapply(pops, function(pop) dstat %>% pull(pop) %>% quantile(probs=c(0.99, 0.999)))

gemma.fdr05.dstat = tabix2df(dstat.f, tabixify(lc.fdr05$chr, lc.fdr05$pos, lc.fdr05$pos), dstat_cols) %>% 
  mutate(snp_id = paste(chr, pos,sep=":")) %>%
  select(-c(chr,pos)) %>%
  merge(., lc.fdr05, by="snp_id")

eqtl_dstat_cors = list()
eqtl_ihs_cors = list()
for (pop in pops){
  
  ihs.f = paste0("../ihs/pop_specific", pop, ".5M.imputed.dose.biallelic.ihs.out.100bins.norm.gz")
  
  gemma.fdr05.ihs = tabix2df(ihs.f, tabixify(gemma.fdr05$chr, gemma.fdr05$pos, gemma.fdr05$pos), ihs_cols) %>%
    mutate(snp_id = paste(chr, pos, sep=":")) %>%
    merge(., gemma.fdr05.dstat, by="snp_id")
  
  pop_sig_genes = gemma.fdr05.dstat %>%
    filter(abs(ihs_norm) > 2.5) %>% 
    pull(gene_id) %>%
    unique
  names(pop_sig_genes) = pop_sig_genes
  
  pop_sig.gemma = gemma %>%
    filter(gene_id %in% pop_sig_genes)
  
  pop_sig.gemma.dstat = tabix2df(dstat.f,
                           tabixify(pop_sig.gemma$chr,
                                    pop_sig.gemma$pos,
                                    pop_sig.gemma$pos),
                           dstat_cols) %>%
    mutate(snp_id = paste(chr, pos, sep=":")) %>%
    merge(pop_sig.gemma, ., by="snp_id") %>%
    merge(., gene_name, by="gene_id")
  
  locus_plots = ggplot(pop_sig.gemma.dstat, aes(x=!!sym(pop), y=-log10(p.value))) +
    geom_point() +
    geom_hline(yintercept=-log10(fdr_thresh), color="red",  linetype="dashed") +
    geom_vline(xintercept=threshs[[pop]][2], color="red",  linetype="dashed") +
    geom_vline(xintercept=threshs[[pop]][1], color="blue", linetype="dashed") +
    theme_classic() +
    facet_wrap(~ name, scales="free")
  
  locus_cors = lapply(pop_sig_genes, function(gene) pop_sig.gemma.dstat %>%
                        filter(gene_id == gene) %>%
                        cor.test(formula=as.formula(paste0("~ -log10(p.value) + ", pop)), data=., alternative="greater"))
  
  eqtl_dstat_cors[[pop]] = list(locus_cor = locus_cors,
                                loc_compare =  locus_plots)
  }

```

Maps using geom_map and scatterpie

```{r}

library(maps)
library(mapdata)
library(scatterpie)

sample_info = read_csv("../misc/Table_TishkoffDB.csv") %>%
  as.data.frame %>%
  rename(sample=lab_id)

pops_1kg = read_tsv("/local1/derek/data/1kgenomes/1khg_sample_pop.tsv", col_names=c("sample","pop"))

onekg_lat_long = read_tsv("/local1/derek/data/1kgenomes/pop_locations.tsv")

long_lat = merge(pop_samps, sample_info) %>%
  dplyr::select(c(pop, long=Longitude_Consensus, lat=Latitude_Consensus)) %>%
  unique

# get frequency from eQTL pops
# frq.df = get_gtypes("6:133030710-133030710") %>%
#   merge(pop_samps) %>%
#   group_by(pop) %>%
#   summarise(alt = sum(gtype1 + gtype2),
#             ref = 2 * n() - alt)
frq.df = get_gtypes("11:61137147-61137147") %>%
  merge(pop_samps) %>%
  group_by(pop) %>%
  summarise(alt = sum(gtype1 + gtype2),
            ref = 2 * n() - alt)

# get frequency from 1KG pops
frq.1kg.cols = readLines("/local1/derek/data/1kgenomes/vcf/1kg_cols.txt")
frq.1kg.df = tabix2df("/local1/derek/data/1kgenomes/vcf/ALL.chr11.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz", "11:61137147-61137147", frq.1kg.cols) %>%
  dplyr::select(-c(3,6:9)) %>%
  unite("ID", CHROM, POS, sep=":") %>%
  gather(key=sample, value=gtype, -c(ID, REF, ALT)) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  merge(pops_1kg) %>%
  group_by(pop) %>%
  summarise(alt = sum(gtype1 + gtype2),
            ref = 2 * n() - alt) %>%
  merge(onekg_lat_long) %>%
  dplyr::rename(lat=latitude, long=longitude)

onekg_data = httr::GET('http://popgen.uchicago.edu/ggv_api/freq_table?data="1000genomes_phase3_table&rsID=rs1834640')

frq.5M.cols = readLines("/local1/derek/data/misc/5M_cols.txt")
frq.5M.df = tabix2df("/local1/derek/data/Illumina5M/imputed/11.5M.imputed.vcf.gz", "11:61137147-61137147", frq.5M.cols) %>%
  select(-c(1,2,6:9)) %>%
  gather(key=sample, value=genotype, -c(ID, REF, ALT)) %>%
  separate(genotype, into=c("gtype","dose"), sep=":") %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  merge(sample_info %>% dplyr::select(sample, pop=PopName_Consensus, long=Longitude_Consensus, lat=Latitude_Consensus)) %>%
  filter(pop %in% c("Agaw","Amhara","Argobba","Burunge","Dizi","Hadza","Hamer","Herero","Ju||Hoan","Kafa","Maasai","Mbukushu","Mursi","Naro","Sabue","Sandawe","Sheko","Tsamai","Weyto","!Xoo")) %>%
  group_by(pop) %>%
  summarise(alt = sum(gtype1 + gtype2),
            ref = 2 * n() - alt,
            long = mean(long),
            lat = mean(lat))



eafrica <- map_data("world2Hires", regions =  c('ethiopia','kenya', 'tanzania', 'uganda'))
ggplot(eafrica, aes(long, lat)) +
  geom_map(map=eafrica, aes(map_id=region), fill="grey97", color="grey") +
  geom_scatterpie(data = merge(long_lat, frq.df) %>% mutate(total=ref + alt), 
                  aes(long, lat, r = sqrt(total)/10),
                  cols = c("alt","ref"),
                  alpha = 0.5)

ggplot(eafrica, aes(long, lat)) +
  geom_map(map=eafrica, aes(map_id=region), fill="grey97", color="grey") +
  geom_point(data=long_lat, aes(x=long, y=lat, color=pop, size=2)) +
  scale_colofr_brewer("Paired")


world <- map_data("world")

#### plot of map with frequencies ####

polar_repel = function(x, y, dr){
  
  # center
  x_bar = mean(x)
  y_bar = mean(y)
  
  # convert to polar
  r = sqrt((x - x_bar)^2 + (y - y_bar)^2)
  theta = atan((y - y_bar)/(x - x_bar))
  
  x_prime = (r + dr) * cos(theta) + x_bar
  y_prime = (r + dr) * sin(theta) + y_bar
  
  data.frame(x_new = x_prime, y_new = y_prime)
  
}

repel_clusts = function(x, y, clusts, drs){
  
  x_new = x
  y_new = y
  
  for (i in 1:length(clusts)){
    
    clust = clusts[[i]]
    dr = drs[i]
    
    new_coords = polar_repel(x[clust], y[clust], dr)
    
    x_new[clust] = new_coords$x_new
    y_new[clust] = new_coords$y_new
  }
  
  data.frame(x_new = x_new,
             y_new = y_new)
  
}


eafr_clust = c(18,31,43,33,38,37,42,44,34,45,32,40,30,29,28,46)
safr_clust = c(36,39,35,27,41)
wafr_clust = c(9,26)

afr_clusts = list(eafr_clust, safr_clust, wafr_clust)

world_plot_data = frq.1kg.df  %>%
  select(-location) %>%
  rbind(frq.5M.df) %>%
  mutate(total=ref + alt)

world_plot_data_repel.coords = repel_clusts(x = world_plot_data$long,
                                            y = world_plot_data$lat,
                                            clusts = afr_clusts,
                                            drs = c(10,10,10))

world_plot_data_repel = world_plot_data
world_plot_data_repel$long = world_plot_data_repel.coords$x_new
world_plot_data_repel$lat = world_plot_data_repel.coords$y_new




ggplot(world, aes(long, lat)) +
  geom_map(map=world, aes(map_id=region), fill="grey97", color="grey") +
  geom_scatterpie(data = world_plot_data_repel, 
                  aes(long, lat, r = 5),
                  cols = c("alt","ref"),
                  alpha = 0.5) +
  coord_fixed()



dist_mat = mapply(function(x,y) sqrt(((abs(x - frq.5M.df$long) - 360) %% 360)^2 + (y - frq.5M.df$lat)^2), frq.5M.df$long, frq.5M.df$lat)

```

TAPBPL haplotypes

```{r}

# get frequency from 1KG pops
geno.1kg.cols = readLines("/local1/derek/data/1kgenomes/vcf/1kg_cols.txt")
geno.1kg.df = tabix2df("/local1/derek/data/1kgenomes/vcf/ALL.chr12.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz", "12:6461229-6660225", frq.1kg.cols) %>%
  dplyr::select(-c(3,6:9)) %>%
  unite("ID", CHROM, POS, sep=":") %>%
  gather(key=sample, value=gtype, -c(ID, REF, ALT)) %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  gather(key=allele, value=gtype, c(gtype1, gtype2)) %>%
  group_by(ID) %>%
  mutate(frq = sum(gtype)/n()) %>%
  unite("sample", sample, allele, sep=":") %>%
  pivot_wider(names_from=sample, values_from=gtype)

# onekg_data = httr::GET('http://popgen.uchicago.edu/ggv_api/freq_table?data="1000genomes_phase3_table&rsID=rs1834640')

frq.5M.cols = readLines("/local1/derek/data/misc/5M_cols.txt")
frq.5M.df = tabix2df("/local1/derek/data/Illumina5M/imputed//12.5M.imputed.vcf.gz", "12:6461229-6660225", frq.5M.cols) %>%
  dplyr::select(-c(1,2,6:9)) %>%
  gather(key=sample, value=genotype, -c(ID, REF, ALT)) %>%
  separate(genotype, into=c("gtype","dose"), sep=":") %>%
  separate(gtype, into=c("gtype1","gtype2"), sep="\\|", convert=T) %>%
  merge(sample_info %>% dplyr::select(sample, pop=PopName_Consensus, long=Longitude_Consensus, lat=Latitude_Consensus)) %>%
  filter(pop %in% c("Agaw","Amhara","Argobba","Burunge","Dizi","Hadza","Hamer","Herero","Ju||Hoan","Kafa","Maasai","Mbukushu","Mursi","Naro","Sabue","Sandawe","Sheko","Tsamai","Weyto","!Xoo")) %>%
  group_by(pop) %>%
  summarise(alt = sum(gtype1 + gtype2),
            ref = 2 * n() - alt,
            long = mean(long),
            lat = mean(lat))

```
